---
title: "HNSCC_initialise_dataset"
author: "Dillon Corvino"
date: "03/02/2020"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

### Dataset information


```r
# HNSCC scRNAseq/scTCRseq dataset
# Cells are HNSCC TILs sorted on Lymphocytes/Live/CD3+/CD4-/CD8+
# Human samples
# Each sample is a pool of 4 patients
# samples 1 and 2 are unstimulated 
# samples 3 and 4 are stimulated 
# Patients in sample 1 match patients in sample 3 and sample 2 pairs with 4
# Data acquired was transcript expression, ADT (antibody expression), and TCRA & B sequences
# This analysis uses just Transcript and TCR data


# General information
# CellRanger was used for sequence alignment/QC/counts/TCR calls - performed by Ross (QIMR, Brisbane, Aus)
# DC was provided the output from CellRanger (counts matrix) and used this as input for analysis within this script
```

### Environment


```r
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
```



```r
gc() # Free memory
```

```
##            used  (Mb) gc trigger   (Mb) limit (Mb) max used   (Mb)
## Ncells 12801007 683.7   20968521 1119.9         NA 20968521 1119.9
## Vcells 23771085 181.4   70589344  538.6      30720 70589096  538.6
```

```r
#library("rstudioapi")

# Set working directory to source file location
#setwd(dirname(getActiveDocumentContext()$path))
setwd("..")


# pipeline variables
quick.load <- FALSE
long.compute <- FALSE

source("R/Load_packages.R")
```



```r
source("R/Setup.R")
source("R/Colour_scheme.R")

# Custom functions
source("R/scRNAseq_Functions.R")
source("R/annotate_seurat_heatmap.R")
source("R/Enhanced_volcano_custom_defaults.R")
```


### Reading data


```r
# Load dataset from output generated by Ross, file is annotated as filtered_feature_bc_matrix

Sample1.data <- Read10X(data.dir = "Data/10X_GEX_data/Sample1_US/")
```

```
## Error in Read10X(data.dir = "Data/10X_GEX_data/Sample1_US/"): could not find function "Read10X"
```

```r
Sample2.data <- Read10X(data.dir = "Data/10X_GEX_data/Sample2_US/")
```

```
## Error in Read10X(data.dir = "Data/10X_GEX_data/Sample2_US/"): could not find function "Read10X"
```

```r
Sample3.data <- Read10X(data.dir = "Data/10X_GEX_data/Sample3_Stim/")
```

```
## Error in Read10X(data.dir = "Data/10X_GEX_data/Sample3_Stim/"): could not find function "Read10X"
```

```r
Sample4.data <- Read10X(data.dir = "Data/10X_GEX_data/Sample4_Stim/")
```

```
## Error in Read10X(data.dir = "Data/10X_GEX_data/Sample4_Stim/"): could not find function "Read10X"
```

```r
Sample1.seurat <- CreateSeuratObject(counts = Sample1.data, min.cells = 3, min.features = 200)
```

```
## Error in CreateSeuratObject(counts = Sample1.data, min.cells = 3, min.features = 200): could not find function "CreateSeuratObject"
```

```r
Sample2.seurat <- CreateSeuratObject(counts = Sample2.data, min.cells = 3, min.features = 200)
```

```
## Error in CreateSeuratObject(counts = Sample2.data, min.cells = 3, min.features = 200): could not find function "CreateSeuratObject"
```

```r
Sample3.seurat <- CreateSeuratObject(counts = Sample3.data, min.cells = 3, min.features = 200)
```

```
## Error in CreateSeuratObject(counts = Sample3.data, min.cells = 3, min.features = 200): could not find function "CreateSeuratObject"
```

```r
Sample4.seurat <- CreateSeuratObject(counts = Sample4.data, min.cells = 3, min.features = 200)
```

```
## Error in CreateSeuratObject(counts = Sample4.data, min.cells = 3, min.features = 200): could not find function "CreateSeuratObject"
```

```r
# Add metadata
Sample1.seurat@meta.data$group <- "S1"
```

```
## Error in Sample1.seurat@meta.data$group <- "S1": object 'Sample1.seurat' not found
```

```r
Sample2.seurat@meta.data$group <- "S2"
```

```
## Error in Sample2.seurat@meta.data$group <- "S2": object 'Sample2.seurat' not found
```

```r
Sample3.seurat@meta.data$group <- "S3"
```

```
## Error in Sample3.seurat@meta.data$group <- "S3": object 'Sample3.seurat' not found
```

```r
Sample4.seurat@meta.data$group <- "S4"
```

```
## Error in Sample4.seurat@meta.data$group <- "S4": object 'Sample4.seurat' not found
```

### Merge Seurat objects


```r
######################
# Merge samples
######################

# Unstimulated samples 1 & 2
US.seurat <- merge(x = Sample1.seurat, 
                   y = Sample2.seurat,
                   add.cell.ids = c("S1", "S2"), 
                   merge.data = FALSE, # Should normed data be merged as well or just raw data slot
                   project = "Unstimulated")
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'merge': object 'Sample1.seurat' not found
```

```r
x <- ncol(Sample1.seurat) + ncol(Sample2.seurat)
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'ncol': object 'Sample1.seurat' not found
```

```r
paste0("expected number of cells after merge is ", ncol(US.seurat) == x)
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'ncol': object 'US.seurat' not found
```

```r
Sample1.seurat # 14,695 genes and 3,042 cells
```

```
## Error in eval(expr, envir, enclos): object 'Sample1.seurat' not found
```

```r
Sample2.seurat # 14,706 genes and 3,536 cells
```

```
## Error in eval(expr, envir, enclos): object 'Sample2.seurat' not found
```

```r
US.seurat # 15,428 genes and 6,578 cells
```

```
## Error in eval(expr, envir, enclos): object 'US.seurat' not found
```

```r
head(US.seurat@meta.data)
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': object 'US.seurat' not found
```

```r
# Stimulated samples 3 & 4
Stim.seurat <- merge(x = Sample3.seurat, 
                     y = Sample4.seurat,
                     add.cell.ids = c("S3", "S4"), 
                     merge.data = FALSE, # Should normed data be merged as well or just raw data slot
                     project = "Stimulated")
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'merge': object 'Sample3.seurat' not found
```

```r
x <- ncol(Sample3.seurat) + ncol(Sample4.seurat)
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'ncol': object 'Sample3.seurat' not found
```

```r
paste0("expected number of cells after merge is ", ncol(Stim.seurat) == x)
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'ncol': object 'Stim.seurat' not found
```

```r
Sample3.seurat # 14,740 genes and 3,885 cells
```

```
## Error in eval(expr, envir, enclos): object 'Sample3.seurat' not found
```

```r
Sample4.seurat # 14,926 genes and 3,279 cells
```

```
## Error in eval(expr, envir, enclos): object 'Sample4.seurat' not found
```

```r
Stim.seurat # 15,619 genes and 7,164 cells
```

```
## Error in eval(expr, envir, enclos): object 'Stim.seurat' not found
```

```r
head(Stim.seurat@meta.data)
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': object 'Stim.seurat' not found
```

```r
# Add a condition metadata column
US.seurat@meta.data$condition <- "US"
```

```
## Error in US.seurat@meta.data$condition <- "US": object 'US.seurat' not found
```

```r
Stim.seurat@meta.data$condition <- "Stim"
```

```
## Error in Stim.seurat@meta.data$condition <- "Stim": object 'Stim.seurat' not found
```

```r
#total.seurat <- merge(x = US.seurat, 
#                   y = Stim.seurat,
#                   merge.data = FALSE, # Should normed data be merged as well or just raw data slot
#                   project = "merged_seurat")
```

### Remove large variables that are no longer needed


```r
rm(Sample1.data, Sample2.data, Sample3.data, Sample4.data, Sample1.seurat, Sample2.seurat, Sample3.seurat, Sample4.seurat)
```

## QC and Normalisation {.tabset}

### QC & Normalisation


```r
# QC metrics commonly used The number of unique genes detected in each cell.  Low-quality cells or empty droplets will often have very few genes Cell
# doublets or multiplets may exhibit an aberrantly high gene count

# Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)

# The percentage of reads that map to the mitochondrial genome Low-quality / dying cells often exhibit extensive mitochondrial contamination


################################################ Get percentage Mitochondria gene expression

# Unstimulated
US.seurat[["percent.mito"]] <- PercentageFeatureSet(US.seurat, pattern = "^MT-")
```

```
## Error in PercentageFeatureSet(US.seurat, pattern = "^MT-"): could not find function "PercentageFeatureSet"
```

```r
# Stimulated
Stim.seurat[["percent.mito"]] <- PercentageFeatureSet(Stim.seurat, pattern = "^MT-")
```

```
## Error in PercentageFeatureSet(Stim.seurat, pattern = "^MT-"): could not find function "PercentageFeatureSet"
```

```r
# total
total.seurat[["percent.mito"]] <- PercentageFeatureSet(total.seurat, pattern = "^MT-")
```

```
## Error in PercentageFeatureSet(total.seurat, pattern = "^MT-"): could not find function "PercentageFeatureSet"
```

```r
################################## Plot mitochondria percentage

# nFeature & nCount & Mito (US)
VlnPlot(object = US.seurat, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), pt.size = 0.1)
```

```
## Error in VlnPlot(object = US.seurat, features = c("nFeature_RNA", "nCount_RNA", : could not find function "VlnPlot"
```

```r
dev.copy(pdf, "output/QC/QC_nFeature_nCount_RNA_percentMito_US.pdf")
```

```
## Error in device(...): cannot open file 'output/QC/QC_nFeature_nCount_RNA_percentMito_US.pdf'
```

```r
dev.off()
```

```
## null device 
##           1
```

```r
# Mito (US)
VlnPlot(US.seurat, features = "percent.mito", y.max = 20, pt.size = 0.1)
```

```
## Error in VlnPlot(US.seurat, features = "percent.mito", y.max = 20, pt.size = 0.1): could not find function "VlnPlot"
```

```r
dev.copy(pdf, "output/QC/QC_RNA_percentMito_US.pdf")
```

```
## Error in dev.copy(pdf, "output/QC/QC_RNA_percentMito_US.pdf"): cannot copy from the null device
```

```r
dev.off()
```

```
## Error in dev.off(): cannot shut down device 1 (the null device)
```

```r
# nFeature & nCount & Mito (Stim)
VlnPlot(object = Stim.seurat, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), pt.size = 0.1)
```

```
## Error in VlnPlot(object = Stim.seurat, features = c("nFeature_RNA", "nCount_RNA", : could not find function "VlnPlot"
```

```r
dev.copy(pdf, "output/QC/QC_nFeature_nCount_RNA_percentMito_Stim.pdf")
```

```
## Error in dev.copy(pdf, "output/QC/QC_nFeature_nCount_RNA_percentMito_Stim.pdf"): cannot copy from the null device
```

```r
dev.off()
```

```
## Error in dev.off(): cannot shut down device 1 (the null device)
```

```r
# Mito (Stim)
VlnPlot(Stim.seurat, features = "percent.mito", y.max = 20, pt.size = 0.1)
```

```
## Error in VlnPlot(Stim.seurat, features = "percent.mito", y.max = 20, pt.size = 0.1): could not find function "VlnPlot"
```

```r
dev.copy(pdf, "output/QC/QC_RNA_percentMito_Stim.pdf")
```

```
## Error in dev.copy(pdf, "output/QC/QC_RNA_percentMito_Stim.pdf"): cannot copy from the null device
```

```r
dev.off()
```

```
## Error in dev.off(): cannot shut down device 1 (the null device)
```

```r
######################################## Count vs Mito & Count vs Feature

# Unstimulated
plot1 <- FeatureScatter(US.seurat, feature1 = "nCount_RNA", feature2 = "percent.mito")
```

```
## Error in FeatureScatter(US.seurat, feature1 = "nCount_RNA", feature2 = "percent.mito"): could not find function "FeatureScatter"
```

```r
plot2 <- FeatureScatter(US.seurat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```

```
## Error in FeatureScatter(US.seurat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA"): could not find function "FeatureScatter"
```

```r
plot1 | plot2
```

```
## Error in eval(expr, envir, enclos): object 'plot1' not found
```

```r
dev.copy(pdf, "output/QC/QC_RNA_percentMito_nCount_nFeature_corr_US.pdf")
```

```
## Error in dev.copy(pdf, "output/QC/QC_RNA_percentMito_nCount_nFeature_corr_US.pdf"): cannot copy from the null device
```

```r
dev.off()
```

```
## Error in dev.off(): cannot shut down device 1 (the null device)
```

```r
# Stimulated
plot1 <- FeatureScatter(Stim.seurat, feature1 = "nCount_RNA", feature2 = "percent.mito")
```

```
## Error in FeatureScatter(Stim.seurat, feature1 = "nCount_RNA", feature2 = "percent.mito"): could not find function "FeatureScatter"
```

```r
plot2 <- FeatureScatter(Stim.seurat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```

```
## Error in FeatureScatter(Stim.seurat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA"): could not find function "FeatureScatter"
```

```r
plot1 | plot2
```

```
## Error in eval(expr, envir, enclos): object 'plot1' not found
```

```r
dev.copy(pdf, "output/QC/QC_RNA_percentMito_nCount_nFeature_corr_Stim.pdf")
```

```
## Error in dev.copy(pdf, "output/QC/QC_RNA_percentMito_nCount_nFeature_corr_Stim.pdf"): cannot copy from the null device
```

```r
dev.off()
```

```
## Error in dev.off(): cannot shut down device 1 (the null device)
```

```r
############################ Structure and metadata

# Unstimulated
head(US.seurat@meta.data)
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': object 'US.seurat' not found
```

```r
str(US.seurat@meta.data)
```

```
## Error in str(US.seurat@meta.data): object 'US.seurat' not found
```

```r
# Stimulated
head(Stim.seurat@meta.data)
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': object 'Stim.seurat' not found
```

```r
str(Stim.seurat@meta.data)
```

```
## Error in str(Stim.seurat@meta.data): object 'Stim.seurat' not found
```

```r
################### Filter cells

# Seurat recommends removing cells with <200 or >2,500 genes or >10% mito genes

# Unstimulated
US.seurat.filt <- subset(US.seurat, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mito < 10)
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'subset': object 'US.seurat' not found
```

```r
# Post-filtering metrics

US.seurat  # 15,429 genes across 6,584 cells
```

```
## Error in eval(expr, envir, enclos): object 'US.seurat' not found
```

```r
US.seurat.filt  # 15,429 genes across 5,785 cells
```

```
## Error in eval(expr, envir, enclos): object 'US.seurat.filt' not found
```

```r
# Therefore
6584 - 5785
```

```
## [1] 799
```

```r
# 799 cells were deleted

# Stimulated
Stim.seurat.filt <- subset(Stim.seurat, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mito < 10)
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'subset': object 'Stim.seurat' not found
```

```r
# Post-filtering metrics

Stim.seurat  # 15,618 genes across 7,167 cells
```

```
## Error in eval(expr, envir, enclos): object 'Stim.seurat' not found
```

```r
Stim.seurat.filt  # 15,618 genes across 6,042 cells
```

```
## Error in eval(expr, envir, enclos): object 'Stim.seurat.filt' not found
```

```r
# Therefore
7167 - 6042
```

```
## [1] 1125
```

```r
# 1125 cells were deleted


# total
total.seurat.filt <- subset(total.seurat, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mito < 10)
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'subset': object 'total.seurat' not found
```

```r
# Post-filtering metrics

total.seurat  # 16,295 genes across 13,751 cells
```

```
## Error in eval(expr, envir, enclos): object 'total.seurat' not found
```

```r
total.seurat.filt  # 16,295 genes across 11,827 cells
```

```
## Error in eval(expr, envir, enclos): object 'total.seurat.filt' not found
```

```r
########################################## Visualise QC metrics after filtering

# Unstimulated

# Before Filtering
VlnPlot(object = US.seurat, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), pt.size = 0.1)
```

```
## Error in VlnPlot(object = US.seurat, features = c("nFeature_RNA", "nCount_RNA", : could not find function "VlnPlot"
```

```r
# After filtering
VlnPlot(object = US.seurat.filt, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), pt.size = 0.1)
```

```
## Error in VlnPlot(object = US.seurat.filt, features = c("nFeature_RNA", : could not find function "VlnPlot"
```

```r
dev.copy(pdf, "output/QC/QC_nFeature_nCount_RNA_percentMito_Post_filtering_US.pdf")
```

```
## Error in dev.copy(pdf, "output/QC/QC_nFeature_nCount_RNA_percentMito_Post_filtering_US.pdf"): cannot copy from the null device
```

```r
dev.off()
```

```
## Error in dev.off(): cannot shut down device 1 (the null device)
```

```r
# Stimulated

# Before Filtering
VlnPlot(object = Stim.seurat, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), pt.size = 0.1)
```

```
## Error in VlnPlot(object = Stim.seurat, features = c("nFeature_RNA", "nCount_RNA", : could not find function "VlnPlot"
```

```r
# After filtering
VlnPlot(object = Stim.seurat.filt, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), pt.size = 0.1)
```

```
## Error in VlnPlot(object = Stim.seurat.filt, features = c("nFeature_RNA", : could not find function "VlnPlot"
```

```r
dev.copy(pdf, "output/QC/QC_nFeature_nCount_RNA_percentMito_Post_filtering_Stim.pdf")
```

```
## Error in dev.copy(pdf, "output/QC/QC_nFeature_nCount_RNA_percentMito_Post_filtering_Stim.pdf"): cannot copy from the null device
```

```r
dev.off()
```

```
## Error in dev.off(): cannot shut down device 1 (the null device)
```

```r
################################################ Normalise, scale & find variable genes

################# Unstimulated

# Normalise
US.seurat.filt <- NormalizeData(object = US.seurat.filt, assay = "RNA", normalization.method = "LogNormalize", scale.factor = 10000, verbose = TRUE)
```

```
## Error in NormalizeData(object = US.seurat.filt, assay = "RNA", normalization.method = "LogNormalize", : could not find function "NormalizeData"
```

```r
# Find variable genes
US.seurat.filt <- FindVariableFeatures(object = US.seurat.filt, verbose = TRUE)
```

```
## Error in FindVariableFeatures(object = US.seurat.filt, verbose = TRUE): could not find function "FindVariableFeatures"
```

```r
# Scale data
US.seurat.filt <- ScaleData(object = US.seurat.filt, assay = "RNA", vars.to.regress = c("nCount_RNA", "percent.mito"), model.use = "linear", do.scale = TRUE, 
    do.center = TRUE, verbose = TRUE)
```

```
## Error in ScaleData(object = US.seurat.filt, assay = "RNA", vars.to.regress = c("nCount_RNA", : could not find function "ScaleData"
```

```r
################# Stimulated

# Normalise
Stim.seurat.filt <- NormalizeData(object = Stim.seurat.filt, assay = "RNA", normalization.method = "LogNormalize", scale.factor = 10000, verbose = TRUE)
```

```
## Error in NormalizeData(object = Stim.seurat.filt, assay = "RNA", normalization.method = "LogNormalize", : could not find function "NormalizeData"
```

```r
# Find variable genes
Stim.seurat.filt <- FindVariableFeatures(object = Stim.seurat.filt, verbose = TRUE)
```

```
## Error in FindVariableFeatures(object = Stim.seurat.filt, verbose = TRUE): could not find function "FindVariableFeatures"
```

```r
# Scale data
Stim.seurat.filt <- ScaleData(object = Stim.seurat.filt, assay = "RNA", vars.to.regress = c("nCount_RNA", "percent.mito"), model.use = "linear", do.scale = TRUE, 
    do.center = TRUE, verbose = TRUE)
```

```
## Error in ScaleData(object = Stim.seurat.filt, assay = "RNA", vars.to.regress = c("nCount_RNA", : could not find function "ScaleData"
```

```r
################# total

# Normalise
total.seurat.filt <- NormalizeData(object = total.seurat.filt, assay = "RNA", normalization.method = "LogNormalize", scale.factor = 10000, verbose = TRUE)
```

```
## Error in NormalizeData(object = total.seurat.filt, assay = "RNA", normalization.method = "LogNormalize", : could not find function "NormalizeData"
```

```r
# Find variable genes
total.seurat.filt <- FindVariableFeatures(object = total.seurat.filt, verbose = TRUE)
```

```
## Error in FindVariableFeatures(object = total.seurat.filt, verbose = TRUE): could not find function "FindVariableFeatures"
```

```r
# Scale data
total.seurat.filt <- ScaleData(object = total.seurat.filt, assay = "RNA", vars.to.regress = c("nCount_RNA", "percent.mito"), model.use = "linear", do.scale = TRUE, 
    do.center = TRUE, verbose = TRUE)
```

```
## Error in ScaleData(object = total.seurat.filt, assay = "RNA", vars.to.regress = c("nCount_RNA", : could not find function "ScaleData"
```

```r
############################### Remove unneeded variables

rm(US.seurat, Stim.seurat, total.seurat)
```

## Combine treatment conditions {.tabset}

### Integrate data


```r
# Set activate assay to RNA
DefaultAssay(US.seurat.filt) <- "RNA"
```

```
## Error in DefaultAssay(US.seurat.filt) <- "RNA": object 'US.seurat.filt' not found
```

```r
DefaultAssay(Stim.seurat.filt) <- "RNA"
```

```
## Error in DefaultAssay(Stim.seurat.filt) <- "RNA": object 'Stim.seurat.filt' not found
```

```r
# Create list of seurat objects
seurat.list <- list(US = US.seurat.filt, Stim = Stim.seurat.filt)
```

```
## Error in eval(expr, envir, enclos): object 'US.seurat.filt' not found
```

```r
# Find Integration Anchors
immune.anchors <- FindIntegrationAnchors(object.list = seurat.list, anchor.features = 2000, scale = TRUE, normalization.method = "LogNormalize", reduction = "cca", 
    l2.norm = TRUE, dims = 1:30, k.anchor = 5, k.filter = 200, k.score = 30, max.features = 200, nn.method = "rann", eps = 0, verbose = TRUE)
```

```
## Error in FindIntegrationAnchors(object.list = seurat.list, anchor.features = 2000, : could not find function "FindIntegrationAnchors"
```

```r
# Integrate dataset
seurat.combined <- IntegrateData(anchorset = immune.anchors, new.assay.name = "integrated", normalization.method = "LogNormalize", dims = 1:30, k.weight = 100, 
    sd.weight = 1, verbose = TRUE)
```

```
## Error in IntegrateData(anchorset = immune.anchors, new.assay.name = "integrated", : could not find function "IntegrateData"
```

```r
# Basic overview of data
seurat.combined  # 11,827 cells and 18,295 genes 
```

```
## Error in eval(expr, envir, enclos): object 'seurat.combined' not found
```

```r
US.seurat.filt  # 5,785 cells
```

```
## Error in eval(expr, envir, enclos): object 'US.seurat.filt' not found
```

```r
Stim.seurat.filt  # 6,042 cells
```

```
## Error in eval(expr, envir, enclos): object 'Stim.seurat.filt' not found
```

```r
# Save files
saveRDS(US.seurat.filt, file = "Exported_RDS_files/US_seurat_filt.rds")
```

```
## Error in saveRDS(US.seurat.filt, file = "Exported_RDS_files/US_seurat_filt.rds"): object 'US.seurat.filt' not found
```

```r
saveRDS(Stim.seurat.filt, file = "Exported_RDS_files/Stim_seurat_filt.rds")
```

```
## Error in saveRDS(Stim.seurat.filt, file = "Exported_RDS_files/Stim_seurat_filt.rds"): object 'Stim.seurat.filt' not found
```

```r
# Remove unnecessary data
rm(US.seurat.filt)
rm(Stim.seurat.filt)


# Format metadata in integrated seurat object
seurat.combined@meta.data$condition <- factor(seurat.combined@meta.data$condition, levels = c("US", "Stim"))
```

```
## Error in factor(seurat.combined@meta.data$condition, levels = c("US", : object 'seurat.combined' not found
```

```r
# Remove unneeded data
rm(immune.anchors, seurat.list)
```

## Clustering and dim reduction {.tabset}

### PCA and cluster identificaton


```r
DefaultAssay(seurat.combined) <- "integrated"
```

```
## Error in DefaultAssay(seurat.combined) <- "integrated": object 'seurat.combined' not found
```

```r
# Scale data
seurat.combined <- ScaleData(seurat.combined, verbose = TRUE)
```

```
## Error in ScaleData(seurat.combined, verbose = TRUE): could not find function "ScaleData"
```

```r
# Run PCA
seurat.combined <- RunPCA(seurat.combined, npcs = 30, verbose = TRUE)
```

```
## Error in RunPCA(seurat.combined, npcs = 30, verbose = TRUE): could not find function "RunPCA"
```

```r
# Find Neighbors
seurat.combined <- FindNeighbors(seurat.combined, reduction = "pca", k.param = 20, dims = 1:20)
```

```
## Error in FindNeighbors(seurat.combined, reduction = "pca", k.param = 20, : could not find function "FindNeighbors"
```

```r
# Find clusters
seurat.combined <- FindClusters(seurat.combined, random.seed = 42, resolution = 0.4)  # 0.4 is cluster resolution decided upon
```

```
## Error in FindClusters(seurat.combined, random.seed = 42, resolution = 0.4): could not find function "FindClusters"
```

### Run UMAP


```r
# Create output directory
if(!dir.exists("output/figures/UMAP_no_clusts_removed")){
  dir.create("output/figures/UMAP_no_clusts_removed", 
             recursive = T)
}


# Run UMAP
seurat.combined <- RunUMAP(object = seurat.combined,
                           reduction = "pca",
                           dims = 1:20,
                           umap.method = "uwot",
                           n.neighbors = 30, # 5 to 50
                           min.dist = 0.3, # Sensible values are in the range 0.001 to 0.5
                           seed.use = 42)
```

```
## Error in RunUMAP(object = seurat.combined, reduction = "pca", dims = 1:20, : could not find function "RunUMAP"
```

```r
# Plot UMAP projection
UMAPPlot(object = seurat.combined,
         label = TRUE, 
         label.size = 6) + ggtitle("Integrated (n.neigh = 30 & min.dist = 0.3)") + NoLegend()
```

```
## Error in UMAPPlot(object = seurat.combined, label = TRUE, label.size = 6): could not find function "UMAPPlot"
```

```r
dev.copy(pdf, "output/figures/UMAP_no_clusts_removed/UMAP_clusts_clust.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
# UMAP projection grouped by condition
UMAPPlot(object = seurat.combined,
         label = FALSE, 
         cols = Condition.cols,
         group.by = "condition") + ggtitle("UMAP vis by Condition") 
```

```
## Error in UMAPPlot(object = seurat.combined, label = FALSE, cols = Condition.cols, : could not find function "UMAPPlot"
```

```r
dev.copy(pdf, "output/figures/UMAP_no_clusts_removed/UMAP_condition.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
p1 <- DimPlot(seurat.combined, 
              reduction = "umap", 
              cols = Condition.cols,
              group.by = "condition")
```

```
## Error in DimPlot(seurat.combined, reduction = "umap", cols = Condition.cols, : could not find function "DimPlot"
```

```r
p2 <- DimPlot(seurat.combined, 
              reduction = "umap", 
              label = TRUE)
```

```
## Error in DimPlot(seurat.combined, reduction = "umap", label = TRUE): could not find function "DimPlot"
```

```r
p1|p2
```

```
## Error in eval(expr, envir, enclos): object 'p1' not found
```

```r
dev.copy(pdf, "output/figures/UMAP_no_clusts_removed/UMAP_clusters_conditions_sidebyside.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

## Remove uninformative clusters {.tabset}

### Visualise and justify removing clusters


```r
# Based on resolution of 0.4 for cluster identification Cluster #15 and 13 are to be removed 15 = Myeloid population 13 = ??


# Create output directory
if (!dir.exists("output/figures/Cluster_removal_validation")) {
    dir.create("output/figures/Cluster_removal_validation", recursive = T)
}


# Show overall number and frequency of clusters

x <- table(seurat.combined@meta.data$seurat_clusters)
```

```
## Error in eval(quote(list(...)), env): object 'seurat.combined' not found
```

```r
y <- prop.table(table(seurat.combined@meta.data$seurat_clusters)) * 100
```

```
## Error in eval(quote(list(...)), env): object 'seurat.combined' not found
```

```r
barplot(x, main = "# of cells per cluster", xlab = "Cluster #", ylab = "# of cells", cex.names = 0.8, ylim = c(0, 2000))
```

```
## Error in barplot(x, main = "# of cells per cluster", xlab = "Cluster #", : object 'x' not found
```

```r
dev.copy(pdf, "output/figures/Cluster_removal_validation/Number_cells_per_cluster.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
barplot(y, main = "% of cells per cluster", xlab = "Cluster #", ylab = "% of cells", cex.names = 0.8, ylim = c(0, 20))
```

```
## Error in barplot(y, main = "% of cells per cluster", xlab = "Cluster #", : object 'y' not found
```

```r
dev.copy(pdf, "output/figures/Cluster_removal_validation/Percent_cells_per_cluster.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
########################################################################################## Get differentially expressed genes for clusters to be removed and show heatmaps

# Set assay to RNA
DefaultAssay(seurat.combined) <- "RNA"
```

```
## Error in DefaultAssay(seurat.combined) <- "RNA": object 'seurat.combined' not found
```

```r
################ Positive only

# Find markers for clust 13
Clust.13.Markers <- FindMarkers(seurat.combined, assay = "RNA", ident.1 = "13", only.pos = TRUE, min.diff.pct = 0.1, return.thresh = 0.05)
```

```
## Error in FindMarkers(seurat.combined, assay = "RNA", ident.1 = "13", only.pos = TRUE, : could not find function "FindMarkers"
```

```r
# Find markers for clust 15
Clust.15.Markers <- FindMarkers(seurat.combined, assay = "RNA", ident.1 = "15", only.pos = TRUE, min.diff.pct = 0.1, return.thresh = 0.05)
```

```
## Error in FindMarkers(seurat.combined, assay = "RNA", ident.1 = "15", only.pos = TRUE, : could not find function "FindMarkers"
```

```r
################### Heatmap vis

# Downsample for heatmap vis seurat.combined.small <- subset(seurat.combined, downsample = 300)

# Heatmap of Clust 13 markers
sig.genes <- Clust.13.Markers[Clust.13.Markers$p_val_adj < 0.1, ]
```

```
## Error in eval(expr, envir, enclos): object 'Clust.13.Markers' not found
```

```r
# Heatmap of Clust 15 markers
sig.genes <- Clust.15.Markers[Clust.15.Markers$p_val_adj < 0.1, ]
```

```
## Error in eval(expr, envir, enclos): object 'Clust.15.Markers' not found
```

```r
# Plot DEGs as heatmap for vis justification for removal
```

### Use Clustifyr to call cluster identities

**Note: What is up with the mouse.ref here?**


```r
# Create output directories
if(!dir.exists("output/figures/Cell_type_annotation/Clustifyr")){
  dir.create("output/figures/Cell_type_annotation/Clustifyr", 
             recursive = T)
}


mouse.ref <- cbmc_ref
rownames(mouse.ref) <- str_to_title(rownames(mouse.ref))


res <- clustify(input = seurat.combined,
                cluster_col = "seurat_clusters",
                ref_mat = cbmc_ref,
                seurat_out = FALSE,
                query_genes = seurat.combined@assays$integrated@var.features)
```

```
## Error in clustify(input = seurat.combined, cluster_col = "seurat_clusters", : object 'seurat.combined' not found
```

```r
res2 <- cor_to_call(cor_mat = res,                  # matrix correlation coefficients
                    cluster_col = "seurat_clusters") # name of column in meta.data containing cell clusters
```

```
## Error in cor_to_call(cor_mat = res, cluster_col = "seurat_clusters"): object 'res' not found
```

```r
print(plot_cor_heatmap(cor_mat = res))
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'print': object 'res' not found
```

```r
dev.copy(pdf, "output/figures/Cell_type_annotation/Clustifyr/Heatmap_cluster_annotations.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
# Remove unneeded data
rm(mouse.ref, res, res2)
```

### Use various SingleR databases to call cluster identities


```r
# Create output directory
if (!dir.exists("output/figures/Cell_type_annotation/SingleR")) {
    dir.create("output/figures/Cell_type_annotation/SingleR", recursive = T)
}

################################################################ Use a couple databased to get broad overview of clusters

################################## Use HumanPrimaryCellAtlasData Info: Human, Microarray, 713 samples, 37 main labels, 157 fine lables, Non-specific focus

HumanPrimaryCellAtlasData.data <- celldex::HumanPrimaryCellAtlasData()
```

```
## Error in loadNamespace(name): there is no package called 'celldex'
```

```r
# First calcualte cell ID using all available labels
SingleR.pred <- SingleR(test = seurat.combined@assays$integrated@data, ref = HumanPrimaryCellAtlasData.data, method = "cluster", clusters = seurat.combined@meta.data$seurat_clusters, 
    labels = HumanPrimaryCellAtlasData.data$label.fine)
```

```
## Error in rownames(x): object 'seurat.combined' not found
```

```r
# Plot
plotScoreHeatmap(SingleR.pred, show_colnames = TRUE)
```

```
## Error in rownames(results): object 'SingleR.pred' not found
```

```r
dev.copy(pdf, "output/figures/Cell_type_annotation/SingleR/Heatmap_cluster_annotations_all_labels_HumanPrimaryCellAtlasData_db.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
########################## Use BlueprintEncodeData Info: Human, RNAseq, 259 samples, 24 main labels, 43 fine lables, Non-specific focus

BlueprintEncodeData.data <- celldex::BlueprintEncodeData()
```

```
## Error in loadNamespace(name): there is no package called 'celldex'
```

```r
# First calcualte cell ID using all available labels
SingleR.pred <- SingleR(test = seurat.combined@assays$integrated@data, ref = BlueprintEncodeData.data, method = "cluster", clusters = seurat.combined@meta.data$seurat_clusters, 
    labels = BlueprintEncodeData.data$label.fine)
```

```
## Error in rownames(x): object 'seurat.combined' not found
```

```r
# Plot
plotScoreHeatmap(SingleR.pred, show_colnames = TRUE)
```

```
## Error in rownames(results): object 'SingleR.pred' not found
```

```r
dev.copy(pdf, "output/figures/Cell_type_annotation/SingleR/Heatmap_cluster_annotations_all_labels_BlueprintEncodeData_db.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
########################################## Use NovershternHematopoieticData Info: Human, Microarray, 211 samples, 17 main labels, 38 fine lables, Hematopoietic & Immune focus

NovershternHematopoieticData.data <- celldex::NovershternHematopoieticData()
```

```
## Error in loadNamespace(name): there is no package called 'celldex'
```

```r
# First calcualte cell ID using all available labels
SingleR.pred <- SingleR(test = seurat.combined@assays$integrated@data, ref = NovershternHematopoieticData.data, method = "cluster", clusters = seurat.combined@meta.data$seurat_clusters, 
    labels = NovershternHematopoieticData.data$label.fine)
```

```
## Error in rownames(x): object 'seurat.combined' not found
```

```r
# Plot
plotScoreHeatmap(SingleR.pred, show_colnames = TRUE)
```

```
## Error in rownames(results): object 'SingleR.pred' not found
```

```r
dev.copy(pdf, "output/figures/Cell_type_annotation/SingleR/Heatmap_cluster_annotations_all_labels_NovershternHematopoietic_db.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
##################### Use ImmGenData Info: Mouse, Microarray, 830 samples, 20 main labels, 253 fine lables, Hematopoietic & Immune focus

ImmGen.data <- celldex::ImmGenData()
```

```
## Error in loadNamespace(name): there is no package called 'celldex'
```

```r
# Convert GeneID to match human
ImmGen.data.counts <- ImmGen.data@assays@data$logcounts
```

```
## Error in eval(expr, envir, enclos): object 'ImmGen.data' not found
```

```r
rownames(ImmGen.data.counts) <- toupper(rownames(ImmGen.data.counts))
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'toupper': error in evaluating the argument 'x' in selecting a method for function 'rownames': object 'ImmGen.data.counts' not found
```

```r
# First calcualte cell ID using all available labels
SingleR.pred <- SingleR(test = seurat.combined@assays$integrated@data, ref = ImmGen.data.counts, method = "cluster", clusters = seurat.combined@meta.data$seurat_clusters, 
    labels = ImmGen.data$label.fine)
```

```
## Error in rownames(x): object 'seurat.combined' not found
```

```r
# Plot
plotScoreHeatmap(SingleR.pred, show_colnames = TRUE)
```

```
## Error in rownames(results): object 'SingleR.pred' not found
```

```r
dev.copy(pdf, "output/figures/Cell_type_annotation/SingleR/Heatmap_cluster_annotations_all_labels_ImmGen_db.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
################################################################ Most useful database is probably the MonacoImmuneData

######################### Use MonacoImmuneData Info: Human, RNAseq, 114 samples, 11 main labels, 29 fine lables, Immune cell focus
Immune.data <- celldex::MonacoImmuneData()
```

```
## Error in loadNamespace(name): there is no package called 'celldex'
```

```r
# First calcualte cell ID using all available labels
SingleR.pred <- SingleR(test = seurat.combined@assays$integrated@data, ref = Immune.data, method = "cluster", clusters = seurat.combined@meta.data$seurat_clusters, 
    labels = Immune.data$label.fine)
```

```
## Error in rownames(x): object 'seurat.combined' not found
```

```r
# Plot
plotScoreHeatmap(SingleR.pred, show_colnames = TRUE)
```

```
## Error in rownames(results): object 'SingleR.pred' not found
```

```r
dev.copy(pdf, "output/figures/Cell_type_annotation/SingleR/Heatmap_cluster_annotations_all_labels_Monaco_db.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
# Restrict labels to biologicaly relevant
Immune.data.counts <- Immune.data@assays@data$logcounts
```

```
## Error in eval(expr, envir, enclos): object 'Immune.data' not found
```

```r
pattern <- c("CD8", "MAIT", "gd", "Intermediate")

labels.keep <- grepl(paste0(pattern, collapse = "|"), Immune.data$label.fine)
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'grepl': object 'Immune.data' not found
```

```r
Immune.data.counts <- Immune.data.counts[, labels.keep]
```

```
## Error in eval(expr, envir, enclos): object 'Immune.data.counts' not found
```

```r
SingleR.pred <- SingleR(test = seurat.combined@assays$integrated@data, ref = Immune.data.counts, method = "cluster", clusters = seurat.combined@meta.data$seurat_clusters, 
    labels = Immune.data$label.fine[labels.keep])
```

```
## Error in rownames(x): object 'seurat.combined' not found
```

```r
# Plot
plotScoreHeatmap(SingleR.pred, show_colnames = TRUE)
```

```
## Error in rownames(results): object 'SingleR.pred' not found
```

```r
dev.copy(pdf, "output/figures/Cell_type_annotation/SingleR/Heatmap_cluster_annotations_limited_labels_Monaco_db.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
# Remove unneeded datasets
rm(HumanPrimaryCellAtlasData.data, BlueprintEncodeData.data, NovershternHematopoieticData.data, ImmGen.data, ImmGen.data.counts, Immune.data, Immune.data.counts, 
    SingleR.pred)
```

### Myeloid and Mitochondrial populations


```r
# Remove cluster 13 and 15
seurat.combined <- subset(seurat.combined, idents = c(0:12, 14))
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'subset': object 'seurat.combined' not found
```

```r
# Check clusters are removed
levels(seurat.combined@meta.data$seurat_clusters)
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'levels': object 'seurat.combined' not found
```

```r
# Save old cluster IDs
seurat.combined@meta.data$seurat_clusters_old <- seurat.combined@meta.data$seurat_clusters
```

```
## Error in eval(expr, envir, enclos): object 'seurat.combined' not found
```

## Recalculate UMAP and Visualise {.tabset}

### Recalculate UMAP


```r
###########################
# Recalculate UMAP
###########################

DefaultAssay(seurat.combined) <- "integrated"
```

```
## Error in DefaultAssay(seurat.combined) <- "integrated": object 'seurat.combined' not found
```

```r
# Scale data
seurat.combined <- ScaleData(seurat.combined, 
                             verbose = TRUE)
```

```
## Error in ScaleData(seurat.combined, verbose = TRUE): could not find function "ScaleData"
```

```r
# Run PCA
seurat.combined <- RunPCA(seurat.combined, 
                          npcs = 30, 
                          verbose = TRUE)
```

```
## Error in RunPCA(seurat.combined, npcs = 30, verbose = TRUE): could not find function "RunPCA"
```

```r
# Find Neighbors
seurat.combined <- FindNeighbors(seurat.combined, 
                                 reduction = "pca", 
                                 k.param = 20,
                                 dims = 1:20)
```

```
## Error in FindNeighbors(seurat.combined, reduction = "pca", k.param = 20, : could not find function "FindNeighbors"
```

```r
# Find clusters
seurat.combined <- FindClusters(seurat.combined, 
                                random.seed = 42,
                                resolution = 0.4)
```

```
## Error in FindClusters(seurat.combined, random.seed = 42, resolution = 0.4): could not find function "FindClusters"
```

```r
# Run UMAP
seurat.combined <- RunUMAP(object = seurat.combined,
                           reduction = "pca",
                           dims = 1:20,
                           umap.method = "uwot",
                           n.neighbors = 30, # 5 to 50
                           min.dist = 0.3, # Sensible values are in the range 0.001 to 0.5
                           seed.use = 42)
```

```
## Error in RunUMAP(object = seurat.combined, reduction = "pca", dims = 1:20, : could not find function "RunUMAP"
```

### Vis new UMAP


```r
# Create output directory
if (!dir.exists("output/figures/UMAP")) {
    dir.create("output/figures/UMAP", recursive = T)
}


Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters_new
```

```
## Error in eval(expr, envir, enclos): object 'seurat.combined' not found
```

```r
# Plot UMAP projection
UMAPPlot(object = seurat.combined, label = TRUE, label.size = 6) + ggtitle("UMAP new projection and clust IDs")
```

```
## Error in UMAPPlot(object = seurat.combined, label = TRUE, label.size = 6): could not find function "UMAPPlot"
```

```r
dev.copy(pdf, "output/figures/UMAP/UMAP_new_clust_IDs_numbered.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
# Plot UMAP projection with old clust IDs
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters_old
```

```
## Error in eval(expr, envir, enclos): object 'seurat.combined' not found
```

```r
UMAPPlot(object = seurat.combined, label = TRUE, label.size = 6) + ggtitle("UMAP new projection - original clust IDs")
```

```
## Error in UMAPPlot(object = seurat.combined, label = TRUE, label.size = 6): could not find function "UMAPPlot"
```

```r
dev.copy(pdf, "output/figures/UMAP/UMAP_original_clust_IDs_numbered.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

### Rename clusters


```r
# Name clusters

# Save cluster IDs - new ids = those after removal of clusters
seurat.combined@meta.data$seurat_clusters_new <- seurat.combined@meta.data$seurat_clusters
```

```
## Error in eval(expr, envir, enclos): object 'seurat.combined' not found
```

```r
# Set Idents to cluster IDs following cluster removal 
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters
```

```
## Error in eval(expr, envir, enclos): object 'seurat.combined' not found
```

```r
# Rename classes.
seurat.combined <- RenameIdents(object = seurat.combined,
                                `0` = "Cytotoxic",
                                `1` = "Naive_like_3",
                                `2` = "Exhausted_1", 
                                `3` = "Type_I_IFN",
                                `4` = "Naive_like_1_CM", # Central memory?
                                `5` = "Naive_like_2_SC", # Stem-cell-like
                                `6` = "Stimulated_1", # Derived from cytotoxic ?
                                `7` = "Stimulated_exhausted",
                                `8` = "gd_T_non_g9d2", 
                                `9` = "Exhausted_2",
                                `10` = "MAIT", 
                                `11` = "TRM", 
                                `12` = "gd_T_g9d2", 
                                `13` = "Proliferative")
```

```
## Error in RenameIdents(object = seurat.combined, `0` = "Cytotoxic", `1` = "Naive_like_3", : could not find function "RenameIdents"
```

```r
# Relevel idents variable to group clusters into order of future modules 

# Module: Naive / Stem-like 
# Naïve_like_1_CM
# Naïve_like_2_SC
# Naïve_like_3

# Module: Activated
# Cytotoxic
# Type_I_IFN
# Stimulated_1

# Module: Exhaustion
# Stimulated_exhausted
# Exhausted_1, 
# Exhausted_2
# TRM

# Module: Innate
# gd_T_g9d2
# gd_T_non_g9d2 
# MAIT 

# Module: Cycling
# Proliferative


# reset levels of factor variable 
seurat.combined@active.ident <- factor(seurat.combined@active.ident, 
                                       levels = c("Naive_like_1_CM",
                                                  "Naive_like_2_SC", 
                                                  "Naive_like_3", 
                                                  "Cytotoxic",
                                                  "Type_I_IFN",
                                                  "Stimulated_1",
                                                  "Stimulated_exhausted",
                                                  "Exhausted_1",
                                                  "Exhausted_2", 
                                                  "TRM", 
                                                  "gd_T_g9d2", 
                                                  "gd_T_non_g9d2", 
                                                  "MAIT",
                                                  "Proliferative"))
```

```
## Error in factor(seurat.combined@active.ident, levels = c("Naive_like_1_CM", : object 'seurat.combined' not found
```

```r
# Change "seurat_clusters" metadata label to new ident names as this slot is used in downstream plotting
seurat.combined@meta.data$seurat_clusters <- Idents(seurat.combined)
```

```
## Error in Idents(seurat.combined): could not find function "Idents"
```

### UMAP with Named clusters


```r
# Create output directory
if (!dir.exists("output/figures/UMAP")) {
    dir.create("output/figures/UMAP", recursive = T)
}


########################################## Plot UMAPs with cluster names


Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters
```

```
## Error in eval(expr, envir, enclos): object 'seurat.combined' not found
```

```r
# UMAP named clusters
UMAPPlot(object = seurat.combined, pt.size = 1, label = FALSE, cols = clust.cols) + ggtitle("UMAP named clusters")
```

```
## Error in UMAPPlot(object = seurat.combined, pt.size = 1, label = FALSE, : could not find function "UMAPPlot"
```

```r
dev.copy(pdf, "output/figures/UMAP/UMAP_clusters.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
# UMAP named clusters - labeled
UMAPPlot(object = seurat.combined, label = TRUE, pt.size = 1, label.size = 6, cols = clust.cols) + NoLegend() + ggtitle("UMAP named clusters")
```

```
## Error in UMAPPlot(object = seurat.combined, label = TRUE, pt.size = 1, : could not find function "UMAPPlot"
```

```r
dev.copy(pdf, "output/figures/UMAP/UMAP_clusters_labeled.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
############# Condition

# Split by condition
UMAPPlot(object = seurat.combined, label = FALSE, pt.size = 1, split.by = "condition", label.size = 6, cols = clust.cols) + ggtitle("UMAP split by condition")
```

```
## Error in UMAPPlot(object = seurat.combined, label = FALSE, pt.size = 1, : could not find function "UMAPPlot"
```

```r
dev.copy(pdf, "output/figures/UMAP/UMAP_splitby_conditions.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
# Split by condition // No legend
UMAPPlot(object = seurat.combined, label = FALSE, pt.size = 1, split.by = "condition", label.size = 6, cols = clust.cols) + ggtitle("UMAP split by condition") + 
    NoLegend()
```

```
## Error in UMAPPlot(object = seurat.combined, label = FALSE, pt.size = 1, : could not find function "UMAPPlot"
```

```r
dev.copy(pdf, "output/figures/UMAP/UMAP_splitby_conditions_Nolegend.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
# Group by condition
UMAPPlot(object = seurat.combined, label = FALSE, pt.size = 1, group.by = "condition", cols = Condition.cols, label.size = 6) + ggtitle("UMAP group by condition")
```

```
## Error in UMAPPlot(object = seurat.combined, label = FALSE, pt.size = 1, : could not find function "UMAPPlot"
```

```r
dev.copy(pdf, "output/figures/UMAP/UMAP_groupby_conditions.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
############# Sample

# Split by Sample
UMAPPlot(object = seurat.combined, label = FALSE, pt.size = 1, split.by = "group", label.size = 6, cols = clust.cols) + ggtitle("UMAP split by Sample")
```

```
## Error in UMAPPlot(object = seurat.combined, label = FALSE, pt.size = 1, : could not find function "UMAPPlot"
```

```r
dev.copy(pdf, "output/figures/UMAP/UMAP_splitby_sample.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

```r
# Group by Sample
UMAPPlot(object = seurat.combined, label = FALSE, group.by = "group", pt.size = 1, label.size = 6) + ggtitle("UMAP group by Sample")
```

```
## Error in UMAPPlot(object = seurat.combined, label = FALSE, group.by = "group", : could not find function "UMAPPlot"
```

```r
dev.copy(pdf, "output/figures/UMAP/UMAP_groupby_sample.pdf")
```

```
## pdf 
##   3
```

```r
dev.off()
```

```
## quartz_off_screen 
##                 2
```

### Saving dim reduction Embeddings to metadata


```r
# UMAP embeddings
seurat.combined <- AddMetaData(seurat.combined, seurat.combined@reductions$umap@cell.embeddings[, 1], "UMAP_1")
```

```
## Error in AddMetaData(seurat.combined, seurat.combined@reductions$umap@cell.embeddings[, : could not find function "AddMetaData"
```

```r
seurat.combined <- AddMetaData(seurat.combined, seurat.combined@reductions$umap@cell.embeddings[, 2], "UMAP_2")
```

```
## Error in AddMetaData(seurat.combined, seurat.combined@reductions$umap@cell.embeddings[, : could not find function "AddMetaData"
```

```r
# Save PCA (Dim 1 and 2) embeddings
seurat.combined <- AddMetaData(seurat.combined, seurat.combined@reductions$pca@cell.embeddings[, 1], "PCA_1")
```

```
## Error in AddMetaData(seurat.combined, seurat.combined@reductions$pca@cell.embeddings[, : could not find function "AddMetaData"
```

```r
seurat.combined <- AddMetaData(seurat.combined, seurat.combined@reductions$pca@cell.embeddings[, 2], "PCA_2")
```

```
## Error in AddMetaData(seurat.combined, seurat.combined@reductions$pca@cell.embeddings[, : could not find function "AddMetaData"
```

```r
head(seurat.combined@meta.data)
```

```
## Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': object 'seurat.combined' not found
```

### Create Cluster by condition metadata


```r
###################################################### Create additional Metadata for downstream analysis Now that clusters are cleaned and renamed Generate a joint cluster & condition_ID variable

# Using named clusters
seurat.combined@meta.data$condition_clust <- paste(seurat.combined@meta.data$condition, seurat.combined@meta.data$seurat_clusters, sep = "_")
```

```
## Error in eval(quote(list(...)), env): object 'seurat.combined' not found
```

## Export large data {.tabset}

### Export seurat object


```r
if (!quick.load) {
    saveRDS(seurat.combined, file = "Exported_RDS_files/seurat_combined_no_imputation.rds")
}
```

```
## Error in eval(expr, envir, enclos): object 'quick.load' not found
```

### Save large data tables


```r
if (long.compute) {
    
    
    # Create output directory
    if (!dir.exists("output/tables/Large_dataframes")) {
        dir.create("output/tables/Large_dataframes", recursive = T)
    }
    
    
    # Raw data
    save.data.frame.function(seurat.combined@assays$RNA@counts, path = "output/tables/Large_dataframes/", title = "Raw_dataframe")
    
    # Normed data
    save.data.frame.function(seurat.combined@assays$RNA@data, path = "output/tables/Large_dataframes/", title = "filtered_dataframe")
    
    # Scaled data
    save.data.frame.function(seurat.combined@assays$RNA@scale.data, path = "output/tables/Large_dataframes/", title = "scaled_dataframe")
    
    # PCA embeddings
    save.data.frame.function(seurat.combined@reductions$pca@cell.embeddings, path = "output/tables/Large_dataframes/", title = "RNA_PCA")
    
    # UMAP embeddings
    save.data.frame.function(seurat.combined@reductions$umap@cell.embeddings, path = "output/tables/Large_dataframes/", title = "RNA_UMAP")
    
    # Metadata
    save.data.frame.function(seurat.combined@meta.data, path = "output/tables/Large_dataframes/", title = "Meta_data_dataframe")
    
    # Integrated normed values
    save.data.frame.function(seurat.combined@assays$integrated@data, path = "output/tables/Large_dataframes/", title = "filtered_integrated_dataframe")
    
    # Integrated scaled data
    save.data.frame.function(seurat.combined@assays$integrated@scale.data, path = "output/tables/Large_dataframes/", title = "scaled_integrated_dataframe")
    
}
```

## Imputation {.tabset}

### Imputation calculation


```r
###################### Impute values
if (long.compute) {
    seurat.combined <- RunALRA(seurat.combined, genes.use = rownames(seurat.combined))
}
```

```
## Error in eval(expr, envir, enclos): object 'long.compute' not found
```

```r
# 8.37% of the values became negative in the scaling process and were set to zero The matrix went from 29.63% nonzero to 38.95% nonzero


#################### Calculate k value
if (long.compute) {
    
    Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters
    seurat.combined.small <- subset(seurat.combined, downsample = 100)
    
    ALRA.out <- RunALRA(seurat.combined.small, k.only = TRUE)
    
    ggouts <- ALRAChooseKPlot(ALRA.out)
    
    ggouts
    dev.copy(pdf, "output/QC/Imputation_K_plot.pdf")
    dev.off()
}
```

```
## Error in eval(expr, envir, enclos): object 'long.compute' not found
```

```r
# set default assay back to RNA
DefaultAssay(seurat.combined) <- "RNA"
```

```
## Error in DefaultAssay(seurat.combined) <- "RNA": object 'seurat.combined' not found
```

```r
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters
```

```
## Error in eval(expr, envir, enclos): object 'seurat.combined' not found
```

## Export large data after imputation {.tabset}

### Export seurat object with imputation


```r
if (!quick.load) {
    saveRDS(seurat.combined, file = "Exported_RDS_files/seurat_combined.rds")
}
```

```
## Error in eval(expr, envir, enclos): object 'quick.load' not found
```

```r
if (long.compute) {
    
    
    # Create output directory
    if (!dir.exists("output/tables/Large_dataframes")) {
        dir.create("output/tables/Large_dataframes", recursive = T)
    }
    
    
    # Imputed values
    save.data.frame.function(seurat.combined@assays$alra@counts, path = "output/tables/Large_dataframes/", title = "Imputed_counts_dataframe")
    
}
```

```
## Error in eval(expr, envir, enclos): object 'long.compute' not found
```



