---
title: "HNSCC_scRNAseq_analysis"
author: "Dillon Corvino"
date: "03/02/2020"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

### Dataset information


```r
# HNSCC scRNAseq/scTCRseq dataset Cells are HNSCC TILs sorted on Lymphocytes/Live/CD3+/CD4-/CD8+ Human samples Each sample is a pool of 4 patients samples
# 1 and 2 are unstimulated samples 3 and 4 are stimulated Patients in sample 1 match patients in sample 3 and sample 2 pairs with 4 Data acquired was
# transcript expression, ADT (antibody expression), and TCRA & B sequences This analysis uses just Transcript and TCR data


# General information CellRanger was used for sequence alignment/QC/counts/TCR calls - performed by Ross (QIMR, Brisbane, Aus) DC was provided the output
# from CellRanger (counts matrix) and used this as input for analysis within this script
```

### Environment


```r
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, eval = TRUE, tidy = TRUE)

# Environment Set up
rm(list = ls())  #Clean workspace
cat("\f")  #Clean Console
```



```r
gc()  # Free memory
```

```
##             used   (Mb) gc trigger   (Mb) limit (Mb)   max used    (Mb)
## Ncells  14932032  797.5   23281922 1243.4         NA   23281922  1243.4
## Vcells 165645922 1263.8  539999387 4119.9      30720 1318351562 10058.3
```

```r
# Set working directory to source file location
setwd("..")


source("R/Setup.R")
```



```
## Error in library("clustifyr"): there is no package called 'clustifyr'
```

```r
source("R/Colour_scheme.R")


# pipeline variables
quick.load <- TRUE
long.compute <- TRUE


# load saved seurat object with cluster annotations and imputation performed
if (quick.load) {
    seurat.combined <- readRDS("Exported_RDS_files/seurat_combined.rds")
}
```


### Reproducibility

```r
# Only run once to initialise renv::init()

# Run snapshot to update renv.lock file renv::snapshot()


# use to restore environment renv::restore()
```
 

### Reading data


```r
# Load dataset from output generated by Ross, file is annotated as filtered_feature_bc_matrix

Sample1.data <- Read10X(data.dir = "Data/10X_GEX_data/Sample1_US/")
```

```
## Error in Read10X(data.dir = "Data/10X_GEX_data/Sample1_US/"): Directory provided does not exist
```

```r
Sample2.data <- Read10X(data.dir = "Data/10X_GEX_data/Sample2_US/")
```

```
## Error in Read10X(data.dir = "Data/10X_GEX_data/Sample2_US/"): Directory provided does not exist
```

```r
Sample3.data <- Read10X(data.dir = "Data/10X_GEX_data/Sample3_Stim/")
```

```
## Error in Read10X(data.dir = "Data/10X_GEX_data/Sample3_Stim/"): Directory provided does not exist
```

```r
Sample4.data <- Read10X(data.dir = "Data/10X_GEX_data/Sample4_Stim/")
```

```
## Error in Read10X(data.dir = "Data/10X_GEX_data/Sample4_Stim/"): Directory provided does not exist
```

```r
Sample1.seurat <- CreateSeuratObject(counts = Sample1.data, min.cells = 3, min.features = 200)
```

```
## Error in CreateSeuratObject(counts = Sample1.data, min.cells = 3, min.features = 200): object 'Sample1.data' not found
```

```r
Sample2.seurat <- CreateSeuratObject(counts = Sample2.data, min.cells = 3, min.features = 200)
```

```
## Error in CreateSeuratObject(counts = Sample2.data, min.cells = 3, min.features = 200): object 'Sample2.data' not found
```

```r
Sample3.seurat <- CreateSeuratObject(counts = Sample3.data, min.cells = 3, min.features = 200)
```

```
## Error in CreateSeuratObject(counts = Sample3.data, min.cells = 3, min.features = 200): object 'Sample3.data' not found
```

```r
Sample4.seurat <- CreateSeuratObject(counts = Sample4.data, min.cells = 3, min.features = 200)
```

```
## Error in CreateSeuratObject(counts = Sample4.data, min.cells = 3, min.features = 200): object 'Sample4.data' not found
```

```r
# Add metadata
Sample1.seurat@meta.data$group <- "S1"
```

```
## Error in Sample1.seurat@meta.data$group <- "S1": object 'Sample1.seurat' not found
```

```r
Sample2.seurat@meta.data$group <- "S2"
```

```
## Error in Sample2.seurat@meta.data$group <- "S2": object 'Sample2.seurat' not found
```

```r
Sample3.seurat@meta.data$group <- "S3"
```

```
## Error in Sample3.seurat@meta.data$group <- "S3": object 'Sample3.seurat' not found
```

```r
Sample4.seurat@meta.data$group <- "S4"
```

```
## Error in Sample4.seurat@meta.data$group <- "S4": object 'Sample4.seurat' not found
```





