---
title: "01_Manuscript_Figure_1"
author: "Dillon Corvino"
date: "07/03/2022"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

### Dataset information

```{r Dataset_Info}

# HNSCC scRNAseq/scTCRseq dataset
# Cells are HNSCC TILs sorted on Lymphocytes/Live/CD3+/CD4-/CD8+
# Human samples
# Each sample is a pool of 4 patients
# samples 1 and 2 are unstimulated 
# samples 3 and 4 are stimulated 
# Patients in sample 1 match patients in sample 3 and sample 2 pairs with 4
# Data acquired was transcript expression, ADT (antibody expression), and TCRA & B sequences
# This analysis uses just Transcript and TCR data


# General information
# CellRanger was used for sequence alignment/QC/counts/TCR calls - performed by Ross (QIMR, Brisbane, Aus)
# DC was provided the output from CellRanger (counts matrix) and used this as input for analysis within this script

```

### Environment

```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Create output directories & load custom functions & colour scheme
source("scripts/Setup.R", local = knitr::knit_global())

# pipeline variables 
long.compute <- TRUE
```


### Reading data

```{r reading_data}

########################
# Load all datasets 
########################

# complete dataset
seurat.combined <- LoadH5Seurat("saves/seurat_combined.h5Seurat")

# set idents and assay to use
Idents(seurat.combined) <- seurat.combined@meta.data$Clusters_l1
DefaultAssay(seurat.combined) <- "RNA"

```



## Figure 1 Overview of dataset {.tabset}


### Set up for fig 1
```{r Fig_1_setup}

# Create output directory

output.dir <- "results/Manuscript/Figure1/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}

# load required packages
library("Nebulosa")


```


### Nebulosa plots of key genes
```{r Nebulosa_plots_key_genes}

output.nebulosa <- "results/Manuscript/Figure1/Nebulosa_RNA/"

# Plot nebulosa density plots of key markers 
if(!dir.exists(paste0(output.nebulosa))){
  dir.create(paste0(output.nebulosa), 
             recursive = T)
}


# genes of ineterest
goi <- c("EOMES", "GZMB", "PRF1", "IFNG", "LAG3", "HAVCR2", "MKI67", "TCF7", "TOX", "IL7R", "PDCD1", "PRF1", "SELL", "NKG7", "TNF", "TIGIT", "CD226", "CCR7", "KLRB1", "RORA", "ISG15", "TRDC", "TRAV1-2", "ZNF683")

for(i in 1:length(goi)){
  
  print(Plot_Density_Custom(seurat.combined, 
                            features = goi[i],
                            custom_palette = batlow.pal,
                            joint = FALSE, 
                            pt.size = 1, 
                            reduction = "umap"))
  
  dev.copy(pdf, paste0(output.nebulosa, "Nebulosa_RNA_", goi[i], ".pdf"))
  dev.off()
  
    
}


Plot_Density_Joint_Only(seurat.combined,
                        features = c("TRAV1-2", "KLRB1"),
                        pt.size = 1,
                        custom_palette = batlow.pal)

dev.copy(pdf, paste0(output.nebulosa, "Nebulosa_RNA_TRAV_and_KLRB1.pdf"))
dev.off()

```


### Stacked VlnPlots, Dot plot chemokines and US vs Stim umap plot
```{r various_fig1_plots}

# Cell cycle scoring
cc.genes$s.genes
cc.genes$g2m.genes

s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes


#######################
# Cell cycle analysis
#######################

seurat.combined <- CellCycleScoring(seurat.combined, 
                                    s.features = s.genes, 
                                    g2m.features = g2m.genes, 
                                    set.ident = FALSE)


seurat.combined@meta.data$S.Score



gene_list_plot <- list(goi.1 = c("CD28", "CD226", "TCF7", "TBX21"),
                       goi.2 = c("GNLY", "PRF1", "GZMK", "XCL1"),
                       goi.3 = c("TOX", "CTLA4", "LAG3", "TNFRSF9"),
                       goi.4 = c("KLRC1", "CD300A", "RORA", "G2M.Score"))

for(i in seq_along(gene_list_plot)){

  
  print(Stacked_VlnPlot(seurat_object = seurat.combined,
                assay = "RNA",
                pt.size = 0,
                features = gene_list_plot[[i]],
                x_lab_rotate = TRUE,
                 plot_spacing = 0.15,
                colors_use = clust.cols))


  dev.copy(pdf, paste0(output.dir, "Stacked_VlnPlots_", i, ".pdf"))
  dev.off()
  
  
  }




# Chemokine receptor plots


goi <- c("CCR6", "CCR7", "CXCR3", "CXCR4", "CXCR6", "CXCL13")


DotPlot_scCustom(seurat_object = seurat.combined, 
                 features = goi,
                 flip_axes = F,
                 x_lab_rotate = TRUE)

dev.copy(pdf, paste0(output.dir, "Dotplot_chemokine_receptors.pdf"))
dev.off()


# Umap plot stim vs unstim
UMAPPlot(seurat.combined, 
         shuffle = T,
         pt.size = 1,
         cols = condition.cols,
         group.by = "condition")

dev.copy(pdf, paste0(output.dir, "/UMAP_condition_coloured.pdf"))
dev.off()

```
 
 
### Key stats of the dataset
```{r dataset_stats}

# Get basic stats on cells per cluster and per sample and plot these values 

cells.sample.clust <- table(seurat.combined@meta.data$group, seurat.combined@meta.data$Clusters_l1)
cells.cluster <- colSums(cells.sample.clust)


output.df <- cells.sample.clust*NA

for(i in 1:length(cells.cluster)){
output.df[,i] <- (cells.sample.clust[,i]/cells.cluster[i]) * 100
}


colSums(output.df)

# Plot freq of sample per cluster
barplot(output.df, 
        las = 2,
        col = sample.cols, 
        cex.names = 0.4)

dev.copy(pdf, paste0(output.dir, "Freq_of_sample_per_cluster.pdf"))
dev.off()

# ggplot version 
output.df <- as.data.frame(output.df)

ggplot(output.df) +
  aes(x = Var2, fill = Var1, weight = Freq) +
  geom_bar() +
  scale_fill_manual(
    values = c(S1 = "lightcyan",
    S2 = "lightblue2",
    S3 = "lightsalmon",
    S4 = "lightsalmon3")
  ) +
  theme_minimal()

# freq of cells per cluster

total.cells <- sum(cells.cluster)
freq.clust <- (cells.cluster/total.cells)*100

z <- order(freq.clust, decreasing = T)



barplot(freq.clust[z], 
        las = 2,
        ylim = c(0, 20),
        col = clust.cols[z], 
        cex.names = 0.4)

dev.copy(pdf, paste0(output.dir, "Freq_of_cells_per_cluster.pdf"))
dev.off()

# ggplot version

z.df <- as.data.frame(freq.clust[z])
z.df$names <- rownames(z.df)

ggplot(z.df) +
  aes(x = names, fill = names, weight = `freq.clust[z]`) +
  geom_bar() +
  scale_fill_manual(
    values = c(Cytotoxic = "#A6CEE3",
    Exhausted_1 = "#3385BB",
    Exhausted_2 = "#84BF96",
    gd_T_g9d2 = "#6DBD57",
    gd_T_non_g9d2 = "#7F9D55",
    MAIT = "#F57C7C",
    Naive_like_1_CM = "#E42622",
    Naive_like_2_SC = "#FBB268",
    Naive_like_3 = "#FE8D19",
    Proliferative = "#DE9E83",
    Stimulated_1 = "#9D7BBA",
    Stimulated_exhausted = "#977899",
    TRM = "#F3E587",
    Type_I_IFN = "#B15928")
  ) +
  theme_minimal()
```

## DEG analysis 

```{r}
# DEG calculation clusters
if(long.compute){
  
  
  Idents(seurat.combined) <- seurat.combined@meta.data$Clusters_l1
  DefaultAssay(seurat.combined) <- "RNA"
  
  seurat.combined <- FindVariableFeatures(seurat.combined)
  
  seurat.combined <- ScaleData(seurat.combined)
  
  DEG.wilcox <- FindAllMarkers(seurat.combined,
                               logfc.threshold = 0.25,
                               min.pct = 0.1,
                               test.use = "wilcox", 
                               only.pos = FALSE)
  

  write.csv(DEG.wilcox, paste0(output.dir, "DEG_RNA_clusters_wilcox.csv"))
  
  print(paste0("Number of sig differentially expressed genes = ", sum(DEG.wilcox$p_val_adj < 0.05)))
}

# load data
DEG.wilcox <- read.csv(paste0(output.dir, "DEG_RNA_clusters_wilcox.csv"))
DEG.wilcox <- DEG.wilcox[,-1]

# prepare dataset
Idents(seurat.combined) <- seurat.combined@meta.data$Clusters_l1
DefaultAssay(seurat.combined) <- "RNA"



# Subset data
small.seurat <- subset(seurat.combined, downsample = 300)

# Average dataset
average.seurat <- AverageExpression(seurat.combined, return.seurat = TRUE)


# Top DEGs

# DEGs > 0.5 log2FC
x <- DEG.wilcox %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_log2FC > 0.5) %>%
  dplyr::select(gene) 

x <- unique(x$gene)
x

# Average expression
DoHeatmap(average.seurat, 
          features = x,
          group.colors = clust.cols,
          assay = "RNA",
          raster = FALSE,
          draw.lines = FALSE) + scico::scale_fill_scico(palette = "batlow", 
                                                        direction = 1, 
                                                        na.value = "white")

dev.copy(pdf, 
         paste0(output.dir, "Top_log2fc_DEGs_RNA_clusters_l1_heatmap_average.pdf"))
dev.off()


# Top 10 DEGs per cluster

x <- DEG.wilcox %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::group_by(cluster) %>%
  dplyr::top_n(10, avg_log2FC) %>%
  dplyr::select(gene) 

x <- unique(x$gene)
x


# Average expression
DoHeatmap(average.seurat, 
                    group.colors = clust.cols,
          features = x,
          assay = "RNA",
          raster = FALSE,
          draw.lines = FALSE) + scico::scale_fill_scico(palette = "batlow", 
                                                                     direction = 1, 
                                                                     na.value = "white") 

dev.copy(pdf, 
         paste0(output.dir, "Top_10_DEGs_RNA_clusters_l1_heatmap_average.pdf"), )
dev.off()
```

### Visualise Innate-only UMAP
```{r vis_Innate_UMAP}



# Plot UMAP projection
UMAPPlot(object = Innate.seurat,
         group.by = "Clusters_l1",
         cols = innate.cols,
         label = TRUE, 
         pt.size = 1,
         label.size = 6)  + NoLegend()

dev.copy(pdf, paste0(output.Supfig1, "/UMAP_innate_clusters_named.pdf"))
dev.off()


UMAPPlot(object = Innate.seurat,
         group.by = "Clusters_l1",
         split.by = "condition",
         cols = innate.cols,
         label = TRUE, 
         pt.size = 1,
         label.size = 6)  + NoLegend()

dev.copy(pdf, paste0(output.Supfig1, "/UMAP_innate_clusters_named_condition_split.pdf"))
dev.off()

```


### Innate cell DEG and heatmap
```{r Innate_cell_DEG_and_Heatmap}


if(long.compute){

# Run DEG analysis
DEG.markers <- FindAllMarkers(Innate.seurat, 
                              slot = "data", 
                              logfc.threshold = 0.25,
                              test.use = "wilcox",
                              only.pos = TRUE)

# Filter for significance
DEG.markers <- 
  DEG.markers %>%
  filter(p_val_adj < 0.05)


write.csv(DEG.markers, paste0(output.Supfig1, "/DEG_Innate_Clusters_only.csv"))

}

DEG.markers <- read.csv(paste0(output.Supfig1, "/DEG_Innate_Clusters_only.csv"))
rownames(DEG.markers) <- DEG.markers$X


# Get top DEGs
goi <- DEG.markers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1) %>%
  dplyr::pull(gene)


goi <- unique(goi)

DefaultAssay(Innate.seurat) <- "RNA"

# Plot heatmap
DoHeatmap(Innate.seurat, 
          features = goi,
          raster = FALSE,
          group.colors = innate.cols,
          angle = 0,
          hjust = 0.50,
          draw.lines = T,
          ) + scico::scale_fill_scico(palette = "batlow", 
                                                    direction = 1, 
                                                    na.value = "white")

dev.copy(pdf, paste0(output.Supfig1, "/Heatmap_DEGs_innate_clusters.pdf"))
dev.off()



# Specifically annotate genes

small.seurat <- subset(Innate.seurat, downsample = 300)

genes.id <- c("GZMH", "NKG7", "GZMB", "PRF1", "GZMM", "GZMA", "XCL1", "TCF7", "SELL", "CD69", "BATF", "CCR6", "CXCR6", "GPR65", "GPR183")

# Plot heatmap downsampled single cell values
annotated.heatmap(small.seurat, 
                  cluster.id = "Clusters_l1",
                  assay.use = "RNA",
                  average.expression = FALSE,
                  goi = goi, 
                  genes.to.label = genes.id, 
                  col_order = names(innate.cols), 
                  col.colours = innate.cols, 
                  range.val = c(-2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2)) # The batlow palette looks odd if you have                                                                             just three colors...
dev.copy(pdf, paste0(output.Supfig1, "/Heatmap_DEGs_innate_clusters_annotated.pdf"))
dev.off()

#

if(long.compute){
  
  Idents(Innate.seurat) <- Innate.seurat@meta.data$condition_clust
  
 DEG.var <- FindMarkers(Innate.seurat,
                         ident.1 = "US_gd_T_g9d2", 
                         ident.2 = "Stim_gd_T_g9d2",
                         test.use = "wilcox",
                         logfc.threshold = 0.25,
                         min.pct = 0.1,
                         only.pos = FALSE,
                         verbose = FALSE)
  


 
 
 }




  # Set idents to cluster ID
  Idents(Innate.seurat) <- Innate.seurat@meta.data$Clusters_l1
  
  # Subset for cluster ID
  subset.seurat <- subset(Innate.seurat, idents = "gd_T_g9d2")
  
  # Set idents to condition var
  Idents(subset.seurat) <- subset.seurat@meta.data$condition
  
  # Get top 10 genes
  sig.genes <- DEG.var %>%
    #filter(p_val_adj < 0.05) %>%
    mutate(abs.FC = abs(avg_log2FC)) %>%
    top_n(10, abs.FC)
  
  genes.to.label <- rownames(sig.genes)
  
  
  
  # Average expression
  avg.subset.seurat <- log1p(AverageExpression(subset.seurat, verbose = FALSE)$RNA)
  avg.subset.seurat <- as.data.frame(avg.subset.seurat)
  
  
  # Generate plot
  p1 <- ggplot(avg.subset.seurat, 
               aes(US, Stim)) + 
    geom_point() + 
    ggtitle(paste0("Cluster "))+
  theme_minimal()
  
  p1 <-  LabelPoints(plot = p1,
                     points = genes.to.label,
                     repel = TRUE)
  
  print(plot(p1))
  
  
  
  dev.copy(pdf, paste0(fig.dir, "ScatterPlot_Stim_vs_US_cluster_", clust.var[i], ".pdf"))
  dev.off()
  


```




### Getting TCR data for Innate clusters
```{r TCR_for_Innate_clusters}


library(gplots)

# read in unfiltered TCR data, as most MAIT cells are low abundance clones lost in filtering
seurat.tcr <- readRDS("saves/seurat_tcr.rds")

unique(seurat.tcr@meta.data$seurat_clusters) # Gamma/delta T-cells are not found in the filtered tcr dataset

# filter for MAIT cells
MAIT.TCR <- subset(seurat.tcr, idents = c("MAIT"))

# remove large dataset
rm(seurat.tcr)


# 290 MAIT cells in dataset
# 130 cells have a TCR associated
# 130/290 = ~44% of MAIT cells


########################
# Heatmap functions
########################

hclust.func <- function(x){
  hclust(x, method = 'ward.D2')
}

dist.func <- function(x){
  dist(x,method = 'binary')
}

library(RColorBrewer)


col.palette <- colorRampPalette(brewer.pal(8, "RdYlBu"))(600)


######################################################
# Plot TCR usage as a log2() of cell number
######################################################

# CTaa

x <- table(MAIT.TCR@meta.data$CTaa, MAIT.TCR@meta.data$condition_clust)

heatmap.2(as.matrix(log2(x+1)), 
          scale = "none",
          trace = "none",
          Colv = TRUE,
          cexRow = 0.5, 
          margins = c(8, 15),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette))     

dev.copy(pdf, paste0(output.fig2, "/Heatmap_CTaa.pdf"))
dev.off()

# CT gene

x <- table(MAIT.TCR@meta.data$CTgene, MAIT.TCR@meta.data$condition_clust)

heatmap.2(as.matrix(log2(x+1)), 
          scale = "none",
          trace = "none",
          Colv = TRUE,
          cexRow = 0.5, 
          margins = c(8, 15),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette))     

dev.copy(pdf, paste0(output.fig2, "/Heatmap_CT_gene.pdf"))
dev.off()

# TRAV gene

x <- table(MAIT.TCR@meta.data$TRAV, MAIT.TCR@meta.data$condition_clust)


heatmap.2(as.matrix(log2(x+1)), 
          scale = "none",
          trace = "none",
          Colv = TRUE,
          cexRow = 0.5, 
          margins = c(8, 15),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette))     

dev.copy(pdf, paste0(output.fig2, "/Heatmap_TRA_Vgene.pdf"))
dev.off()


# TRAJ gene


x <- table(MAIT.TCR@meta.data$TRAJ, MAIT.TCR@meta.data$condition_clust)


heatmap.2(as.matrix(log2(x+1)), 
          scale = "none",
          trace = "none",
          Colv = TRUE,
          cexRow = 0.5, 
          margins = c(8, 15),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette))     

dev.copy(pdf, paste0(output.fig2, "/Heatmap_TRA_Jgene.pdf"))
dev.off()



# TRBV gene
x <- table(MAIT.TCR@meta.data$TRBV, MAIT.TCR@meta.data$condition_clust)

heatmap.2(as.matrix(log2(x+1)), 
          scale = "none",
          trace = "none",
          Colv = TRUE,
          cexRow = 0.5, 
          margins = c(8, 15),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette))     

dev.copy(pdf, paste0(output.fig2, "/Heatmap_TRB_Vgene.pdf"))
dev.off()

# TRBJ gene
x <- table(MAIT.TCR@meta.data$TRBJ, MAIT.TCR@meta.data$condition_clust)

heatmap.2(as.matrix(log2(x+1)), 
          scale = "none",
          trace = "none",
          Colv = TRUE,
          cexRow = 0.5, 
          margins = c(8, 15),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette))     

dev.copy(pdf, paste0(output.fig2, "/Heatmap_TRB_Jgene.pdf"))
dev.off()

# remove large dataset
rm(MAIT.TCR)


```



### Innate cell TF analysis 

```{r TF_analysis}


# Searching for TFs
TF.ref.list <- read.delim("Data/TFcheckpoint.txt") # http://www.tfcheckpoint.org/index.php/browse

dim(TF.ref.list) # 1020, 11
colnames(TF.ref.list)
head(TF.ref.list)

Cluster.markers <- read.csv("results/Manuscript/Sup_Figure1/DEG_Innate_Clusters_only.csv", row.names = 1)


# Up-regulated TFs
sig.genes <- Cluster.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_log2FC > 0.5) %>%
  dplyr::select(gene)

sig.genes <- unique(sig.genes$gene)

TF.logic <- sig.genes %in% TF.ref.list$Gene_symbol

TF.DEG <- sig.genes[TF.logic]




########################################
# Plot TFs differentially regulated 
########################################

Innate.seurat <- SeuratDisk::LoadH5Seurat("saves/Innate_seurat.h5Seurat")

Idents(Innate.seurat) <- Innate.seurat@meta.data$Clusters_l1
DefaultAssay(Innate.seurat) <- "RNA"

#seurat.combined.small <- subset(seurat.combined, downsample = 300)

average.seurat <- AverageExpression(Innate.seurat,
                                    assay = "RNA",
                                    slot = "data",
                                    verbose = TRUE,
                                    return.seurat = TRUE)

DoHeatmap(average.seurat, 
          features = TF.DEG, 
          draw.lines = FALSE, 
          raster = FALSE)

dev.copy(pdf, paste0(output.fig2, "/Heatmap_of_TF_diff_expressed.pdf"))
dev.off()



```

### Scillus

```{r Scillus_GSEA}


#remotes::install_github("xmc811/Scillus")
library("Scillus")


# Scillus package would not compile
# I took the functions and made a new .R file
# Some of the functions called and used in this script required modifications to work outside of the package environment


#source("~/Documents/Work/Sciebo/Scripts/Packages/Scillus_package/Scillus_functions.R")
#load("~/Documents/Work/Sciebo/Scripts/Packages/Scillus_package/pathways.hallmark.rda")



######################
# Cluster analysis
######################

# set idents to cluster
Idents(Innate.seurat) <- Innate.seurat@meta.data$Clusters_l1


# get a broad list of DEG markers 
Cluster.markers <- FindAllMarkers(Innate.seurat,
                                  assay = "RNA",
                                  only.pos = TRUE,
                                  min.diff.pct = 0.1,
                                  logfc.threshold = 0,
                                  return.thresh = 0.01)




# Threshold marker list by P value and fold change for GO analysis 

GO.markers <- Cluster.markers[Cluster.markers$p_val_adj < 0.05, ]
GO.markers <- GO.markers[abs(GO.markers$avg_log2FC) > 0.25, ]
dim(Cluster.markers) # 8,635 genes
dim(GO.markers) # 4,890 genes

table(GO.markers$cluster) # number of genes DEG per cluster // used to justify using topn = 100

# Plot Gene Ontology analysis
# CC, BP, MF

# To plot graph for a specific cluster use below code
#plot_cluster_go(Module.markers, cluster_name = "Activated", org = "human", ont = "MF")

# CC 
pdf(paste0(output.fig2, "/GO_Clusters_CC.pdf"), width = 20, height = 20)

plot_all_cluster_go(GO.markers,
                    topn = 100,
                    org = "human", 
                    ont = "CC")

dev.off()


# BP 
pdf(paste0(output.fig2, "/GO_Clusters_BP.pdf"), width = 20, height = 20)

plot_all_cluster_go(GO.markers,
                    topn = 100,
                    org = "human", 
                    ont = "BP")

dev.off()

# MF 
pdf(paste0(output.fig2, "/GO_Clusters_MF.pdf"), width = 20, height = 20)

plot_all_cluster_go(GO.markers,
                    topn = 100,
                    org = "human", 
                    ont = "MF")
dev.off()




```








