---
title: "02_Figure_1"
author: "Dillon Corvino"
date: "07/03/2022"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

### Environment

```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Create output directories & load custom functions & colour scheme
source("scripts/Setup.R", local = knitr::knit_global())

# pipeline variables 
long.compute <- TRUE
```


### Reading data

```{r reading_data}

########################
# Load dataset
########################

# complete dataset
seurat.combined <- LoadH5Seurat("saves/seurat_combined.h5Seurat")

# set idents and assay to use
Idents(seurat.combined) <- seurat.combined@meta.data$Clusters_l1
DefaultAssay(seurat.combined) <- "RNA"

```


### Set up output directory for fig 1 and supplementary
```{r Fig_1_setup}

# Create output directory

output.dir <- "results/Figure_1/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}


output.Supfig1 <- "results/Supplementary_Figure_1/"

if(!dir.exists(paste0(output.Supfig1))){
  dir.create(paste0(output.Supfig1), 
             recursive = T)
}


```


## Figure 1 Overview of dataset {.tabset}

### Figure 1B
```{r Figure_1B}

UMAPPlot(seurat.combined,
         group.by = "Clusters_l1", 
         cols = clust.cols,
         label = TRUE, 
         pt.size = 1) + NoLegend()

dev.copy(pdf, paste0(output.dir, "Figure_1B.pdf"))
dev.off()

```


### Figure 1 C-G
```{r Figure_1C_G}

# highlight clusters 

Idents(seurat.combined) <- seurat.combined@meta.data$Clusters_l1

# Figure 1C  
clust.i <- c("Naive_like_1", "Naive_like_2", "Naive_like_3")

scCustomize::Cluster_Highlight_Plot(seurat.combined, 
                                    cluster_name = clust.i, 
                                    highlight_color = clust.cols[grepl(paste0(clust.i, collapse = "|"), names(clust.cols))])

dev.copy(pdf, paste0(output.dir, "Figure_1C_Highlight_clusters.pdf"))
dev.off()

# Figure 1D  
clust.i <- c("Effector", "ISG", "Stimulated_1")

scCustomize::Cluster_Highlight_Plot(seurat.combined, 
                                    cluster_name = clust.i, 
                                    highlight_color = clust.cols[grepl(paste0(clust.i, collapse = "|"), names(clust.cols))])

dev.copy(pdf, paste0(output.dir, "Figure_1D_Highlight_clusters.pdf"))
dev.off()

# Figure 1E  
clust.i <- c("Stimulated_exhausted", "Exhausted_1", "Exhausted_2")

scCustomize::Cluster_Highlight_Plot(seurat.combined, 
                                    cluster_name = clust.i, 
                                    highlight_color = clust.cols[grepl(paste0(clust.i, collapse = "|"), names(clust.cols))])

dev.copy(pdf, paste0(output.dir, "Figure_1E_Highlight_clusters.pdf"))
dev.off()

# Figure 1F  
clust.i <- c("TRM", "Proliferative")

scCustomize::Cluster_Highlight_Plot(seurat.combined, 
                                    cluster_name = clust.i, 
                                    highlight_color = clust.cols[grepl(paste0(clust.i, collapse = "|"), names(clust.cols))])

dev.copy(pdf, paste0(output.dir, "Figure_1F_Highlight_clusters.pdf"))
dev.off()

# Figure 1G  
clust.i <- c("gd_T_g9d2", "gd_T_non_g9d2", "MAIT")

scCustomize::Cluster_Highlight_Plot(seurat.combined, 
                                    cluster_name = clust.i, 
                                    highlight_color = clust.cols[grepl(paste0(clust.i, collapse = "|"), names(clust.cols))])

dev.copy(pdf, paste0(output.dir, "Figure_1G_Highlight_clusters.pdf"))
dev.off()




#Figure 1 C-G density expression plots
goi <- c("IL7R", "CCR7", "SELL", "GZMK", "ISG15", "IFNG", "XCL1", "HAVCR2", "PDCD1", "TOX", "ZNF683", "MKI67", "TRDC")


for(i in 1:length(goi)){
  
  print(Plot_Density_Custom(seurat.combined, 
                            features = goi[i],
                            custom_palette = batlow.pal,
                            joint = FALSE, 
                            pt.size = 1, 
                            reduction = "umap"))
  
  dev.copy(pdf, paste0(output.dir, "Figure_1C_G_", goi[i], "_expression_density.pdf"))
  dev.off()
  
  
}

# Figure 1G joint expression density plot
Plot_Density_Joint_Only(seurat.combined,
                        features = c("TRAV1-2", "KLRB1"),
                        pt.size = 1,
                        custom_palette = batlow.pal)

dev.copy(pdf, paste0(output.dir, "Figure_1G_TRAV1_2_and_KLRB1_expression_density.pdf"))
dev.off()






```

### Figure 1 H
```{r Figure_1H}

# Get basic stats on cells per cluster and per sample and plot these values 

cells.sample.clust <- table(seurat.combined@meta.data$group, seurat.combined@meta.data$Clusters_l1)
cells.cluster <- colSums(cells.sample.clust)


# freq of cells per cluster

total.cells <- sum(cells.cluster)
freq.clust <- (cells.cluster/total.cells)*100

z <- order(freq.clust, decreasing = T)

# write to file 
write.csv(freq.clust, paste0(output.dir, "Figure_1H_values.csv"))

barplot(freq.clust[z], 
        las = 2,
        ylim = c(0, 20),
        col = clust.cols[z], 
        cex.names = 0.4)

dev.copy(pdf, paste0(output.dir, "Figure_1H.pdf"))
dev.off()


```

## Sup Figure 1 dataset overview and Innate clusters {.tabset}

### Read in Innate dataset
```{r read_innate_dataset}

# Load innate dataset
Innate.seurat <- SeuratDisk::LoadH5Seurat("saves/Innate_seurat.h5Seurat")

Idents(Innate.seurat) <- Innate.seurat@meta.data$Clusters_l1
DefaultAssay(Innate.seurat) <- "RNA"

```

### Sup Figure 1A
```{r Sup_Figure_1A}

# Get basic stats on cells per cluster and per condition and plot these values 

cells.conditon.clust <- table(seurat.combined@meta.data$condition, seurat.combined@meta.data$Clusters_l1)

# remove innate cluster columns 
#pattern.val <- c("^gd_T_.*", "^MAIT$")
#remove.logic <- grepl(paste0(pattern.val, collapse = "|"), colnames(cells.conditon.clust))
#cells.conditon.clust <- cells.conditon.clust[ , !remove.logic]



cells.cluster <- colSums(cells.conditon.clust)


output.df <- cells.conditon.clust*NA

for(i in 1:length(cells.cluster)){
output.df[,i] <- (cells.conditon.clust[,i]/cells.cluster[i]) * 100
}


colSums(output.df)

# Plot freq of sample per cluster
barplot(output.df, 
        las = 2,
        col = condition.cols, 
        cex.names = 0.7)

dev.copy(pdf, paste0(output.Supfig1, "Sup_Figure_1A.pdf"))
dev.off()

```

### Sup Figure 1B
```{r Sup_Figure_1B}

# set object defaults 
DefaultAssay(seurat.combined) <- "RNA"
Idents(seurat.combined) <- seurat.combined@meta.data$Clusters_l1



# Functional markers
goi <- c("GZMA", "GZMB", "GZMH", "GZMK", "GNLY")

Stacked_VlnPlot(seurat.combined, 
                features = goi, 
                colors_use = clust.cols, 
                x_lab_rotate = TRUE)

dev.copy(pdf, paste0(output.Supfig1, "Sup_Figure_1B_1.pdf"))
dev.off()

goi <- c("CRTAM", "TNF", "TNFSF14", "EOMES", "IFIT1")

Stacked_VlnPlot(seurat.combined, 
                features = goi, 
                colors_use = clust.cols, 
                x_lab_rotate = TRUE)

dev.copy(pdf, paste0(output.Supfig1, "Sup_Figure_1B_2.pdf"))
dev.off()

# Exhaustion markers
goi <- c("TCF7", "CTLA4", "ENTPD1", "TIGIT", "TNFSF10")

Stacked_VlnPlot(seurat.combined, 
                features = goi, 
                colors_use = clust.cols, 
                x_lab_rotate = TRUE)

dev.copy(pdf, paste0(output.Supfig1, "Sup_Figure_1B_3.pdf"))
dev.off()


# Activation markers
goi <- c("CD226", "ICOS", "TNFRSF9", "CD28",  "ITGAE")

Stacked_VlnPlot(seurat.combined, 
                features = goi, 
                colors_use = clust.cols, 
                x_lab_rotate = TRUE)

dev.copy(pdf, paste0(output.Supfig1, "Sup_Figure_1B_4.pdf"))
dev.off()


```

### Sup Figure 1C
```{r Sup_Figure_1C}

# Plot UMAP projection
UMAPPlot(object = Innate.seurat,
         group.by = "Clusters_l1",
         cols = innate.cols,
         label = TRUE, 
         pt.size = 1,
         label.size = 6)  + NoLegend()

dev.copy(pdf, paste0(output.Supfig1, "Sup_Figure_1C.pdf"))
dev.off()

```

### Sup_Figure_1D
```{r Sup_Figure_1D}


# plot gene expression of TCR gamma/delta genes + MAIT genes
average.seurat <- AverageExpression(Innate.seurat,
                                    assay = "RNA",
                                    slot = "data",
                                    verbose = TRUE,
                                    return.seurat = TRUE)

# Sup Figure 1D

goi <- c("TRGV9", "TRDV2", "TRGC1", "TRGC2", "TRDC", "TRAV1-2", "KLRB1")

DoHeatmap(average.seurat, 
          features = goi, 
          group.colors = innate.cols,
          draw.lines = FALSE, 
          raster = FALSE)+ scico::scale_fill_scico(palette = "batlow", 
                                                   direction = 1, 
                                                   na.value = "white")

dev.copy(pdf, paste0(output.Supfig1, "Sup_Figure_1D.pdf"))
dev.off()


```

### Sup Figure 1E
```{r Sup_Figure_1E}


# Vln Plots key genes 

# Sup Figure 1E

list.goi <- list("CD3" = c("CD3D", "CD3E", "CD3G"),
                 "CD8" = c("CD8A", "CD8B", "CD4"))

for(i in seq_along(list.goi)){
  print(
    Stacked_VlnPlot(Innate.seurat, 
                    features = list.goi[[i]], 
                    colors_use = innate.cols,
                    pt.size = 0)
  )
  
  dev.copy(pdf, paste0(output.Supfig1, "Sup_Figure_1E_", names(list.goi)[i], ".pdf"))
  dev.off()
  
}


```

### Sup Figure 1F 
```{r Sup_Figure_1F}


source("scripts/functions/annotate_seurat_heatmap_function.R")


if(long.compute){
  
  # Run DEG analysis
  DEG.markers <- FindAllMarkers(Innate.seurat, 
                                slot = "data", 
                                logfc.threshold = 0.25,
                                test.use = "wilcox",
                                only.pos = TRUE)
  
  # Filter for significance
  DEG.markers <- 
    DEG.markers %>%
    filter(p_val_adj < 0.05)
  
  
  write.csv(DEG.markers, paste0(output.Supfig1, "DEG_Innate_Clusters_only.csv"))
  
}

DEG.markers <- read.csv(paste0(output.Supfig1, "DEG_Innate_Clusters_only.csv"))
rownames(DEG.markers) <- DEG.markers$X




# Downsample seurat object for heatmap plots
DefaultAssay(Innate.seurat) <- "RNA"

Idents(Innate.seurat) <- Innate.seurat@meta.data$Clusters_l1

small.seurat <- subset(Innate.seurat, downsample = 200)


# Get top DEGs
goi <- DEG.markers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1) %>%
  dplyr::pull(gene)


goi <- unique(goi)


# Specifically annotate genes
genes.id <- c("GZMH", "NKG7", "GZMB", "PRF1", "GZMM", "GZMA", "ZNF683", "XCL1", "TCF7", "CD27", "KLRD1", "SELL", "BACH2", "TIGIT", "GPR183", "CCR6", "CXCR6", "NFKBIA", "GPR65", "BATF", "CDKN1A", "CD69","MYC")

# Plot heatmap downsampled single cell values
annotated.heatmap(small.seurat, 
                  cluster.id = "Clusters_l1",
                  assay.use = "RNA",
                  average.expression = FALSE,
                  goi = goi, 
                  genes.to.label = genes.id, 
                  col_order = names(innate.cols), 
                  col.colours = innate.cols)

dev.copy(pdf, paste0(output.Supfig1, "Sup_Figure_1F.pdf"))
dev.off()



```

### Sup Figure 1G
```{r Sup_Figure_1G}


# Searching for TFs
TF.ref.list <- read.delim("Data/TFcheckpoint.txt") # http://www.tfcheckpoint.org/index.php/browse

dim(TF.ref.list) # 1020, 11
colnames(TF.ref.list)
head(TF.ref.list)


Cluster.markers <- read.csv(paste0(output.Supfig1, "DEG_Innate_Clusters_only.csv"))
rownames(Cluster.markers) <- Cluster.markers$X


# Up-regulated TFs
sig.genes <- Cluster.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_log2FC > 0.5) %>%
  dplyr::select(gene)

sig.genes <- unique(sig.genes$gene)

TF.logic <- sig.genes %in% TF.ref.list$gene_symbol

TF.DEG <- sig.genes[TF.logic]


########################################
# Plot TFs differentially regulated 
########################################

Idents(Innate.seurat) <- Innate.seurat@meta.data$Clusters_l1
DefaultAssay(Innate.seurat) <- "RNA"

#seurat.combined.small <- subset(seurat.combined, downsample = 300)

average.seurat <- AverageExpression(Innate.seurat,
                                    assay = "RNA",
                                    slot = "data",
                                    verbose = TRUE,
                                    return.seurat = TRUE)

DoHeatmap(average.seurat, 
          features = TF.DEG, 
          group.colors = innate.cols,
          draw.lines = FALSE, 
          raster = FALSE)+ scico::scale_fill_scico(palette = "batlow", 
                                                   direction = 1, 
                                                   na.value = "white")

dev.copy(pdf, paste0(output.Supfig1, "Sup_Figure_1G.pdf"))
dev.off()


```



### Reviewer plots
```{r reviewer_plots}


output.dir <- "results/reviewer_plots/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}


DefaultAssay(seurat.combined) <- "RNA"
Idents(seurat.combined) <- seurat.combined@meta.data$Clusters_l1

grep("IL10.", rownames(seurat.combined), value = T)

# Functional markers
goi <- c("IL10", "IL10RA", "IL10RB")

Stacked_VlnPlot(seurat.combined, 
                features = goi, 
                pt.size = 0,
                colors_use = clust.cols, 
                x_lab_rotate = TRUE)

dev.copy(pdf, paste0(output.dir, "VlnPlot_IL10_genes.pdf"))
dev.off()



#SeuratExtend::feature_percent(seurat.combined, feature = "IL10RA")



input.dir <- "results/Figure_2/"

# load data
DEG.wilcox <- read.csv(paste0(input.dir, "Figure_2B_DEGs.csv"))
DEG.wilcox <- DEG.wilcox[,-1]


DEG.wilcox %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(gene %in% goi)





```



