---
title: "02_Manuscript_Figure_2"
author: "Dillon Corvino"
date: "07/03/2022"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

### Dataset information

```{r Dataset_Info}

# HNSCC scRNAseq/scTCRseq dataset
# Cells are HNSCC TILs sorted on Lymphocytes/Live/CD3+/CD4-/CD8+
# Human samples
# Each sample is a pool of 4 patients
# samples 1 and 2 are unstimulated 
# samples 3 and 4 are stimulated 
# Patients in sample 1 match patients in sample 3 and sample 2 pairs with 4
# Data acquired was transcript expression, ADT (antibody expression), and TCRA & B sequences
# This analysis uses just Transcript and TCR data


# General information
# CellRanger was used for sequence alignment/QC/counts/TCR calls - performed by Ross (QIMR, Brisbane, Aus)
# DC was provided the output from CellRanger (counts matrix) and used this as input for analysis within this script

```

### Environment

```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Create output directories & load custom functions & colour scheme
source("scripts/Setup.R", local = knitr::knit_global())

# pipeline variables
long.compute <- FALSE

# load required packages
require(tidyverse)
require(Seurat)
library(ggplot2)



# Create output directory
output.dir <- "results/Manuscript/Figure2/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}

```


### Reading data

```{r reading_data}

########################
# Load all datasets 
########################

# CD8 dataset
CD8.seurat <- LoadH5Seurat("saves/CD8_seurat.h5Seurat")

Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"

# CD8 TCR data

# TCR with no minimum clonotype filter
CD8.seurat.tcr <- LoadH5Seurat("saves/CD8_seurat_tcr.h5Seurat")

Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$Clusters_l1
DefaultAssay(CD8.seurat.tcr) <- "RNA"

# TCR data with a minimum clonotype filter
CD8.seurat.tcr.filt <- LoadH5Seurat("saves/CD8_seurat_tcr_filt.h5Seurat")

Idents(CD8.seurat.tcr.filt) <- CD8.seurat.tcr.filt@meta.data$Clusters_l1
DefaultAssay(CD8.seurat.tcr.filt) <- "RNA"

```


## Figure 2 {.tabset}

### Visualise CD8 only UMAP
```{r vis_CD8_UMAP}


UMAPPlot(object = CD8.seurat,
         group.by = "Clusters_l1",
         label = TRUE, 
         cols = CD8.cols,
         pt.size = 1,
         label.size = 6)  + 
  NoLegend()

dev.copy(pdf, paste0(output.dir, "UMAP_CD8_clusters.pdf"))
dev.off()


Idents(CD8.seurat) <- CD8.seurat@meta.data$condition
US.seurat <- subset(CD8.seurat, ident = "US")
Stim.seurat <- subset(CD8.seurat, ident = "Stim")

UMAPPlot(object = US.seurat,
         group.by = "Clusters_l1",
         label = TRUE, 
         cols = CD8.cols,
         pt.size = 1,
         label.size = 6)  +
    ggtitle("Unstimulated") + 
  NoLegend()


dev.copy(pdf, paste0(output.dir, "UMAP_CD8_clusters_Unstimulated.pdf"))
dev.off()


UMAPPlot(object = Stim.seurat,
         group.by = "Clusters_l1",
         label = TRUE, 
         cols = CD8.cols,
         pt.size = 1,
         label.size = 6)  + 
  ggtitle("Stimulated") + 
  NoLegend()

dev.copy(pdf, paste0(output.dir, "UMAP_CD8_clusters_Stimulated.pdf"))
dev.off()



UMAPPlot(object = CD8.seurat,
         group.by = "Clusters_l1",
         split.by = "condition",
         cols = CD8.cols,
         pt.size = 1,
         label = TRUE, 
         label.size = 6) + NoLegend()

dev.copy(pdf, paste0(output.dir, "UMAP_CD8_clusters_named_condition_split.pdf"))
dev.off()





for(i in seq_along(condition.cols)){

  
  temp.cols <- c("Gray", "Gray")
  temp.cols[i] <- paste0(condition.cols)[i]

print(UMAPPlot(object = CD8.seurat,
         group.by = "condition",
         cols = temp.cols,
         pt.size = 1,
         shuffle = T,
         label = TRUE, 
         label.size = 6) + 
    NoLegend())

dev.copy(pdf, paste0(output.dir, "UMAP_CD8_clusters_", names(condition.cols[i]), "_coloured.pdf"))
dev.off()

}


UMAPPlot(object = CD8.seurat,
         group.by = "condition",
         cols = temp.cols,
         pt.size = 1,
         shuffle = T,
         label = TRUE, 
         label.size = 6) + 
    NoLegend()
dev.copy(pdf, paste0(output.dir, "UMAP_CD8_condition_combined_coloured.pdf"))
dev.off()

temp.cols <- condition.cols
names(temp.cols)[2] <-  "Stim"
temp.cols[2] <- "#ee8262"


library(RColorBrewer)


```


### Visualise key gene expression
```{r vis_key_genes}

# set object defaults 
DefaultAssay(CD8.seurat) <- "RNA"
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1



# Functional markers
goi <- c("GZMA", "GZMB", "GZMM", "GZMH", "GZMK")

Stacked_VlnPlot(CD8.seurat, 
                features = goi, 
                colors_use = CD8.cols, 
                x_lab_rotate = TRUE)

dev.copy(pdf, paste0(output.dir, "Stacked_VlnPlot_Functional_markers_1.pdf"))
dev.off()

goi <- c("PRF1", "GNLY", "IFNG", "TNF", "XCL1")

Stacked_VlnPlot(CD8.seurat, 
                features = goi, 
                colors_use = CD8.cols, 
                x_lab_rotate = TRUE)

dev.copy(pdf, paste0(output.dir, "Stacked_VlnPlot_Functional_markers_2.pdf"))
dev.off()

# Exhaustion markers
goi <- c("TCF7", "TOX", "HAVCR2", "CTLA4", "PDCD1", "ENTPD1")

Stacked_VlnPlot(CD8.seurat, 
                features = goi, 
                colors_use = CD8.cols, 
                x_lab_rotate = TRUE)

dev.copy(pdf, paste0(output.dir, "Stacked_VlnPlot_Exhaustion_markers_1.pdf"))
dev.off()

goi <- c("EOMES", "TBX21")

Stacked_VlnPlot(CD8.seurat, 
                features = goi, 
                colors_use = CD8.cols, 
                x_lab_rotate = TRUE)

dev.copy(pdf, paste0(output.dir, "Stacked_VlnPlot_Exhaustion_markers_2.pdf"))
dev.off()


# Activation markers
goi <- c("CD226", "ICOS", "TNFRSF9", "CD28",  "CD69")

Stacked_VlnPlot(CD8.seurat, 
                features = goi, 
                colors_use = CD8.cols, 
                x_lab_rotate = TRUE)

dev.copy(pdf, paste0(output.dir, "Stacked_VlnPlot_Activation_markers.pdf"))
dev.off()



```

### Dotplot of chemokine receptors
```{r chemokine_dotplot}

# Set dataset defaults 
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"

# identify chemokine genes 
#pattern.var <- c("CCL.*", "CXCR.*", "CXCL.*", "CCR")
#goi <- grep(paste0(pattern.var, collapse = "|"), rownames(CD8.seurat), value = T)
# used as a unbised method to check expression of all chemokine genes and then select the ones that showed expression 

goi <- c("CCL3", "CCL4", "CCL4L2", "CCR7", "CXCR4", "CXCR3", "CXCR6", "CXCL13")

scCustomize::DotPlot_scCustom(CD8.seurat, 
                              colors_use = batlow.pal,
                              x_lab_rotate = T,
                              features = goi)

dev.copy(pdf, paste0(output.dir, "DotPlot_Chemokine_genes.pdf"))
dev.off()


```



### Calculate DEGs without innate clusters
```{r DEG_calculation}

if(long.compute){
  
  Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
  DefaultAssay(CD8.seurat) <- "RNA"
  
# Run DEG analysis
DEG.markers <- FindAllMarkers(CD8.seurat, 
                              slot = "data", 
                              logfc.threshold = 0.25,
                              test.use = "wilcox",
                              only.pos = FALSE)

# Filter for significance
DEG.markers <- 
  DEG.markers %>%
  filter(p_val_adj < 0.05)


write.csv(DEG.markers, paste0(output.dir, "DEG_CD8_Clusters_only.csv"))
}

```

### Visualise DEGs in common and unique between stim_1 and stim_exhaust
```{r Vis_Stim_cluster_DEGs}

# Create output directory
output.dir <- "results/Manuscript/Figure2/DEG_plots/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}


# set up seurat object
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"


# set input directory
input.dir <- "results/Manuscript/Figure2/"

# load data 
DEG.markers <- read.csv(paste0(input.dir, "DEG_CD8_Clusters_only.csv"))


# first plot large heatmap showing top DEGs per cluster
goi <- DEG.markers %>%
  group_by(cluster) %>%
  filter(p_val_adj < 0.05) %>%
  filter(avg_log2FC > 1) %>%
  pull(gene)

goi <- unique(goi)


# Plot as an average of expression
temp.seurat <- AverageExpression(CD8.seurat, 
                                 features = goi,
                                 return.seurat = TRUE)

DoHeatmap(temp.seurat, 
          group.colors = CD8.cols,
          features = goi, 
          raster = FALSE,
          draw.lines = FALSE) + 
  scico::scale_fill_scico(palette = "batlow", 
                          direction = 1, 
                          na.value = "white")

dev.copy(pdf, paste0(output.dir, "Heatmap_large_DEGs_logfc_greater_1_across_clusters_average.pdf"))
dev.off()


##################################################
# Venn overlap of Stim_1 and Stim_Exhaust clusters
##################################################


# Load library
# install.packages("VennDiagram")
library(VennDiagram)
library(grDevices)


############################
# up-regulated genes
############################

stim.1.goi.up <- DEG.markers %>%
  filter(cluster %in% c("Stimulated_1")) %>%
  filter(p_val_adj < 0.05) %>%
  filter(avg_log2FC > 0) %>%
  arrange(desc(avg_log2FC)) %>%
  pull(gene)

stim.exh.goi.up <- DEG.markers %>%
  filter(cluster %in% c("Stimulated_exhausted")) %>%
  filter(p_val_adj < 0.05) %>%
  filter(avg_log2FC > 0) %>%
  arrange(desc(avg_log2FC)) %>%
  pull(gene)




# Venn overlap of up-regulated genes
venn <- venn.diagram(x = list(stim.exh.goi.up, stim.1.goi.up),
                     category.names = c("Stimulated_exhausted", "Stimulated_1"),
                     filename = NULL,
                     output=F,
                     
                     # Circles
                     lwd = 2,
                     lty = 'blank',
                     fill = condition.cols,
                     
                     # Numbers
                     cex = .6,
                     fontface = "bold",
                     fontfamily = "sans",
                     
                     # Set names
                     cat.cex = 0.6,
                     cat.fontface = "bold",
                     cat.default.pos = "outer",
                     cat.fontfamily = "sans")

pdf(file = paste0(output.dir,"Up_regulated_Venn.pdf"))
grid.draw(venn)
dev.off()


# calculate union and unique gois
unique.stim.1.up <- stim.1.goi.up[!stim.1.goi.up %in% stim.exh.goi.up]
length(unique.stim.1.up)

unique.stim.exh.up <- stim.exh.goi.up[!stim.exh.goi.up %in% stim.1.goi.up]
length(unique.stim.exh.up)

overlap.goi.up <- stim.1.goi.up[stim.1.goi.up %in% stim.exh.goi.up]
length(overlap.goi.up)


# arrange genes for plotting

goi <- unique(c(unique.stim.1.up, overlap.goi.up, unique.stim.exh.up))


small.seurat <- ScaleData(small.seurat,
                          features = goi)


DoHeatmap(small.seurat, 
          group.colors = CD8.cols,
          raster = FALSE,
          features = goi) + 
  scico::scale_fill_scico(palette = "batlow", 
                          direction = 1, 
                          na.value = "white")

dev.copy(pdf, paste0(output.dir, "Heatmap_upregulated_DEGs_shared_and_unique_stim_clusters.pdf"))
dev.off()


temp.seurat <- AverageExpression(CD8.seurat, 
                                 features = goi,
                                 return.seurat = TRUE)


DoHeatmap(temp.seurat, 
          group.colors = CD8.cols,
          features = goi, 
          raster = FALSE,
          draw.lines = FALSE) + 
  scico::scale_fill_scico(palette = "batlow", 
                          direction = 1, 
                          na.value = "white")

dev.copy(pdf, paste0(output.dir, "Heatmap_upregulated_DEGs_shared_and_unique_stim_clusters_average.pdf"))
dev.off()



############################
# Down-regulated genes
############################

stim.1.goi.dn <- DEG.markers %>%
  filter(cluster %in% c("Stimulated_1")) %>%
  filter(p_val_adj < 0.05) %>%
  filter(avg_log2FC < 0) %>%
  arrange(avg_log2FC) %>%
  pull(gene)

stim.exh.goi.dn <- DEG.markers %>%
  filter(cluster %in% c("Stimulated_exhausted")) %>%
  filter(p_val_adj < 0.05) %>%
  filter(avg_log2FC < 0) %>%
  arrange(avg_log2FC) %>%
  pull(gene)


# Venn overlap of up-regulated genes
venn <- venn.diagram(x = list(stim.exh.goi.dn, stim.1.goi.dn),
                     category.names = c("Stimulated_exhausted", "Stimulated_1"),
                     filename = NULL,
                     output=F,
                     
                     # Circles
                     lwd = 2,
                     lty = 'blank',
                     fill = condition.cols,
                     
                     # Numbers
                     cex = .6,
                     fontface = "bold",
                     fontfamily = "sans",
                     
                     # Set names
                     cat.cex = 0.6,
                     cat.fontface = "bold",
                     cat.default.pos = "outer",
                     cat.fontfamily = "sans")

pdf(file = paste0(output.dir,"Dn_regulated_Venn.pdf"))
grid.draw(venn)
dev.off()


# calculate union and unique gois
unique.stim.1.dn <- stim.1.goi.dn[!stim.1.goi.dn %in% stim.exh.goi.dn]
length(unique.stim.1.dn)

unique.stim.exh.dn <- stim.exh.goi.dn[!stim.exh.goi.dn %in% stim.1.goi.dn]
length(unique.stim.exh.dn)

overlap.goi.dn <- stim.1.goi.dn[stim.1.goi.dn %in% stim.exh.goi.dn]
length(overlap.goi.dn)


# arrange genes for plotting

goi <- unique(c(unique.stim.1.dn, overlap.goi.dn, unique.stim.exh.dn))


small.seurat <- ScaleData(small.seurat,
                          features = goi)


DoHeatmap(small.seurat, 
          group.colors = CD8.cols,
          raster = FALSE,
          features = goi) + 
  scico::scale_fill_scico(palette = "batlow", 
                          direction = 1, 
                          na.value = "white")

dev.copy(pdf, paste0(output.dir, "Heatmap_dnregulated_DEGs_shared_and_unique_stim_clusters.pdf"))
dev.off()


temp.seurat <- AverageExpression(CD8.seurat, 
                                 features = goi,
                                 return.seurat = TRUE)


DoHeatmap(temp.seurat, 
          group.colors = CD8.cols,
          features = goi, 
          raster = FALSE,
          draw.lines = FALSE) + 
  scico::scale_fill_scico(palette = "batlow", 
                          direction = 1, 
                          na.value = "white")

dev.copy(pdf, paste0(output.dir, "Heatmap_dnregulated_DEGs_shared_and_unique_stim_clusters_average.pdf"))
dev.off()


##########################################
# Create a list for downstream plotting 
##########################################

stim.cluster.degs <- list("unique_stim_1_up" = unique.stim.1.up,
                          "unique_stim_exh_up" = unique.stim.exh.up,
                          "overlap_up" = overlap.goi.up, 
                          "unique_stim_1_dn" = unique.stim.1.dn, 
                          "unique_stim_exh_dn" = unique.stim.exh.dn,
                          "overlap_dn" = overlap.goi.dn)


```




##annotated heatmaps

```{r Stim_clusters}

# Source custom function // loaded in custom functions chunk 
source("scripts/functions/annotate_seurat_heatmap_function.R")

# Create output directory
output.dir <- "results/Manuscript/Figure2/annotated_heatmaps/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}

# set up seurat object
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"


######################
# Stim cluster DEGs
######################

goi <- unique(c(stim.cluster.degs$unique_stim_1_up, stim.cluster.degs$overlap_up, stim.cluster.degs$unique_stim_exh_up))


label.var <- list("Stim_1" = c("XCL1", "XCL2", "MYC", "CD69", "TNFSF9", "CDKN1A", "GPR183", "HLA-E", "CD44", "CD6"), 
                  "overlap" = c("IFNG", "TNF", "VISR", "CRTAM", "TNFSF14", "TNFRSF9", "GZMB", "ICOS", "IL21R", "CD226"), 
                  "Stim_Exhausted" = c("LAG3", "TIGIT", "HAVCR2", "NKG7", "PDCD1", "CD96", "ENTPD1", "GNLY", "PRF1", "TNFSF4",
                                       "BHLHE40", "TOX", "AHR", "XBP1", "IL2RG", "TNFRSF25"))


label.var <- as.vector(unlist(label.var))


pdf(paste0(output.dir, "Stim_cluster_unique_and_overlap_annotated.pdf"))

# Plot heatmap
annotated.heatmap(CD8.seurat, 
                  cluster.id = "Clusters_l1",
                  assay.use = "RNA",
                  average.expression = TRUE,
                  goi = goi, 
                  genes.to.label = label.var, 
                  col_order = names(CD8.cols), 
                  col.colours = CD8.cols)

dev.off()

```









### Visualise DEGs in common and unique between ISG and Cytotoxic
```{r Vis_cyto_ISG_DEGs}

# Create output directory
output.dir <- "results/Manuscript/Figure2/ISG_cyto/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}

# set input directory
input.dir <- "results/Manuscript/Figure2/"

# set up seurat object
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"

# load data 
DEG.markers <- read.csv(paste0(input.dir, "DEG_CD8_Clusters_only.csv"))


##################################################
# Venn overlap of ISG and Cyto clusters
##################################################


# Load library
# install.packages("VennDiagram")
library(VennDiagram)
library(grDevices)

unique(DEG.markers$cluster)
############################
# up-regulated genes
############################

ISG.goi.up <- DEG.markers %>%
  filter(cluster %in% c("Type_I_IFN")) %>%
  filter(p_val_adj < 0.05) %>%
  filter(avg_log2FC > 0) %>%
  arrange(desc(avg_log2FC)) %>%
  pull(gene)

Cytotoxic.goi.up <- DEG.markers %>%
  filter(cluster %in% c("Cytotoxic")) %>%
  filter(p_val_adj < 0.05) %>%
  filter(avg_log2FC > 0) %>%
  arrange(desc(avg_log2FC)) %>%
  pull(gene)


# Venn overlap of up-regulated genes
venn <- venn.diagram(x = list(ISG.goi.up, Cytotoxic.goi.up),
                     category.names = c("Type_I_IFN", "Cytotoxic"),
                     filename = NULL,
                     output=F,
                     
                     # Circles
                     lwd = 2,
                     lty = 'blank',
                     fill = condition.cols,
                     
                     # Numbers
                     cex = .6,
                     fontface = "bold",
                     fontfamily = "sans",
                     
                     # Set names
                     cat.cex = 0.6,
                     cat.fontface = "bold",
                     cat.default.pos = "outer",
                     cat.fontfamily = "sans")

pdf(file = paste0(output.dir,"Up_regulated_Venn.pdf"))
grid.draw(venn)
dev.off()


# calculate union and unique gois
unique.ISG.up <- ISG.goi.up[!ISG.goi.up %in% Cytotoxic.goi.up]
length(unique.ISG.up)

unique.cyto.up <- Cytotoxic.goi.up[!Cytotoxic.goi.up %in% ISG.goi.up]
length(unique.cyto.up)

overlap.goi.up <- ISG.goi.up[ISG.goi.up %in% Cytotoxic.goi.up]
length(overlap.goi.up)


# arrange genes for plotting

goi <- unique(c(unique.ISG.up, overlap.goi.up, unique.cyto.up))


small.seurat <- ScaleData(small.seurat,
                          features = goi)


DoHeatmap(small.seurat, 
          group.colors = CD8.cols,
          raster = FALSE,
          features = goi) + 
  scico::scale_fill_scico(palette = "batlow", 
                          direction = 1, 
                          na.value = "white")

dev.copy(pdf, paste0(output.dir, "Heatmap_upregulated_DEGs_shared_and_unique_ISG_cyto.pdf"))
dev.off()


temp.seurat <- AverageExpression(CD8.seurat, 
                                 features = goi,
                                 return.seurat = TRUE)


DoHeatmap(temp.seurat, 
          group.colors = CD8.cols,
          features = goi, 
          raster = FALSE,
          draw.lines = FALSE) + 
  scico::scale_fill_scico(palette = "batlow", 
                          direction = 1, 
                          na.value = "white")

dev.copy(pdf, paste0(output.dir, "Heatmap_upregulated_DEGs_shared_and_unique_ISG_cyto_averaged.pdf"))
dev.off()



############################
# Down-regulated genes
############################

ISG.goi.dn <- DEG.markers %>%
  filter(cluster %in% c("Type_I_IFN")) %>%
  filter(p_val_adj < 0.05) %>%
  filter(avg_log2FC < 0) %>%
  arrange(avg_log2FC) %>%
  pull(gene)

Cytotoxic.goi.dn <- DEG.markers %>%
  filter(cluster %in% c("Cytotoxic")) %>%
  filter(p_val_adj < 0.05) %>%
  filter(avg_log2FC < 0) %>%
  arrange(avg_log2FC) %>%
  pull(gene)


# Venn overlap of up-regulated genes
venn <- venn.diagram(x = list(ISG.goi.dn, Cytotoxic.goi.dn),
                     category.names = c("Type_I_IFN", "Cytotoxic"),
                     filename = NULL,
                     output=F,
                     
                     # Circles
                     lwd = 2,
                     lty = 'blank',
                     fill = condition.cols,
                     
                     # Numbers
                     cex = .6,
                     fontface = "bold",
                     fontfamily = "sans",
                     
                     # Set names
                     cat.cex = 0.6,
                     cat.fontface = "bold",
                     cat.default.pos = "outer",
                     cat.fontfamily = "sans")

pdf(file = paste0(output.dir,"Dn_regulated_Venn.pdf"))
grid.draw(venn)
dev.off()


# calculate union and unique gois
unique.ISG.dn <- ISG.goi.dn[!ISG.goi.dn %in% Cytotoxic.goi.dn]
length(unique.ISG.dn)

unique.cyto.dn <- Cytotoxic.goi.dn[!Cytotoxic.goi.dn %in% ISG.goi.dn]
length(unique.cyto.dn)

overlap.goi.dn <- ISG.goi.dn[ISG.goi.dn %in% Cytotoxic.goi.dn]
length(overlap.goi.dn)


# arrange genes for plotting

goi <- unique(c(unique.ISG.dn, overlap.goi.dn, unique.cyto.dn))


small.seurat <- ScaleData(small.seurat,
                          features = goi)


DoHeatmap(small.seurat, 
          group.colors = CD8.cols,
          raster = FALSE,
          features = goi) + 
  scico::scale_fill_scico(palette = "batlow", 
                          direction = 1, 
                          na.value = "white")

dev.copy(pdf, paste0(output.dir, "Heatmap_dnregulated_DEGs_shared_and_unique_ISG_cyto.pdf"))
dev.off()


temp.seurat <- AverageExpression(CD8.seurat, 
                                 features = goi,
                                 return.seurat = TRUE)


DoHeatmap(temp.seurat, 
          group.colors = CD8.cols,
          features = goi, 
          raster = FALSE,
          draw.lines = FALSE) + 
  scico::scale_fill_scico(palette = "batlow", 
                          direction = 1, 
                          na.value = "white")

dev.copy(pdf, paste0(output.dir, "Heatmap_dnregulated_DEGs_shared_and_unique_ISG_cyto_averaged.pdf"))
dev.off()



##########################################
# Create a list for downstream plotting 
##########################################

ISG.cyto.cluster.degs <- list("unique_ISG_up" = unique.ISG.up,
                          "unique_Cyto_up" = unique.cyto.up,
                          "overlap_up" = overlap.goi.up, 
                          "unique_ISG_dn" = unique.ISG.dn, 
                          "unique_Cyto_dn" = unique.cyto.dn,
                          "overlap_dn" = overlap.goi.dn)



```




##annotated heatmaps

```{r annotated_heatmap_ISG_cyto_clusters}

# Source custom function // loaded in custom functions chunk 
source("scripts/functions/annotate_seurat_heatmap_function.R")

# Create output directory
output.dir <- "results/Manuscript/Figure2/annotated_heatmaps/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}

# set up seurat object
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"



######################
# ISG and Cyto DEGs
######################

goi <- unique(c(ISG.cyto.cluster.degs$unique_ISG_up, ISG.cyto.cluster.degs$overlap_up, ISG.cyto.cluster.degs$unique_Cyto_up))
ISG.cyto.cluster.degs$unique_Cyto_up


label.var <- list("ISG" = c("IFI6", "ISG15", "MX1", "ISG20", "LY6E", "OAS1", "STAT2", "GPR183", "TAP1", "CCR7", "SELL", "TNFRSF14", "IL32"), 
                  "overlap" = c("GZMK", "CXCR3", "EOMES", "GZMM", "KLRG1", "TRAT1"), 
                  "Cyto" = c("CD27", "CXCR4", "GZMA", "CD5", "NKG7", "CCR5", "TCF7"))


label.var <- as.vector(unlist(label.var))


pdf(paste0(output.dir, "ISG_Cyto_unique_and_overlap_annotated.pdf"))

# Plot heatmap
annotated.heatmap(CD8.seurat, 
                  cluster.id = "Clusters_l1",
                  assay.use = "RNA",
                  average.expression = TRUE,
                  goi = goi, 
                  genes.to.label = label.var, 
                  col_order = names(CD8.cols), 
                  col.colours = CD8.cols)

dev.off()

```






### Visualise GO terms enriched in shared and unique genes between stim clusters
```{r stim_cluster_GO_enrichment}

# Create output directory
output.dir <- "results/Manuscript/Figure2/GO_plots/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}


# load packages
library("enrichR")
#install.packages("SCpubr")
library("SCpubr")
 
# set serach to human genes
enrichR::setEnrichrSite(site = "Enrichr")

dbs <- enrichR::listEnrichrDbs()
dbs <- sort(dbs$libraryName)

# Choose the dataset to query against.
dbs_use <- c("GO_Biological_Process_2021", 
             "GO_Cellular_Component_2021",
             "GO_Molecular_Function_2021", 
             "KEGG_2021_Human",
             "MSigDB_Hallmark_2020",
             "Reactome_2022",
             "WikiPathway_2021_Human",
             "Azimuth_Cell_Types_2021")



############################################################
# Perform enrichment analysis for Upregulated genes 
############################################################

x <- list("Unique_to_stim_1" = unique.stim.1, 
          "Unique_to_stim_exh" = unique.stim.exh, 
          "Shared_stim_clusters" = overlap.goi)

for(dataset in seq_along(x)){
  
dataset.name <- names(x)[dataset]
# List of genes to use as input.
up.genes <- x[[dataset]]

# Retrieve the enriched terms.
up.enriched.terms <- enrichR::enrichr(up.genes, dbs_use)

# Plot enriched terms 
upregulated.plots <- SCpubr::do_TermEnrichmentPlot(enriched_terms = up.enriched.terms, 
                                                   nchar_wrap = 20,
                                                   font.size = 10, 
                                                   font.type = "sans", 
                                                   text_labels_size = 5,
                                                   colors.use = c("salmon2", "gray"),
                                                   nterms = 10)


# Plot and save all plots generated 
for(i in seq_len(length(upregulated.plots))){
  
  # upregulated plots
  name.var <- names(upregulated.plots)[i]
  
  ggsave(filename = paste0(output.dir, "Upregulated_", dataset.name, "_",  name.var, ".pdf"),
         plot = upregulated.plots[[i]], 
         width = 20, 
         height = 20, 
         units = "in",
         device = "pdf")
  
}

}


















############################################################
# Perform enrichment analysis for Downregulated genes 
############################################################

# List of genes to use as input.
dn.genes <- unique(IFN.sig.dn$gene_symbol)

# Retrieve the enriched terms.
dn.enriched.terms <- enrichR::enrichr(dn.genes, dbs_use)

# Plot enriched terms 
downregulated.plots <- SCpubr::do_TermEnrichmentPlot(enriched_terms = dn.enriched.terms, 
                                                   nchar_wrap = 20,
                                                   font.size = 5, 
                                                   font.type = "sans", 
                                                   text_labels_size = 1,
                                                   colors.use = c("royalblue", "gray"),
                                                   nterms = 10)

# Plot and save all plots generated 
for(i in seq_len(length(upregulated.plots))){
  
  
  # upregulated plots
  name.var <- names(upregulated.plots)[i]
  
  ggsave(filename = paste0(output.dir, "Upregulated_",  name.var, ".pdf"),
         plot = upregulated.plots[[i]], 
         width = 20, 
         height = 20, 
         units = "in",
         device = "pdf")
  
  # downregulated plots
  name.var <- names(downregulated.plots)[i]
  
  ggsave(filename = paste0(output.dir, "Downregulated_",  name.var, ".pdf"),
         plot = downregulated.plots[[i]], 
         width = 20, 
         height = 20, 
         units = "in",
         device = "pdf")
  
}








```


### Subset data for Cytotoxic, IFN, Stim-1 and DEG calculation and visualisation
```{r DEG_calculation}

# Create output directory
output.dir <- "results/Manuscript/Figure2/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}


Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"

Idents(CD8.seurat)
temp.seurat <- subset(CD8.seurat, idents = c("Cytotoxic", "Type_I_IFN", "Stimulated_1"))


if(long.compute){
# Run DEG analysis
DEG.markers <- FindAllMarkers(temp.seurat, 
                              slot = "data", 
                              logfc.threshold = 0.25,
                              test.use = "wilcox",
                              only.pos = TRUE)

# Filter for significance
DEG.markers <- 
  DEG.markers %>%
  filter(p_val_adj < 0.05)


write.csv(DEG.markers, paste0(output.dir, "DEG_Cytotoxic_IFN_Stim_1_Clusters_only.csv"))
}

DEG.markers <- read.csv(paste0(output.dir, "DEG_Cytotoxic_IFN_Stim_1_Clusters_only.csv"))

small.seurat <- subset(temp.seurat, downsample = 300)
small.seurat <- ScaleData(small.seurat)



goi <- DEG.markers %>%
  group_by(cluster) %>%
  #filter(avg_log2FC > 1) %>%
  top_n(10, avg_log2FC) %>%
  pull(gene)

goi <- unique(goi)


DoHeatmap(small.seurat, 
          features = goi) + scico::scale_fill_scico(palette = "batlow", 
                                                    direction = 1, 
                                                    na.value = "white")




```

### Subset data for Exhausted, Stim_exhausted and DEG calculation and visualisation
```{r DEG_calculation}


# Create output directory
output.dir <- "results/Manuscript/Figure2/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}


Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"


Idents(CD8.seurat)
temp.seurat <- subset(CD8.seurat, idents = c("Exhausted_1", "Exhausted_2", "Stimulated_exhausted"))


if(long.compute){
# Run DEG analysis
DEG.markers <- FindAllMarkers(temp.seurat, 
                              slot = "data", 
                              logfc.threshold = 0.25,
                              test.use = "wilcox",
                              only.pos = TRUE)

# Filter for significance
DEG.markers <- 
  DEG.markers %>%
  filter(p_val_adj < 0.05)


write.csv(DEG.markers, paste0(output.dir, "DEG_Exhausted_stim_exhasted_Clusters_only.csv"))
}


small.seurat <- subset(temp.seurat, downsample = 300)

goi <- DEG.markers %>%
  group_by(cluster) %>%
  #filter(avg_log2FC > 1) %>%
  top_n(10, avg_log2FC) %>%
  pull(gene)

goi <- unique(goi)



DoHeatmap(small.seurat, 
          features = goi) + scico::scale_fill_scico(palette = "batlow", 
                                                    direction = 1, 
                                                    na.value = "white")


```

### DEG and vis Stim-1 vs Stim-exhausted
```{r DEG_calculation}


# Create output directory
output.dir <- "results/Manuscript/Figure2/Stim_vs_StimExh/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}


Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"


if(long.compute){

  # Run DEG analysis
  DEG.markers <- FindMarkers(CD8.seurat, 
                             ident.1 = "Stimulated_1",
                             ident.2 = "Stimulated_exhausted",
                             slot = "data", 
                             logfc.threshold = 0.25,
                             test.use = "wilcox",
                             only.pos = FALSE)
  
# Filter for significance
DEG.markers <- 
  DEG.markers %>%
  filter(p_val_adj < 0.05)


write.csv(DEG.markers, paste0(output.dir, "DEG_Stim_vs_StimEx.csv"))
}


# read in dataset
DEG.markers <- read.csv(paste0(output.dir, "DEG_Stim_vs_StimEx.csv"))


goi <- DEG.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::mutate(abs.log = abs(avg_log2FC)) %>%
  dplyr::filter(abs.log > 0.5) %>%
  dplyr::arrange(desc(avg_log2FC)) %>%
  pull(X)

goi <- unique(goi)


temp.seurat <- AverageExpression(CD8.seurat, 
                                 features = goi,
                                 return.seurat = TRUE)


DoHeatmap(temp.seurat, 
          group.colors = CD8.cols,
          features = goi, 
          raster = FALSE,
          draw.lines = FALSE) + 
  scico::scale_fill_scico(palette = "batlow", 
                          direction = 1, 
                          na.value = "white")


dev.copy(pdf, paste0(output.dir, "Heatmap_DEGs_between_stim_clusters_logfc_greater_abs_0_5_averaged.pdf"))
dev.off()

########################
# Perform GO analysis 
########################

GO.input <- DEG.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::mutate(abs.log = abs(avg_log2FC)) %>%
  dplyr::filter(abs.log > 0.5) %>%
  dplyr::arrange(desc(avg_log2FC))

GO.input.stim1 <- GO.input %>%
  dplyr::filter(avg_log2FC > 0.5) %>%
  pull(X)

GO.input.stimexh <- GO.input %>%
  dplyr::filter(avg_log2FC < -0.5) %>%
  pull(X)



# load packages
library("enrichR")
#install.packages("SCpubr")
library("SCpubr")
 
# set serach to human genes
enrichR::setEnrichrSite(site = "Enrichr")

dbs <- enrichR::listEnrichrDbs()
dbs <- sort(dbs$libraryName)

# Choose the dataset to query against.
dbs_use <- c("GO_Biological_Process_2021", 
             "GO_Cellular_Component_2021",
             "GO_Molecular_Function_2021", 
             "KEGG_2021_Human",
             "MSigDB_Hallmark_2020",
             "Reactome_2022",
             "WikiPathway_2021_Human",
             "Azimuth_Cell_Types_2021")


x <- list("Stim1" = GO.input.stim1, 
          "StimExh" = GO.input.stimexh)

for(dataset in seq_along(x)){
  
dataset.name <- names(x)[dataset]
# List of genes to use as input.
up.genes <- x[[dataset]]

# Retrieve the enriched terms.
up.enriched.terms <- enrichR::enrichr(up.genes, dbs_use)

# Plot enriched terms 
upregulated.plots <- SCpubr::do_TermEnrichmentPlot(enriched_terms = up.enriched.terms, 
                                                   nchar_wrap = 20,
                                                   font.size = 10, 
                                                   font.type = "sans", 
                                                   text_labels_size = 5,
                                                   colors.use = c("salmon2", "gray"),
                                                   nterms = 10)


# Plot and save all plots generated 
for(i in seq_len(length(upregulated.plots))){
  
  # upregulated plots
  name.var <- names(upregulated.plots)[i]
  
  ggsave(filename = paste0(output.dir, "Upregulated_", dataset.name, "_",  name.var, ".pdf"),
         plot = upregulated.plots[[i]], 
         width = 20, 
         height = 20, 
         units = "in",
         device = "pdf")
  
}

}






```



### Transcription_factor inference
```{r TF_analysis_inference}

# Create output directory
output.dir <- "results/Manuscript/Figure2/TF_inference/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}



# install and load packages
#BiocManager::install("decoupleR") 
library("decoupleR")

# Retrieve prior knowledge network.
network <- decoupleR::get_dorothea(organism = "human",
                                   levels = c("A", "B", "C"))

# Run weighted means algorithm.
activities <- decoupleR::run_wmean(mat = as.matrix(CD8.seurat@assays$RNA@data),
                                   network = network,
                                   .source = "source",
                                   .targe = "target",
                                   .mor = "mor",
                                   times = 100,
                                   minsize = 5)

# General heatmap.
out <- SCpubr::do_TFActivityPlot(sample = CD8.seurat,
                                 cluster_rows = FALSE,
                                 flip = FALSE,
                                 activities = activities, 
                                 n_tfs = 50)
p <- out$heatmaps$average_scores
p

dev.copy(pdf, paste0(output.dir, "/Heatmap_of_TF_activity_top50.pdf"))
dev.off()

# load custom functions to interact with dorothea output 
source("scripts/functions/dorothea_functions.R")


# Add dorothea data to seurat object
CD8.seurat <- add.dorothea.data(CD8.seurat, activities)

# get a table of activity scores 
tf.data <- extract.top.TFs.dorothea(CD8.seurat, 
                         group.var = "Clusters_l1", 
                         return.table = TRUE,
                         n_tfs = 50)

# filter table for top transcription factors
goi <- tf.data %>%
  filter(mean > 1) %>%
  arrange(group.var, mean) %>%
  pull(source)

length(goi)

# Average seurat object
average.seurat <- AverageExpression(CD8.seurat, 
                                    return.seurat = TRUE)

# plot heatmap of top predicted transcripton factor activity scores
DoHeatmap(average.seurat, 
          group.colors = CD8.cols,
          features = goi, 
          raster = FALSE,
          draw.lines = FALSE) + 
  scico::scale_fill_scico(palette = "batlow", 
                          direction = 1, 
                          na.value = "white")


dev.copy(pdf, paste0(output.dir, "/Heatmap_of_TF_activity_top_predicted.pdf"))
dev.off()


```















### GO CD8 dataset
```{r GO_analysis_CD8_Dataset}

if(long.compute){

  library(Scillus)
  
# get a broad list of DEG markers 
Cluster.markers <- FindAllMarkers(CD8.seurat,
                                  assay = "RNA",
                                  only.pos = TRUE,
                                  min.diff.pct = 0.1,
                                  logfc.threshold = 0,
                                  return.thresh = 0.01)



# Filter for significance
Cluster.markers <- 
  Cluster.markers %>%
  filter(p_val_adj < 0.05) %>%
  mutate(abs.FC = abs(avg_log2FC)) %>%
  filter(abs.FC > 0.25)


write.csv(Cluster.markers, paste0(output.dir, "DEG_CD8_clusters_for_GO.csv"))




table(Cluster.markers$cluster) # number of genes DEG per cluster // used to justify using topn = 100

# Plot Gene Ontology analysis
# CC, BP, MF

# CC 
pdf(paste0(output.dir, "GO_Clusters_CC.pdf"), width = 20, height = 20)

plot_all_cluster_go(Cluster.markers,
                    topn = 100,
                    org = "human", 
                    ont = "CC")

dev.off()


# BP 
pdf(paste0(output.dir, "GO_Clusters_BP.pdf"), width = 20, height = 20)

plot_all_cluster_go(Cluster.markers,
                    topn = 100,
                    org = "human", 
                    ont = "BP")

dev.off()



# MF 
pdf(paste0(output.dir, "GO_Clusters_MF.pdf"), width = 20, height = 20)

plot_all_cluster_go(Cluster.markers,
                    topn = 100,
                    org = "human", 
                    ont = "MF")

dev.off()


# Specifically plot Stim-1 and StimEx clusters
# To plot graph for a specific cluster use below code


# Stimulated-1 CC
pdf(paste0(output.dir, "GO_Stimulated_1_CC.pdf"), width = 20, height = 20)

plot_cluster_go(Cluster.markers, 
                cluster_name = "Stimulated_1",
                org = "human",
                ont = "CC")

dev.off()

# Stimulated-1 BP
pdf(paste0(output.dir, "GO_Stimulated_1_BP.pdf"), width = 20, height = 20)

plot_cluster_go(Cluster.markers, 
                cluster_name = "Stimulated_1",
                org = "human",
                ont = "BP")

dev.off()

# Stimulated-1 MF
pdf(paste0(output.dir, "GO_Stimulated_1_MF.pdf"), width = 20, height = 20)

plot_cluster_go(Cluster.markers, 
                cluster_name = "Stimulated_1", #"Stimulated_exhausted"
                org = "human",
                ont = "MF")

dev.off()


# Stimulated_exhausted CC
pdf(paste0(output.dir, "GO_Stimulated_Exhausted_CC.pdf"), width = 20, height = 20)

plot_cluster_go(Cluster.markers, 
                cluster_name = "Stimulated_exhausted",
                org = "human",
                ont = "CC")

dev.off()

# Stimulated_exhausted BP
pdf(paste0(output.dir, "GO_Stimulated_Exhausted_BP.pdf"), width = 20, height = 20)

plot_cluster_go(Cluster.markers, 
                cluster_name = "Stimulated_exhausted",
                org = "human",
                ont = "BP")

dev.off()

# Stimulated_exhausted MF
pdf(paste0(output.dir, "GO_Stimulated_Exhausted_MF.pdf"), width = 20, height = 20)

plot_cluster_go(Cluster.markers, 
                cluster_name = "Stimulated_exhausted",
                org = "human",
                ont = "MF")

dev.off()

}
```









 
### Annotated_heatmaps
```{r annotated_heatmap}


##################################################################
# Use complexheatmap to generate a manuscript ready figure
##################################################################

# Source custom function // loaded in custom functions chunk 
source("scripts/functions/annotate_seurat_heatmap_function.R")


# Make small seurat object for heatmap plotting
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
small.seurat <- subset(CD8.seurat, downsample = 100)
small.seurat <- ScaleData(small.seurat)



# Ensure markers table is loaded
DEG.markers <- read.csv(paste0(output.dir, "DEG_CD8_Clusters_only.csv"))


# Get top cluster markers
genes.stim <- DEG.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_log2FC > 1) %>%
  dplyr::filter(cluster == "Stimulated_1") %>% 
  dplyr::pull(gene)


genes.exh <- DEG.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_log2FC > 1) %>%
  dplyr::filter(cluster == "Stimulated_exhausted") %>% 
  dplyr::pull(gene)


genes.to.plot <- unique(c(genes.stim, genes.exh))





DoHeatmap(small.seurat, genes.to.plot) + scico::scale_fill_scico(palette = "batlow", 
                                                    direction = 1, 
                                                    na.value = "white")



# Genes to be labeled if present in plotted heatmap
goi <- c("XCL1", "XCL2", "TNF", "TNFRSF9", "CRTAM", "IL2", "CD69", "TIGIT", "GZMB", "LAG3", "HAVCR2", "NKG7", "CTLA4", "GNLY", "CD226", "IFNG")


#install.packages("magick")

pdf(paste0(output.dir, "DEG_CD8_Clusters_only_annotated_sig_Stim_clusts.pdf"))
# Plot heatmap averaged values
annotated.heatmap(small.seurat, 
                  cluster.id = "Clusters_l1",
                  assay.use = "RNA",
                  average.expression = FALSE,
                  goi = genes.to.plot, 
                  genes.to.label = goi, 
                  col_order = names(CD8.cols), 
                  col.colours = CD8.cols, 
                  range.val = c(-2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2))

dev.off()




# Get top cluster markers
genes.to.plot <- DEG.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::group_by(cluster) %>%
  dplyr::top_n(50, avg_log2FC) %>% 
  dplyr::pull(gene)


genes.to.plot <- unique(genes.to.plot)



# Genes to be labeled if present in plotted heatmap
goi <- c("ISG15", "IL7R", "GZMK", "MKI67", "XCL2", "HAVCR2", "LTB", "JUNB", "NKG7", "PRF1", "CCR7", "IFNG")


#install.packages("magick")

pdf(paste0(output.dir, "DEG_CD8_Clusters_only_annotated.pdf"))
# Plot heatmap averaged values
annotated.heatmap(CD8.seurat, 
                  cluster.id = "Clusters_l1",
                  assay.use = "RNA",
                  average.expression = TRUE,
                  goi = genes.to.plot, 
                  genes.to.label = goi, 
                  col_order = names(CD8.cols), 
                  col.colours = CD8.cols, 
                  range.val = c(-2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2))

dev.off()



# IFN Signature 



# Get top cluster markers
genes.to.plot <- DEG.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(cluster == "Type_I_IFN") %>%
    dplyr::filter(avg_log2FC > 1) %>%
  dplyr::pull(gene)


genes.to.plot <- unique(genes.to.plot)



# Genes to be labeled if present in plotted heatmap
goi <- c("IFI6", "ISG20", "IFIT1", "OAS1", "IFIT2", "IRF7", "OAS2", "IFI35", "GZMK")



pdf(paste0(output.dir, "DEG_CD8_Clusters_only_annotated_IFNsig.pdf"))

# Plot heatmap averaged values
annotated.heatmap(CD8.seurat, 
                  cluster.id = "Clusters_l1",
                  assay.use = "RNA",
                  average.expression = FALSE,
                  goi = genes.to.plot, 
                  genes.to.label = goi, 
                  col_order = names(CD8.cols), 
                  col.colours = CD8.cols, 
                  range.val = c(-2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2))

dev.off()


```


### Distribution of cells across stimulation condition
```{r stim_condition_distribution}

# Get basic stats on cells per cluster and per condition and plot these values 

cells.conditon.clust <- table(CD8.seurat@meta.data$condition, CD8.seurat@meta.data$Clusters_l1)

# remove innate cluster columns 
pattern.val <- c("^gd_T_.*", "^MAIT$")
remove.logic <- grepl(paste0(pattern.val, collapse = "|"), colnames(cells.conditon.clust))
cells.conditon.clust <- cells.conditon.clust[ , !remove.logic]



cells.cluster <- colSums(cells.conditon.clust)


output.df <- cells.conditon.clust*NA

for(i in 1:length(cells.cluster)){
output.df[,i] <- (cells.conditon.clust[,i]/cells.cluster[i]) * 100
}


colSums(output.df)

# Plot freq of sample per cluster
barplot(output.df, 
        las = 2,
        col = condition.cols, 
        cex.names = 0.4)

dev.copy(pdf, paste0(output.dir, "Freq_of_condition_per_cluster.pdf"))
dev.off()

```

### Transcription factor analysis
```{r TF_analysis}

# Searching for TFs
TF.ref.list <- read.delim("data/TFCheckpoint.txt") # http://www.tfcheckpoint.org

dim(TF.ref.list) # 3479, 36
colnames(TF.ref.list)
head(TF.ref.list)

unique(TF.ref.list$evidence.type)

# subset for transcription factors 
TF.ref.list <- TF.ref.list %>%
  dplyr::filter(DbTF == "yes") %>%
  dplyr::filter(evidence.type == "Experimental")

dim(TF.ref.list) # 825, 36

Cluster.markers <- read.csv(paste0(output.dir, "DEG_CD8_Clusters_only.csv"), row.names = 1)

# annotate genes as TF or not
tf.logic <- Cluster.markers$gene %in% TF.ref.list$gene_symbol
Cluster.markers$TF <- tf.logic


# Up-regulated TFs
sig.genes <- Cluster.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_log2FC > 0.5) %>%
  dplyr::filter(TF == TRUE) %>%
  dplyr::group_by(cluster) %>%
  dplyr::top_n(10, avg_log2FC) %>%
  dplyr::select(gene)
  
TF.DEG <- unique(sig.genes$gene)
TF.DEG


########################################
# Plot TFs differentially regulated 
########################################

Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"

average.seurat <- AverageExpression(CD8.seurat,
                                    assay = "RNA",
                                    slot = "data",
                                    verbose = TRUE,
                                    return.seurat = TRUE)
DoHeatmap(average.seurat, 
          features = TF.DEG, 
          draw.lines = FALSE, 
          group.colors = CD8.cols,
          raster = FALSE) + scico::scale_fill_scico(palette = "batlow", 
                                       direction = 1, 
                                       na.value = "white")

dev.copy(pdf, paste0(output.dir, "Heatmap_of_TF_diff_expressed.pdf"))
dev.off()


```


### CD8 seurat TCR data analysis
```{r CD8_TCR_analysis}

# Key questions
# Why do some cells move between clusters after stimulation 
# TCR recogniton between stim_1 and Stim_exhausted, is one bystander and the other tumor-specific, what is the clonal overlap between populations 


#install.packages("ggraph")
library(ggraph)

#devtools::install_github("ncborcherding/scRepertoire@dev")
library("scRepertoire")




# Show the distribution of clonotypes 

Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$cloneType

clonetypes <- levels(Idents(CD8.seurat.tcr))
clonotype.names <- c("Hyperexpanded", "Large", "Medium", "Small", "Single")




for(i in seq_along(clonetypes)){
  print(scCustomize::Cluster_Highlight_Plot(CD8.seurat.tcr, 
                                            cluster_name = paste0(clonetypes[i]), 
                                            highlight_color = "navy",
                                            background_color = "lightgray",
                                            split.by = "condition",
                                            pt.size = 1) + 
          NoLegend() + 
          ggtitle(paste0(clonetypes[i])))
  
  dev.copy(pdf, paste0(output.dir, "UMAP_highlighted_clonotype_", clonotype.names[i], "_split_condition.pdf"))
  dev.off()
}








# use .filt dataset which has single clonetypes removed


# subset data
Idents(CD8.seurat.tcr.filt) <- CD8.seurat.tcr.filt@meta.data$condition
US.seurat <- subset(CD8.seurat.tcr.filt, idents = "US")
Stim.seurat <- subset(CD8.seurat.tcr.filt, idents = "Stim")

# Reset Idents for plotting
Idents(CD8.seurat.tcr.filt) <- CD8.seurat.tcr.filt@meta.data$Clusters_l1
Idents(US.seurat) <- US.seurat@meta.data$Clusters_l1
Idents(Stim.seurat) <- Stim.seurat@meta.data$Clusters_l1


# First show the clonal overlap between clusters 
clonalOverlap(US.seurat, 
              cloneCall = "aa",
              chain = "both",
              method = "overlap")

dev.copy(pdf, paste0(output.dir, "Clonal_overlap_US_only.pdf"))
dev.off()

clonalOverlap(Stim.seurat,
              cloneCall = "aa",
              chain = "both",
              method = "overlap")

dev.copy(pdf, paste0(output.dir, "Clonal_overlap_Stim_only.pdf"))
dev.off()

clonalOverlap(CD8.seurat.tcr.filt,
              cloneCall = "aa",
              chain = "both",
              method = "overlap")

dev.copy(pdf, paste0(output.dir, "Clonal_overlap.pdf"))
dev.off()






# Plot bargraph of clonetype frequency per cluster

Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$cloneType

cells.clonetype.per.clust<- table(CD8.seurat.tcr@meta.data$cloneType, CD8.seurat.tcr@meta.data$Clusters_l1)

cells.cluster <- colSums(cells.clonetype.per.clust)

output.df <- cells.clonetype.per.clust*NA

for(i in 1:length(cells.cluster)){
output.df[,i] <- (cells.clonetype.per.clust[,i]/cells.cluster[i]) * 100
}


colSums(output.df)

clonetype.cols <- c("salmon4", "salmon", "orange", "cyan3", "Lightblue")
names(clonetype.cols) <- c("Hyperexpanded", "Large", "Medium", "Small", "Single")



# Plot freq of sample per cluster
barplot(output.df, 
        las = 2,
        legend = T,
        col = clonetype.cols, 
        cex.names = 0.5)

dev.copy(pdf, paste0(output.dir, "Freq_of_clonetype_per_cluster.pdf"))
dev.off()



# Overview of clonotype overlap 

UMAPPlot(CD8.seurat.tcr, 
         group.by = "Clonotype_overlap")


Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$Clonotype_overlap

overlap.status <- levels(Idents(CD8.seurat.tcr))



for(i in seq_along(overlap.status)){
  print(scCustomize::Cluster_Highlight_Plot(CD8.seurat.tcr, 
                                            cluster_name = paste0(overlap.status[i]), 
                                            highlight_color = "navy",
                                            background_color = "lightgray",
                                            pt.size = 1) + 
          NoLegend() + 
          ggtitle(paste0(overlap.status[i])))
  
  dev.copy(pdf, paste0(output.dir, "UMAP_highlighted_clonotype_", overlap.status[i], ".pdf"))
  dev.off()
}











# subset data for shared clonotypes only
Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$Clonotype_overlap
temp.seurat <- subset(CD8.seurat.tcr, idents = "Shared")
Idents(temp.seurat) <- temp.seurat@meta.data$Clusters_l1


scRepertoire::clonalNetwork(temp.seurat, 
                                    reduction = "umap", 
                                    identity = "Clusters_l1",
                                    filter.proportion = 0.05,
                                    filter.identity = "Stimulated_1",
                                    cloneCall = "aa") + ggtitle("Shared clonetypes")

dev.copy(pdf, "temp.pdf")
dev.off()


seurat.list <- expression2List(CD8.seurat.tcr.filt, split.by = "Clusters_l1")


clonalProportion(seurat.list,
                 split = split.var,
                 cloneCall = "aa") 


clonalHomeostasis(tcr.clusters, 
                  cloneType = homeostasis.var,
                  cloneCall = "aa") + 
  theme(axis.text.x = element_text(angle = 90))



# for clonalProportion split
split.var <- c(1, 5, 10, 20, 50, 500, 5000, 1e+05)

# for clonalHomeostasis split
homeostasis.var <- c(Rare = 0.001, 
                     very_small = 0.002,
                     Small = 0.005, 
                     Medium = 0.01, 
                     Large = 0.05, 
                     very_Large = 0.1, 
                     Hyperexpanded = 1)






UMAPPlot(object = CD8.seurat.tcr,
         group.by = "Clusters_l1",
         label = TRUE, 
         cols = CD8.cols,
         pt.size = 1,
         label.size = 6)  + NoLegend()








# visualise clusters
p1 <- UMAPPlot(CD8.seurat.tcr, 
         group.by = "Clusters_l1",
         label = T) + NoLegend() + ggtitle("")

# clusters to plot
clust.ids <- as.character(unique(CD8.seurat.tcr@meta.data$Clusters_l1))

# plot all clusters
for(i in seq_along(clust.ids)){

  p2 <- scRepertoire::clonalNetwork(US.seurat, 
                                    reduction = "umap", 
                                    identity = "Clusters_l1",
                                    filter.proportion = 0.01,
                                    filter.identity = clust.ids[i],
                                    cloneCall = "aa") + ggtitle("US")
  
  p3 <- scRepertoire::clonalNetwork(Stim.seurat, 
                                    reduction = "umap", 
                                    identity = "Clusters_l1",
                                    filter.proportion = 0.01,
                                    filter.identity = clust.ids[i],
                                    cloneCall = "aa") + ggtitle("Stim")
  

  print(p1 + p2/p3)
  
 #ggsave(paste0(output.dir, "output_tcr/figures/clonalnetworks/ClonalNetwork_", clust.ids[i], ".pdf"))

  
}


clonalOverlay(CD8.seurat.tcr, reduction = "umap", 
              freq.cutpoint = 1) + 
                guides(color = "none")



scRepertoire::clonalNetwork(CD8.seurat.tcr, 
                                    reduction = "umap", 
                                    identity = "Clusters_l1",
                                    filter.proportion = 0.01,
                                    filter.identity = "Cytotoxic",
                                    cloneCall = "aa") + ggtitle("Stim")


scRepertoire::clonalNetwork(CD8.seurat.tcr, 
                                    reduction = "umap", 
                                    identity = "Clusters_l1",
                                    #filter.proportion = 0.01,
                                    filter.clones =  1,
                                    cloneCall = "aa") + ggtitle("Stim")



CD8.seurat.tcr@meta.data$cloneType

occupiedscRepertoire(CD8.seurat.tcr, x.axis = "Clusters_l1", proportion = T)

?occupiedscRepertoire
CD8.seurat.tcr@meta.data

temp.seurat <- clusterTCR(seurat.list,
                          chain = "TRB",
                          group.by = "group", 
                          sequence = "aa", 
                          threshold = 0.85)

DimPlot(seurat, group.by = "TRB_cluster") + 
  NoLegend()
?clusterTCR


head(CD8.seurat.tcr@meta.data)
StartracDiversity(CD8.seurat.tcr,
                  type = "condition", 
            sample = "group", by = "overall")




clonotypeBias(CD8.seurat.tcr, cloneCall = "aa", split.by = "group", group.by = "Clusters_l1",
n.boots = 20, min.expand =10)

### Misc code for exploring functions of scRep package

seurat.list <- expression2List(CD8.seurat.tcr, split.by = "Clusters_l1")
length(seurat.list) #now listed by cluster









scRepertoire::clonalNetwork(US.seurat, 
                                    reduction = "umap", 
                                    identity = "Clusters_l1",
                                    filter.proportion = 0.01,
                                    filter.identity = "Cytotoxic",
                                    cloneCall = "aa") + ggtitle("US")





```


### Heatmap of top clonotypes
```{r Heatmap_top_clonotypes}

library(gplots)
library(RColorBrewer)
########################
# Heatmap functions
########################

hclust.func <- function(x){
  hclust(x, method = 'ward.D2')
}

dist.func <- function(x){
  dist(x,method = 'binary')
}

col.palette <- colorRampPalette(brewer.pal(8, "RdYlBu"))(600)





# Get data of how many cells in each cluster are assigned to what clonotypes
clonotype.data <- table(CD8.seurat.tcr@meta.data$CTaa, CD8.seurat.tcr@meta.data$Clusters_l1)


# Plot the top 50 clonotypes
aa.keep <- sort(rowSums(clonotype.data), decreasing = TRUE)[1:50]

keep.vec <- rownames(clonotype.data) %in% names(aa.keep)

plot.data <- clonotype.data[keep.vec, ]
dim(plot.data)

heatmap.2(as.matrix(log2(plot.data+1)), 
          scale = "none",
          trace = "none",
          ColSideColors = CD8.cols,
          Colv = TRUE,
          cexRow = 0.8, 
          margins = c(8, 15),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette))     

dev.copy(pdf, paste0(output.dir, "Heatmap_top_50_clonotypes.pdf"))
dev.off()







############################################
# Plot top clonotypes in each cluster 
############################################

for(i in seq_along(colnames(clonotype.data))){
  temp <- as.data.frame(clonotype.data)
  
  temp <- temp %>%
    dplyr::filter(Var2 == paste0(colnames(clonotype.data)[i])) %>%
    dplyr::top_n(10)
  
  clonotypes.to.plot <- temp$Var1
  
  keep.vec <- rownames(clonotype.data) %in% clonotypes.to.plot
  
  plot.data <- clonotype.data[keep.vec, ]
  
  heatmap.2(as.matrix(log2(plot.data+1)), 
            scale = "none",
            trace = "none",
            ColSideColors = CD8.cols,
            Colv = FALSE,
            cexRow = 0.8, 
            margins = c(8, 15),
            cexCol = 0.8, 
            distfun = dist.func,
            hclustfun = hclust.func,
            srtCol = 90,
            col = rev(col.palette))
  
  dev.copy(pdf, paste0(output.dir, "Heatmap_top_10_clonotypes_in_", colnames(clonotype.data)[i], ".pdf"))
  dev.off()
  
}



```


```{r remedy001}

# CD8 dataset
  Trex.seurat <- LoadH5Seurat("data/TREX_.h5Seurat")

Trex.seurat

Idents(Trex.seurat) <- Trex.seurat@meta.data$Clusters_l1
DefaultAssay(Trex.seurat) <- "RNA"

seuratObj <- Trex::quietTCRgenes(seuratObj)
seuratObj <- RunPCA(seuratObj)

DimPlot(Trex.seurat, 
        reduction = "umap")



```


## DEG analysis 
```{r}
# DEG calculation clusters
if(long.compute){
  
  
  Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
  DefaultAssay(CD8.seurat) <- "RNA"
  
  CD8.seurat <- FindVariableFeatures(CD8.seurat)
  
  CD8.seurat <- ScaleData(CD8.seurat)
  
  DEG.wilcox <- FindAllMarkers(CD8.seurat,
                               logfc.threshold = 0.25,
                               min.pct = 0.1,
                               test.use = "wilcox", 
                               only.pos = FALSE)
  

  write.csv(DEG.wilcox, paste0(output.dir, "DEG_RNA_clusters_wilcox.csv"))
  
  print(paste0("Number of sig differentially expressed genes = ", sum(DEG.wilcox$p_val_adj < 0.05)))
}

# load data
DEG.wilcox <- read.csv(paste0(output.dir, "DEG_RNA_clusters_wilcox.csv"))
DEG.wilcox <- DEG.wilcox[,-1]

# prepare dataset
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"



# Subset data
small.seurat <- subset(CD8.seurat, downsample = 300)

# Average dataset
average.seurat <- AverageExpression(CD8.seurat, return.seurat = TRUE)


# Top DEGs

# DEGs > 0.5 log2FC
x <- DEG.wilcox %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_log2FC > 0.5) %>%
  dplyr::select(gene) 

x <- unique(x$gene)
x

# Average expression
DoHeatmap(average.seurat, 
          features = x,
          group.colors = clust.cols,
          assay = "RNA",
          raster = FALSE,
          draw.lines = FALSE) + scico::scale_fill_scico(palette = "batlow", 
                                                        direction = 1, 
                                                        na.value = "white")

dev.copy(pdf, 
         paste0(output.dir, "Top_log2fc_DEGs_RNA_clusters_l1_heatmap_average.pdf"))
dev.off()


# Top 10 DEGs per cluster

x <- DEG.wilcox %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::group_by(cluster) %>%
  dplyr::top_n(10, avg_log2FC) %>%
  dplyr::select(gene) 

x <- unique(x$gene)
x


# Average expression
DoHeatmap(average.seurat, 
                    group.colors = clust.cols,
          features = x,
          assay = "RNA",
          raster = FALSE,
          draw.lines = FALSE) + scico::scale_fill_scico(palette = "batlow", 
                                                                     direction = 1, 
                                                                     na.value = "white") 

dev.copy(pdf, 
         paste0(output.dir, "Top_10_DEGs_RNA_clusters_l1_heatmap_average.pdf"), )
dev.off()
```






