---
title: "02_Manuscript_Figure_X_TCR_analysis"
author: "Dillon Corvino"
date: "07/03/2022"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

### Dataset information

```{r Dataset_Info}

# HNSCC scRNAseq/scTCRseq dataset
# Cells are HNSCC TILs sorted on Lymphocytes/Live/CD3+/CD4-/CD8+
# Human samples
# Each sample is a pool of 4 patients
# samples 1 and 2 are unstimulated 
# samples 3 and 4 are stimulated 
# Patients in sample 1 match patients in sample 3 and sample 2 pairs with 4
# Data acquired was transcript expression, ADT (antibody expression), and TCRA & B sequences
# This analysis uses just Transcript and TCR data


# General information
# CellRanger was used for sequence alignment/QC/counts/TCR calls - performed by Ross (QIMR, Brisbane, Aus)
# DC was provided the output from CellRanger (counts matrix) and used this as input for analysis within this script

```

### Environment

```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Create output directories & load custom functions & colour scheme
source("scripts/Setup.R", local = knitr::knit_global())

# pipeline variables
long.compute <- FALSE

# load required packages
require(tidyverse)
require(Seurat)
library(ggplot2)



# Create output directory
output.dir <- "results/Manuscript/FigureTCR/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}

```


### Reading data

```{r reading_data}

########################
# Load all datasets 
########################

# CD8 dataset
#CD8.seurat <- LoadH5Seurat("saves/CD8_seurat.h5Seurat")

#Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
#DefaultAssay(CD8.seurat) <- "RNA"

# CD8 TCR data

# TCR with no minimum clonotype filter
CD8.seurat.tcr <- LoadH5Seurat("saves/CD8_seurat_tcr.h5Seurat")

Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$Clusters_l1
DefaultAssay(CD8.seurat.tcr) <- "RNA"

# TCR data with a minimum clonotype filter
CD8.seurat.tcr.filt <- LoadH5Seurat("saves/CD8_seurat_tcr_filt.h5Seurat")

Idents(CD8.seurat.tcr.filt) <- CD8.seurat.tcr.filt@meta.data$Clusters_l1
DefaultAssay(CD8.seurat.tcr.filt) <- "RNA"

```


## Figure X {.tabset}


### CD8 seurat TCR data analysis
```{r CD8_TCR_analysis}

#install.packages("ggraph")
library("ggraph")

#devtools::install_github("ncborcherding/scRepertoire@dev")
library("scRepertoire")





# Plot bargraph of clonetype frequency per cluster

Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$cloneType

cells.clonetype.per.clust<- table(CD8.seurat.tcr@meta.data$cloneType, CD8.seurat.tcr@meta.data$Clusters_l1)

cells.cluster <- colSums(cells.clonetype.per.clust)

output.df <- cells.clonetype.per.clust*NA

for(i in 1:length(cells.cluster)){
output.df[,i] <- (cells.clonetype.per.clust[,i]/cells.cluster[i]) * 100
}


colSums(output.df)

clonetype.cols <- c("salmon4", "salmon", "orange", "cyan3", "Lightblue")
names(clonetype.cols) <- c("Hyperexpanded", "Large", "Medium", "Small", "Single")



# Plot freq of sample per cluster
barplot(output.df, 
        las = 2,
        legend = T,
        col = clonetype.cols, 
        cex.names = 0.5)

dev.copy(pdf, paste0(output.dir, "Freq_of_clonetype_per_cluster.pdf"))
dev.off()

# ggplot version 

output.df.gg <- as.data.frame(output.df)

ggplot(output.df.gg) +
  aes(x = Var2, y = Freq, fill = Var1) +
  geom_col() +
  scale_fill_manual(
    values = c(`Hyperexpanded (20 < X <= 150)` = "salmon4",
    `Large (10 < X <= 20)` = "salmon",
    `Medium (5 < X <= 10)` = "orange",
    `Small (1 < X <= 5)` = "cyan3",
    `Single (1)` = "Lightblue")
  ) +
  labs(x = " ", y = "Frequency [%]") +
  theme_minimal()

ggsave(paste0(output.dir, "Freq_of_clonetype_per_cluster_ggversion.pdf"))

# convert seurat object to list in order to use functions 
seurat.list <- expression2List(CD8.seurat.tcr.filt, split.by = "Clusters_l1")



# Clonal overlap 
clonalOverlap(seurat.list,
              cloneCall = "aa",
              method = "overlap")

dev.copy(pdf, paste0(output.dir, "Clonal_overlap.pdf"))
dev.off()



# Clonal Diversity 
clonalDiversity(seurat.list, 
                cloneCall = "aa")
dev.copy(pdf, paste0(output.dir, "Clonal_diversity.pdf"))
dev.off()




# Show the distribution of clonotypes 

Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$cloneType

clonetypes <- levels(Idents(CD8.seurat.tcr))
clonotype.names <- c("Hyperexpanded", "Large", "Medium", "Small", "Single")




for(i in seq_along(clonetypes)){
  print(scCustomize::Cluster_Highlight_Plot(CD8.seurat.tcr, 
                                            cluster_name = paste0(clonetypes[i]), 
                                            highlight_color = "navy",
                                            background_color = "lightgray",
                                            split.by = "condition",
                                            pt.size = 1) + 
          NoLegend() + 
          ggtitle(paste0(clonetypes[i])))
  
  dev.copy(pdf, paste0(output.dir, "UMAP_highlighted_clonotype_", clonotype.names[i], "_split_condition.pdf"))
  dev.off()
}


# Overview of clonotype overlap 

UMAPPlot(CD8.seurat.tcr, 
         group.by = "Clonotype_overlap")


Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$Clonotype_overlap

overlap.status <- levels(Idents(CD8.seurat.tcr))


for(i in seq_along(overlap.status)){
  print(scCustomize::Cluster_Highlight_Plot(CD8.seurat.tcr, 
                                            cluster_name = paste0(overlap.status[i]), 
                                            highlight_color = "navy",
                                            background_color = "lightgray",
                                            pt.size = 1) + 
          NoLegend() + 
          ggtitle(paste0(overlap.status[i])))
  
  dev.copy(pdf, paste0(output.dir, "UMAP_highlighted_clonotype_", overlap.status[i], ".pdf"))
  dev.off()
}




##########

## clonotype overlap venn
# Load library
# install.packages("VennDiagram")
library(VennDiagram)
library(grDevices)


extract_list_clonotypes <- function(seurat_object = CD8.seurat.tcr, overlap_var = "") {

# Subset for overlap_var  
seurat_subset <- subset(x = seurat_object, subset = condition == overlap_var)

# get metadata 
seurat_subset_meta <- seurat_subset[[]]

# get vector with clonotypes
clonotype <- unique(seurat_subset_meta$CTaa)
return(clonotype)
}

# Generate 2 sets 
Stim <- extract_list_clonotypes (seurat_object = CD8.seurat.tcr, overlap_var = "Stim")
US <- extract_list_clonotypes (seurat_object = CD8.seurat.tcr, overlap_var = "US")

# For filtered result
Stim_filt <- extract_list_clonotypes (seurat_object = CD8.seurat.tcr.filt, overlap_var = "Stim")
US_filt <- extract_list_clonotypes (seurat_object = CD8.seurat.tcr.filt, overlap_var = "US")

# Chart non filt
venn <- venn.diagram(
        x = list(US, Stim),
        category.names = c("US", "Stim" ),
        filename = paste0(output.dir,"venn_condition_CTaa_overlap.tiff"),
        output=F,

        # Output features
        type="tiff" , 
        height = 1200 , 
        width = 1200 , 
        resolution = 300,
        compression = "lzw",
        
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = condition.cols,
        
        # Numbers
        cex = .6,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 0.6,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.fontfamily = "sans")

# Chart filt

dir.create(paste0(output.dir, "Venn_filt"))
setwd(paste0(output.dir, "Venn_filt"))

venn_filt <- venn.diagram(
        x = list(US_filt, Stim_filt),
        category.names = c("US", "Stim" ),
        filename = paste0(output.dir,"/Venn_filt/","venn_condition_CTaa_overlap_filt.tiff"),
        output=F,

        # Output features
        type="tiff" , 
        height = 1200 , 
        width = 1200 , 
        resolution = 300,
        compression = "lzw",
        
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = condition.cols,
        
        # Numbers
        cex = .6,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 0.6,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.fontfamily = "sans")

setwd(working.dir)

##########



# Umap visualise top clonotypes
clonotype.data <- table(CD8.seurat.tcr@meta.data$CTaa, CD8.seurat.tcr@meta.data$Clusters_l1)
aa.keep <- sort(rowSums(clonotype.data), decreasing = TRUE)[1:5]

names(aa.keep)



Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$CTaa


for(i in seq_along(names(aa.keep))){
  print(scCustomize::Cluster_Highlight_Plot(CD8.seurat.tcr, 
                                            cluster_name = paste0(names(aa.keep)[i]), 
                                            highlight_color = "navy",
                                            background_color = "lightgray",
                                            pt.size = 1) + 
          NoLegend() + 
          ggtitle(paste0(names(aa.keep)[i])))
  
  dev.copy(pdf, paste0(output.dir, "UMAP_highlighted_clonotype_", names(aa.keep)[i], ".pdf"))
  dev.off()
}




# Umap visualise top clonotypes in IFN cluster
clonotype.data <- table(CD8.seurat.tcr@meta.data$CTaa, CD8.seurat.tcr@meta.data$Clusters_l1)
clonotype.data <- as.data.frame(clonotype.data)

aa.keep <- clonotype.data %>% 
  filter(Var2 == "Type_I_IFN") %>%
  top_n(10, Freq)

aa.keep <- aa.keep$Var1

Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$CTaa


for(i in seq_along(aa.keep)){
  print(scCustomize::Cluster_Highlight_Plot(CD8.seurat.tcr, 
                                            cluster_name = paste0(aa.keep[i]), 
                                            highlight_color = "navy",
                                            background_color = "lightgray",
                                            pt.size = 1) + 
          NoLegend() + 
          ggtitle(paste0(aa.keep[i])))
  
  dev.copy(pdf, paste0(output.dir, "UMAP_highlighted_clonotype_Type_IFN_top_10_", aa.keep[i], ".pdf"))
  dev.off()
}





```


### Heatmap of top clonotypes
```{r Heatmap_top_clonotypes}

library(gplots)
library(RColorBrewer)
########################
# Heatmap functions
########################

hclust.func <- function(x){
  hclust(x, method = 'ward.D2')
}

dist.func <- function(x){
  dist(x,method = 'binary')
}

col.palette <- colorRampPalette(brewer.pal(8, "RdYlBu"))(600)





# Get data of how many cells in each cluster are assigned to what clonotypes
clonotype.data <- table(CD8.seurat.tcr@meta.data$CTaa, CD8.seurat.tcr@meta.data$Clusters_l1)


# Plot the top 50 clonotypes
aa.keep <- sort(rowSums(clonotype.data), decreasing = TRUE)[1:50]

keep.vec <- rownames(clonotype.data) %in% names(aa.keep)

plot.data <- clonotype.data[keep.vec, ]
dim(plot.data)

heatmap.2(as.matrix(log2(plot.data+1)), 
          scale = "none",
          trace = "none",
          ColSideColors = CD8.cols,
          Colv = TRUE,
          cexRow = 0.8, 
          margins = c(8, 15),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette))     

dev.copy(pdf, paste0(output.dir, "Heatmap_top_50_clonotypes.pdf"))
dev.off()







############################################
# Plot top clonotypes in each cluster 
############################################

for(i in seq_along(colnames(clonotype.data))){
  temp <- as.data.frame(clonotype.data)
  
  temp <- temp %>%
    dplyr::filter(Var2 == paste0(colnames(clonotype.data)[i])) %>%
    dplyr::top_n(10)
  
  clonotypes.to.plot <- temp$Var1
  
  keep.vec <- rownames(clonotype.data) %in% clonotypes.to.plot
  
  plot.data <- clonotype.data[keep.vec, ]
  
  heatmap.2(as.matrix(log2(plot.data+1)), 
            scale = "none",
            trace = "none",
            ColSideColors = CD8.cols,
            Colv = FALSE,
            cexRow = 0.8, 
            margins = c(8, 15),
            cexCol = 0.8, 
            distfun = dist.func,
            hclustfun = hclust.func,
            srtCol = 90,
            col = rev(col.palette))
  
  dev.copy(pdf, paste0(output.dir, "Heatmap_top_10_clonotypes_in_", colnames(clonotype.data)[i], ".pdf"))
  dev.off()
  
}



```


```{r remedy001}

# CD8 dataset
  Trex.seurat <- LoadH5Seurat("data/TREX_.h5Seurat")

Trex.seurat

Idents(Trex.seurat) <- Trex.seurat@meta.data$Clusters_l1
DefaultAssay(Trex.seurat) <- "RNA"

seuratObj <- Trex::quietTCRgenes(seuratObj)
seuratObj <- RunPCA(seuratObj)

DimPlot(Trex.seurat, 
        reduction = "umap")



```








