---
title: "HNSCC_DEG_analysis"
author: "Dillon Corvino"
date: "03/02/2020"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

### Environment

```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
#setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Load packages
source("R/Load_packages.R", local = knitr::knit_global())

# Create output directories & load custom functions & colour scheme
source("R/Setup.R", local = knitr::knit_global())


# pipeline variables
quick.load <- TRUE
long.compute <- TRUE

# load saved seurat object with cluster annotations and imputation performed
if(quick.load){
  seurat.combined <- readRDS("Exported_RDS_files/seurat_combined.rds")
}

```

## DEG analysis {.tabset}

### DEGs between clusters

```{r DEGs_between_clusters}

# Create output directory
if(!dir.exists("output/tables/DEG_Clusters")){
  dir.create("output/tables/DEG_Clusters", 
             recursive = T)
}

# Set assay to RNA
DefaultAssay(seurat.combined) <- "RNA"

# Set Idents 
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters

################
# Positive only
################

# Find markers
Cluster.markers.pos <- FindAllMarkers(seurat.combined,
                                      assay = "RNA",
                                      only.pos = TRUE,
                                      min.diff.pct = 0.1,
                                      return.thresh = 0.05)

# Write data to file
write.csv(Cluster.markers.pos, "output/tables/DEG_Clusters/Cluster_markers_unfiltered_pos.csv")

# Filter by sig Padj val
sum(Cluster.markers.pos$p_val_adj < 0.05) # 2,919 genes 

write.csv(Cluster.markers.pos[Cluster.markers.pos$p_val_adj < 0.05, ], "output/tables/DEG_Clusters/Cluster_markers_sig_only_pos.csv")


################
# Pos and Neg
################

# Find markers
Cluster.markers <- FindAllMarkers(seurat.combined,
                                  assay = "RNA",
                                  only.pos = FALSE,
                                  min.diff.pct = 0.1,
                                  return.thresh = 0.05)

# Write data to file
write.csv(Cluster.markers, "output/tables/DEG_Clusters/Cluster_markers_unfiltered.csv")

# Filter by sig Padj val
sum(Cluster.markers$p_val_adj < 0.05) # 4,890 genes 

write.csv(Cluster.markers[Cluster.markers$p_val_adj < 0.05, ], "output/tables/DEG_Clusters/Cluster_markers_sig_only.csv")


########################
# Get top markers
########################

# Get top markers for each cluster - use Positive markers only
sig.markers <- Cluster.markers.pos[Cluster.markers.pos$p_val_adj < 0.05, ]

Cluster.vect <- unique(sig.markers$cluster)

for(i in 1:length(Cluster.vect)){
  
  
  print(paste0("Getting markers for cluster ", Cluster.vect[i]))
  
  input.df <- sig.markers[sig.markers$cluster == Cluster.vect[i], ]
  
  assign(paste0("Cluster_", Cluster.vect[i]), top_n(input.df, 40,  avg_log2FC))
  
  
  x <- eval(parse(text = paste0("Cluster_", Cluster.vect[i])))
  write.csv(x, paste0("output/tables/DEG_Clusters/Top_40_markers_for_cluster_", Cluster.vect[i], ".csv"))
  
}


```

### DEGs of clusters between conditions

```{r DEGs_of_clusts_between_conditions}

# Create output directories
if(!dir.exists("output/tables/DEG_Clusters_btn_conditions")){
  dir.create("output/tables/DEG_Clusters_btn_conditions", 
             recursive = T)
}


if(!dir.exists("output/figures/DEG_Clusters_btn_conditions")){
  dir.create("output/figures/DEG_Clusters_btn_conditions", 
             recursive = T)
}


# Set assay to RNA
DefaultAssay(seurat.combined) <- "integrated"

# Get vector of clusters
clust.var <- seurat.combined@meta.data$seurat_clusters
clust.var <- levels(clust.var)

# Plot just the top 10 up and 10 downregulated DEGs
for(i in 1:length(clust.var)){
  
  print(paste0("Analysis for cluster ", clust.var[i]))
  
  # Set idents to the condition_cluster
  Idents(seurat.combined) <- seurat.combined@meta.data$condition_clust
  
  # Calculating DEGs for each cluster across condition
  DEG.var <- FindMarkers(seurat.combined,
                         ident.1 = paste0("Stim_", clust.var[i]), 
                         ident.2 = paste0("US_", clust.var[i]),
                         test.use = "wilcox",
                         logfc.threshold = 0.25,
                         min.pct = 0.1,
                         only.pos = FALSE,
                         verbose = FALSE)
  
  # create column for gene ID
  DEG.var$gene <- rownames(DEG.var)
  
  # Write data to file
  write.csv(DEG.var, paste0("output/tables/DEG_Clusters_btn_conditions/Stim_vs_US_DEG_clust_", clust.var[i], ".csv"))
  
  #########################################
  # Get top DEGs for clusters Stim vs. US
  #########################################
  
  print(paste0("Getting top markers for cluster ", clust.var[i], " Stim vs. US"))
  
  sig.markers <- DEG.var[DEG.var$p_val_adj < 0.05, ]
  
  assign(paste0("Condition_Cluster_DEGs_", clust.var[i]), top_n(sig.markers, 40,  avg_log2FC))
  
  x <- eval(parse(text = paste0("Condition_Cluster_DEGs_", clust.var[i])))
  write.csv(x, paste0("output/tables/DEG_Clusters_btn_conditions/Top_40_markers_for_Stim_vs_US_cluster_", clust.var[i], ".csv"))
  
  ############################################
  # plotting DEG per cluster across condition
  ############################################
  
  # Set idents to cluster ID
  Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters
  
  # Subset for cluster ID
  subset.seurat <- subset(seurat.combined, idents = paste0(clust.var[i]))
  
  # Set idents to condition var
  Idents(subset.seurat) <- subset.seurat@meta.data$condition
  
  # Get top 10 genes
  sig.genes <- DEG.var[DEG.var$p_val_adj < 0.05, ]
  
  # up-regulated genes
  genes.to.label.1 <- top_n(sig.genes, 10, avg_log2FC)$gene
  
  # Down-regulated genes
  genes.to.label.2 <- top_n(sig.genes, -10, avg_log2FC)$gene
  
  # Create vector of genes
  genes.to.label <- unique(c(genes.to.label.1, genes.to.label.2))
  
  # Average expression
  avg.subset.seurat <- log1p(AverageExpression(subset.seurat, verbose = FALSE)$RNA)
  avg.subset.seurat <- as.data.frame(avg.subset.seurat)
  

  # Generate plot
  p1 <- ggplot(avg.subset.seurat, 
               aes(US, Stim)) + 
    geom_point() + 
    ggtitle(paste0("Cluster ", clust.var[i]))
  
  p1 <-  LabelPoints(plot = p1,
                     points = genes.to.label,
                     repel = TRUE)
  
  print(plot(p1))
  dev.copy(pdf, paste0("output/figures/DEG_Clusters_btn_conditions/ScatterPlot_Stim_vs_US_cluster_", clust.var[i], ".pdf"))
  dev.off()
  
  # Remove large variable
  rm(subset.seurat)
  rm(DEG.var)
  
}










# Plot All DEGs just to visualise the number of DEGs per cluster detected
for(i in 1:length(clust.var)){
  
  print(paste0("Analysis for cluster ", clust.var[i]))
  
  # Set idents to the condition_cluster
  Idents(seurat.combined) <- seurat.combined@meta.data$condition_clust
  
  # Calculating DEGs for each cluster across condition
  DEG.var <- FindMarkers(seurat.combined,
                         ident.1 = paste0("Stim_", clust.var[i]), 
                         ident.2 = paste0("US_", clust.var[i]),
                         test.use = "wilcox",
                         logfc.threshold = 0.25,
                         min.pct = 0.1,
                         only.pos = FALSE,
                         verbose = FALSE)
  
  # create column for gene ID
  DEG.var$gene <- rownames(DEG.var)
  
  ############################################
  # plotting DEG per cluster across condition
  ############################################
  
  # Set idents to cluster ID
  Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters
  
  # Subset for cluster ID
  subset.seurat <- subset(seurat.combined, idents = paste0(clust.var[i]))
  
  # Set idents to condition var
  Idents(subset.seurat) <- subset.seurat@meta.data$condition
  
  # Get all significant DEGs
  genes.to.label <- DEG.var[DEG.var$p_val_adj < 0.05, "gene"]
  
  
  # Average expression
  avg.subset.seurat <- log1p(AverageExpression(subset.seurat, verbose = FALSE)$RNA)
   avg.subset.seurat <- as.data.frame(avg.subset.seurat)

  
  # Generate plot
  p1 <- ggplot(avg.subset.seurat, 
               aes(US, Stim)) + 
    geom_point() + 
    ggtitle(paste0("Cluster ", clust.var[i]))
  
  p1 <-  LabelPoints(plot = p1,
                     points = genes.to.label,
                     repel = TRUE)
  
  print(plot(p1))
  dev.copy(pdf, paste0("output/figures/DEG_Clusters_btn_conditions/ScatterPlot_Stim_vs_US_cluster_", clust.var[i], "_all_Sig_genes.pdf"))
  dev.off()
  
  # Remove large variable
  rm(subset.seurat)
  rm(DEG.var)
  
}





```


### Visualising gene expression accross conditions
```{r Visualising_condition_differences}

# Generate output directories
if(!dir.exists("output/figures/DEG_Clusters_btn_conditions")){
  dir.create("output/figures/DEG_Clusters_btn_conditions", 
             recursive = T)
}

if(!dir.exists("output/figures/DEG_Clusters_btn_conditions/visualising_gene_expression")){
  dir.create("output/figures/DEG_Clusters_btn_conditions/visualising_gene_expression", 
             recursive = T)
}


#cluster.markers.df <- read.csv("output/tables/DEG_Clusters/Cluster_markers_unfiltered.csv")

DefaultAssay(seurat.combined) <- "integrated"


goi <- c("GZMB", "IFNG", "CD27", "CD28", "PDCD1", "CCL4", "PRF1", "TIGIT", "IL2RA", "CCL4L2")

for(i in seq_along(goi)){
  
  print(VlnPlot(seurat.combined, 
        pt.size = 0, 
        log = FALSE,
        feature = goi[i], 
        cols = Condition.cols,
        split.by = "condition"))
  
  dev.copy(pdf, paste0("output/figures/DEG_Clusters_btn_conditions/visualising_gene_expression/VlnPlot_", goi[i], ".pdf"))
  dev.off()
}


  print(VlnPlot(seurat.combined, 
        pt.size = 0, 
        log = FALSE,
        feature = "LAG3", 
        cols = Condition.cols,
        group.by = "condition"))




```


### singleCellHaystack DEG

```{r singleCellHaystack}

if(long.compute){
  
  source("R/Analysis/SingleCellHaystack_analysis.R", local = knitr::knit_global())

  
}

```



## Analysis of Stim Stimulated_1 and Stim Stimulated_Exhausted clusters {.tabset} // INCOMPLETE

### DEGs between stimulation specific Stim_1 and Stimulated_Exhausted clusters

```{r DEG_stim_clusters}

# Aim 
# Identify and visualise the difference between Stimulated clusters Stimulated_1 and Stimulated_exhausted

# Also, should look at the overlap of clonotypes betwen these two clusters



# Output directory
if(!dir.exists("output/figures/Stim_1_vs_Stim_exhausted")){
  dir.create("output/figures/Stim_1_vs_Stim_exhausted", 
             recursive = T)}


# Run DEG analysis
DefaultAssay(seurat.combined) <- "RNA"

Idents(seurat.combined) <- seurat.combined@meta.data$condition_clust
unique(Idents(seurat.combined))


DEG.markers <- FindMarkers(seurat.combined, 
                           ident.1 = "Stim_Stimulated_1", 
                           ident.2 = "Stim_Stimulated_exhausted", 
                           slot = "data", 
                           logfc.threshold = 0,
                           test.use = "wilcox",
                           only.pos = FALSE)

DEG.markers <- 
  DEG.markers %>%
  filter(p_val_adj < 0.05)

write.csv(DEG.markers, "output/tables/DEG_Stim_stim1_vs_Stim_Stim_exhausted.csv")



  
#################################
# Create visualisations
#################################

###########
# Volcano
###########

# Format data
volcano.data <- DEG.markers %>%
  dplyr::select(avg_log2FC, p_val_adj)

colnames(volcano.data) <- c("logFC", "FDR")

keyvals <- colour.points(volcano.data)


# get list of top 10 up and 10 downregulated genes
goi.up <- volcano.data %>%
  dplyr::filter(FDR < 0.05) %>%
  top_n(15, logFC)

goi.dn <- volcano.data %>%
  dplyr::filter(FDR < 0.05) %>%
  top_n(-15, logFC)

# combine vectors
goi <- unique(c(rownames(goi.up), rownames(goi.dn)))


custom.enhanced.volcano(volcano.data, 
                        selectLab = goi, 
                        colCustom = keyvals, 
                        xlim = c(-3.5, 3.5),
                        title = "Stim1 vs Stim Exhausted")


dev.copy(pdf, "output/figures/Stim_1_vs_Stim_exhausted/Volcano.pdf")
dev.off()

#############
# Heatmaps
#############

# All clusters 
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters

seurat.combined.small <- subset(seurat.combined, downsample = 300)

# All genes significantly different
goi <- DEG.markers %>%
  arrange(desc(avg_log2FC))


DoHeatmap(seurat.combined.small, 
          features = rownames(goi)) + 
  NoLegend()

dev.copy(pdf, "output/figures/Stim_1_vs_Stim_exhausted/Heatmap_all_clusters_allsig_genes.pdf")
dev.off()

# All genes > or < 1 fold change 
goi <- DEG.markers %>%
  dplyr::filter(avg_log2FC > 1 | avg_log2FC < -1) %>%
  arrange(desc(avg_log2FC))


DoHeatmap(seurat.combined.small, 
          features = rownames(goi)) + 
  NoLegend()

dev.copy(pdf, "output/figures/Stim_1_vs_Stim_exhausted/Heatmap_all_clusters.pdf")
dev.off()


######################################################
# Do DEG analysis across clonally related clusters 
######################################################

DefaultAssay(seurat.combined) <- "RNA"
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters

temp.seurat <- subset(seurat.combined, idents = c("Exhausted_1", "Exhausted_2",
                                                  "Cytotoxic", "Type_I_IFN",
                                                  "Stimulated_exhausted", "Stimulated_1"))

DefaultAssay(temp.seurat) <- "RNA"

# Set Idents 
Idents(temp.seurat) <- temp.seurat@meta.data$seurat_clusters

# Find markers
DEG.markers <- FindAllMarkers(temp.seurat,
                              assay = "RNA",
                              only.pos = FALSE,
                              min.diff.pct = 0.1,
                              return.thresh = 0.05)


DEG.markers <- 
  DEG.markers %>%
  filter(p_val_adj < 0.05)

length(DEG.markers$gene) # 2,640 genes 

write.csv(DEG.markers, "output/tables/DEG_Cytotoxic_TypeI_Stim1_StimExhausted_Exhausted_all_sig.csv")


##############################
# Visualise data
##############################

temp.seurat.small <- subset(temp.seurat, downsample = 300)

goi <- DEG.markers %>%
  group_by(cluster) %>%
  top_n(50, avg_log2FC) %>%
  pull(gene)

goi <- unique(goi)

DoHeatmap(temp.seurat.small, 
          goi) + 
  NoLegend()

dev.copy(pdf, "output/figures/Stim_1_vs_Stim_exhausted/Heatmap_clonally_related_clusters_DEG_top_50.pdf")
dev.off()





########################
# Annotated heatmap 
########################

goi <- DEG.markers %>%
  group_by(cluster) %>%
  top_n(100, avg_log2FC) %>%
  arrange(desc(avg_log2FC), .by_group = TRUE) %>%
  pull(gene)

goi <- unique(goi)


# Genes to be labeled if present in plotted heatmap
genes.id <- c("EOMES", "TCF7", "CCR7",
              "IL7R", "IFI6", "ISG15", "XCL1", "TNF", "IL2", "IFNG", "LAG3",
              "TIGIT", "CD226", "GNLY", "PDCD1", "PRF1", "GATA3", "CTLA4", "CTSW")
  
# Colour annotation for column labels 
names(clust.cols) <- clust.names

clust.cols.temp <- clust.cols[names(clust.cols) %in% c("Cytotoxic", "Type_I_IFN", "Stimulated_1",
                                                       "Stimulated_exhausted", "Exhausted_1", "Exhausted_2")]


cluster.order <- c("Cytotoxic", "Type_I_IFN", "Stimulated_1", "Stimulated_exhausted", "Exhausted_1", "Exhausted_2")

# Plot heatmap downsampled single cell values
annotated.heatmap(temp.seurat.small, 
                  cluster.id = "seurat_clusters",
                  assay.use = "integrated",
                  average.expression = FALSE,
                  goi = goi, 
                  genes.to.label = genes.id, 
                  col_order = cluster.order, 
                  col.colours = clust.cols.temp, 
                  range.val = c(-2, 0, 2))

dev.copy(pdf, "output/figures/DEG_cytotoxic_and_stim_clusters/Heatmap_highlighted_DEGs.pdf")
dev.off()


rm(temp.seurat, temp.seurat.small)


#####################################
# Plot Nebulosa density UMAPs 
#####################################

goi <- DEG.markers %>%
  group_by(cluster) %>%
  top_n(10, avg_log2FC) %>%
  pull(gene)

goi <- unique(goi)

DefaultAssay(seurat.combined) <- "RNA"

for(i in seq_along(goi)){
  
  print(Nebulosa::plot_density(seurat.combined, 
                               goi[i], 
                               pal = "inferno"))
  
  dev.copy(pdf, paste0("output/figures/DEG_cytotoxic_and_stim_clusters/Density_UMAP_", goi[i], ".pdf"))
  dev.off()
}




```



## Visualise DEGs {.tabset}

### Heatmaps and dotplots of Cluster DEGs

```{r Heatmaps_and_dotplots_of_Cluster_DEGs}

# Output directory
# Heatmaps
if(!dir.exists("output/figures/Cluster_DEG_Heatmaps")){
  dir.create("output/figures/Cluster_DEG_Heatmaps", 
             recursive = T)
}

# Dotplots
if(!dir.exists("output/figures/Cluster_DEG_dotplots")){
  dir.create("output/figures/Cluster_DEG_dotplots", 
             recursive = T)
}



# Set assay to RNA
DefaultAssay(seurat.combined) <- "integrated"

# Set Idents 
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters

# Downsample 
seurat.combined.small <- subset(seurat.combined, downsample = 300)


# Get average expression
average.seurat <- AverageExpression(seurat.combined,
                                    assay = "integrated",
                                    slot = "data",
                                    verbose = TRUE,
                                    return.seurat = TRUE)


################################################
# Visualise top markers for each Cluster
################################################
# as single cell heatmap, averaged heatmap, and dotplot


# get vector of clusters to iterate over
Cluster.vect <- unique(seurat.combined.small@meta.data$seurat_clusters)

#sum(rownames(average.seurat) %in% x$gene)
for(i in 1:length(Cluster.vect)){
  
  print(paste0("Evaluating cluster = ", Cluster.vect[i]))
  
  x <- eval(parse(text = paste0("Cluster_", Cluster.vect[i])))
  
  # Single cell heatmap
  print(DoHeatmap(seurat.combined.small, 
                  features = x$gene, 
                  size = 4, 
                  angle = 90, 
                  raster = FALSE) + 
          NoLegend()
  )
  
  dev.copy(pdf, paste0("output/figures/Cluster_DEG_Heatmaps/Heatmap_topDEGs_clust_", Cluster.vect[i], ".pdf"))
  dev.off()
  
  
  # Average expression heatmap
  print(DoHeatmap(average.seurat, 
                  features = x$gene, 
                  size = 4, 
                  angle = 90, 
                  draw.lines = FALSE,
                  raster = FALSE) + 
          NoLegend()
  )
  
  dev.copy(pdf, paste0("output/figures/Cluster_DEG_Heatmaps/Heatmap_topDEGs_clust_", Cluster.vect[i], "_average.pdf"))
  dev.off()
  
  
  print(DotPlot(seurat.combined, 
                assay = "RNA",
                features = x$gene, 
                dot.scale = 8) + 
          RotatedAxis() +  
          theme(text = element_text(size = 4)) + 
          NoLegend()
  )
  
  dev.copy(pdf, paste0("output/figures/Cluster_DEG_dotplots/DotPlot_TopDEGs_clust_", Cluster.vect[i], ".pdf"))
  dev.off()
  
  
}


```

### Large heatmaps of top DEGs by foldchange // Average and single cell

```{r Large_Heatmaps_of_top_DEGs}

# Output directory
if(!dir.exists("output/figures/Large_Summary_heatmaps")){
  dir.create("output/figures/Large_Summary_heatmaps", 
             recursive = T)
}


###########################
# Get top overall DEGs
###########################

# Set assay to integrated
DefaultAssay(seurat.combined) <- "integrated"

# Set Idents 
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters

# Downsample 
seurat.combined.small <- subset(seurat.combined, downsample = 300)

# Get average expression
average.seurat <- AverageExpression(seurat.combined,
                                    assay = "integrated",
                                    slot = "data",
                                    verbose = TRUE,
                                    return.seurat = TRUE)



# Get top cluster markers
sig.genes <- Cluster.markers[Cluster.markers$p_val_adj < 0.05, ]

clusts <- unique(sig.genes$cluster)
output.vect <- NA
top.n <- 40

for(i in 1:length(clusts)){
  
  temp.gene <- sig.genes[sig.genes$cluster == paste0(clusts[i]), ]
  temp.top <- dplyr::top_n(temp.gene, top.n, avg_log2FC)$gene
  
  output.vect <- c(output.vect, temp.top)
  output.vect <- unique(output.vect)
}


DoHeatmap(average.seurat, 
          features = output.vect, 
          size = 4, 
          angle = 90, 
          draw.lines = FALSE,
          raster = FALSE) + 
  NoLegend()

dev.copy(pdf, "output/figures/Large_Summary_heatmaps/Heatmap_top_up_genes_bycluster_Averaged.pdf")
dev.off()

DoHeatmap(seurat.combined.small, 
          features = output.vect, 
          size = 4, 
          angle = 90, 
          draw.lines = TRUE,
          raster = FALSE) + 
  NoLegend()

dev.copy(pdf, "output/figures/Large_Summary_heatmaps/Heatmap_top_up_genes_bycluster.pdf")
dev.off()


```
 
 
### Large heatmap annotated

```{r large_annotated_heatmap}

# Output directory
if(!dir.exists("output/figures/Large_Summary_heatmaps")){
  dir.create("output/figures/Large_Summary_heatmaps", 
             recursive = T)
}


##################################################################
# Use complexheatmap to generate a manuscript ready figure
##################################################################

# Source custom function // loaded in custom functions chunk 
#source("R/annotate_seurat_heatmap.R")

######################################################################
# Generate a vector of the top DEGs to be plotted in the heatmap
######################################################################

# Ensure markers table is loaded
Cluster.markers <- read.csv("output/tables/DEG_Clusters/Cluster_markers_sig_only.csv")


# Get top cluster markers
sig.genes <- Cluster.markers[Cluster.markers$p_val_adj < 0.05, ]

clusts <- unique(sig.genes$cluster)
output.vect <- NA
top.n <- 40

for(i in 1:length(clusts)){
  
  temp.gene <- sig.genes[sig.genes$cluster == paste0(clusts[i]), ]
  temp.top <- dplyr::top_n(temp.gene, top.n, avg_log2FC)$gene
  
  output.vect <- c(output.vect, temp.top)
  output.vect <- unique(output.vect)
}

top.genes <- output.vect

# Genes to be labeled if present in plotted heatmap
genes.id <- c("KLRB1", "TRAV1-2", "IL4I1", "IL18RAP", "TRBV21-1", "TRBV2", "TRGV9", "TRDV2",  "DPP4", 
                 "IFNG", "TNF", "TCF7", "ENTPD1", "PDCD1", "TIGIT", "LAYN", "EOMES", "GZMA", "GZMB", "GZMK",
                 "CXCR3", "TOX", "ZNF683", "CD226", "CXCL13", "CD160", "TBX21", "LAG3", "MKI67", "GNLY", "NKG7", "SELL", "IL7R")

# Colour annotation for column labels 
names(clust.cols) <- clust.names

# Plot heatmap averaged values
annotated.heatmap(seurat.combined, 
                  cluster.id = "seurat_clusters",
                  assay.use = "integrated",
                  average.expression = TRUE,
                  goi = top.genes, 
                  genes.to.label = genes.id, 
                  col_order = clust.names, 
                  col.colours = clust.cols, 
                  range.val = c(-2, 0, 2))

dev.copy(pdf, "output/figures/Large_Summary_heatmaps/Heatmap_top_up_genes_bycluster_highlighted_geneIDs_averaged.pdf")
dev.off()


# Plot heatmap downsampled single cell values
annotated.heatmap(seurat.combined, 
                  cluster.id = "seurat_clusters",
                  assay.use = "integrated",
                  average.expression = FALSE,
                  goi = top.genes, 
                  genes.to.label = genes.id, 
                  col_order = clust.names, 
                  col.colours = clust.cols, 
                  range.val = c(-2, 0, 2))

dev.copy(pdf, "output/figures/Large_Summary_heatmaps/Heatmap_top_up_genes_bycluster_highlighted_geneIDs.pdf")
dev.off()
```

### DEGs and selected genes Vlnplots / Clusters

```{r Vln_feature_plots}

# Create output directories
if(!dir.exists("output/figures/Top_DEG_plots/FeaturePlots")){
  dir.create("output/figures/Top_DEG_plots/FeaturePlots", 
             recursive = T)
}

if(!dir.exists("output/figures/Top_DEG_plots/VlnPlots")){
  dir.create("output/figures/Top_DEG_plots/VlnPlots", 
             recursive = T)
}

# Set assay
DefaultAssay(seurat.combined) <- "RNA"


# Custom list of genes to plot


goi <- c("IFNG", "TNF", "ENTPD1",
         "PDCD1", "TIGIT", "LAYN", 
         "EOMES", "GZMA", "GZMB", 
         "GZMK", "CXCR3", "TOX",
         "CASZ1", "CD226", "CXCL13",
         "CD160", "TBX21",
         "TRBC1", "CD8A", "CD8B", 
         "CD4",
         "FOXP3", "IL2RA", "CTLA4", 
         "CD14", "LYZ", 
         "PRF1", "NKG7", "GNLY", 
         "FCGR3A",
         "KLRB1", "KLRC1", 
         "MKI67", 
         "KIT", 
         "HAVCR2", "LAG3", "TCF7",
         "ICOS", 
         "CCR7", "IL7R", "CD7", "CD69",  
         "ITGAX", "ITGAM", "IL3RA",
         "CD86", "CD83", "IL4I1")



################################################################
# Plot Feature plot and VlnPlot for top DEGs for Clusters
################################################################


# Generate vector of genes to plot
x <- read.csv("output/tables/DEG_Clusters/Cluster_markers_sig_only.csv")

Top.cluster.genes <- x %>% 
  dplyr::group_by(cluster) %>%
  dplyr::top_n(10, avg_log2FC) %>%
  dplyr::ungroup() %>% 
  dplyr::select(gene)

Top.cluster.genes <- unique(Top.cluster.genes$gene)


# set idents 
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters


# Add custom genes 
Top.cluster.genes <- unique(c(Top.cluster.genes, goi))

for(i in 1:length(Top.cluster.genes)){
  
  print(paste0("Plotting ", Top.cluster.genes[i]))
  
  # Plot VlnPlot
  print(VlnPlot(seurat.combined, 
                features = Top.cluster.genes[i], 
                log = TRUE,
                cols = clust.cols,
                pt.size = 0) + 
          NoLegend())
  
  dev.copy(pdf, paste0("output/figures/Top_DEG_plots/VlnPlots/VlnPlot_", Top.cluster.genes[i], ".pdf"))
  dev.off()
  
  # Plot feature plot on UMAP projection
  print(FeaturePlot(seurat.combined,
                    features = Top.cluster.genes[i],
                    order = TRUE,
                    pt.size = 1,
                    reduction = "umap") + 
          NoLegend())
  
  dev.copy(pdf, paste0("output/figures/Top_DEG_plots/FeaturePlots/FeaturePlot_", Top.cluster.genes[i], ".pdf"))
  dev.off()
}


```

### DEGs and selected genes Vlnplots / Clusters / Imputed

```{r Vln_feature_plots_imputed}

# Create output directories
if(!dir.exists("output/figures/Top_DEG_plots_Imputed/FeaturePlots")){
  dir.create("output/figures/Top_DEG_plots_Imputed/FeaturePlots", 
             recursive = T)
}

# Create output directories
if(!dir.exists("output/figures/Top_DEG_plots_Imputed/VlnPlots")){
  dir.create("output/figures/Top_DEG_plots_Imputed/VlnPlots", 
             recursive = T)
}

# Set assay
DefaultAssay(seurat.combined) <- "alra"



# Custom list of genes to plot


goi <- c("IFNG", "TNF", "ENTPD1",
         "PDCD1", "TIGIT", "LAYN", 
         "EOMES", "GZMA", "GZMB", 
         "GZMK", "CXCR3", "TOX",
         "CASZ1", "CD226", "CXCL13",
         "CD160", "TBX21",
         "TRBC1", "CD8A", "CD8B", 
         "CD4",
         "FOXP3", "IL2RA", "CTLA4", 
         "CD14", "LYZ", 
         "PRF1", "NKG7", "GNLY", 
         "FCGR3A",
         "KLRB1", "KLRC1", 
         "MKI67", 
         "KIT", 
         "HAVCR2", "LAG3", "TCF7",
         "ICOS", 
         "CCR7", "IL7R", "CD7", "CD69",  
         "ITGAX", "ITGAM", "IL3RA",
         "CD86", "CD83", "IL4I1")




################################################################
# Plot Feature plot and VlnPlot for top DEGs for Clusters
################################################################

# Generate vector of genes to plot
x <- read.csv("output/tables/DEG_Clusters/Cluster_markers_sig_only.csv")

Top.cluster.genes <- x %>% 
  dplyr::group_by(cluster) %>%
  dplyr::top_n(10, avg_log2FC) %>%
  dplyr::ungroup() %>% 
  dplyr::select(gene)

Top.cluster.genes <- unique(Top.cluster.genes$gene)


# set idents 
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters


# Add custom genes 
Top.cluster.genes <- unique(c(Top.cluster.genes, goi))

for(i in 1:length(Top.cluster.genes)){
  
  print(paste0("Plotting ", Top.cluster.genes[i]))
  
  # Plot VlnPlot
  print(VlnPlot(seurat.combined, 
                features = Top.cluster.genes[i], 
                log = TRUE,
                cols = clust.cols,
                pt.size = 0) + 
          NoLegend())
  
  dev.copy(pdf, paste0("output/figures/Top_DEG_plots_Imputed/VlnPlots/VlnPlot_", Top.cluster.genes[i], "_imputed.pdf"))
  dev.off()
  
  # Plot feature plot on UMAP projection
  print(FeaturePlot(seurat.combined,
                    features = Top.cluster.genes[i],
                    order = TRUE,
                    pt.size = 1,
                    reduction = "umap") + 
          NoLegend())
  
  dev.copy(pdf, paste0("output/figures/Top_DEG_plots_Imputed/FeaturePlots/FeaturePlot_", Top.cluster.genes[i], "_imputed.pdf"))
  dev.off()
}




```

### Volcano DEGs

```{r Volcano_plots}

# Generate output directories
if(!dir.exists("output/figures/VolcanoPlots/Clusters")){
  dir.create("output/figures/VolcanoPlots/Clusters", 
             recursive = TRUE)
}

##############################
# Install and load package
##############################

# edited enhancedvolcano() function to have the default variables for aesthetics I want // loaded in custom functions chunk
#source("R/Enhanced_volcano_custom_defaults.R")



################
# Clusters
################

cluster.markers.df <- read.csv("output/tables/DEG_Clusters/Cluster_markers_unfiltered.csv")


cluster.var <- unique(cluster.markers.df$cluster)


for(i in 1:length(cluster.var)){
  
  # format data
  volcano.data <- clean.data(cluster.markers.df,
                             group.id = cluster.var[i])
  
  keyvals <- colour.points(volcano.data)
  
  
  # get list of top 10 up and 10 downregulated genes
  goi.up <- volcano.data %>%
    dplyr::filter(FDR < 0.05) %>%
    top_n(10, logFC)
  
  goi.dn <- volcano.data %>%
    dplyr::filter(FDR < 0.05) %>%
    top_n(-10, logFC)
  
  # combine vectors
  goi <- unique(c(rownames(goi.up), rownames(goi.dn)))
  
  
  
  print(custom.enhanced.volcano(volcano.data, 
                                selectLab = goi, 
                                colCustom = keyvals, 
                                title = paste0(cluster.var[i])))
  
  
  dev.copy(pdf, paste0("output/figures/VolcanoPlots/Clusters/Volcano_", cluster.var[i], ".pdf"))
  dev.off()
  
  
}


```



