---
title: "04_Manuscript_Figure_4"
author: "Dillon Corvino"
date: "07/03/2022"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

### Dataset information

```{r Dataset_Info}

# HNSCC scRNAseq/scTCRseq dataset
# Cells are HNSCC TILs sorted on Lymphocytes/Live/CD3+/CD4-/CD8+
# Human samples
# Each sample is a pool of 4 patients
# samples 1 and 2 are unstimulated 
# samples 3 and 4 are stimulated 
# Patients in sample 1 match patients in sample 3 and sample 2 pairs with 4
# Data acquired was transcript expression, ADT (antibody expression), and TCRA & B sequences
# This analysis uses just Transcript and TCR data


# General information
# CellRanger was used for sequence alignment/QC/counts/TCR calls - performed by Ross (QIMR, Brisbane, Aus)
# DC was provided the output from CellRanger (counts matrix) and used this as input for analysis within this script

```

### Environment

```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Create output directories & load custom functions & colour scheme
source("scripts/Setup.R", local = knitr::knit_global())

# pipeline variables
long.compute <- FALSE



```


### Reading data

```{r reading_data}

# Read in CD8 only dataset
CD8.seurat <- LoadH5Seurat("saves/CD8_seurat.h5Seurat")
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"

```

## Type I IFN score {.tabset}

### Info
```{r info}
# analysis derived from Patrick Gunther from aimed analytics 
# original code from patrick is found in bald_analysis.rmd

```

### Define signature
```{r define_signature}
# Create output directory

output.dir <- "results/Manuscript/Figure4/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = TRUE)
}

if(long.compute){
DefaultAssay(CD8.seurat) <- "RNA"
inf_cluster_signature <- FindMarkers(CD8.seurat,
                                     ident.1 = "Type_I_IFN",
                                     logfc.threshold = 0,
                                     min.pct = 0)

inf_cluster_signature <-
  inf_cluster_signature%>%
  as_tibble(rownames = "gene_symbol") %>%
  mutate(DE=case_when(
    avg_log2FC>0 & p_val_adj <=.1 ~ "up",
    avg_log2FC<0 & p_val_adj <=.1 ~ "down",
    TRUE ~ "notDE"))

write_csv(x = inf_cluster_signature,
          file = paste0(output.dir, "ifncluster_DE_result.csv"))



inf_cluster_signature %>%
  mutate(label=if_else(abs(avg_log2FC)>1 & p_val_adj<.1,
                       gene_symbol,"")) %>%
  ggplot(aes(avg_log2FC,y=-log10(p_val))) +
  geom_point(aes(color=avg_log2FC)) +
  geom_text_repel(aes(label=label),max.time = 2,max.overlaps = 20,segment.colour = "grey",size=3) +
  theme_minimal() +
  scale_color_distiller(palette = "RdBu",limits=c(-1,1),oob=scales::squish)


inf_cluster_signature %>%
  group_by(DE) %>% tally()

}




# 
# 
# # All markers
# 
# all_markers <- FindAllMarkers(seurat_combined,logfc.threshold = .4)
# 
# all_markers %>%
#   group_by(cluster) %>%
#   top_n(n = 5, wt = avg_log2FC) -> top10
# 
# seurat_combined <- ScaleData(seurat_combined,features = rownames(seurat_combined))
# 
# DoHeatmap(seurat_combined, features = top10$gene,
#           group.colors = pals::cols25()) +
#   NoLegend() +
#   scale_fill_viridis_c(option = "A")


#write_csv(x = all_markers,
#          file = paste0("/work_dir/results/exports/DE_result_allcluster_",Sys.Date(),".csv"))

#save.image(file = paste0("/work_dir/results/saves/",Sys.Date(),"_after_signature.RData"))

```



## Validation bald dataset

```{r}
ifn_pos_signature <- inf_cluster_signature %>%
  filter(avg_log2FC>=log2(2) , p_val_adj<=.1) %>%
  pull(gene_symbol)

ifn_signatures <- list(ifn_pos_signature=ifn_pos_signature)

```


### UCell
```{r}
DefaultAssay(seurat_combined) <- "RNA"
seurat_combined <- AddModuleScore_UCell(seurat_combined,
                                     features = ifn_signatures,
                                     ncores = 2)

# Viz data
featnames <- paste0(names(ifn_signatures), "_UCell")
FeaturePlot(seurat_combined, features = featnames,
            pt.size = 0.1, max.cutoff = "q99",
            ncol = 1,cols = viridis::magma(n = 50))

VlnPlot(seurat_combined, features = featnames,
        pt.size = 0, group.by = "seurat_clusters") +
  scale_fill_manual(values = pals::cols25()) +
  geom_boxplot(width=.1,fill="white")

plot_density(seurat_combined, featnames)

```



