---
title: "04_Manuscript_Figure_4"
author: "Dillon Corvino"
date: "07/03/2022"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

### Dataset information

```{r Dataset_Info}

# HNSCC scRNAseq/scTCRseq dataset
# Cells are HNSCC TILs sorted on Lymphocytes/Live/CD3+/CD4-/CD8+
# Human samples
# Each sample is a pool of 4 patients
# samples 1 and 2 are unstimulated 
# samples 3 and 4 are stimulated 
# Patients in sample 1 match patients in sample 3 and sample 2 pairs with 4
# Data acquired was transcript expression, ADT (antibody expression), and TCRA & B sequences
# This analysis uses just Transcript and TCR data


# General information
# CellRanger was used for sequence alignment/QC/counts/TCR calls - performed by Ross (QIMR, Brisbane, Aus)
# DC was provided the output from CellRanger (counts matrix) and used this as input for analysis within this script

```

### Environment

```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Create output directories & load custom functions & colour scheme
source("scripts/Setup.R", local = knitr::knit_global())

# pipeline variables
long.compute <- FALSE

# load required packages
require(tidyverse)
require(Seurat)
require(useful)
require(ggrepel)
require(clusterProfiler)
require(enrichplot)
require(org.Hs.eg.db)
require(UCell)
require(Nebulosa)
require(ggsci)
library(harmony)
require(irGSEA)
#source("scripts/aimed_analysis/functions.R")



# Create output directory
output.dir <- "results/Manuscript/Figure4/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = TRUE)
}
```




## Type I IFN score HNSCC dataset {.tabset}



### Reading data
```{r reading_data}

# Read in CD8 only dataset
CD8.seurat <- LoadH5Seurat("saves/CD8_seurat.h5Seurat")
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"

```


### Define signature
```{r define_signature}

# Compute DEGs and define the Type_I_IFN cluster signature
if(long.compute){
  
  DefaultAssay(CD8.seurat) <- "RNA"
  
  # Calculate DEGs
  IFN.clust.DEGs <- FindMarkers(CD8.seurat,
                                ident.1 = "Type_I_IFN",
                                logfc.threshold = 0,
                                min.pct = 0)
  # Annotate DEGs
  IFN.clust.DEGs <-
    IFN.clust.DEGs%>%
    as_tibble(rownames = "gene_symbol") %>%
    mutate(DE = case_when(
      avg_log2FC > 0 & p_val_adj <= 0.1 ~ "up",
      avg_log2FC < 0 & p_val_adj <= 0.1 ~ "down",
      TRUE ~ "notDE"))
  
  # Export results
  write_csv(x = IFN.clust.DEGs,
            file = paste0(output.dir, "IFN_clust_DEGs.csv"))
  
  
  # Define signature as top DEGs only
  IFN.sig <- IFN.clust.DEGs %>%
    filter(p_val_adj <= 0.05) %>%
    top_n(20, avg_log2FC)
  
  
  # Export results
  write_csv(x = IFN.sig,
            file = paste0(output.dir, "IFN_signature.csv"))
  
  
}


# create signatures
#IFN.clust.DEGs <- read.csv(paste0(output.dir, "IFN_clust_DEGs.csv"))

#Sig.10 <- IFN.clust.DEGs %>%
#  filter(p_val_adj <= 0.05) %>%
#  top_n(10, avg_log2FC) %>%
#  pull(gene_symbol)


# Hard code the signature to prevent any alternations with statistical drift 
Sig.10 <- c("IFI6", "ISG15", "IFIT3", "MX1", "ISG20", "IFITM1", "LY6E", "IFIT1", "MX2", "OAS1")

Sig.10 <- paste0(Sig.10, "+")



Signature.list <- list(Top_10 = Sig.10)

signature.names <- paste0(names(Signature.list), "_UCell")




# extract hallmark pathways
# x <- msigdbr::msigdbr(species = "Homo sapiens",
#                       category = "H")
# 
# Hallmark.IFNa.response <- x[x$gs_name == "HALLMARK_INTERFERON_ALPHA_RESPONSE", ]$gene_symbol
# Hallmark.IFNa.response <- unique(Hallmark.IFNa.response)
# 
# 
# Hallmark.IFNg.response <- x[x$gs_name == "HALLMARK_INTERFERON_GAMMA_RESPONSE", ]$gene_symbol
# Hallmark.IFNg.response <- unique(Hallmark.IFNg.response)
# 
# 
# # create list of signatures
# Signature.list <- list(Positive_sig = IFN.sig, 
#                        IFNa_Hallmark = Hallmark.IFNa.response,
#                        IFNg_Hallmark = Hallmark.IFNg.response)
# 
# signature.names <- paste0(names(Signature.list), "_UCell")

```


### Type 1 or Type 2 IFN
```{r TypeI_or_II_IFN}


#https://www.nature.com/articles/nri1604

# Prepare dataset for heatmap
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"

average.seurat <- AverageExpression(CD8.seurat, return.seurat = T)

DoHeatmap(average.seurat, 
          draw.lines = F,
          raster = FALSE,
          features = c("STAT1", "STAT2", "IRF9", "OAS1"), 
          group.colors = CD8.cols) + scico::scale_fill_scico(palette = "batlow", 
                                                    direction = 1, 
                                                    na.value = "white")

dev.copy(pdf, paste0(output.dir, "Heatmap_Type1_vs_Type2_IFN.pdf"))
dev.off()


#c("STAT1", "STAT2", "STAT3", "TYK2", "IRF9", "JAK1", "JAK2", "IFNAR1", "IFNAR2", "IFNGR1", "IFNGR2", "SOCS1", "SOCS3", "OAS1")
#c("OAS1", "OAS2", "OAS3", "OASL")

#goi <- grep("STAT.*", rownames(average.seurat), value = T)


```

### Heatmap of IFN genes across clusters 
```{r IFN gene expression}

genes <- unique(rownames(CD8.seurat@assays[["RNA"]]))

IFN_genes <- genes[grepl("^IFN", genes)]
IFN_genes_no_receptors <- c("IFNG-AS1", "IFNG", "IFNL1")

# Prepare dataset for heatmap
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"

average.seurat <- AverageExpression(CD8.seurat, return.seurat = T)

DoHeatmap(average.seurat, 
          draw.lines = F,
          raster = FALSE,
          features = c(IFN_genes_no_receptors), 
          group.colors = CD8.cols) + scico::scale_fill_scico(palette = "batlow", 
                                                    direction = 1, 
                                                    na.value = "white")

dev.copy(pdf, paste0(output.dir, "Heatmap_IFN_genes.pdf"))
dev.off()

```

 
### Visualise Type I IFN cluster 
```{r heatmap_gene_expression}


 #extract hallmark pathways
 x <- msigdbr::msigdbr(species = "Homo sapiens",
                       category = "H")
 
 Hallmark.IFNa.response <- x[x$gs_name == "HALLMARK_INTERFERON_ALPHA_RESPONSE", ]$gene_symbol
 Hallmark.IFNa.response <- unique(Hallmark.IFNa.response)
 
 
 Hallmark.IFNg.response <- x[x$gs_name == "HALLMARK_INTERFERON_GAMMA_RESPONSE", ]$gene_symbol
 Hallmark.IFNg.response <- unique(Hallmark.IFNg.response)

 Hallmark.IFN.both <- unique(c(Hallmark.IFNa.response, Hallmark.IFNg.response))
 
 # Prepare dataset for heatmap
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"

average.seurat <- AverageExpression(CD8.seurat, return.seurat = T)

DoHeatmap(average.seurat, 
          draw.lines = F,
          raster = FALSE,
          features = Hallmark.IFNa.response, 
          group.colors = CD8.cols) + scico::scale_fill_scico(palette = "batlow", 
                                                    direction = 1, 
                                                    na.value = "white")

dev.copy(pdf, paste0(output.dir, "HNSCC_Hallmark_IFNa_Averaged.pdf"))
dev.off()



# Heatmap of IFN cluster DEGs
IFN.clust.DEGs <- read.csv(paste0(output.dir, "IFN_clust_DEGs.csv"))

DEGs.UP <- IFN.clust.DEGs %>%
  filter(p_val_adj <= 0.05) %>%
  filter(DE == "up") %>%
  pull(gene_symbol)



grep("IFN.*", DEGs.UP, value = T)


annotated.heatmap(CD8.seurat, 
                  cluster.id = "Clusters_l1",
                              assay.use = "RNA",
                              average.expression = TRUE,
                              goi = DEGs.UP, 
                              genes.to.label = c("ISG15", "IFI6", "ISG20", "IRF7", "IRF2", "IRF9", "IRF2", "IFI35", "IFI44"), 
                              col_order = names(CD8.cols), 
                              col.colours = CD8.cols, 
                              range.val = c(-2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2))

dev.copy(pdf, paste0(output.dir, "HNSCC_IFN_cluster_DEGs_highlighted_average.pdf"))
dev.off()


```


### UCell HNSCC dataset
```{r Ucell_HNSCC_dataset}

# Calculate Ucell scores for signatures
DefaultAssay(CD8.seurat) <- "RNA"
CD8.seurat <- AddModuleScore_UCell(CD8.seurat,
                                     features = Signature.list,
                                     ncores = 4)



# Visualise results
VlnPlot(CD8.seurat,
        pt.size = 0, 
        cols = CD8.cols,
        same.y.lims = T,
        features = signature.names[1],
        group.by = "Clusters_l1") + 
  NoLegend()

dev.copy(pdf, paste0(output.dir, "HNSCC_VlnPlot_UCell_scores_top10.pdf"))
dev.off()


 
Plot_Density_Custom(CD8.seurat, 
                    features =  signature.names[1],
                    custom_palette = batlow.pal,
                    joint = FALSE, 
                    pt.size = 1, 
                    reduction = "umap")

dev.copy(pdf, paste0(output.dir, "HNSCC_Nebulosa_top10.pdf"))
dev.off()



```

### UCell HNSCC dataset noIFN cluster
```{r Ucell_HNSCC_dataset}

# Calculate Ucell scores for signatures
DefaultAssay(CD8.seurat) <- "RNA"
keep.var <- levels(Idents(CD8.seurat))
keep.var <- keep.var[-5]

temp.seurat <- subset(CD8.seurat, idents = keep.var)
levels(Idents(temp.seurat))

temp.seurat <- AddModuleScore_UCell(temp.seurat,
                                     features = Signature.list,
                                     ncores = 4)


# Visualise results
VlnPlot(temp.seurat,
        pt.size = 0, 
        cols = CD8.cols,
        same.y.lims = T,
        features = signature.names[1],
        group.by = "Clusters_l1") + 
  NoLegend()

dev.copy(pdf, paste0(output.dir, "HNSCC_VlnPlot_UCell_scores_top10_no_IFN.pdf"))
dev.off()

 

Plot_Density_Custom(temp.seurat, 
                    features =  signature.names[1],
                    custom_palette = batlow.pal,
                    joint = FALSE, 
                    pt.size = 1, 
                    reduction = "umap")

dev.copy(pdf, paste0(output.dir, "HNSCC_Nebulosa_top10_no_IFN.pdf"))
dev.off()

rm(temp.seurat)


```

## Type I IFN score Cillo dataset {.tabset}

### read in cillo data

```{r read_dataset_cillo}

Cillo.CD8.Seurat <- LoadH5Seurat("data/external_datasets/Cillo_TIL_CD8_only.h5Seurat")


head(Cillo.CD8.Seurat@meta.data)
Idents(Cillo.CD8.Seurat) <- Cillo.CD8.Seurat@meta.data$cd8_cells_clusterID

DimPlot(Cillo.CD8.Seurat, 
        reduction = "tsne", 
        group.by = "Pt_ID")



# Recreate Cillo figure 2C
averaged.seurat <- AverageExpression(Cillo.CD8.Seurat, return.seurat = TRUE)

goi <- c("KLRG1", "JUNB", "IL7R", "CD69", "FOSB", "CCR7", "CD27", "EOMES", "KLRD1", "GZMH", "KLRC2", "GZMB", 
         "HIST1H4C", "TUBA1B", "STMN1", "MKI67", "TIGIT", "HAVCR2", "CTLA4", "ENTPD1", "TNFRSF4", "PDCD1", "TNFRSF18", "IFI6", "ISG15", "IFI27", "IFIT3", "IFIT1")

Idents(averaged.seurat) <- factor(averaged.seurat@active.ident, levels = c(1:8))

DoHeatmap(averaged.seurat, 
          features = goi, 
          draw.lines = F)

# Plot where cluster 2 which is cillo identified IFN-driven cluster
library(scCustomize)

scCustomize::Cluster_Highlight_Plot(Cillo.CD8.Seurat, 
                                    reduction = "umap",
                                    pt.size = 1,
                                    cluster_name = "2")

dev.copy(pdf, paste0(output.dir, "Cillo_UMAP_cluster_2_highlighted.pdf"))
dev.off()

```
 
### UCell Cillo
```{r cillo_ucell}

# Calculate Ucell scores for signatures
DefaultAssay(Cillo.CD8.Seurat) <- "RNA"
Cillo.CD8.Seurat <- AddModuleScore_UCell(Cillo.CD8.Seurat,
                                     features = Signature.list,
                                     ncores = 4)

# Visualise results
VlnPlot(Cillo.CD8.Seurat,
        pt.size = 0, 
        same.y.lims = T,
        features = signature.names[1],
        group.by = "cd8_cells_clusterID") + 
  NoLegend()

dev.copy(pdf, paste0(output.dir, "Cillo_VlnPlot_UCell_scores_top10.pdf"))
dev.off()


Plot_Density_Custom(Cillo.CD8.Seurat, 
                    features =  signature.names[1],
                    custom_palette = batlow.pal,
                    joint = FALSE, 
                    pt.size = 1, 
                    reduction = "umap")

dev.copy(pdf, paste0(output.dir, "Cillo_Nebulosa_top10.pdf"))
dev.off()


# UMAPPlot(Cillo.CD8.Seurat)
# 
# 
# Idents(Cillo.CD8.Seurat) <- Cillo.CD8.Seurat@meta.data$cd8_cells_clusterID 
# Cluster_Highlight_Plot(seurat_object = Cillo.CD8.Seurat,
#                        cluster_name = "2",
#                        reduction = "umap",
#                        highlight_color = "navy",
#     background_color = "lightgray")
# 


```

## Type I IFN score pancancer dataset {.tabset}

### Prepare dataset / read and filter for CD8 / or load dataset

```{r read_dataset_pancancer}

if(long.compute){
  
  # Due to memory limits, the dataset is written in as pancancer.cd8 rather than subsetting for cd8 and creating a new variable. 
  #i.e use only one variable to limit the memory requirments 
pancancer.cd8 <- LoadH5Seurat("data/external_datasets/Utility_dataset.h5Seurat")

Idents(pancancer.cd8) <- pancancer.cd8@meta.data$functional.cluster
pancancer.cd8 <- subset(pancancer.cd8, idents = c("CD8_EffectorMemory", "CD8_NaiveLike", "CD8_EarlyActiv", "CD8_Tex", "CD8_Tpex"))

# still many cells of mixed annotation by consensus.major ID
Idents(pancancer.cd8) <- pancancer.cd8@meta.data$consensus.major
pancancer.cd8 <- subset(pancancer.cd8, idents = "T_cell")

# further filtering for CD8 T cell annotation
Idents(pancancer.cd8) <- pancancer.cd8@meta.data$consensus.Tcell
pancancer.cd8 <- subset(pancancer.cd8, idents = "CD8")

# further filtering for annotation
Idents(pancancer.cd8) <- pancancer.cd8@meta.data$HPCA.pruned.labels
keep.var <- grep(".*CD8.*", unique(Idents(pancancer.cd8)), value =T)

pancancer.cd8 <- subset(pancancer.cd8, idents = keep.var)

# Type annotation
table(pancancer.cd8@meta.data$Type)/1000
# Juxta = 262 cells only
# Blood = 12.993K
# LN = 23.25K
# Met = 5.711K
# Normal = 41.119K
# Tumor = 173.74K

# Therefore remove Juxta cells 
# also remove Met and LN as these are a bit complicated to include in analysise.
# i.e the biological interpretation of Met tissue is highly dependent on each manuscript and what they mean by met and if LN is healthy or diseased
# also remove blood to keep it simple, normal and tumor tissue


Idents(pancancer.cd8) <- pancancer.cd8@meta.data$Type
pancancer.cd8 <- subset(pancancer.cd8, idents = c("Tumor", "Normal"))


table(pancancer.cd8@meta.data$Type, pancancer.cd8@meta.data$Tissue)


library(harmony)
# Normalise and harmonize dataset and calculate umap projection
pancancer.cd8 <- 
  pancancer.cd8 %>%
  NormalizeData() %>%
  FindVariableFeatures() %>%
  ScaleData() %>%
  RunPCA() %>%
  harmony::RunHarmony(reduction = "pca",
                      group.by.vars = c("Cohort"), 
                      plot_convergence = TRUE, 
                      verbose = TRUE) %>%
  RunUMAP(reduction = "harmony", 
          dims = 1:20) %>%
  FindNeighbors(reduction = "harmony",
                dims = 1:20) %>%
  FindClusters(resolution = 0.5)


# save filtered dataset
SaveH5Seurat(pancancer.cd8, 
             "data/external_datasets/pancancer_cd8_normed.h5Seurat",
             overwrite = TRUE)


}else{
  
  pancancer.cd8 <- LoadH5Seurat("data/external_datasets/pancancer_cd8_normed.h5Seurat")

}

# quick visualisation of the normed and subseted CD8 dataset

UMAPPlot(pancancer.cd8, 
         group.by = "HPCA.pruned.labels")

UMAPPlot(pancancer.cd8, 
         group.by = "seurat_clusters")





```
 
### apply ucell score to pancancer

```{r ucell_pancancer}


if(!exists("pancancer.cd8")){
  
    pancancer.cd8 <- LoadH5Seurat("data/external_datasets/pancancer_cd8_normed.h5Seurat")

}

if(long.compute){

 cores <- parallel::detectCores(all.tests = FALSE, logical = TRUE)

# Calculate Ucell scores for signatures
DefaultAssay(pancancer.cd8) <- "RNA"
pancancer.cd8 <- UCell::AddModuleScore_UCell(pancancer.cd8,
                                             features = Signature.list,
                                             ncores = cores)

# Export object with UCell score
SaveH5Seurat(pancancer.cd8, 
             "data/external_datasets/pancancer_cd8_normed.h5Seurat", 
             overwrite = TRUE)

}


colnames(pancancer.cd8@meta.data)

# Visualise the results


# Density plots 
Plot_Density_Custom(pancancer.cd8, 
                    features =  "Top_10_UCell",
                    custom_palette = batlow.pal,
                    joint = FALSE, 
                    pt.size = 1, 
                    reduction = "umap")

dev.copy(pdf, paste0(output.dir, "Pancancer_Nebulosa_UCell_scores_top10.pdf"))
dev.off()

# Cluster distribution

UMAPPlot(pancancer.cd8, 
         group.by = "seurat_clusters")

dev.copy(pdf, paste0(output.dir, "Pancancer_UMAP_clusters.pdf"))
dev.off()


temp.cols <- c(rep("Gray", length(unique(pancancer.cd8@meta.data$seurat_clusters))))
temp.cols[8] <- "Red"

UMAPPlot(pancancer.cd8, 
         cols = temp.cols,
         group.by = "seurat_clusters")

dev.copy(pdf, paste0(output.dir, "Pancancer_UMAP_clusters_highlighted.pdf"))
dev.off()




# VlnPlots
VlnPlot(pancancer.cd8,
        pt.size = 0, 
        same.y.lims = T,
        features = "Top_10_UCell",
        group.by = "seurat_clusters") + 
  NoLegend()

dev.copy(pdf, paste0(output.dir, "Pancancer_VlnPlot_UCell_scores_top10.pdf"))
dev.off()

# Cluster number 7 = IFN cluster

#########################
# Get statistics
#########################
#HPCA.pruned.labels
#Type
#Tissue
#cloneType
#functional.cluster


# extract normed values 


# Function for getting normed values
get.norm <- function(input.var, cluster.ID = 7){
  
var.by.cluster <- table(pancancer.cd8@meta.data[, input.var],
                        pancancer.cd8@meta.data$seurat_clusters)

total.cells.var <- rowSums(var.by.cluster)


normed.vals <- sweep(var.by.cluster, 
                     MARGIN = 1, # margin 1 = by row, margin 2 = by column
                     STATS = total.cells.var,
                     FUN = '/')

normed.vals <- normed.vals*100
print(normed.vals)
IFN.clust.vals <- normed.vals[,paste0(cluster.ID)]

return(IFN.clust.vals)

}


# set up output variables
file.name <- "pancancer_CD8_barplot_Normalised_cluster_7_"

# By Type
variable.val <- "Type"

x <- get.norm(variable.val)

barplot(x, 
        las = 2,
        col = c("gray", "blue"), 
        cex.names = 0.6, 
        ylim = c(0, 5))

dev.copy(pdf, paste0(output.dir, file.name, variable.val, ".pdf"))
dev.off()

# Export stats
write.csv(x,  paste0(output.dir, file.name, variable.val, ".csv"))



# By Tissue
variable.val <- "Tissue"

x <- get.norm(variable.val)

barplot(x, 
        las = 2,
        col = c("gray", "blue"), 
        cex.names = 0.6, 
        ylim = c(0, 25))

dev.copy(pdf, paste0(output.dir, file.name, variable.val, ".pdf"))
dev.off()

# Export stats
write.csv(x,  paste0(output.dir, file.name, variable.val, ".csv"))


# By cloneType
variable.val <- "cloneType"

x <- get.norm(variable.val)

barplot(x, 
        las = 2,
        col = c("gray", "blue"), 
        cex.names = 0.6, 
        ylim = c(0, 10))

dev.copy(pdf, paste0(output.dir, file.name, variable.val, ".pdf"))
dev.off()

# Export stats
write.csv(x,  paste0(output.dir, file.name, variable.val, ".csv"))



# By functional.cluster
variable.val <- "functional.cluster"

x <- get.norm(variable.val)

barplot(x, 
        las = 2,
        col = c("gray", "blue"), 
        cex.names = 0.6, 
        ylim = c(0, 10))

dev.copy(pdf, paste0(output.dir, file.name, variable.val, ".pdf"))
dev.off()

# Export stats
write.csv(x,  paste0(output.dir, file.name, variable.val, ".csv"))



# By HPCA.pruned.labels
variable.val <- "HPCA.pruned.labels"

x <- get.norm(variable.val)

barplot(x, 
        las = 2,
        col = c("gray", "blue"), 
        cex.names = 0.6, 
        ylim = c(0, 15))

dev.copy(pdf, paste0(output.dir, file.name, variable.val, ".pdf"))
dev.off()

# Export stats
write.csv(x,  paste0(output.dir, file.name, variable.val, ".csv"))











###############################
# Cluster-wise normalisation
###############################

# Function for getting normed values with cluster-wise normalisation
get.norm.bycluster <- function(input.var, cluster.ID = 9){
  
var.by.cluster <- table(pancancer.cd8@meta.data[, input.var],
                        pancancer.cd8@meta.data$seurat_clusters)

total.cells.clust <- colSums(var.by.cluster)


normed.vals <- sweep(var.by.cluster, 
                     MARGIN = 2, # margin 1 = by row, margin 2 = by column
                     STATS = total.cells.clust,
                     FUN = '/')

normed.vals <- normed.vals*100
print(normed.vals)

return(normed.vals)

}



# set up output variables
file.name <- "pancancer_CD8_barplot_Normalised_by_cluster_"


# By cloneType
variable.val <- "cloneType"

x <- get.norm.bycluster(variable.val)

barplot(x, 
        las = 2,
        col = c("gray", "blue", "red", "orange", "Green"), 
        cex.names = 0.6, 
        legend = T,
        ylim = c(0, 100))



dev.copy(pdf, paste0(output.dir, file.name, variable.val, ".pdf"))
dev.off()

# Export stats
write.csv(x,  paste0(output.dir, file.name, variable.val, ".csv"))




# By HPCA.pruned.labels
variable.val <- "HPCA.pruned.labels"

x <- get.norm.bycluster(variable.val)

barplot(x, 
        las = 2,
        col = c("gray", "blue", "red", "orange", "Green"), 
        cex.names = 0.6, 
        legend = T,
        ylim = c(0, 100))



dev.copy(pdf, paste0(output.dir, file.name, variable.val, ".pdf"))
dev.off()

# Export stats
write.csv(x,  paste0(output.dir, file.name, variable.val, ".csv"))

# By functional.cluster
variable.val <- "functional.cluster"

x <- get.norm.bycluster(variable.val)

barplot(x, 
        las = 2,
        col = c("gray", "blue", "red", "orange", "Green"), 
        cex.names = 0.6, 
        legend = T,
        ylim = c(0, 100))



dev.copy(pdf, paste0(output.dir, file.name, variable.val, ".pdf"))
dev.off()

# Export stats
write.csv(x,  paste0(output.dir, file.name, variable.val, ".csv"))



# By Tissue
variable.val <- "Tissue"

x <- get.norm.bycluster(variable.val)

barplot(x, 
        las = 2,
        col = c("gray", "blue", "red", "orange", "Green"), 
        cex.names = 0.6, 
        legend = T,
        ylim = c(0, 100))



dev.copy(pdf, paste0(output.dir, file.name, variable.val, ".pdf"))
dev.off()

# Export stats
write.csv(x,  paste0(output.dir, file.name, variable.val, ".csv"))

# By Type
variable.val <- "Type"

x <- get.norm.bycluster(variable.val)

barplot(x, 
        las = 2,
        col = c("gray", "blue", "red", "orange", "Green"), 
        cex.names = 0.6, 
        legend = T,
        ylim = c(0, 100))



dev.copy(pdf, paste0(output.dir, file.name, variable.val, ".pdf"))
dev.off()

# Export stats
write.csv(x,  paste0(output.dir, file.name, variable.val, ".csv"))


###################################
# Umap overview plots of clusters
###################################

# By tissue type, downsampled 
x <- table(pancancer.cd8@meta.data$Tissue)
min(x)

Idents(pancancer.cd8) <- pancancer.cd8@meta.data$Tissue
temp.seurat <- subset(pancancer.cd8, downsample = 4000)

Idents(temp.seurat) <- temp.seurat@meta.data$seurat_clusters

scCustomize::DimPlot_scCustom(temp.seurat, 
                              pt.size = 1,
                              num_columns = 4,
                              split_seurat = T,
                              split.by = "Tissue")

dev.copy(pdf, paste0(output.dir, "pancancer_umap_by_tissue_downsampled_bytissue_4K_cells_all_except_colorectal_remove_this.pdf"))
dev.off()

# By disease/sample type downsampled
x <- table(pancancer.cd8@meta.data$Type)
min(x)

Idents(pancancer.cd8) <- pancancer.cd8@meta.data$Type
temp.seurat <- subset(pancancer.cd8, downsample = 5000)

Idents(temp.seurat) <- temp.seurat@meta.data$seurat_clusters

scCustomize::DimPlot_scCustom(temp.seurat, 
                              pt.size = 1,
                              num_columns = 4,
                              split_seurat = T,
                              split.by = "Type")

dev.copy(pdf, paste0(output.dir, "pancancer_umap_by_type_downsampled_by_type.pdf"))
dev.off()


```


