---
title: "04_Manuscript_Pancancer_IFN"
author: "Dillon Corvino"
date: "07/03/2022"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

### Dataset information

```{r Dataset_Info}

# HNSCC scRNAseq/scTCRseq dataset
# Cells are HNSCC TILs sorted on Lymphocytes/Live/CD3+/CD4-/CD8+
# Human samples
# Each sample is a pool of 4 patients
# samples 1 and 2 are unstimulated 
# samples 3 and 4 are stimulated 
# Patients in sample 1 match patients in sample 3 and sample 2 pairs with 4
# Data acquired was transcript expression, ADT (antibody expression), and TCRA & B sequences
# This analysis uses just Transcript and TCR data


# General information
# CellRanger was used for sequence alignment/QC/counts/TCR calls - performed by Ross (QIMR, Brisbane, Aus)
# DC was provided the output from CellRanger (counts matrix) and used this as input for analysis within this script

```

### Environment

```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Create output directories & load custom functions & colour scheme
source("scripts/Setup.R", local = knitr::knit_global())

# pipeline variables
long.compute <- FALSE

# load required packages
require(tidyverse)
require(Seurat)
require(useful)
require(ggrepel)
require(clusterProfiler)
require(enrichplot)
require(org.Hs.eg.db)
require(UCell)
require(Nebulosa)
require(ggsci)
library(harmony)
require(irGSEA)
#source("scripts/aimed_analysis/functions.R")


# Create output directory
output.dir <- "results/Manuscript/Pancancer_IFN/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = TRUE)
}
```


## Type I IFN score pancancer dataset {.tabset}

### Prepare dataset / read and filter for CD8 / or load dataset

```{r read_dataset_pancancer}

if(long.compute){
  
  # Due to memory limits, the dataset is written in as pancancer.cd8 rather than subsetting for cd8 and creating a new variable. 
  #i.e use only one variable to limit the memory requirments 
pancancer.cd8 <- LoadH5Seurat("data/external_datasets/Utility_dataset.h5Seurat")

Idents(pancancer.cd8) <- pancancer.cd8@meta.data$functional.cluster
pancancer.cd8 <- subset(pancancer.cd8, idents = c("CD8_EffectorMemory", "CD8_NaiveLike", "CD8_EarlyActiv", "CD8_Tex", "CD8_Tpex"))

# still many cells of mixed annotation by consensus.major ID
Idents(pancancer.cd8) <- pancancer.cd8@meta.data$consensus.major
pancancer.cd8 <- subset(pancancer.cd8, idents = "T_cell")

# further filtering for CD8 T cell annotation
Idents(pancancer.cd8) <- pancancer.cd8@meta.data$consensus.Tcell
pancancer.cd8 <- subset(pancancer.cd8, idents = "CD8")

# further filtering for annotation
Idents(pancancer.cd8) <- pancancer.cd8@meta.data$HPCA.pruned.labels
keep.var <- grep(".*CD8.*", unique(Idents(pancancer.cd8)), value =T)

pancancer.cd8 <- subset(pancancer.cd8, idents = keep.var)

# Type annotation
table(pancancer.cd8@meta.data$Type)/1000
# Juxta = 262 cells only
# Blood = 12.993K
# LN = 23.25K
# Met = 5.711K
# Normal = 41.119K
# Tumor = 173.74K

# Therefore remove Juxta cells 
# also remove Met and LN as these are a bit complicated to include in analysise.
# i.e the biological interpretation of Met tissue is highly dependent on each manuscript and what they mean by met and if LN is healthy or diseased
# also remove blood to keep it simple, normal and tumor tissue


Idents(pancancer.cd8) <- pancancer.cd8@meta.data$Type
pancancer.cd8 <- subset(pancancer.cd8, idents = c("Tumor", "Normal"))


table(pancancer.cd8@meta.data$Type, pancancer.cd8@meta.data$Tissue)


library(harmony)
# Normalise and harmonize dataset and calculate umap projection
pancancer.cd8 <- 
  pancancer.cd8 %>%
  NormalizeData() %>%
  FindVariableFeatures() %>%
  ScaleData() %>%
  RunPCA() %>%
  harmony::RunHarmony(reduction = "pca",
                      group.by.vars = c("Cohort"), 
                      plot_convergence = TRUE, 
                      verbose = TRUE) %>%
  RunUMAP(reduction = "harmony", 
          dims = 1:20) %>%
  FindNeighbors(reduction = "harmony",
                dims = 1:20) %>%
  FindClusters(resolution = 0.5)


# save filtered dataset
SaveH5Seurat(pancancer.cd8, 
             "data/external_datasets/pancancer_cd8_normed.h5Seurat",
             overwrite = TRUE)


}else{
  
  pancancer.cd8 <- LoadH5Seurat("data/external_datasets/pancancer_cd8_normed.h5Seurat")

}

# quick visualisation of the normed and subseted CD8 dataset

UMAPPlot(pancancer.cd8, 
         group.by = "HPCA.pruned.labels")

UMAPPlot(pancancer.cd8, 
         group.by = "seurat_clusters")





```
 
### apply ucell score to pancancer

```{r ucell_pancancer}

if(!exists("pancancer.cd8")){
  
    pancancer.cd8 <- LoadH5Seurat("data/external_datasets/pancancer_cd8_normed.h5Seurat")

}

if(long.compute){

 cores <- parallel::detectCores(all.tests = FALSE, logical = TRUE)

# Calculate Ucell scores for signatures
DefaultAssay(pancancer.cd8) <- "RNA"
pancancer.cd8 <- UCell::AddModuleScore_UCell(pancancer.cd8,
                                             features = Signature.list,
                                             ncores = cores)

# Export object with UCell score
SaveH5Seurat(pancancer.cd8, 
             "data/external_datasets/pancancer_cd8_normed.h5Seurat", 
             overwrite = TRUE)

}

```
 
 
### Visualise IFN cluster 
```{r vis_IFN_cluster}

colnames(pancancer.cd8@meta.data)

# Visualise the results

# Density plots 
Plot_Density_Custom(pancancer.cd8, 
                    features =  "Top_10_UCell",
                    custom_palette = batlow.pal,
                    joint = FALSE, 
                    pt.size = 1, 
                    reduction = "umap")

dev.copy(pdf, paste0(output.dir, "Pancancer_Nebulosa_UCell_scores_top10.pdf"))
dev.off()

# Cluster distribution

UMAPPlot(pancancer.cd8, 
         group.by = "seurat_clusters")

dev.copy(pdf, paste0(output.dir, "Pancancer_UMAP_clusters.pdf"))
dev.off()


# VlnPlots
VlnPlot(pancancer.cd8,
        pt.size = 0, 
        same.y.lims = T,
        features = "Top_10_UCell",
        group.by = "seurat_clusters") + 
  NoLegend()

dev.copy(pdf, paste0(output.dir, "Pancancer_VlnPlot_UCell_scores_top10.pdf"))
dev.off()



# Specifically colour IFN cluster
Idents(pancancer.cd8) <- pancancer.cd8@meta.data$seurat_clusters

scCustomize::Cluster_Highlight_Plot(pancancer.cd8, 
                                    cluster_name = "7", 
                                    highlight_color = "Red", 
                                    background_color = "lightgray", 
                                    pt.size = 0.1, 
                                    raster = TRUE)

dev.copy(pdf, paste0(output.dir, "Pancancer_UMAP_clusters_highlighted.pdf"))
dev.off()


# Cluster number 7 = IFN cluster


```
 
 
### Extract statistics on IFN cluster 
```{r extract_stats}

######################################################################################################
# Extract and calculate freq of IFN cluster across normal and disease in various tumor types
######################################################################################################

IFN.cluster <- "7"


# select tumors that have both normal and disease 
table(pancancer.cd8@meta.data$Type, pancancer.cd8@meta.data$Tissue)

# subset for tumor type
Idents(pancancer.cd8) <- pancancer.cd8@meta.data$Tissue

tissue.vec <- c("Colorectal", "Endometrial", "Esophagus", "Lung", "Renal")


for(i in seq_along(tissue.vec)){
  
  # subset for tumor type
  temp.seurat <- subset(pancancer.cd8, idents = tissue.vec[i])
  
  # get number of sequenced cells
  sequenced.cell.num <- table(temp.seurat@meta.data$Type)
  
  # get values for number of cells in each cluster across normal or disease
  cluster.vals <- table(temp.seurat@meta.data$seurat_clusters, 
                        temp.seurat@meta.data$Type)
  
  # extract cell numbers for cluster 7 only
  IFN.clust.vals <- cluster.vals[grepl(paste0(IFN.cluster), rownames(cluster.vals)), ]
  
  # calculate the freq of normal or disease that is within the IFN cluster
  IFN.freq <- (IFN.clust.vals/sequenced.cell.num)*100
  
  # plot values
  barplot(IFN.freq, 
          las = 2,
          col = c("gray", "blue"), 
          cex.names = 0.6, 
          ylim = c(0, 25))
  
  dev.copy(pdf, paste0(output.dir, tissue.vec[i], "_barplot_IFN_cluster_freq_across_disease_status.pdf"))
  dev.off()
  
  # write data to file
  write.csv(IFN.freq, paste0(output.dir, tissue.vec[i], "_IFN_cluster_freq_across_disease_status.csv"))
  
  
  # freq of within IFN cluster 
  total.IFN <- sum(IFN.freq)
  
  normed.IFN <- (IFN.freq/total.IFN)*100
  
  
  
  barplot(normed.IFN, 
          las = 2,
          col = c("gray", "blue"), 
          cex.names = 0.6, 
          ylim = c(0, 100))
  
  dev.copy(pdf, paste0(output.dir, tissue.vec[i], "_barplot_freq_across_disease_status_within_IFN_cluster.pdf"))
  dev.off()
  
  # write data to file
  write.csv(normed.IFN, paste0(output.dir, tissue.vec[i], "_freq_within_IFN_cluster_across_disease_status.csv"))
  
  rm(temp.seurat)
  gc()
  
}


#################################################################################
# within sequenced cells from the tumor, which tumor has the most IFN cluster
#################################################################################

Idents(pancancer.cd8) <- pancancer.cd8@meta.data$Type

tumor.seurat <- subset(pancancer.cd8, idents = "Tumor")

Idents(tumor.seurat) <- tumor.seurat@meta.data$Tissue

tumor.seurat # 54,971 cells and 173,744 cells

 

tumor.by.cluster <- table(tumor.seurat@meta.data$Tissue, tumor.seurat@meta.data$seurat_clusters)

logic.vec <- grepl(paste0(IFN.cluster), colnames(tumor.by.cluster))
IFN.by.cluster <- tumor.by.cluster[, logic.vec]


cells.sequenced <- table(tumor.seurat@meta.data$Tissue)

cells.sequenced

IFN.freq <- (IFN.by.cluster/cells.sequenced)*100

max(IFN.freq)

barplot(IFN.freq, 
        las = 2,
        col = c("gray", "blue"), 
        cex.names = 0.6, 
        ylim = c(0, 25))

dev.copy(pdf, paste0(output.dir, "barplot_IFN_cluster_freq_across_tumor_types.pdf"))
dev.off()

# write data to file
write.csv(IFN.freq, paste0(output.dir, "IFN_cluster_freq_across_tumor_types.csv"))

# clean environment 
rm(tumor.seurat)
gc()



```


### Extract and focus analysis on just high IFN cluster expressing tumors
```{r ISG_high_tumors}


if(long.compute){
  # Subset for ISG cluster enriched tumors
  Idents(pancancer.cd8) <- pancancer.cd8@meta.data$Tissue
  table(pancancer.cd8@meta.data$Tissue, pancancer.cd8@meta.data$Type)
  
  tissue.vec <- c("Ovarian", "Esophagus", "Breast")
  
  ISG.tumors.seurat <- subset(pancancer.cd8, idents = tissue.vec)
  
  # take just cells from tumor tissue
  Idents(ISG.tumors.seurat) <- ISG.tumors.seurat@meta.data$Type
  
  ISG.tumors.seurat <- subset(ISG.tumors.seurat, idents = "Tumor")
  
  # check subsetting correctly performed 
  table(ISG.tumors.seurat@meta.data$Tissue, ISG.tumors.seurat@meta.data$Type)
  

  
  
  library(harmony)
  # Normalise and harmonize dataset and calculate umap projection
  
  # store upper level seurat clusters incase needed later
  ISG.tumors.seurat$orig.clusters <- ISG.tumors.seurat@meta.data$seurat_clusters

  
  
  ISG.tumors.seurat <- 
    ISG.tumors.seurat %>%
    Seurat::NormalizeData() %>%
    Seurat::FindVariableFeatures() %>%
    Seurat::ScaleData() %>%
    Seurat::RunPCA() %>%
    harmony::RunHarmony(reduction = "pca",
                        group.by.vars = c("Cohort", "Sample"), 
                        plot_convergence = TRUE, 
                        verbose = TRUE)  %>%
    Seurat::RunUMAP(reduction = "harmony", 
          dims = 1:20, 
          min.dist = 0.3, 
          spread = 1,
          n.neighbors = 5, # 5 - 50 , larger = more global
          densmap = F,
          seed.use = 42)  %>%
  FindNeighbors(reduction = "harmony",
                dims = 1:20) %>%
  FindClusters(resolution = 0.5)
  
  
  
  # Rename clusters
ISG.tumors.seurat@meta.data$IFN_clusters <- ISG.tumors.seurat@meta.data$seurat_clusters

Idents(ISG.tumors.seurat) <- ISG.tumors.seurat@meta.data$IFN_clusters
ISG.tumors.seurat <- ISG.tumors.seurat %>%
  RenameIdents('0' = "other",
               '1' = "other",
               '2' = "other", 
               '3' = "other",
               '4' = "other", 
               '5' = "ISG", 
               '6' = "other", 
               '7' = "other",
               '8' = "other",
               '9' = "other",
               '10' = "other")
Idents(ISG.tumors.seurat)

ISG.tumors.seurat@meta.data$IFN_status <- Idents(ISG.tumors.seurat)

    # save normed and recalculated umap
  SaveH5Seurat(ISG.tumors.seurat, 
               "data/external_datasets/ISG_tumors_seurat.h5Seurat",
               overwrite = TRUE)
  
  
}else if(!exists("ISG.tumors.seurat")){
  
  ISG.tumors.seurat <- LoadH5Seurat("data/external_datasets/ISG_tumors_seurat.h5Seurat")
  
  
}

scCustomize::DimPlot_scCustom(ISG.tumors.seurat)

unique(ISG.tumors.seurat@meta.data$Tissue)
scCustomize::DimPlot_scCustom(ISG.tumors.seurat, 
                              split.by = "Tissue")

scCustomize::DimPlot_scCustom(ISG.tumors.seurat, 
                              split.by = "Tissue", 
                              group.by = "Sample")


Idents(ISG.tumors.seurat) <- ISG.tumors.seurat@meta.data$seurat_clusters


# Density plots 
Plot_Density_Custom(ISG.tumors.seurat, 
                    features =  "Top_10_UCell",
                    custom_palette = batlow.pal,
                    joint = FALSE, 
                    pt.size = 1, 
                    reduction = "umap")

dev.copy(pdf, paste0(output.dir, "Breast_Esoph_Ovar_Nebulosa_UCell_scores_top10.pdf"))
dev.off()

scCustomize::Cluster_Highlight_Plot(ISG.tumors.seurat, 
                                            cluster_name = "5", 
                                            highlight_color = "navy",
                                            background_color = "lightgray",
                                            pt.size = 0.5) + NoLegend()

dev.copy(pdf, paste0(output.dir, "Breast_Esoph_Ovar_highlight_clust_5.pdf"))
dev.off()


VlnPlot(ISG.tumors.seurat, 
        pt.size = 0,
        features =  "Top_10_UCell")

dev.copy(pdf, paste0(output.dir, "Breast_Esoph_Ovar_vlnPlot.pdf"))
dev.off()


scCustomize::DimPlot_scCustom(temp)


##################################################################################################
# Vln plot of Ucell scores for T1 and T2 IFN calculated on original dataset by Nick Borcherding
##################################################################################################

VlnPlot(ISG.tumors.seurat,
        feature = "T1_Interferon_UCell",
        pt.size = 0,
        split.by = "Tissue",
        group.by = "IFN_status")

dev.copy(pdf, paste0(output.dir, "Breast_Esoph_Ovar_vlnPlot_T1_Ucell.pdf"))
dev.off()

VlnPlot(ISG.tumors.seurat,
        feature = "T2_Interferon_UCell",
        pt.size = 0,
        split.by = "Tissue",
        group.by = "IFN_status")

dev.copy(pdf, paste0(output.dir, "Breast_Esoph_Ovar_vlnPlot_T2_Ucell.pdf"))
dev.off()


















library("SCpubr")

# Apply statistical tests.
p <- SCpubr::do_BoxPlot(sample = ISG.tumors.seurat,
                        feature = "T2_Interferon_UCell",
                        assay = "RNA",
                        group.by = "IFN_status",
                        split.by = "Tissue",
                        use_test = TRUE,
                        use_silhouette = FALSE,
                        map_signif_level = TRUE, # show star or pvalue
                        comparisons = list(c("Breast", "T_cell"),
                                           c("Epithelial", "Endothelial"),
                                           c("iCAF", "Endothelial")))
p





temp.seurat <- ISG.tumors.seurat

temp.seurat@meta.data$Tissue_IFN_status <- paste0(temp.seurat@meta.data$Tissue, "_", temp.seurat@meta.data$IFN_status)




SCpubr::do_BoxPlot(sample = temp.seurat,
                   feature = "T2_Interferon_UCell",
                   assay = "RNA",
                   group.by = "Tissue_IFN_status", 
                   use_test = TRUE,
                   map_signif_level = F,
                   comparisons = list(c("Breast_ISG", "Breast_other"),
                                      c("Esophagus_ISG", "Esophagus_other"),
                                      c("Ovarian_ISG", "Ovarian_other")))





# Show the distribution of clonotypes 

Idents(ISG.tumors.seurat) <- ISG.tumors.seurat@meta.data$cloneType

clonetypes <- levels(Idents(ISG.tumors.seurat))
clonotype.names <- c("Hyperexpanded", "Large", "Medium", "Small", "Rare")




for(i in seq_along(clonetypes)){
  print(scCustomize::Cluster_Highlight_Plot(ISG.tumors.seurat, 
                                            cluster_name = paste0(clonetypes[i]), 
                                            highlight_color = "navy",
                                            background_color = "lightgray",
                                            split.by = "Tissue",
                                            pt.size = 1) + 
          NoLegend() + 
          ggtitle(paste0(clonetypes[i])))
  
  #dev.copy(pdf, paste0(output.dir, "UMAP_highlighted_clonotype_", clonotype.names[i], "_split_condition.pdf"))
  #dev.off()
}


Idents(ISG.tumors.seurat) <- ISG.tumors.seurat@meta.data$seurat_clusters

scCustomize::Cluster_Highlight_Plot(ISG.tumors.seurat, 
                                            cluster_name = "7", 
                                            highlight_color = "navy",
                                            background_color = "lightgray",
                                            split.by = "Tissue",
                                            pt.size = 1)  + NoLegend()








# Umap visualise top clonotypes in IFN cluster
clonotype.data <- table(temp@meta.data$CTaa, temp@meta.data$seurat_clusters)
clonotype.data <- as.data.frame(clonotype.data)

aa.keep <- clonotype.data %>% 
  filter(Var2 == "ISG") %>%
  top_n(30, Freq)

aa.keep <- aa.keep$Var1

Idents(ISG.tumors.seurat) <- ISG.tumors.seurat@meta.data$CTaa


for(i in seq_along(aa.keep)){
  print(scCustomize::Cluster_Highlight_Plot(ISG.tumors.seurat, 
                                            cluster_name = paste0(aa.keep[i]), 
                                            highlight_color = "navy",
                                            split.by = "Tissue",
                                            background_color = "lightgray",
                                            pt.size = 1) + 
          NoLegend() + 
          ggtitle(paste0(aa.keep[i])))
  
  #dev.copy(pdf, paste0(output.dir, "UMAP_highlighted_clonotype_Type_IFN_top_10_", aa.keep[i], ".pdf"))
  #dev.off()
  
}




table(ISG.tumors.seurat@meta.data$Tissue) #

table(ISG.tumors.seurat@meta.data$seurat_clusters, ISG.tumors.seurat@meta.data$cloneType)







IFN.cluster <- "7"
Idents(ISG.tumors.seurat) <- ISG.tumors.seurat@meta.data$Tissue

# subset for tumor type
temp.seurat <- subset(ISG.tumors.seurat, idents = tissue.vec[2])


table(temp.seurat@meta.data$seurat_clusters, temp.seurat@meta.data$cloneType)




# get number of cells per clometype
clonetype.cell.num <- table(temp.seurat@meta.data$cloneType)

# get values for number of cells in each cluster across normal or disease
cluster.vals <- table(temp.seurat@meta.data$seurat_clusters, 
                      temp.seurat@meta.data$cloneType)

# extract cell numbers for cluster 7 only
IFN.clust.vals <- cluster.vals[grepl(paste0(IFN.cluster), rownames(cluster.vals)), ]

# calculate the freq of normal or disease that is within the IFN cluster
IFN.freq <- (IFN.clust.vals/sequenced.cell.num)*100





#install.packages("ggraph")
library("ggraph")

#devtools::install_github("ncborcherding/scRepertoire@dev")
library("scRepertoire")

tissue.vec <- c("Ovarian", "Esophagus", "Breast")
clonetype.cols <- c("salmon4", "salmon", "orange", "cyan3", "Lightblue")
names(clonetype.cols) <- c("Hyperexpanded", "Large", "Medium", "Small", "Rare")


for(i in seq_along(tissue.vec)){
  

# Plot bargraph of clonetype frequency per cluster
Idents(ISG.tumors.seurat) <- ISG.tumors.seurat@meta.data$Tissue

# subset for tumor type
temp.seurat <- subset(ISG.tumors.seurat, idents = tissue.vec[i])

Idents(temp.seurat) <- temp.seurat@meta.data$cloneType

cells.clonetype.per.clust<- table(temp.seurat@meta.data$cloneType, temp.seurat@meta.data$IFN_status)
cells.cluster <- colSums(cells.clonetype.per.clust)

output.df <- cells.clonetype.per.clust*NA

for(i in 1:length(cells.cluster)){
output.df[,i] <- (cells.clonetype.per.clust[,i]/cells.cluster[i]) * 100
}


temp.df <- output.df[, 2]
# Plot freq of sample per cluster
barplot(temp.df, 
        las = 2,
        legend = T,
        col = clonetype.cols, 
        cex.names = 0.5)

write.csv(output.df, paste0(output.dir, tissue.vec[1], "clontype_freq.csv"))


}







colSums(output.df)


#temp.cols <- c(rep("gray", 11))
#temp.cols[6] <- "red"



sum(temp.df)


#dev.copy(pdf, paste0(output.dir, "Freq_of_clonetype_per_cluster.pdf"))
#dev.off()

# ggplot version 

output.df.gg <- as.data.frame(output.df)
output.df.gg %>%
  dplyr::filter(Var2 == "ISG") %>%
ggplot() +
  aes(x = Var1, y = Freq, fill = Var1) +
  geom_col() +
  scale_fill_manual(
    values = c(`Hyperexpanded (0.1 < X <= 1)` = "salmon4",
    `Large (0.01 < X <= 0.1)` = "salmon",
    `Medium (0.001 < X <= 0.01)` = "orange",
    `Small (1e-04 < X <= 0.001)` = "cyan3",
    `Rare (0 < X <= 1e-04)` = "Lightblue")
  ) +
  labs(x = " ", y = "Frequency [%]") +
  theme_minimal()

ggsave(paste0(output.dir, "Freq_of_clonetype_per_cluster_ggversion.pdf"))




# convert seurat object to list in order to use functions 
seurat.list <- expression2List(temp.seurat, group = "IFN_clusters")



# Clonal overlap 
clonalOverlap(seurat.list,
              cloneCall = "aa",
              method = "overlap")

dev.copy(pdf, paste0(output.dir, "Clonal_overlap.pdf"))
dev.off()

?clonalDiversity

# Clonal Diversity 
clonalDiversity(seurat.list, 
                exportTable = T,
                cloneCall = "aa")


dev.copy(pdf, paste0(output.dir, "Clonal_diversity.pdf"))
dev.off()







colnames(temp.seurat@meta.data)



x <- table(temp.seurat@meta.data$Frequency)
x <- temp.seurat@meta.data
x <- names(x)
x <- as.numeric(x)
x <- x*100
hist(x, freq = T, breaks = 100)


?hist

  


# plot values
  barplot(IFN.freq, 
          las = 2,
          col = c("gray", "blue"), 
          cex.names = 0.6, 
          ylim = c(0, 25))
  
  dev.copy(pdf, paste0(output.dir, tissue.vec[i], "_barplot_IFN_cluster_freq_across_disease_status.pdf"))
  dev.off()
  
  # write data to file
  write.csv(IFN.freq, paste0(output.dir, tissue.vec[i], "_IFN_cluster_freq_across_disease_status.csv"))
  


```



### Heatmap of top clonotypes
```{r Heatmap_top_clonotypes}

library(gplots)
library(RColorBrewer)
########################
# Heatmap functions
########################

hclust.func <- function(x){
  hclust(x, method = 'ward.D2')
}

dist.func <- function(x){
  dist(x,method = 'binary')
}

col.palette <- colorRampPalette(brewer.pal(8, "RdYlBu"))(600)


# Plot bargraph of clonetype frequency per cluster
Idents(ISG.tumors.seurat) <- ISG.tumors.seurat@meta.data$Tissue

# subset for tumor type
temp.seurat <- subset(ISG.tumors.seurat, idents = "Esophagus")



# Get data of how many cells in each cluster are assigned to what clonotypes
clonotype.data <- table(temp.seurat@meta.data$CTaa, temp.seurat@meta.data$IFN_clusters)
colnames(clonotype.data)[6]
aa.keep <- sort(clonotype.data[,6], decreasing = TRUE)[1:50]

# Plot the top 50 clonotypes
aa.keep <- sort(rowSums(clonotype.data), decreasing = TRUE)[1:10]

keep.vec <- rownames(clonotype.data) %in% names(aa.keep)

plot.data <- clonotype.data[keep.vec, ]

rownames(plot.data)
plot.data <- plot.data[-44, ]

heatmap.2(as.matrix(log2(plot.data+1)), 
          scale = "none",
          trace = "none",
          ColSideColors = CD8.cols,
          Colv = TRUE,
          cexRow = 0.8, 
          margins = c(8, 15),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette))     

dev.copy(pdf, paste0(output.dir, "Heatmap_top_50_clonotypes.pdf"))
dev.off()







############################################
# Plot top clonotypes in each cluster 
############################################

for(i in seq_along(colnames(clonotype.data))){
  temp <- as.data.frame(clonotype.data)
  
  temp <- temp %>%
    dplyr::filter(Var2 == paste0(colnames(clonotype.data)[i])) %>%
    dplyr::top_n(10)
  
  clonotypes.to.plot <- temp$Var1
  
  keep.vec <- rownames(clonotype.data) %in% clonotypes.to.plot
  
  plot.data <- clonotype.data[keep.vec, ]
  
  heatmap.2(as.matrix(log2(plot.data+1)), 
            scale = "none",
            trace = "none",
            ColSideColors = CD8.cols,
            Colv = FALSE,
            cexRow = 0.8, 
            margins = c(8, 15),
            cexCol = 0.8, 
            distfun = dist.func,
            hclustfun = hclust.func,
            srtCol = 90,
            col = rev(col.palette))
  
  dev.copy(pdf, paste0(output.dir, "Heatmap_top_10_clonotypes_in_", colnames(clonotype.data)[i], ".pdf"))
  dev.off()
  
}



```

