---
title: "05_Manuscript_Figure_4"
author: "Dillon Corvino"
date: "07/03/2022"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

### Dataset information

```{r Dataset_Info}

# HNSCC scRNAseq/scTCRseq dataset
# Cells are HNSCC TILs sorted on Lymphocytes/Live/CD3+/CD4-/CD8+
# Human samples
# Each sample is a pool of 4 patients
# samples 1 and 2 are unstimulated 
# samples 3 and 4 are stimulated 
# Patients in sample 1 match patients in sample 3 and sample 2 pairs with 4
# Data acquired was transcript expression, ADT (antibody expression), and TCRA & B sequences
# This analysis uses just Transcript and TCR data


# General information
# CellRanger was used for sequence alignment/QC/counts/TCR calls - performed by Ross (QIMR, Brisbane, Aus)
# DC was provided the output from CellRanger (counts matrix) and used this as input for analysis within this script

```

### Environment

```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Create output directories & load custom functions & colour scheme
source("scripts/Setup.R", local = knitr::knit_global())

# pipeline variables
long.compute <- FALSE

# load required packages
require(tidyverse)
require(Seurat)
require(useful)
require(clusterProfiler)
require(enrichplot)
require(UCell)

#source("scripts/aimed_analysis/functions.R")

# Create output directory
output.dir <- "results/Manuscript/Figure4/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = TRUE)
}
```



## Figure 4 {.tabset}

### Reading data
```{r reading_data}

# Read in CD8 only dataset
CD8.seurat <- LoadH5Seurat("saves/CD8_seurat.h5Seurat")
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"

```


### Define signature
```{r define_signature}

# Compute DEGs and define the Type_I_IFN cluster signature
if(long.compute){
  
  DefaultAssay(CD8.seurat) <- "RNA"
  
  # Calculate DEGs
  IFN.clust.DEGs <- FindMarkers(CD8.seurat,
                                ident.1 = "Type_I_IFN",
                                logfc.threshold = 0,
                                min.pct = 0)
  # Annotate DEGs
  IFN.clust.DEGs <-
    IFN.clust.DEGs%>%
    as_tibble(rownames = "gene_symbol") %>%
    mutate(DE = case_when(
      avg_log2FC > 0 & p_val_adj <= 0.1 ~ "up",
      avg_log2FC < 0 & p_val_adj <= 0.1 ~ "down",
      TRUE ~ "notDE"))
  
  # Export results
  write_csv(x = IFN.clust.DEGs,
            file = paste0(output.dir, "IFN_clust_DEGs.csv"))
  
  
  # Define signature as top DEGs only
  IFN.sig <- IFN.clust.DEGs %>%
    filter(p_val_adj <= 0.05) %>%
    top_n(20, avg_log2FC)
  
  
  # Export results
  write_csv(x = IFN.sig,
            file = paste0(output.dir, "IFN_signature.csv"))
  
  
}


# create signatures
#IFN.clust.DEGs <- read.csv(paste0(output.dir, "IFN_clust_DEGs.csv"))

#Sig.10 <- IFN.clust.DEGs %>%
#  filter(p_val_adj <= 0.05) %>%
#  top_n(10, avg_log2FC) %>%
#  pull(gene_symbol) 


# Hard code the signature to prevent any alternations with statistical drift 
Sig.10 <- c("IFI6", "ISG15", "IFIT3", "MX1", "ISG20", "IFITM1", "LY6E", "IFIT1", "MX2", "OAS1")

Sig.10.up <- paste0(Sig.10, "+")

Signature.list <- list(Top_10 = Sig.10.up)

signature.names <- paste0(names(Signature.list), "_UCell")

```

### Figure 4 A & B
```{r Figure_4A_B}


  # read in data
input.dir <- "results/Manuscript/Figure4/"

IFN.clust.DEGs <- read.csv(paste0(input.dir, "IFN_clust_DEGs.csv"))


# Prepare dataset for heatmap
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"


average.seurat <- AverageExpression(CD8.seurat, return.seurat = T)


goi <- IFN.clust.DEGs %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_log2FC > 0.25) %>%
  dplyr::filter(DE == "up") %>%
  dplyr::arrange(desc(avg_log2FC)) %>%
  dplyr::pull(gene_symbol)


DoHeatmap(average.seurat, 
          draw.lines = FALSE,
          raster = FALSE,
          features = goi, 
          group.colors = CD8.cols) + 
  scico::scale_fill_scico(palette = "batlow", 
                          direction = 1, 
                          na.value = "white")

dev.copy(pdf, paste0(output.dir, "Figure_4A.pdf"))
dev.off()


# Plot top 10 DEG / i.e the ISG cluster signature

DoHeatmap(average.seurat, 
          draw.lines = FALSE,
          raster = FALSE,
          features = Sig.10, 
          group.colors = CD8.cols) + 
  scico::scale_fill_scico(palette = "batlow", 
                          direction = 1, 
                          na.value = "white")

dev.copy(pdf, paste0(output.dir, "Figure_4B.pdf"))
dev.off()


```

### Figure 4C
```{r Figure_4C}

# install required packages
#BiocManager::install("decoupleR") 
#BiocManager::install("OmnipathR") 
#BiocManager::install("ComplexHeatmap")

library("decoupleR")

# Retrieve prior knowledge network.
network <- decoupleR::get_progeny(organism = "human")


# Run weighted means algorithm.
activities <- decoupleR::run_wmean(mat = as.matrix(CD8.seurat@assays$RNA@data),
                                   network = network,
                                   .source = "source",
                                   .targe = "target",
                                   .mor = "weight",
                                   times = 100,
                                   minsize = 5)


# General heatmap.
out <- SCpubr::do_PathwayActivityPlot(sample = CD8.seurat,
                                      cluster_rows = FALSE,
                                      activities = activities)
p <- out$heatmaps$average_scores
p
dev.copy(pdf, paste0(output.dir, "Figure_4C.pdf"))
dev.off()

```



### Figure 4D
```{r Figure_4D}


# load packages
library("enrichR")
#install.packages("SCpubr")
library("SCpubr")
 
# set serach to human genes
enrichR::setEnrichrSite(site = "Enrichr")

dbs <- enrichR::listEnrichrDbs()
dbs <- sort(dbs$libraryName)

# Choose the dataset to query against.
dbs_use <- c("GO_Biological_Process_2021")


# read in IFN clust DEGS
IFN.clust.DEGs <- read.csv(paste0(output.dir, "IFN_clust_DEGs.csv"))

# filter DEGs for sig up genes 
IFN.sig.up <- IFN.clust.DEGs %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(DE == "up") 

# filter DEGs for sig down genes 
IFN.sig.dn <- IFN.clust.DEGs %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(DE == "down") 




############################################################
# Perform enrichment analysis for Upregulated genes 
############################################################

# List of genes to use as input.
up.genes <- unique(IFN.sig.up$gene_symbol)

# Retrieve the enriched terms.
up.enriched.terms <- enrichR::enrichr(up.genes, dbs_use)

# Plot enriched terms 
upregulated.plots <- SCpubr::do_TermEnrichmentPlot(enriched_terms = up.enriched.terms, 
                                                   nchar_wrap = 20,
                                                   font.size = 5, 
                                                   font.type = "sans", 
                                                   text_labels_size = 1,
                                                   colors.use = c("salmon2", "gray"),
                                                   nterms = 10)






############################################################
# Perform enrichment analysis for Downregulated genes 
############################################################

# List of genes to use as input.
dn.genes <- unique(IFN.sig.dn$gene_symbol)

# Retrieve the enriched terms.
dn.enriched.terms <- enrichR::enrichr(dn.genes, dbs_use)

# Plot enriched terms 
downregulated.plots <- SCpubr::do_TermEnrichmentPlot(enriched_terms = dn.enriched.terms, 
                                                   nchar_wrap = 20,
                                                   font.size = 5, 
                                                   font.type = "sans", 
                                                   text_labels_size = 1,
                                                   colors.use = c("royalblue", "gray"),
                                                   nterms = 10)

# Plot and save all plots generated 
for(i in seq_len(length(upregulated.plots))){
  
  
  # upregulated plots
  name.var <- names(upregulated.plots)[i]
  
  ggsave(filename = paste0(output.dir, "Upregulated_",  name.var, ".pdf"),
         plot = upregulated.plots[[i]], 
         width = 20, 
         height = 20, 
         units = "in",
         device = "pdf")
  
  # downregulated plots
  name.var <- names(downregulated.plots)[i]
  
  ggsave(filename = paste0(output.dir, "Downregulated_",  name.var, ".pdf"),
         plot = downregulated.plots[[i]], 
         width = 20, 
         height = 20, 
         units = "in",
         device = "pdf")
  
}



```




### Figure 4E
```{r Figure_4E}


# Prepare dataset for heatmap
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"


# signature from Azizi et al 2018, Single-cell map of diverse immune phenotypes in breast tumor microenvironment (https://doi.org/10.1016/j.cell.2018.05.060)

# note signature published in manuscript has FAM46C, this is also called TENT5C in newest HGNC symbol annotation - signature has been adjusted to reflect this
Type1.signature <- c("IRF1", "IFIH1", "IFITM3", "DDX58", "IFI44L", "IFI6", "IFITM2", "NAMPT", "OASL",
                     "RTP4", "TREX1", "ADAR", "TENT5C", "LY6E", "MCOLN2", "APOBEC3G", "IL15", "ISG15", "MX1", "TLR3")

# Manuscript signature is published with C-MYC, P21, PD-L1, these have been adjusted to human HNGC symbols, MYC, CDKN1A, and CD274 respectively
# Additionally, signature simply stated MHC1 and MHC2, these are collected using  grex expression "HLA.*"
Type2.signature <- c("IFNG", "CXCL9", "CXCL10", "STAT1", "CD274", "CDKN1A", "MYC", "SMAD7", "IRF1")
HLA.genes <- grep("^HLA.*", rownames(CD8.seurat), value = T)
Type2.signature <- c(Type2.signature, HLA.genes)


Signature.list <- list(Type1.sig = Type1.signature, 
                       Type2.sig = Type2.signature)


if(long.compute){

 cores <- parallel::detectCores(all.tests = FALSE, logical = TRUE)

# Calculate Ucell scores for signatures
 CD8.seurat <- UCell::AddModuleScore_UCell(CD8.seurat,
                                           features = Signature.list,
                                           ncores = cores)


}

########################
# plot signature scores 
########################

# Type 1 Signature
VlnPlot(CD8.seurat, 
        features = "Type1.sig_UCell",
        cols = CD8.cols,
        pt.size = 0) + 
  NoLegend()

dev.copy(pdf, paste0(output.dir, "Figure_4E_Type1_signature.pdf"))
dev.off()


# Type 2 Signature
VlnPlot(CD8.seurat, 
        features = "Type2.sig_UCell",
        cols = CD8.cols,
        pt.size = 0) + 
  NoLegend()

dev.copy(pdf, paste0(output.dir, "Figure_4E_Type2_signature.pdf"))
dev.off()

```






### TEMP
```{r temp}



# Create output directory
output.dir <- "results/Manuscript/temp/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = TRUE)
}




exhaustion.sig <- c("CD96", "PDCD1", "LAG3", "HAVCR2", "TIGIT", "CTLA4", "CD244", "CD160", "TOX", "TCF7", "ENTPD1", "ITGAE", "TNFRSF9", "TNFRSF18")
# TNFRSF9 = CD137 = 4-1BB
#TNFRSF18 = GITR
effector.sig <- c("GZMA", "GZMB", "NKG7", "PRF1", "IFNG", "XCL1", "SP140")

# SP140


DoHeatmap(temp.seurat, 
          features = exhaustion.sig)



Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
average.seurat <- AverageExpression(CD8.seurat, 
                                    return.seurat = T)


exhaustion.sig <- c("CD96", "PDCD1", "LAG3", "HAVCR2", "TIGIT", "CTLA4", "CD244", "CD160", "TOX", "TCF7", "ENTPD1", "ITGAE", "TNFRSF9", "TNFRSF18", "BTLA", "IL10", "IL27")
effector.sig <- c("GZMA", "GZMB", "NKG7", "PRF1", "IFNG", "XCL1", "SP140")

DoHeatmap(average.seurat, 
          features = effector.sig, 
          draw.lines = F)

VlnPlot(CD8.seurat, feature = "LAG3", pt.size = 0)


```


