---
title: "06_Manuscript_Figure_5"
author: "Dillon Corvino"
date: "07/03/2022"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

### Environment

```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Create output directories & load custom functions & colour scheme
source("scripts/Setup.R", local = knitr::knit_global())

# pipeline variables
long.compute <- FALSE

# load required packages
require(tidyverse)
require(Seurat)
require(useful)
require(UCell)
require(Nebulosa)
library(harmony)

#source("scripts/aimed_analysis/functions.R")


# Create output directory
output.dir <- "results/Manuscript/Figure_5/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = TRUE)
}
```


### Define signature
```{r define_signature}

# Hard code the signature to prevent any alternations with statistical drift 
Sig.10 <- c("IFI6", "ISG15", "IFIT3", "MX1", "ISG20", "IFITM1", "LY6E", "IFIT1", "MX2", "OAS1")

Sig.10.up <- paste0(Sig.10, "+")

Signature.list <- list(Top_10 = Sig.10.up)

```

### Boxplot ggplot theme and function
```{r boxplot_theme_and_function}

boxplot_theme <- function() {
  theme_pubr() +
    theme(
      axis.line = element_line(linewidth = 0.5),
      axis.title.y = element_text(margin = margin(r = 0)),
      axis.text.y = element_text(size = 10, color = c("black", NA)),
      axis.ticks.y = element_line(linewidth = 0.5, color = "black", linetype = "solid"),
      axis.text.x = element_text(angle = 90, hjust = 1),
      plot.margin = margin(1, 5, 1, 5, unit = "cm"),
      plot.title = element_text(hjust = 0.5, size = 30)
    )
}


# Function to generate paired analysis plot
custom.boxplot <- function(data,
                           x_var = "disease",
                           x_lab = "Disease",
                           y_lab = "%", 
                           y_scale_limit = 100,
                           y_scale_breaks = seq(0, 100, by = 10), 
                           paired = TRUE, 
                           pairing.var = "donor_id",
                           title = "Temp_title", 
                           do.stats = FALSE, 
                           return.gg = FALSE) {
  
  library(ggplot2)
  library(lemon)
  library(ggpubr)
  
  # if paired analysis then filter to only keep paired values
  if(paired){
    data <- data %>%
      group_by(.data[[pairing.var]]) %>%
      filter(n() > 1) %>%
      ungroup()
  }
  
  
  # create jitter function 
  pj <- position_jitterdodge(jitter.width = 0.1, 
                             seed = 9,
                             jitter.height = 0,
                             dodge.width = 0)
  
  # Get the n value for each variable
  legend_data <- data %>%
    group_by(.data[[x_var]]) %>%
    summarise(n = n_distinct(.data[[pairing.var]]))
  
  
  # create base plot
  p1 <- ggplot(data, aes_string(x = x_var, y = "prop", color = x_var, fill = x_var)) +
    geom_boxplot(alpha = 0.3, outlier.shape = NA, width = 0.75) +
    labs(x = x_lab, y = y_lab) +
    boxplot_theme() + 
    scale_y_continuous(expand = expansion(add = c(0, 0.01)), limits = c(0, y_scale_limit), breaks = y_scale_breaks) +
    ggtitle(title) + 
    scale_colour_manual(values = seq(1:length(levels(data[[x_var]]))), 
                        labels = paste(levels(data[[x_var]]), " (n=", legend_data$n, ")", sep = ""),
                        aesthetics = c("colour", "fill"))
  
  if(paired){
    
    p1 <- p1 + lemon::geom_pointpath(aes_string(colour = x_var, group = pairing.var),
                                     position = pj, 
                                     linecolor = "gray", 
                                     linesize = 0.5)
    
  }else{
    
    p1 <- p1 +
      geom_point(position = pj, shape = 19, size = 2) 
    
  }
  
  
  if(do.stats){
    p1 <- p1 + ggpubr::stat_compare_means(method = "t.test", 
                                          paired = paired,
                                          vjust = 5)
  }
  
  print(p1)
  
  if(return.gg){
    return(p1)
  }
  
}

temp.function <- function(data, pairing.var = "donor_id", x_var = "disease"){
  
  # Preprocess the data to get the n value for each variable
  
  
  return(legend_data)
}

```


## Figure 5 {.tabset}

### Prepare dataset / read and filter for CD8 / or load dataset

```{r read_dataset_pancancer}

if(long.compute){
  
  # Due to memory limits, the dataset is written in as pancancer.cd8 rather than subsetting for cd8 and creating a new variable. 
  #i.e use only one variable to limit the memory requirments 
  pancancer.cd8 <- LoadH5Seurat("data/external_datasets/Utility_dataset.h5Seurat")
  
  Idents(pancancer.cd8) <- pancancer.cd8@meta.data$functional.cluster
  pancancer.cd8 <- subset(pancancer.cd8, idents = c("CD8_EffectorMemory", "CD8_NaiveLike", "CD8_EarlyActiv", "CD8_Tex", "CD8_Tpex"))
  
  # still many cells of mixed annotation by consensus.major ID
  Idents(pancancer.cd8) <- pancancer.cd8@meta.data$consensus.major
  pancancer.cd8 <- subset(pancancer.cd8, idents = "T_cell")
  
  # further filtering for CD8 T cell annotation
  Idents(pancancer.cd8) <- pancancer.cd8@meta.data$consensus.Tcell
  pancancer.cd8 <- subset(pancancer.cd8, idents = "CD8")
  
  # further filtering for annotation
  Idents(pancancer.cd8) <- pancancer.cd8@meta.data$HPCA.pruned.labels
  keep.var <- grep(".*CD8.*", unique(Idents(pancancer.cd8)), value =T)
  
  pancancer.cd8 <- subset(pancancer.cd8, idents = keep.var)
  
  # Type annotation
  table(pancancer.cd8@meta.data$Type)/1000
  # Juxta = 262 cells only
  # Blood = 12.993K
  # LN = 23.25K
  # Met = 5.711K
  # Normal = 41.119K
  # Tumor = 173.74K
  
  # Therefore remove Juxta cells 
  # also remove Met and LN as these are a bit complicated to include in analysise.
  # i.e the biological interpretation of Met tissue is highly dependent on each manuscript and what they mean by met and if LN is healthy or diseased
  # also remove blood to keep it simple, normal and tumor tissue
  
  
  Idents(pancancer.cd8) <- pancancer.cd8@meta.data$Type
  pancancer.cd8 <- subset(pancancer.cd8, idents = c("Tumor", "Normal"))
  
  
  table(pancancer.cd8@meta.data$Type, pancancer.cd8@meta.data$Tissue)
  
  
  library(harmony)
  # Normalise and harmonize dataset and calculate umap projection
  pancancer.cd8 <- 
    pancancer.cd8 %>%
    NormalizeData() %>%
    FindVariableFeatures() %>%
    ScaleData() %>%
    RunPCA() %>%
    harmony::RunHarmony(reduction = "pca",
                        group.by.vars = c("Cohort"), 
                        plot_convergence = TRUE, 
                        verbose = TRUE) %>%
    RunUMAP(reduction = "harmony", 
            dims = 1:20) %>%
    FindNeighbors(reduction = "harmony",
                  dims = 1:20) %>%
    FindClusters(resolution = 0.5)
  
  
  # save filtered dataset
  SaveH5Seurat(pancancer.cd8, 
               "data/external_datasets/pancancer_cd8_normed.h5Seurat",
               overwrite = TRUE)
  
  
}else{
  
  pancancer.cd8 <- LoadH5Seurat("data/external_datasets/pancancer_cd8_normed.h5Seurat")
  
}

# quick visualisation of the normed and subseted CD8 dataset

UMAPPlot(pancancer.cd8, 
         group.by = "HPCA.pruned.labels")

UMAPPlot(pancancer.cd8,
         pt.size = 1,
         group.by = "seurat_clusters")



```

### apply ucell score to pancancer

```{r ucell_pancancer}

if(!exists("pancancer.cd8")){
  
  pancancer.cd8 <- LoadH5Seurat("data/external_datasets/pancancer_cd8_normed.h5Seurat")
  
}

if(long.compute){
  
  cores <- parallel::detectCores(all.tests = FALSE, logical = TRUE)
  
  # Calculate Ucell scores for signatures
  DefaultAssay(pancancer.cd8) <- "RNA"
  pancancer.cd8 <- UCell::AddModuleScore_UCell(pancancer.cd8,
                                               features = Signature.list,
                                               ncores = cores)
  
  # Export object with UCell score
  SaveH5Seurat(pancancer.cd8, 
               "data/external_datasets/pancancer_cd8_normed.h5Seurat", 
               overwrite = TRUE)
  
}

```


### Figure 5A
```{r Figure_5A}

colnames(pancancer.cd8@meta.data)

# Visualise the results

# Density plots 
Plot_Density_Custom(pancancer.cd8, 
                    features =  "Top_10_UCell",
                    custom_palette = batlow.pal,
                    joint = FALSE, 
                    pt.size = 1, 
                    reduction = "umap")

dev.copy(pdf, paste0(output.dir, "Figure_5A.pdf"))
dev.off()


# subset dataset to enable figure creation as to many points 
Idents(pancancer.cd8) <- pancancer.cd8@meta.data$consensus.major
# 214,863 cells total dataset 
unique(Idents(pancancer.cd8))

temp.seurat <- subset(pancancer.cd8, downsample = 50000)

temp.seurat # 50,000 cells

# Density plots 
Plot_Density_Custom(temp.seurat, 
                    features =  "Top_10_UCell",
                    custom_palette = batlow.pal,
                    joint = FALSE, 
                    pt.size = 1, 
                    reduction = "umap")

dev.copy(pdf, paste0(output.dir, "Figure_5A_downsampled.pdf"))
dev.off()

rm(temp.seurat)



# Cluster distribution

# VlnPlots
VlnPlot(pancancer.cd8,
        pt.size = 0, 
        same.y.lims = T,
        features = "Top_10_UCell",
        group.by = "seurat_clusters") + 
  NoLegend()

#dev.copy(pdf, paste0(output.dir, "Pancancer_VlnPlot_UCell_scores_top10.pdf"))
#dev.off()



# Specifically colour IFN cluster
Idents(pancancer.cd8) <- pancancer.cd8@meta.data$seurat_clusters

scCustomize::Cluster_Highlight_Plot(pancancer.cd8, 
                                    cluster_name = "7", 
                                    highlight_color = "Red", 
                                    background_color = "lightgray", 
                                    pt.size = 0.1, 
                                    raster = TRUE)

#dev.copy(pdf, paste0(output.dir, "Pancancer_UMAP_clusters_highlighted.pdf"))
#dev.off()


# Cluster number 7 = IFN cluster


```


### Figure 5B-E
```{r Figure_5B_E}

################################################################################
# Read in dataset and export metadata or read in already exported metadata
################################################################################


# output.dir
output.dir <- "data/external_datasets/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = TRUE)
}


if(long.compute){
  
  # load pancancer 
  pancancer.cd8 <- LoadH5Seurat("data/external_datasets/pancancer_cd8_normed.h5Seurat")
  
  # extract metadata
  temp.df <- pancancer.cd8@meta.data
  saveRDS(temp.df, file = paste0(output.dir, "pancancer_cd8_metadata.rds"))
  
}else{
  
  # load metadata
  pancancer.cd8.metadata <- readRDS(file = paste0(output.dir, "pancancer_cd8_metadata.rds"))
  
}

##################################################
# Set output dir for figures to be created 
##################################################

# output.dir
output.dir <- "results/Manuscript/Figure_5/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = TRUE)
}


##############################
# Filter pancancer dataset
##############################

# vis signature score and verify cluster identity of ISG population 
# VlnPlot(pancancer.cd8,
#         "Top_10_UCell",
#         group.by = "seurat_clusters", 
#         pt.size = 0)

# cluster 7 = ISG cluster 

# NOTEs: 
# SampleID = the identity of each sample, for example patient 1 might have had tils and healthy tissue taken, these are therefore two different SampleIDs
# Whereas, Sample = patient identifier 
# use Sample as donor_id
x <- unique(pancancer.cd8.metadata$Sample) 
x
length(x)  # 67 donors 

# also note: 
# Can use HPCA.labels but not HPCA.first.labels 
# using HPCA.first.labels some cells are annotated as CD4, B cell etc etc. 
# this is despite dataset being filtered on consensus.major and consensus.Tcell for CD8 T annotations 

# note:
# some patients have low cell numbers of less than 300
# therefore should remove this to limit analysis variation
# i.e "Patient 64" only has 53 cells 

# dataset contains normal and tumor 
# also there are 10 tumor types

###################
# filter dataset 
###################

# filter for CD8 annotation - this is unnecessary as was already done as part of constructing object
# however, this is done to ensure no errors if upstream is changed and also for clarity in understanding how the object is filtered 
# remove donors with low cell counts 
# rename Sample to donor_id
# filter dataset to remove unnecessary columns and data 
# identify cluster 7 cells as ISG cells

cols.keep <- c("rowname.temp", "donor_id",
               "GEO_RNA", "Cohort", 
               "Type", "Tissue",
               "HPCA.labels", "functional.cluster",
               "functional.cluster.conf", "consensus.major", 
               "consensus.Tcell", "CTgene", 
               "CTaa", "cloneType", "seurat_clusters")

# for some reason rownames to column was not working, kept losing string and becoming numerical 
# therefore explicitly do this outside dplyr chain

pancancer.cd8.metadata.filt <- pancancer.cd8.metadata
pancancer.cd8.metadata.filt$rowname.temp <- rownames(pancancer.cd8.metadata.filt)

pancancer.cd8.metadata.filt <- pancancer.cd8.metadata.filt %>% 
  dplyr::filter(consensus.major == "T_cell") %>%
  dplyr::filter(consensus.Tcell == "CD8") %>%
  dplyr::rename(donor_id = Sample) %>%
  dplyr::group_by(donor_id) %>%
  dplyr::mutate(count = n()) %>%
  dplyr::ungroup() %>%
  dplyr::filter(count >= 300) %>%
  dplyr::select(all_of(cols.keep), ends_with("_UCell")) %>%
  tibble::column_to_rownames(var = "rowname.temp") %>%
  dplyr::mutate(seurat_clusters = case_when(seurat_clusters == "7" ~ "ISG_cluster", 
                                            TRUE ~ "Non_ISG")) %>%
  dplyr::rename(ISG_status = seurat_clusters)


#####################################################
# Calculate ISG frequency per donor/tissue/disease
#####################################################


pancancer.cd8.metadata.filt[1:10, ]

disease.freq.table <- pancancer.cd8.metadata.filt %>%
  group_by(donor_id, Type, Tissue, ISG_status) %>%
  summarise(count = n()) %>%
  ungroup() %>%
  group_by(donor_id, Type, Tissue) %>%
  mutate(prop = count / sum(count)) 

# ensure no donors are lost by having no cells in ISG cluster 
additional.values <- disease.freq.table %>% 
  filter(ISG_status == "Non_ISG" & prop == 1) %>% 
  mutate(ISG_status = "ISG_cluster",
         prop = 0,
         count = 0)

# add data together and filter for just ISG values
if(dim(additional.values)[1] != 0 && dim(additional.values)[2] != 0) {
  
  plot.data.cancer <- disease.freq.table %>%
    bind_rows(additional.values) %>%
    filter(ISG_status == "ISG_cluster") %>% 
    mutate(prop = prop * 100) # adjust to 0-100% scale
  
}else{
  plot.data.cancer <- disease.freq.table %>%
    filter(ISG_status == "ISG_cluster") %>% 
    mutate(prop = prop * 100) # adjust to 0-100% scale
}

###############################
# format data for ploting
###############################

# rename columns, remove unnecessary data, and sort in descending order
plot.data.cancer <- plot.data.cancer %>%
  dplyr::ungroup() %>%
  dplyr::rename(disease = Type) %>%
  dplyr::mutate(disease = paste(disease, Tissue, sep = "_")) %>%
  dplyr::select(donor_id, disease, ISG_status, prop) %>%
  dplyr::group_by(disease) %>%
  dplyr::mutate(temp.var = median(prop)) %>%
  dplyr::arrange(desc(temp.var)) %>%
  dplyr::select(!temp.var)



#################################
# Figure 5B - Normal vs Tumor 
#################################

temp.plot <- plot.data.cancer %>%
  dplyr::mutate(disease = case_when(disease = grepl("Tumor.*", disease) ~ "Tumor", 
                                    disease = grepl("Normal.*", disease) ~ "Normal",
                                    TRUE ~ "YO"))



# Convert disease to factor with ordered levels
temp.plot$disease <- factor(temp.plot$disease,
                            levels = rev(unique(temp.plot$disease)))

p1 <- custom.boxplot(data = temp.plot,
                     x_var = "disease",
                     y_scale_limit = 100,
                     y_scale_breaks = seq(0, 100, by = 1), 
                     paired = F, 
                     pairing.var = "donor_id",
                     x_lab = "Disease",
                     y_lab = "%", 
                     title = "Freq of ISG population across disease status for all tissues", 
                     do.stats = TRUE, 
                     return.gg = TRUE)

library(ggforce)

p1 + facet_zoom(ylim = c(0,10), zoom.data = ifelse(prop <= 10, NA, FALSE))


# save plot
ggsave(filename = paste0(output.dir, "Figure_5B.pdf"),
       width = 25, height = 15, units = "in", dpi = 300,
       device = "pdf")


################################################################
# Figure 5C - ISG frequency across tumor types
################################################################

# filter for just the tumor data
tumor.plot <- plot.data.cancer %>%
  dplyr::filter(grepl("^Tumor_.*", disease))

# Convert disease to factor with ordered levels
tumor.plot$disease <- factor(tumor.plot$disease, 
                             levels = unique(tumor.plot$disease))


custom.boxplot(data = tumor.plot,
               x_var = "disease",
               y_scale_limit = 80,
               y_scale_breaks = seq(0, 100, by = 5), 
               paired = F, 
               pairing.var = "donor_id",
               x_lab = "Disease",
               y_lab = "%", 
               title = "Freq of ISG population across tumor types", 
               do.stats = F)

# save plot
ggsave(filename = paste0(output.dir, "Figure_5C.pdf"),
       width = 25, height = 15, units = "in", dpi = 300,
       device = "pdf")



############################################################
# Figure 5D - Normal vs tumor for selected tumor types
############################################################

###############
# Esophagus 
###############

temp.plot <- plot.data.cancer %>%
  dplyr::filter(grepl(".*_Esophagus", disease))

# Convert disease to factor with ordered levels
temp.plot$disease <- factor(temp.plot$disease,
                            levels = rev(unique(temp.plot$disease)))

custom.boxplot(data = temp.plot,
               x_var = "disease",
               y_scale_limit = 70,
               y_scale_breaks = seq(0, 100, by = 5), 
               paired = TRUE, 
               pairing.var = "donor_id",
               x_lab = "Disease",
               y_lab = "%", 
               title = "Freq of ISG population across disease status for Esophagus tissue", 
               do.stats = TRUE)

# save plot
ggsave(filename = paste0(output.dir, "Figure_5D_Esophagus.pdf"),
       width = 25, height = 15, units = "in", dpi = 300,
       device = "pdf")

###############
# Endometrial 
###############

temp.plot <- plot.data.cancer %>%
  dplyr::filter(grepl(".*_Endometrial", disease))

# Convert disease to factor with ordered levels
temp.plot$disease <- factor(temp.plot$disease,
                            levels = rev(unique(temp.plot$disease)))

custom.boxplot(data = temp.plot,
               x_var = "disease",
               y_scale_limit = 10,
               y_scale_breaks = seq(0, 10, by = 1), 
               paired = TRUE, 
               pairing.var = "donor_id",
               x_lab = "Disease",
               y_lab = "%", 
               title = "Freq of ISG population across disease status for Endometrial tissue", 
               do.stats = TRUE)

# save plot
ggsave(filename = paste0(output.dir, "Figure_5D_Endometrial.pdf"),
       width = 25, height = 15, units = "in", dpi = 300,
       device = "pdf")


###############
# Lung 
###############

temp.plot <- plot.data.cancer %>%
  dplyr::filter(grepl(".*_Lung", disease))

# Convert disease to factor with ordered levels
temp.plot$disease <- factor(temp.plot$disease,
                            levels = rev(unique(temp.plot$disease)))

custom.boxplot(data = temp.plot,
               x_var = "disease",
               y_scale_limit = 20,
               y_scale_breaks = seq(0, 20, by = 2), 
               paired = TRUE, 
               pairing.var = "donor_id",
               x_lab = "Disease",
               y_lab = "%", 
               title = "Freq of ISG population across disease status for Lung tissue", 
               do.stats = TRUE)

# save plot
ggsave(filename = paste0(output.dir, "Figure_5D_Lung.pdf"),
       width = 25, height = 15, units = "in", dpi = 300,
       device = "pdf")


##########################################################################
# Figure 5E - ISG frequency across CD8 annotations
##########################################################################

# I.e ask the question, within ISG cells what frequency are effector, naive, etc 
# do this in a donor specific manner
# therefore, for each donor, the sum of frequencies within ISG cells should = 100%, same for non_isg cells

pancancer.cd8.metadata.filt[1:10, ]

# do this analysis just within tumor samples

disease.freq.table <- pancancer.cd8.metadata.filt %>%
  dplyr::filter(Type == "Tumor") %>%
  group_by(donor_id, ISG_status, functional.cluster) %>%
  summarise(count = n())

# ensure factorisation is removed as it contains levels no longer present
disease.freq.table$donor_id <- as.character(disease.freq.table$donor_id)
disease.freq.table$functional.cluster <- as.character(disease.freq.table$functional.cluster)

# Create all possible combinations of donor_id, functional.cluster, and ISG_status
all_combinations <- crossing(donor_id = unique(disease.freq.table$donor_id),
                             functional.cluster = unique(disease.freq.table$functional.cluster),
                             ISG_status = unique(disease.freq.table$ISG_status))

# Merge with the summarized data
plot_data <- merge(all_combinations, disease.freq.table, all.x = TRUE, by = c("donor_id", "functional.cluster", "ISG_status"))

# Replace NA with 0 for count
plot_data$count[is.na(plot_data$count)] <- 0

plot_data <- plot_data %>%
  ungroup() %>%
  group_by(donor_id, ISG_status) %>%
  mutate(prop = count / sum(count))

plot_data$prop <- plot_data$prop*100

# verify that sum = 100% 
plot_data %>%
  dplyr::filter(donor_id == "Patient6") %>%
  dplyr::filter(ISG_status == "ISG_cluster") %>%
  mutate(temp = sum(prop))


# get n value 
legend_data <- plot_data %>%
  group_by(ISG_status, functional.cluster) %>%
  distinct(donor_id) %>%
  summarise(count_2 = n())

legend_data

###############################
# format data for ploting
###############################

# rename columns, remove unnecessary data, and sort in descending order
plot_data <- plot_data %>%
  dplyr::ungroup() %>%
  dplyr::rename(ClusterID = functional.cluster)

# Convert ClusterID to factor with ordered levels
plot_data$ClusterID <- factor(plot_data$ClusterID,
                              levels = c("CD8_NaiveLike", "CD8_EarlyActiv", "CD8_EffectorMemory", "CD8_Tex", "CD8_Tpex"))

# Convert ISG status to factor with ordered levels
plot_data$ISG_status <- factor(plot_data$ISG_status,
                               levels = c("Non_ISG", "ISG_cluster"))



# create jitter function 
pj <- position_jitterdodge(jitter.width = 0.25, 
                           seed = 9,
                           jitter.height = 0,
                           dodge.width = 0.8)


ggplot(plot_data, aes(x = ClusterID, y = prop, fill = ISG_status, color = ISG_status)) +
  geom_boxplot(alpha = 0.3, outlier.shape = NA, width = 0.75, position = position_dodge(width = 0.8)) +
  geom_point(shape = 19, size = 2, position = pj) +
  labs(x = "Functional Cluster", y = "Frequency", title = "Boxplot of Proportion by Functional Cluster and ISG Status") +
  scale_y_continuous(expand = expansion(add = c(0, 0.01)), limits = c(0, 100), breaks = seq(0, 100, by = 10)) +
  boxplot_theme() +
  stat_compare_means(method = "t.test", 
                     label.y = 80)

# save plot
ggsave(filename = paste0(output.dir, "Figure_5E.pdf"),
       width = 25, height = 15, units = "in", dpi = 300,
       device = "pdf")

```


## Figure 5F

### Prepare Cillo dataset

```{r Cillo_dataset_info}

# Manuscript title: "Immune Landscape of Viral- and Carcinogen-Driven Head and Neck Cancer"
# Authors: Anthony R. Cillo, Cornelius H.L. Ku€rten, Tracy Tabib, ..., Tullia C. Bruno, Robert L. Ferris, Dario A.A. Vignali
# Journal: Immunity
# Year: 2020
# DOI: 10.1016/j.immuni.2019.11.014
## Abstract: Head and neck squamous cell carcinoma (HNSCC) arises through exposure to environmental carcino- gens or malignant transformation by human papillomavirus (HPV). Here, we assessed the transcriptional profiles of 131,224 single cells from peripheral and intra-tumoral immune populations from patients with HPV– and HPV+ HNSCC and healthy donors. Immune cells within tumors of HPV and HPV HNSCC displayed a spectrum of transcriptional signatures, with helper CD4+ T cells and B cells being relatively divergent and CD8+ T cells and CD4+ regulatory T cells being relatively similar. Transcriptional results were contextualized through multispectral immuno- fluorescence analyses and evaluating putative cell- cell communication based on spatial proximity. These analyses defined a gene expression signature associated with CD4+ T follicular helper cells that is associated with longer progression-free survival in HNSCC patients. The datasets and analytical ap- proaches herein provide a resource for the further study of the impact of immune cells on viral- and carcinogen-induced cancers.


# Total of 63 samples were analyzed by scRNAseq. 
# 26 HNSCC patients (18 HPVNeg and 8 HPVPos)
# HNSCC patients have paired PBMC and TIL samples
# 6 PBMC from healthy donors
# 5 healthy donor tonsils 


# data was processed using code found in repo https://github.com/BaldLab/Cillo_et_al_HNSCC_dataset
# In brief, data was subsetted for TIL samples only 
# subsequently data was exported or further subsetted for CD8+ T-cells 
# resulting in two RDS files, one containing all data and one containing only CD8+ T cells

```

### Use Cillo CD8 dataset as ISG score validation dataset
```{r validate_ISG_scoring_Cillo_CD8}

if(long.compute){
  ##############################################################################################################
  # Use Cillo CD8 dataset as a validation tool to confirm ISG signature captures the authors ISG population
  ##############################################################################################################
  
  library(SeuratDisk)
  
  # Load data
  Cillo.CD8.seurat <- LoadH5Seurat("data/external_datasets/Cillo_CD8_seurat.h5Seurat")
  
  
  cores <- parallel::detectCores(all.tests = FALSE, logical = TRUE)
  
  # Calculate Ucell scores for signatures
  DefaultAssay(Cillo.CD8.seurat) <- "RNA"
  Cillo.CD8.seurat <- UCell::AddModuleScore_UCell(Cillo.CD8.seurat,
                                                  features = Signature.list,
                                                  ncores = cores)
  
  
  
  # VlnPlots
  VlnPlot(Cillo.CD8.seurat,
          pt.size = 0, 
          same.y.lims = T,
          features = "Top_10_UCell",
          group.by = "cd8_cells_clusterID") + 
    NoLegend()
  
  
  #dev.copy(pdf, paste0(output.dir, "Sup_Figure_3X.pdf"))
  #dev.off()
}
```

### Figure 5F - cell-cell communcation using Cillo dataset
```{r Figure_5F}

################################################
# Read in and format complete Cillo dataset
################################################

# Load data
Cillo.seurat <- LoadH5Seurat("data/external_datasets/Cillo_seurat.h5Seurat")

# formatting 
# having Na values in the cd8 cluster id causes problems downstream, 
# also note that some tconv and tregs and NK cells also have a cd8 cluster id value, this is quite messed up annotation from Cillo 
# therefore, create a simplified annotation

# for only cells annotated as CD8, append the cd8 cluster ID. some cells dont have this, in which case call them misc, and straight away convert cluster 2 into ISG annotation

# also have an annotation where only ISG cd8 cells are specifically annotated and all other CD8s are combined 

Cillo.seurat@meta.data <- Cillo.seurat@meta.data %>%
  mutate(complete_annotation = case_when(
    cell_type == "cd8.cells" & is.na(cd8_cells_clusterID) ~ "cd8.cells_misc",
    cell_type == "cd8.cells" & cd8_cells_clusterID == "2" ~ "CD8_ISG_cells",
    cell_type == "cd8.cells" & !is.na(cd8_cells_clusterID) ~ paste(cell_type, cd8_cells_clusterID, sep = "_"),
    TRUE ~ cell_type
  )) %>%
  mutate(simplified_annotation = case_when(grepl("cd8.cells_.*", complete_annotation) ~ "cd8.cells", 
                                           TRUE ~ complete_annotation))



unique(Cillo.seurat@meta.data$simplified_annotation)
unique(Cillo.seurat@meta.data$complete_annotation)


# plot overview of dataset
UMAPPlot(Cillo.seurat, group.by = "complete_annotation", pt.size = 1, label = T)

UMAPPlot(Cillo.seurat, group.by = "simplified_annotation", pt.size = 1, label = T)


################################################
# Cell-Cell communcation using Liana
################################################

# Install and load packages required
#remotes::install_github('saezlab/liana')
library(liana)
library(tidyverse)
library(magrittr)

if(!require("circlize")){
  install.packages("circlize", quiet = TRUE,
                   repos = "http://cran.us.r-project.org")
}


#############################################
# calculate communcation interactions 
#############################################

if(long.compute){
  # Resource currently included in OmniPathR (and hence `liana`) include:
  show_resources()
  
  
  # Resource currently included in OmniPathR (and hence `liana`) include:
  show_methods()
  
  Cillo.seurat %>% dplyr::glimpse()
  
  # set idents
  Idents(Cillo.seurat) <- Cillo.seurat@meta.data$complete_annotation
  DefaultAssay(Cillo.seurat) <- "SCT"
  
  
  cores <- parallel::detectCores(all.tests = FALSE, logical = TRUE)
  
  
  # Run liana
  liana_cillo <- liana_wrap(Cillo.seurat, 
                            method = c("natmi", "connectome", "logfc", "sca", "cellphonedb"),
                            resource = "Consensus",
                            parallelize = TRUE, 
                            workers = cores)
  
  saveRDS(liana_cillo, file = paste0(output.dir, "liana_cillo.rds"))
  
  
}

#############################################
# load computed data
#############################################

liana_cillo <- readRDS(paste0(output.dir, "liana_cillo.rds"))

# Liana returns a list of results, each element of which corresponds to a method
liana_cillo %>% dplyr::glimpse()

# We can aggregate these results into a tibble with consensus ranks
liana_cillo <- liana_cillo %>%
  liana_aggregate()

dplyr::glimpse(liana_cillo)

#(*) The aggregate consensus rank (aggregate_rank) is obtained using a re-implementation of the RRA method from the RobustRankAggreg package.
#RRA scores can be interpreted as p-values and interactions which are ranked consistently higher than random are assigned low scores/p-values.


# Filter dataset to remove other CD8 groups

liana_trunc <- liana_cillo %>%
  # only keep interactions concordant between methods
  filter(aggregate_rank <= 0.01) %>% # note that these pvals are already corrected 
  filter(!grepl("cd8.cells_.", source) & !grepl("cd8.cells_.", target) )


# Chord diagram of the top 20 interactions for each source

top_20_ints_source <- liana_trunc %>%
  group_by(source) %>%
  top_n(-20, aggregate_rank)

top_20_ints_source
chord_freq(top_20_ints_source)

dev.copy(pdf, paste0(output.dir, "Figure_5F_top_20_source.pdf"))
dev.off()

# Chord diagram of the top 20 interactions for each target

top_20_ints_target <- liana_trunc %>%
  group_by(target) %>%
  top_n(-20, aggregate_rank)

top_20_ints_target
chord_freq(top_20_ints_target)

dev.copy(pdf, paste0(output.dir, "Figure_5F_top_20_target.pdf"))
dev.off()


```




## Supplementary Figure 3 {.tabset}
### Set up output dir
```{r Supplementary_Figure_3_setup}

# Create output directory
output.dir <- "results/Manuscript/Supplementary_Figure_3/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = TRUE)
}


```

### Supplementary Figure 3A - ISG score VlnPlot Ovarian and Esophagus - Pancancer CD8 dataset
```{r Sup_Figure_3A}

if(long.compute){
  # Subset for ISG cluster enriched tumors
  Idents(pancancer.cd8) <- pancancer.cd8@meta.data$Tissue
  table(pancancer.cd8@meta.data$Tissue, pancancer.cd8@meta.data$Type)
  
  tissue.vec <- c("Ovarian", "Esophagus")
  
  ISG.tumors.seurat <- subset(pancancer.cd8, idents = tissue.vec)
  
  # take just cells from tumor tissue
  Idents(ISG.tumors.seurat) <- ISG.tumors.seurat@meta.data$Type
  
  ISG.tumors.seurat <- subset(ISG.tumors.seurat, idents = "Tumor")
  
  # check subsetting correctly performed 
  table(ISG.tumors.seurat@meta.data$Tissue, ISG.tumors.seurat@meta.data$Type)
  
  
  library(harmony)
  # Normalise and harmonize dataset and calculate umap projection
  
  # store upper level seurat clusters incase needed later
  ISG.tumors.seurat$orig.clusters <- ISG.tumors.seurat@meta.data$seurat_clusters
  
  
  ISG.tumors.seurat <- 
    ISG.tumors.seurat %>%
    Seurat::NormalizeData() %>%
    Seurat::FindVariableFeatures() %>%
    Seurat::ScaleData() %>%
    Seurat::RunPCA() %>%
    harmony::RunHarmony(reduction = "pca",
                        group.by.vars = c("Cohort", "Sample"), 
                        plot_convergence = TRUE, 
                        verbose = TRUE)  %>%
    Seurat::RunUMAP(reduction = "harmony", 
                    dims = 1:20, 
                    min.dist = 0.3, 
                    spread = 1,
                    n.neighbors = 5, # 5 - 50 , larger = more global
                    densmap = F,
                    seed.use = 42)  %>%
    FindNeighbors(reduction = "harmony",
                  dims = 1:20) %>%
    FindClusters(resolution = 0.5)
  
  
  # Rename clusters
  ISG.tumors.seurat@meta.data$IFN_clusters <- ISG.tumors.seurat@meta.data$seurat_clusters
  
  Idents(ISG.tumors.seurat) <- ISG.tumors.seurat@meta.data$IFN_clusters
  ISG.tumors.seurat <- ISG.tumors.seurat %>%
    RenameIdents('0' = "other",
                 '1' = "other",
                 '2' = "other", 
                 '3' = "other",
                 '4' = "other", 
                 '5' = "ISG", 
                 '6' = "other", 
                 '7' = "other",
                 '8' = "other",
                 '9' = "other",
                 '10' = "other")
  Idents(ISG.tumors.seurat)
  
  ISG.tumors.seurat@meta.data$IFN_status <- Idents(ISG.tumors.seurat)
  
  # save normed and recalculated umap
  SaveH5Seurat(ISG.tumors.seurat, 
               "data/external_datasets/ISG_tumors_seurat.h5Seurat",
               overwrite = TRUE)
  
  
}else if(!exists("ISG.tumors.seurat")){
  
  ISG.tumors.seurat <- LoadH5Seurat("data/external_datasets/ISG_tumors_seurat.h5Seurat")
  
  
}


##################################################################################################
# Vln plot of Ucell scores for T1 and T2 IFN calculated on original dataset by Nick Borcherding
##################################################################################################

VlnPlot(ISG.tumors.seurat,
        feature = "T1_Interferon_UCell",
        pt.size = 0,
        split.by = "Tissue",
        group.by = "IFN_status")

dev.copy(pdf, paste0(output.dir, "Supplementary_Figure_3A_T1_Ucell.pdf"))
dev.off()

VlnPlot(ISG.tumors.seurat,
        feature = "T2_Interferon_UCell",
        pt.size = 0,
        split.by = "Tissue",
        group.by = "IFN_status")

dev.copy(pdf, paste0(output.dir, "Supplementary_Figure_3A_T2_Ucell.pdf"))
dev.off()

```



## COMBAT COVID dataset
### Subset for CD8 and score for ISG
```{r format_COMBAT_COVID_dataset}

if(long.compute){
  
  # load dataset
  covid.seurat <- readRDS("/Users/ieo/Documents/scRNAseq_datasets/COVID_datasets/COMBAT_Blood_Atlas_of_COVID_19/blood_atlas_covid.rds")
  
  # quick overview of dataset 
  covid.seurat # 37,412 genes and 836,148 cells
  
  # All data is: 
  # Human derived
  # PBMCs
  # three disease = normal, influenza, and COVID-19
  # male and female donors
  # 124 donors?
  
  
  covid.seurat %>% 
    glimpse()
  
  head(covid.seurat@meta.data)
  colnames(covid.seurat@meta.data)
  
  
  # info on disease vs source annotation
  table(covid.seurat@meta.data$Source, covid.seurat@meta.data$disease)
  
  # note major_subset and cell_type annotation is equiv 
  vars.plot <- c("disease", "sex", "cell_type", "major_subset", "minor_subset", "Source")
  
  for(i in seq_along(vars.plot)){
    print(
      UMAPPlot(covid.seurat,
               pt.size = 1,
               group.by = vars.plot[i], 
               label = T, 
               raster = T)
    )
    dev.copy(pdf, paste0(output.dir, "UMAP_", vars.plot[i], ".pdf"))
    dev.off()
    
  }
  
  
  # subset for CD8 T-cells
  unique(covid.seurat@meta.data$major_subset)
  
  Idents(covid.seurat) <- covid.seurat@meta.data$major_subset
  cd8.covid.seurat <- subset(covid.seurat, idents = "CD8")
  
  cd8.covid.seurat # 37,412 genes across 106,025 cells
  
  # remove large dataset 
  rm(covid.seurat)
  gc()
  
  # quick vis
  UMAPPlot(cd8.covid.seurat,
           pt.size = 1,
           group.by = "minor_subset", 
           label = T) + NoLegend()
  
  
  # three cells annotated as "nan" remove these cells
  table(cd8.covid.seurat@meta.data$minor_subset)
  Idents(cd8.covid.seurat) <- cd8.covid.seurat@meta.data$minor_subset
  
  cd8.covid.seurat <- subset(cd8.covid.seurat, idents = "nan", invert = TRUE)
  
  # renorm, and dim reduction reduced dataset
  cd8.covid.seurat <- 
    cd8.covid.seurat %>%
    NormalizeData() %>%
    FindVariableFeatures() %>%
    ScaleData() %>%
    RunPCA() %>%
    RunUMAP(reduction = "pca", 
            dims = 1:20, 
            min.dist = 0.5, 
            spread = 1,
            n.neighbors = 30, # 5 - 50 , larger = more global
            densmap = F,
            seed.use = 42)  %>%
    FindNeighbors(reduction = "pca",
                  dims = 1:20) %>%
    FindClusters(resolution = 0.3)
  
  
  ############################
  # score for ISG signature
  ############################
  
  cores <- parallel::detectCores(all.tests = FALSE, logical = TRUE)
  
  # replace sig list with ensemble IDs
  Signature.list[[1]] <- paste0(Sig.10.ensembl, "+")
  
  
  # Calculate Ucell scores for signatures
  DefaultAssay(cd8.covid.seurat) <- "RNA"
  cd8.covid.seurat <- UCell::AddModuleScore_UCell(cd8.covid.seurat,
                                                  features = Signature.list,
                                                  ncores = cores)
  
  
  
  
  # VlnPlots
  VlnPlot(cd8.covid.seurat,
          pt.size = 0, 
          same.y.lims = T,
          features = "Top_10_UCell",
          group.by = "seurat_clusters") + 
    NoLegend()
  
  dev.copy(pdf, paste0(output.dir, "Vlnplot_ISG_signature.pdf"))
  dev.off()
  
  
  
  # make new metadata to specifically isolate ISG cluster 
  # cluster 6 = ISG 
  
  cd8.covid.seurat@meta.data <- cd8.covid.seurat@meta.data %>%
    mutate(ISG_status = case_when(seurat_clusters == "6" ~ "ISG_cluster",
                                  TRUE ~ "Non_ISG"))
  
  
  # reorder disease factor levels for downstream plotting
  cd8.covid.seurat@meta.data$disease <- factor(cd8.covid.seurat@meta.data$disease, levels = c("normal", "COVID-19", "influenza"))
  
  
  SaveH5Seurat(cd8.covid.seurat,
               file = "data/external_datasets/COMBAT_COVID_CD8.h5Seurat")
  
}else if(quick.load){
  cd8.covid.seurat <- LoadH5Seurat("data/external_datasets/COMBAT_COVID_CD8.h5Seurat")
}


```

### Supplementary Figure 3B
```{r Sup_Figure_3B}


##############################
# Supplementary Figure 3B
##############################

library(scCustomize)

# Density plots 
Plot_Density_Custom(cd8.covid.seurat, 
                    features =  "Top_10_UCell",
                    custom_palette = batlow.pal,
                    joint = FALSE, 
                    pt.size = 1, 
                    reduction = "umap")

dev.copy(pdf, paste0(output.dir, "Supplementary_Figure_3B.pdf"))
dev.off()


```

### Supplementary Figure 3C-E
```{r Supplementary_Figure_3C_E}

#################################################
# Supplementary Figure 3C - ISG freq by disease
#################################################

plot_data <- cd8.covid.seurat@meta.data %>%
  group_by(donor_id, disease, ISG_status) %>%
  summarise(freq = n()) %>%
  ungroup() %>%
  group_by(donor_id, disease) %>%
  mutate(prop = freq / sum(freq)) %>%
  filter(ISG_status == "ISG_cluster")

plot_data$prop <- plot_data$prop*100


# Create a data frame for the t-test groups
t_test_groups <- data.frame(
  group1 = c("normal", "COVID-19"),
  group2 = c("COVID-19", "influenza"), 
  group3 = c("normal", "influenza")
)
# Get the n value for each variable
legend_data <- plot_data %>%
  group_by(disease) %>%
  summarise(n = n_distinct(donor_id))


# plot data
ggplot(plot_data, aes(x = disease, y = prop, color = disease, fill = disease)) +
  geom_boxplot(alpha = 0.3, position = position_dodge(width = 0.75), aes(color = disease, fill = disease), outlier.shape = NA, width = 0.75) +
  geom_point(aes(color = disease), position = position_jitter(width = 0.1), shape = 19, size = 2) + 
  labs(x = "disease", y = "Proportion of ISG_cluster") +
  boxplot_theme() + 
  scale_y_continuous(expand = expansion(add = c(0, 0.01)), limits = c(0, 100), breaks = seq(0, 100, by = 10)) + 
  stat_compare_means(comparisons = t_test_groups, 
                     method = "t.test", 
                     label = "p.signif",
                     hide.ns = FALSE,
                     label.y = c(80, 85, 90)) + # Perform t-test and add statistics
  ggtitle("Proportion of ISG_cluster by Disease Group") + 
  scale_colour_manual(values = seq(1:length(levels(plot_data$disease))), 
                      labels = paste(levels(plot_data$disease), " (n=", legend_data$n, ")", sep = ""),
                      aesthetics = c("colour", "fill"))


ggsave(filename = paste0(output.dir, "Supplementary_Figure_3C.pdf"),
       device = "pdf")


############################################################
# Supplementary Figure 3D - ISG Freq by disease status
############################################################

# note sepsis is annotated as COVID-19 but not sure if this is correct based on paper, looks like sepsis is maybe/maybenot covid attributed
# not sure what COVID_HCW_MILD or COVID_LDN are
# therefore, will remove these and keep only HV, and COVID annotations


plot_data <- cd8.covid.seurat@meta.data %>%
  filter(Source %in% c("HV", "COVID_MILD", "COVID_SEV", "COVID_CRIT", "Sepsis")) %>%
  group_by(donor_id, Source, ISG_status) %>%
  summarise(count = n()) %>%
  ungroup() %>%
  group_by(donor_id, Source) %>%
  mutate(prop = count / sum(count)) %>%
  filter(ISG_status == "ISG_cluster")

plot_data$prop <- plot_data$prop*100

unique(plot_data$Source)

# Create a data frame for the t-test groups
t_test_groups <- data.frame(
  group1 = c("HV", "COVID_MILD"),
  group2 = c("HV", "COVID_SEV"), 
  group2 = c("HV", "COVID_CRIT"), 
  group2 = c("HV", "Sepsis")
)

# Get the n value for each variable
legend_data <- plot_data %>%
  group_by(Source) %>%
  summarise(n = n_distinct(donor_id))

plot_data$Source <- as.character(plot_data$Source)
plot_data$Source <- factor(plot_data$Source, levels = c("HV", "COVID_MILD", "COVID_SEV", "COVID_CRIT", "Sepsis"))

ggplot(plot_data, aes(x = Source, y = prop, color = Source, fill = Source)) +
  geom_boxplot(alpha = 0.3, position = position_dodge(width = 0.75), aes(color = Source, fill = Source), outlier.shape = NA, width = 0.75) +
  geom_point(aes(color = Source), position = position_jitter(width = 0.1), shape = 19, size = 2) + 
  labs(x = "Source", y = "Proportion of ISG_cluster") +
  boxplot_theme() + 
  scale_y_continuous(expand = expansion(add = c(0, 0.01)), limits = c(0, 60), breaks = seq(0, 60, by = 5)) + 
  stat_compare_means(comparisons = t_test_groups, 
                     label = "p.signif",
                     method = "t.test", 
                     label.y = c(55, 45), 
                     hide.ns = TRUE) + 
  ggtitle("Proportion of ISG_cluster by Disease status") + 
  scale_colour_manual(values = seq(1:length(levels(plot_data$Source))), 
                      labels = paste(levels(plot_data$Source), " (n=", legend_data$n, ")", sep = ""),
                      aesthetics = c("colour", "fill"))


ggsave(filename = paste0(output.dir, "Supplementary_Figure_3D.pdf"),
       device = "pdf")





###########################################################
# Supplementary Figure 3E - ISG freq by CD8 annotation 
###########################################################

table(cd8.covid.seurat@meta.data$minor_subset, cd8.covid.seurat@meta.data$ISG_status)



library(dplyr)
library(tidyr)

plot_data <- cd8.covid.seurat@meta.data %>%
  filter(disease == "COVID-19") %>%
  group_by(donor_id, minor_subset, ISG_status) %>%
  summarise(count = n())


# note that some values are 0, as in within ISG or non_ISG classified cells, there are no counts associated with a particular subset annotation. 
# notably, no ISG cells are classified as naive. 
# therefore, with the above summarisation, these are lost and therefore need to be added back 

# ensure factorisation is removed as it contains levels no longer present
plot_data$donor_id <- as.character(plot_data$donor_id)
plot_data$minor_subset <- as.character(plot_data$minor_subset)

# Create all possible combinations of donor_id, minor_subset, and ISG_status
all_combinations <- crossing(donor_id = unique(plot_data$donor_id),
                             minor_subset = unique(plot_data$minor_subset),
                             ISG_status = unique(plot_data$ISG_status))

# Merge with the summarized data
plot_data <- merge(all_combinations, plot_data, all.x = TRUE, by = c("donor_id", "minor_subset", "ISG_status"))

# Replace NA with 0 for count
plot_data$count[is.na(plot_data$count)] <- 0

plot_data <- plot_data %>%
  ungroup() %>%
  group_by(donor_id, ISG_status) %>%
  mutate(prop = count / sum(count))

plot_data$prop <- plot_data$prop*100

# verify that sum = 100% 
plot_data %>%
  dplyr::filter(donor_id == "G05112") %>%
  dplyr::filter(ISG_status == "ISG_cluster") %>%
  mutate(temp = sum(prop))


# Convert ISG status to factor with ordered levels
plot_data$ISG_status <- factor(plot_data$ISG_status,
                               levels = c("Non_ISG", "ISG_cluster"))

# convert minor_subset to a factor for enforced ordering 
unique(plot_data$minor_subset)
plot_data$minor_subset <- factor(plot_data$minor_subset,
                                 levels = c("CD8.NAIVE", "CD8.TCM", "CD8.TCM.CCL5", "CD8.TEFF", "CD8.TEFF.prolif", "CD8.TEM", "CD8.TEMRA", "CD8.mitohi", "CD8.TREG"))


# get n value 
legend_data <- plot_data %>%
  group_by(ISG_status, minor_subset) %>%
  distinct(donor_id) %>%
  summarise(count_2 = n())

legend_data


# create jitter function 
pj <- position_jitterdodge(jitter.width = 0.25, 
                           seed = 9,
                           jitter.height = 0,
                           dodge.width = 0.8)


ggplot(plot_data, aes(x = minor_subset, y = prop, fill = ISG_status, color = ISG_status)) +
  geom_boxplot(alpha = 0.3, outlier.shape = NA, width = 0.75, position = position_dodge(width = 0.8)) +
  geom_point(shape = 19, size = 2, position = pj) +
  labs(x = "Functional Cluster", y = "Frequency", title = "Boxplot of Proportion by Functional Cluster and ISG Status") +
  scale_y_continuous(expand = expansion(add = c(0, 0.01)), limits = c(0, 100), breaks = seq(0, 100, by = 10)) +
  boxplot_theme() +
  stat_compare_means(method = "t.test", 
                     label.y = 80)

ggsave(filename = paste0(output.dir, "Supplementary_Figure_3E.pdf"),
       device = "pdf")


#################################################
# Proportion of ISG_cluster by sex
#################################################
# 
# plot_data <- cd8.covid.seurat@meta.data %>%
#   filter(disease == "COVID-19") %>%
#   group_by(donor_id, sex, ISG_status) %>%
#   summarise(freq = n()) %>%
#   ungroup() %>%
#   group_by(donor_id, sex) %>%
#   mutate(prop = freq / sum(freq)) %>%
#   filter(ISG_status == "ISG_cluster")
# 
# plot_data$prop <- plot_data$prop*100
# 
# # Create a data frame for the t-test groups
# t_test_groups <- data.frame(
#   group1 = c("female", "male")
# )
# 
# ggplot(plot_data, aes(x = sex, y = prop, color = sex, fill = sex)) +
#   geom_boxplot(alpha = 0.3, position = position_dodge(width = 0.75), aes(color = sex, fill = sex), outlier.shape = NA, width = 0.75) +
#   geom_point(aes(color = sex), position = position_jitter(width = 0.1), shape = 19, size = 2) + 
#   labs(x = "sex", y = "Proportion of ISG_cluster") +
#   boxplot_theme() + 
#   scale_y_continuous(expand = expansion(add = c(0, 0.01)), limits = c(0, 60), breaks = seq(0, 60, by = 5)) + 
#   stat_compare_means(comparisons = t_test_groups, 
#                      label = "p.signif",
#                      method = "t.test", 
#                      label.y = c(55), 
#                      hide.ns = F) + 
#   ggtitle("Proportion of ISG_cluster by sex")

#ggsave(filename = paste0(output.dir, "Supplementary_Figure_3X.pdf"),
#      device = "pdf")


```


### Supplementary Figure 3X - Data not shown, but references in manuscript

```{r Supplementary_Figure_3X}

if(long.compute){
  
  
  # data from https://www.nature.com/articles/s41467-023-43181-7
  malaria.seurat <- readRDS("data/external_datasets/annotated_Sabah_data_21Oct2022.rds")
  
  
  Idents(malaria.seurat) <- malaria.seurat@meta.data$celltype
  malaria.seurat.CD8 <- subset(malaria.seurat, idents = "CD8 T cells")
  
  
  
  library(harmony)
  # Normalise and harmonize dataset and calculate umap projection
  
  malaria.seurat.CD8 <- 
    malaria.seurat.CD8 %>%
    Seurat::NormalizeData() %>%
    Seurat::FindVariableFeatures() %>%
    Seurat::ScaleData() %>%
    Seurat::RunPCA() %>%
    harmony::RunHarmony(reduction.use = "pca",
                        group.by.vars = c("ID"), 
                        plot_convergence = TRUE, 
                        verbose = TRUE)  %>%
    Seurat::RunUMAP(reduction = "harmony", 
                    dims = 1:20, 
                    min.dist = 0.3, 
                    spread = 1,
                    n.neighbors = 5, # 5 - 50 , larger = more global
                    densmap = F,
                    seed.use = 42)  %>%
    FindNeighbors(reduction = "harmony",
                  dims = 1:20) %>%
    FindClusters(resolution = 0.5)
  
  UMAPPlot(malaria.seurat.CD8, pt.size = 1, group.by = "ID")
  
  
  # score for ISG signature 
  cores <- parallel::detectCores(all.tests = FALSE, logical = TRUE)
  
  # Calculate Ucell scores for signatures
  DefaultAssay(malaria.seurat.CD8) <- "RNA"
  malaria.seurat.CD8 <- UCell::AddModuleScore_UCell(malaria.seurat.CD8,
                                                    features = Signature.list,
                                                    ncores = cores)
  
  
  
  
  ###########################
  # Visualise ISG signature
  ###########################
  
  UMAPPlot(malaria.seurat.CD8,
           pt.size = 1, 
           group.by = "seurat_clusters",
           label = T)
  
  #dev.copy(pdf, paste0(output.dir, "Supplementary_Figure_3X_clusters.pdf"))
  #dev.off()
  
  VlnPlot(malaria.seurat.CD8, 
          features = "Top_10_UCell",
          pt.size = 0, 
          group.by = "seurat_clusters")
  
  #dev.copy(pdf, paste0(output.dir, "Supplementary_Figure_3X_Vlnplot.pdf"))
  #dev.off()
  
  
  # Density plots 
  Plot_Density_Custom(malaria.seurat.CD8, 
                      features =  "Top_10_UCell",
                      custom_palette = batlow.pal,
                      joint = FALSE, 
                      pt.size = 1, 
                      reduction = "umap")
  
  #dev.copy(pdf, paste0(output.dir, "Supplementary_Figure_3X_Density.pdf"))
  #dev.off()
  
  FeaturePlot_scCustom(malaria.seurat.CD8, 
                       features =  "Top_10_UCell")
  
  #dev.copy(pdf, paste0(output.dir, "Supplementary_Figure_3X_Featureplot.pdf"))
  #dev.off()
  
  
}


```





