---
title: "07_TCGA_survival"
author: "Dillon Corvino"
date: "02/02/2023"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

### Dataset information

```{r Dataset_Info}

```

### Environment

```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Create output directories & load custom functions & colour scheme
source("scripts/Setup.R", local = knitr::knit_global())

# pipeline variables
long.compute <- FALSE

# load required packages
require(tidyverse)

#source("scripts/aimed_analysis/functions.R")


# Create output directory
output.dir <- "results/Manuscript/TCGA_survival/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = TRUE)
}
```

### ISG signature
```{r ISG_signature}


# Hard code the signature to prevent any alternations with statistical drift 
ISG.sig <- c("IFI6", "ISG15", "IFIT3", "MX1", "ISG20", "IFITM1", "LY6E", "IFIT1", "MX2", "OAS1")


```

### Moving average
```{r moving_average}
# Create output directory
output.dir <- "results/Manuscript/Moving_average/"

if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = TRUE)
}



usefulfunctions::moving.average(tcga.datasets = c("hnsc_tcga", "brca_tcga", "esca_tcga"),
                                output.dir = output.dir,
                                primary.goi = "CD8A",
                                goi = ISG.sig)



```


### Surivival analysis
```{r tcga_survival}




usefulfunctions::TCGA.OS(TCGA.dataset = "TCGA-HNSC", 
                         gene.signature = "IFI6", 
                         rda.file.path = "scripts/Output/TCGA-HNSC/Data/TCGA-HNSC_Transcriptome_Profiling_raw_counts.rda")





  


require(TCGAbiolinks)


DGE.and.OS.plot.function(data.set.name = "TCGA-HNSC", 
                                     data.loaded = FALSE, 
                                     #rda.file.path, 
                                     goi = "ENSG00000126709", 
                                     gene.name = "IFI6",
                                     genes.to.plot = NULL,
                                     perform.DEG.analysis = FALSE)



query.raw.expression.hg38 <- GDCquery(project = "TCGA-HNSC", 
                                          data.category = "Transcriptome Profiling",
                                          data.type = "Gene Expression Quantification", 
                                          experimental.strategy = "RNA-Seq")


rm(query.raw.expression.hg38)

# DGE.and.OS.plot.function("TCGA-DLBC", data.loaded = T, 
#                           rda.file.path = "Archive_Output/Data/DLBC/Raw_Counts/TCGA_DLBC_Transcriptome_Profiling_raw_counts.rda",
#                         genes.to.plot = goi.plot.vector)






  # Initialise which TCGA dataset to use
  tcga.cancer <- paste0(movingaverage.datasets[dataset])
  
  
  # Create output directory

  if(!dir.exists(paste0(output.dir))){
    dir.create(paste0(output.dir), recursive = TRUE)}
  
  
  # Get tcga data using CGDS R package
  # create CGDS object and connection to cBioportal
  mycgds = CGDS("http://www.cbioportal.org/", verbose = T)
  
  # test Bioportal connection
  test(mycgds)
  
  # check list of cancer studies at server 
  getCancerStudies(mycgds)[,1:2]
  
  # check status and available data
  getGeneticProfiles(mycgds, tcga.cancer)[,1:2]
  getCaseLists(mycgds, tcga.cancer)[,1:2]
  
  # Get gene expression data using getProfileData
  genetic_profile_id <-paste0(tcga.cancer,"_rna_seq_v2_mrna")
  case_list_id <- paste0(tcga.cancer, "_rna_seq_v2_mrna")
  
  
  # enter genes of interest to retrieve via cBioportal
  # Vector of genes of interest
  gene <- c("CD226", "CD8B", "NCAM1", "IFNG", "GZMB", "PVR", "NECTIN2")
  
  # get gene expression data frame 
  gedf <- getProfileData(mycgds, gene, genetic_profile_id, case_list_id)
  
  # check normalized gene expression data (read counts) and log2 transform (avoid negative values)
  dim(gedf)
  ncol(gedf) == length(gene) # TRUE
  head(gedf)
  gedf[gedf < 1] <- 1
  gedf <- log2(gedf)
  head(gedf)
  n.val <- nrow(gedf)
```


