---
title: "22_Manuscript_output"
author: "Dillon Corvino"
date: "07/03/2022"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

### Dataset information

```{r Dataset_Info}

# HNSCC scRNAseq/scTCRseq dataset
# Cells are HNSCC TILs sorted on Lymphocytes/Live/CD3+/CD4-/CD8+
# Human samples
# Each sample is a pool of 4 patients
# samples 1 and 2 are unstimulated 
# samples 3 and 4 are stimulated 
# Patients in sample 1 match patients in sample 3 and sample 2 pairs with 4
# Data acquired was transcript expression, ADT (antibody expression), and TCRA & B sequences
# This analysis uses just Transcript and TCR data


# General information
# CellRanger was used for sequence alignment/QC/counts/TCR calls - performed by Ross (QIMR, Brisbane, Aus)
# DC was provided the output from CellRanger (counts matrix) and used this as input for analysis within this script

```

### Environment

```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Load packages
#source("R/Load_packages.R", local = knitr::knit_global())
renv::restore() # lockfile

# Create output directories & load custom functions & colour scheme
source("R/Setup.R", local = knitr::knit_global())


long.compute <- FALSE

library("Seurat")
#devtools::install_github(repo = "samuel-marsh/scCustomize")
library("scCustomize")
library("dplyr")


```


### Reading data

```{r reading_data}

########################
# Load all datasets 
########################

# Complete dataset
seurat.combined <- readRDS("Exported_RDS_files/seurat_combined.rds")

# set idents and assay to use
Idents(seurat.combined) <- seurat.combined@meta.data$Clusters_l1
DefaultAssay(seurat.combined) <- "RNA"


# Read in innate only dataset
Innate.seurat <- readRDS("Exported_RDS_files/Innate_seurat.rds")

Idents(Innate.seurat) <- Innate.seurat@meta.data$Clusters_l1
DefaultAssay(Innate.seurat) <- "RNA"

# Read in CD8 only dataset
CD8.seurat <- readRDS("Exported_RDS_files/CD8_seurat.rds")
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"

# CD8 TCR data
CD8.seurat.tcr <- readRDS("Exported_RDS_files/CD8_seurat_tcr.rds")
CD8.seurat.tcr.filt <- readRDS("Exported_RDS_files/CD8_seurat_tcr_filt.rds")


```



## Figure 1 Overview of dataset {.tabset}


### Set up for fig 1
```{r Fig_1_setup}

# Create output directory

output.fig1 <- "output/figures/Manuscript/Figure1"


if(!dir.exists(paste0(output.fig1))){
  dir.create(paste0(output.fig1), 
             recursive = T)
}


library(Nebulosa)


```


### Trying to rename Naive and cytotoxic clusters
```{r naive_vs_cytotoxic}


# CD69, CD27, KLRG1
Nebulosa::plot_density(seurat.combined, 
                       features = "TNFSF10")


unique(Idents(seurat.combined))

temp.seurat <- subset(seurat.combined, idents = c("Naive_like_3", "Naive_like_1_CM", "Cytotoxic", "Naive_like_2_SC"))






if(long.compute){
# Run DEG analysis
DEG.markers <- FindAllMarkers(temp.seurat, 
                              slot = "data", 
                              logfc.threshold = 0.25,
                              test.use = "wilcox",
                              only.pos = TRUE)

# Filter for significance
DEG.markers <- 
  DEG.markers %>%
  filter(p_val_adj < 0.05)


write.csv(DEG.markers, paste0("Temp_DEG_Naive_cytotoxic_Clusters_only.csv"))
}


```



### Nebulosa plots of key genes
```{r Nebulosa_plots_key_genes}


# Plot nebulosa density plots of key markers 
if(!dir.exists(paste0(output.fig1, "/Nebulosa_RNA"))){
  dir.create(paste0(output.fig1, "/Nebulosa_RNA"), 
             recursive = T)
}



goi <- c("EOMES", "GZMB", "PRF1", "IFNG", "LAG3", "HAVCR2", "MKI67", "TCF7", "TOX", "IL7R", "PDCD1", "PRF1", "SELL", "NKG7", "TNF", "TIGIT", "CD226", "CCR7", "KLRB1", "RORA", "ISG15", "TRDC", "TRAV1-2", "ZNF683")

for(i in 1:length(goi)){
  
  print(Plot_Density_Custom(seurat.combined, 
                            features = goi[i],
                            custom_palette = batlow.pal,
                            joint = FALSE, 
                            pt.size = 1, 
                            reduction = "umap"))
  
  dev.copy(pdf, paste0(output.fig1, "/Nebulosa_RNA/Nebulosa_RNA_", goi[i], ".pdf"))
  dev.off()
  
    
}




Plot_Density_Joint_Only(seurat.combined,
                        features = c("TRAV1-2", "KLRB1"),
                        pt.size = 1,
                        custom_palette = batlow.pal)

dev.copy(pdf, paste0(output.fig1, "/Nebulosa_RNA/Nebulosa_RNA_TRAV_and_KLRB1.pdf"))
dev.off()

```


### Stacked VlnPlots, Dot plot chemokines and US vs Stim umap plot
```{r various_fig1_plots}

# Cell cycle scoring
cc.genes$s.genes
cc.genes$g2m.genes

s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes


#######################
# Cell cycle analysis
#######################

seurat.combined <- CellCycleScoring(seurat.combined, 
                                    s.features = s.genes, 
                                    g2m.features = g2m.genes, 
                                    set.ident = FALSE)


seurat.combined@meta.data$S.Score



gene_list_plot <- list(goi.1 = c("CD28", "CD226", "TCF7", "TBX21"),
                       goi.2 = c("GNLY", "PRF1", "GZMK", "XCL1"),
                       goi.3 = c("TOX", "CTLA4", "LAG3", "TNFRSF9"),
                       goi.4 = c("KLRC1", "CD300A", "RORA", "G2M.Score"))

for(i in seq_along(gene_list_plot)){

  
  print(Stacked_VlnPlot(seurat_object = seurat.combined,
                assay = "RNA",
                pt.size = 0,
                features = gene_list_plot[[i]],
                x_lab_rotate = TRUE,
                 plot_spacing = 0.15,
                colors_use = clust.cols))


  dev.copy(pdf, paste0(output.fig1, "/Stacked_VlnPlots_", i, ".pdf"))
  dev.off()
  
  
  }




# Chemokine receptor plots


goi <- c("CCR6", "CCR7", "CXCR3", "CXCR4", "CXCR6", "CXCL13")


DotPlot_scCustom(seurat_object = seurat.combined, 
                 features = goi,
                 flip_axes = F,
                 x_lab_rotate = TRUE)

dev.copy(pdf, paste0(output.fig1, "/Dotplot_chemokine_receptors.pdf"))
dev.off()


# Umap plot stim vs unstim
UMAPPlot(seurat.combined, 
         shuffle = T,
         pt.size = 1,
         cols = condition.cols,
         group.by = "condition")

dev.copy(pdf, paste0(output.fig1, "/UMAP_condition_coloured.pdf"))
dev.off()

```
 
 
### Key stats of the dataset
```{r dataset_stats}

# Get basic stats on cells per cluster and per sample and plot these values 

cells.sample.clust <- table(seurat.combined@meta.data$group, seurat.combined@meta.data$Clusters_l1)
cells.cluster <- colSums(cells.sample.clust)


output.df <- cells.sample.clust*NA

for(i in 1:length(cells.cluster)){
output.df[,i] <- (cells.sample.clust[,i]/cells.cluster[i]) * 100
}


colSums(output.df)

# Plot freq of sample per cluster
barplot(output.df, 
        las = 2,
        col = sample.cols, 
        cex.names = 0.4)

dev.copy(pdf, paste0(output.fig1, "/Freq_of_sample_per_cluster.pdf"))
dev.off()




# freq of cells per cluster

total.cells <- sum(cells.cluster)
freq.clust <- (cells.cluster/total.cells)*100

z <- order(freq.clust, decreasing = T)

barplot(freq.clust[z], 
        las = 2,
        ylim = c(0, 20),
        col = clust.cols[z], 
        cex.names = 0.4)

dev.copy(pdf, paste0(output.fig1, "/Freq_of_cells_per_cluster.pdf"))
dev.off()


```











## Figure 2 Innate clusters overview {.tabset}


### Set up for fig 2
```{r Fig_2_setup}

# Create output directory
output.fig2 <- "output/figures/Manuscript/Figure2"


if(!dir.exists(paste0(output.fig2))){
  dir.create(paste0(output.fig2), 
             recursive = T)
}



# MAIT rearrangment 
# TRAV1-2/TRAJ33



```



### Visualise Innate-only UMAP
```{r vis_Innate_UMAP}



# Plot UMAP projection

UMAPPlot(object = Innate.seurat,
         group.by = "Clusters_l1",
         cols = innate.cols,
         label = TRUE, 
         pt.size = 1,
         label.size = 6)  + NoLegend()

dev.copy(pdf, paste0(output.fig2, "/UMAP_innate_clusters_named.pdf"))
dev.off()




UMAPPlot(object = Innate.seurat,
         group.by = "Clusters_l1",
         split.by = "condition",
         cols = innate.cols,
         label = TRUE, 
         pt.size = 1,
         label.size = 6)  + NoLegend()

dev.copy(pdf, paste0(output.fig2, "/UMAP_innate_clusters_named_condition_split.pdf"))
dev.off()



```


### Getting TCR data for Innate clusters
```{r TCR_for_Innate_clusters}


library(gplots)

# read in unfiltered TCR data, as most MAIT cells are low abundance clones lost in filtering
seurat.tcr <- readRDS("Exported_RDS_files/seurat_tcr.rds")

unique(seurat.tcr@meta.data$seurat_clusters) # Gamma/delta T-cells are not found in the filtered tcr dataset

# filter for MAIT cells
MAIT.TCR <- subset(seurat.tcr, idents = c("MAIT"))

# remove large dataset
rm(seurat.tcr)


# 290 MAIT cells in dataset
# 130 cells have a TCR associated
# 130/290 = ~44% of MAIT cells


########################
# Heatmap functions
########################

hclust.func <- function(x){
  hclust(x, method = 'ward.D2')
}

dist.func <- function(x){
  dist(x,method = 'binary')
}

library(RColorBrewer)


col.palette <- colorRampPalette(brewer.pal(8, "RdYlBu"))(600)


######################################################
# Plot TCR usage as a log2() of cell number
######################################################

# CTaa

x <- table(MAIT.TCR@meta.data$CTaa, MAIT.TCR@meta.data$condition_clust)


heatmap.2(as.matrix(log2(x+1)), 
          scale = "none",
          trace = "none",
          Colv = TRUE,
          cexRow = 0.5, 
          margins = c(8, 15),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette))     

dev.copy(pdf, paste0(output.fig2, "/Heatmap_CTaa.pdf"))
dev.off()

# CT gene

x <- table(MAIT.TCR@meta.data$CTgene, MAIT.TCR@meta.data$condition_clust)

heatmap.2(as.matrix(log2(x+1)), 
          scale = "none",
          trace = "none",
          Colv = TRUE,
          cexRow = 0.5, 
          margins = c(8, 15),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette))     

dev.copy(pdf, paste0(output.fig2, "/Heatmap_CT_gene.pdf"))
dev.off()

# TRAV gene

x <- table(MAIT.TCR@meta.data$TRAV, MAIT.TCR@meta.data$condition_clust)


heatmap.2(as.matrix(log2(x+1)), 
          scale = "none",
          trace = "none",
          Colv = TRUE,
          cexRow = 0.5, 
          margins = c(8, 15),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette))     

dev.copy(pdf, paste0(output.fig2, "/Heatmap_TRA_Vgene.pdf"))
dev.off()


# TRAJ gene


x <- table(MAIT.TCR@meta.data$TRAJ, MAIT.TCR@meta.data$condition_clust)


heatmap.2(as.matrix(log2(x+1)), 
          scale = "none",
          trace = "none",
          Colv = TRUE,
          cexRow = 0.5, 
          margins = c(8, 15),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette))     

dev.copy(pdf, paste0(output.fig2, "/Heatmap_TRA_Jgene.pdf"))
dev.off()



# TRBV gene
x <- table(MAIT.TCR@meta.data$TRBV, MAIT.TCR@meta.data$condition_clust)

heatmap.2(as.matrix(log2(x+1)), 
          scale = "none",
          trace = "none",
          Colv = TRUE,
          cexRow = 0.5, 
          margins = c(8, 15),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette))     

dev.copy(pdf, paste0(output.fig2, "/Heatmap_TRB_Vgene.pdf"))
dev.off()

# TRBJ gene
x <- table(MAIT.TCR@meta.data$TRBJ, MAIT.TCR@meta.data$condition_clust)

heatmap.2(as.matrix(log2(x+1)), 
          scale = "none",
          trace = "none",
          Colv = TRUE,
          cexRow = 0.5, 
          margins = c(8, 15),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette))     

dev.copy(pdf, paste0(output.fig2, "/Heatmap_TRB_Jgene.pdf"))
dev.off()

# remove large dataset
rm(MAIT.TCR)


```


### Innate cell DEG and heatmap
```{r Innate_cell_DEG_and_Heatmap}


# Run DEG analysis
DEG.markers <- FindAllMarkers(Innate.seurat, 
                              slot = "data", 
                              logfc.threshold = 0.25,
                              test.use = "wilcox",
                              only.pos = TRUE)

# Filter for significance
DEG.markers <- 
  DEG.markers %>%
  filter(p_val_adj < 0.05)


write.csv(DEG.markers, paste0(output.fig2, "/DEG_Innate_Clusters_only.csv"))



# Get top DEGs
goi <- DEG.markers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1) %>%
  dplyr::pull(gene)


goi <- unique(goi)

DefaultAssay(Innate.seurat) <- "RNA"

# Plot heatmap
DoHeatmap(Innate.seurat, 
          features = goi,
          raster = FALSE)

dev.copy(pdf, paste0(output.fig2, "/Heatmap_DEGs.pdf"))
dev.off()

























########################################################
# Plot TF heatmap but only annotate specific genes 
########################################################

# Source custom function // loaded in custom functions chunk 
#source("R/annotate_seurat_heatmap.R")

# Colour annotation for column labels 
names(clust.cols) <- clust.names

genes.id <- c("BACH2", "EOMES", "KLF2", "ZEB2", "IRF4", "TCF7", "JUNB", "ATF4", "IRF7", "STAT1")

# Plot heatmap averaged values
annotated.heatmap(seurat.combined, 
                  cluster.id = "seurat_clusters",
                  assay.use = "integrated",
                  average.expression = TRUE,
                  goi = TF.DEG, 
                  genes.to.label = genes.id, 
                  col_order = clust.names, 
                  col.colours = clust.cols, 
                  range.val = c(-2, 0, 2))

dev.copy(pdf, "output/figures/Large_Summary_heatmaps/Heatmap_top_up_genes_bycluster_highlighted_geneIDs_averaged.pdf")
dev.off()


# Plot heatmap downsampled single cell values
annotated.heatmap(seurat.combined, 
                  cluster.id = "seurat_clusters",
                  assay.use = "integrated",
                  average.expression = FALSE,
                  goi = TF.DEG, 
                  genes.to.label = genes.id, 
                  col_order = clust.names, 
                  col.colours = clust.cols, 
                  range.val = c(-2, 0, 2))

dev.copy(pdf, "output/figures/Large_Summary_heatmaps/Heatmap_top_up_genes_bycluster_highlighted_geneIDs.pdf")
dev.off()


```


### Innate cell TF analysis 

```{r TF_analysis}


# Searching for TFs
TF.ref.list <- read.delim("Data/TFcheckpoint.txt") # http://www.tfcheckpoint.org/index.php/browse

dim(TF.ref.list) # 1020, 11
colnames(TF.ref.list)
head(TF.ref.list)

Cluster.markers <- read.csv(paste0(output.fig2, "/DEG_Innate_Clusters_only.csv"), row.names = 1)



# Up-regulated TFs
sig.genes <- Cluster.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_log2FC > 0.5) %>%
  dplyr::select(gene)

sig.genes <- unique(sig.genes$gene)
TF.logic <- sig.genes %in% TF.ref.list$Gene_symbol

TF.DEG <- sig.genes[TF.logic]




########################################
# Plot TFs differentially regulated 
########################################

Idents(Innate.seurat) <- Innate.seurat@meta.data$Clusters_l1
DefaultAssay(Innate.seurat) <- "RNA"

#seurat.combined.small <- subset(seurat.combined, downsample = 300)

average.seurat <- AverageExpression(Innate.seurat,
                                    assay = "RNA",
                                    slot = "data",
                                    verbose = TRUE,
                                    return.seurat = TRUE)

DoHeatmap(average.seurat, 
          features = TF.DEG, 
          draw.lines = FALSE, 
          raster = FALSE)

dev.copy(pdf, paste0(output.fig2, "/Heatmap_of_TF_diff_expressed.pdf"))
dev.off()



```

### Scillus

```{r Scillus_GSEA}


#remotes::install_github("xmc811/Scillus")
library("Scillus")


# Scillus package would not compile
# I took the functions and made a new .R file
# Some of the functions called and used in this script required modifications to work outside of the package environment


#source("~/Documents/Work/Sciebo/Scripts/Packages/Scillus_package/Scillus_functions.R")
#load("~/Documents/Work/Sciebo/Scripts/Packages/Scillus_package/pathways.hallmark.rda")



######################
# Cluster analysis
######################

# set idents to cluster
Idents(Innate.seurat) <- Innate.seurat@meta.data$Clusters_l1


# get a broad list of DEG markers 
Cluster.markers <- FindAllMarkers(Innate.seurat,
                                  assay = "RNA",
                                  only.pos = TRUE,
                                  min.diff.pct = 0.1,
                                  logfc.threshold = 0,
                                  return.thresh = 0.01)




# Threshold marker list by P value and fold change for GO analysis 

GO.markers <- Cluster.markers[Cluster.markers$p_val_adj < 0.05, ]
GO.markers <- GO.markers[abs(GO.markers$avg_log2FC) > 0.25, ]
dim(Cluster.markers) # 8,635 genes
dim(GO.markers) # 4,890 genes

table(GO.markers$cluster) # number of genes DEG per cluster // used to justify using topn = 100

# Plot Gene Ontology analysis
# CC, BP, MF

# To plot graph for a specific cluster use below code
#plot_cluster_go(Module.markers, cluster_name = "Activated", org = "human", ont = "MF")

# CC 
pdf(paste0(output.fig2, "/GO_Clusters_CC.pdf"), width = 20, height = 20)

plot_all_cluster_go(GO.markers,
                    topn = 100,
                    org = "human", 
                    ont = "CC")

dev.off()


# BP 
pdf(paste0(output.fig2, "/GO_Clusters_BP.pdf"), width = 20, height = 20)

plot_all_cluster_go(GO.markers,
                    topn = 100,
                    org = "human", 
                    ont = "BP")

dev.off()

# MF 
pdf(paste0(output.fig2, "/GO_Clusters_MF.pdf"), width = 20, height = 20)

plot_all_cluster_go(GO.markers,
                    topn = 100,
                    org = "human", 
                    ont = "MF")
dev.off()




```










## Figure 3 Stimulation effect {.tabset}


### Set up for fig 3 output
```{r Fig_3_output}

# Create output directory

output.fig3 <- "output/figures/Manuscript/Figure3"


if(!dir.exists(paste0(output.fig3))){
  dir.create(paste0(output.fig3), 
             recursive = T)
}



```




### Visualise CD8 only UMAP
```{r vis_CD8_UMAP}

library(ggplot2)

UMAPPlot(object = CD8.seurat,
         group.by = "Clusters_l1",
         label = TRUE, 
         cols = CD8.cols,
         pt.size = 1,
         label.size = 6)  + NoLegend()

dev.copy(pdf, paste0(output.fig3, "/UMAP_CD8_clusters.pdf"))
dev.off()



Idents(CD8.seurat) <- CD8.seurat@meta.data$condition
US.seurat <- subset(CD8.seurat, ident = "US")
Stim.seurat <- subset(CD8.seurat, ident = "Stim")

UMAPPlot(object = US.seurat,
         group.by = "Clusters_l1",
         label = TRUE, 
         cols = CD8.cols,
         pt.size = 1,
         label.size = 6)  +
    ggtitle("Unstimulated") + 
  NoLegend()


dev.copy(pdf, paste0(output.fig3, "/UMAP_CD8_clusters_Unstimulated.pdf"))
dev.off()


UMAPPlot(object = Stim.seurat,
         group.by = "Clusters_l1",
         label = TRUE, 
         cols = CD8.cols,
         pt.size = 1,
         label.size = 6)  + 
  ggtitle("Stimulated") + 
  NoLegend()

dev.copy(pdf, paste0(output.fig3, "/UMAP_CD8_clusters_Stimulated.pdf"))
dev.off()



UMAPPlot(object = CD8.seurat,
         group.by = "Clusters_l1",
         split.by = "condition",
         cols = CD8.cols,
         pt.size = 1,
         label = TRUE, 
         label.size = 6) + NoLegend()

dev.copy(pdf, paste0(output.fig3, "/UMAP_CD8_clusters_named_condition_split.pdf"))
dev.off()





for(i in seq_along(condition.cols)){

  
  temp.cols <- c("Gray", "Gray")
  temp.cols[i] <- paste0(condition.cols)[i]

print(UMAPPlot(object = CD8.seurat,
         group.by = "condition",
         cols = temp.cols,
         pt.size = 1,
         shuffle = T,
         label = TRUE, 
         label.size = 6) + 
    NoLegend())

dev.copy(pdf, paste0(output.fig3, "/UMAP_CD8_clusters_", names(condition.cols[i]), "_coloured.pdf"))
dev.off()

}


UMAPPlot(object = CD8.seurat,
         group.by = "condition",
         cols = temp.cols,
         pt.size = 1,
         shuffle = T,
         label = TRUE, 
         label.size = 6) + 
    NoLegend()
dev.copy(pdf, paste0(output.fig3, "/UMAP_CD8_condition_combined_coloured.pdf"))
dev.off()

temp.cols <- condition.cols
names(temp.cols)[2] <-  "Stim"
temp.cols[2] <- "#ee8262"


library(RColorBrewer)


```



### Calculate DEGs without innate clusters
```{r DEG_calculation}

if(long.compute){
# Run DEG analysis
DEG.markers <- FindAllMarkers(CD8.seurat, 
                              slot = "data", 
                              logfc.threshold = 0.25,
                              test.use = "wilcox",
                              only.pos = FALSE)

# Filter for significance
DEG.markers <- 
  DEG.markers %>%
  filter(p_val_adj < 0.05)


write.csv(DEG.markers, paste0(output.fig3, "/DEG_CD8_Clusters_only.csv"))
}

```


 
### Annotated_heatmaps

```{r annotated_heatmap}


##################################################################
# Use complexheatmap to generate a manuscript ready figure
##################################################################

# Source custom function // loaded in custom functions chunk 
#source("R/Functions/annotate_seurat_heatmap_function.R")


# Make small seurat object for heatmap plotting
Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
small.seurat <- subset(CD8.seurat, downsample = 100)
small.seurat <- ScaleData(small.seurat)



# Ensure markers table is loaded
DEG.markers <- read.csv(paste0(output.fig3, "/DEG_CD8_Clusters_only.csv"))


# Get top cluster markers
genes.stim <- DEG.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_log2FC > 1) %>%
  dplyr::filter(cluster == "Stimulated_1") %>% 
  dplyr::pull(gene)


genes.exh <- DEG.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_log2FC > 1) %>%
  dplyr::filter(cluster == "Stimulated_exhausted") %>% 
  dplyr::pull(gene)


genes.to.plot <- unique(c(genes.stim, genes.exh))





DoHeatmap(small.seurat, genes.to.plot)



# Genes to be labeled if present in plotted heatmap
goi <- c("XCL1", "XCL2", "TNF", "TNFRSF9", "CRTAM", "IL2", "CD69", "TIGIT", "GZMB", "LAG3", "HAVCR2", "NKG7", "CTLA4", "GNLY", "CD226", "IFNG")


#install.packages("magick")

pdf(paste0(output.fig3, "/DEG_CD8_Clusters_only_annotated_sig_Stim_clusts.pdf"))
# Plot heatmap averaged values
annotated.heatmap(small.seurat, 
                  cluster.id = "Clusters_l1",
                  assay.use = "RNA",
                  average.expression = FALSE,
                  goi = genes.to.plot, 
                  genes.to.label = goi, 
                  col_order = names(CD8.cols), 
                  col.colours = CD8.cols, 
                  range.val = c(-2, 0, 2))

dev.off()




# Get top cluster markers
genes.to.plot <- DEG.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::group_by(cluster) %>%
  dplyr::top_n(50, avg_log2FC) %>% 
  dplyr::pull(gene)


genes.to.plot <- unique(genes.to.plot)



# Genes to be labeled if present in plotted heatmap
goi <- c("ISG15", "IL7R", "GZMK", "MKI67", "XCL2", "HAVCR2", "LTB", "JUNB", "NKG7", "PRF1", "CCR7", "IFNG")


#install.packages("magick")

pdf(paste0(output.fig3, "/DEG_CD8_Clusters_only_annotated.pdf"))
# Plot heatmap averaged values
annotated.heatmap(CD8.seurat, 
                  cluster.id = "Clusters_l1",
                  assay.use = "RNA",
                  average.expression = TRUE,
                  goi = genes.to.plot, 
                  genes.to.label = goi, 
                  col_order = names(CD8.cols), 
                  col.colours = CD8.cols, 
                  range.val = c(-2, 0, 2))

dev.off()



# IFN Signature 



# Get top cluster markers
genes.to.plot <- DEG.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(cluster == "Type_I_IFN") %>%
    dplyr::filter(avg_log2FC > 1) %>%
  dplyr::pull(gene)


genes.to.plot <- unique(genes.to.plot)



# Genes to be labeled if present in plotted heatmap
goi <- c("IFI6", "ISG20", "IFIT1", "OAS1", "IFIT2", "IRF7", "OAS2", "IFI35", "GZMK")



pdf(paste0(output.fig3, "/DEG_CD8_Clusters_only_annotated_IFNsig.pdf"))

# Plot heatmap averaged values
annotated.heatmap(CD8.seurat, 
                  cluster.id = "Clusters_l1",
                  assay.use = "RNA",
                  average.expression = FALSE,
                  goi = genes.to.plot, 
                  genes.to.label = goi, 
                  col_order = names(CD8.cols), 
                  col.colours = CD8.cols, 
                  range.val = c(-2, 0, 2))

dev.off()


```




### Distribution of cells across stimulation condition
```{r stim_condition_distribution}

# Get basic stats on cells per cluster and per condition and plot these values 

cells.conditon.clust <- table(CD8.seurat@meta.data$condition, CD8.seurat@meta.data$Clusters_l1)

# remove innate cluster columns 
pattern.val <- c("^gd_T_.*", "^MAIT$")
remove.logic <- grepl(paste0(pattern.val, collapse = "|"), colnames(cells.conditon.clust))
cells.conditon.clust <- cells.conditon.clust[ , !remove.logic]



cells.cluster <- colSums(cells.conditon.clust)


output.df <- cells.conditon.clust*NA

for(i in 1:length(cells.cluster)){
output.df[,i] <- (cells.conditon.clust[,i]/cells.cluster[i]) * 100
}


colSums(output.df)

# Plot freq of sample per cluster
barplot(output.df, 
        las = 2,
        col = condition.cols, 
        cex.names = 0.4)

dev.copy(pdf, paste0(output.fig3, "/Freq_of_condition_per_cluster.pdf"))
dev.off()




```





### CD8 seurat TCR data analysis
```{r CD8_TCR_analysis}

# Key questions
# Why do some cells move between clusters after stimulation 
# TCR recogniton between stim_1 and Stim_exhausted, is one bystander and the other tumor-specific, what is the clonal overlap between populations 


#install.packages("ggraph")
library(ggraph)

#devtools::install_github("ncborcherding/scRepertoire@dev")
library("scRepertoire")




# Show the distribution of clonotypes 

Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$cloneType

clonetypes <- levels(Idents(CD8.seurat.tcr))
clonotype.names <- c("Hyperexpanded", "Large", "Medium", "Small", "Single")




for(i in seq_along(clonetypes)){
  print(scCustomize::Cluster_Highlight_Plot(CD8.seurat.tcr, 
                                            cluster_name = paste0(clonetypes[i]), 
                                            highlight_color = "navy",
                                            background_color = "lightgray",
                                            split.by = "condition",
                                            pt.size = 1) + 
          NoLegend() + 
          ggtitle(paste0(clonetypes[i])))
  
  dev.copy(pdf, paste0(output.fig3, "/UMAP_highlighted_clonotype_", clonotype.names[i], "_split_condition.pdf"))
  dev.off()
}








# use .filt dataset which has single clonetypes removed


# subset data
Idents(CD8.seurat.tcr.filt) <- CD8.seurat.tcr.filt@meta.data$condition
US.seurat <- subset(CD8.seurat.tcr.filt, idents = "US")
Stim.seurat <- subset(CD8.seurat.tcr.filt, idents = "Stim")

# Reset Idents for plotting
Idents(CD8.seurat.tcr.filt) <- CD8.seurat.tcr.filt@meta.data$Clusters_l1
Idents(US.seurat) <- US.seurat@meta.data$Clusters_l1
Idents(Stim.seurat) <- Stim.seurat@meta.data$Clusters_l1


# First show the clonal overlap between clusters 
clonalOverlap(US.seurat, 
              cloneCall = "aa",
              chain = "both",
              method = "overlap")

dev.copy(pdf, paste0(output.fig3, "/Clonal_overlap_US_only.pdf"))
dev.off()

clonalOverlap(Stim.seurat,
              cloneCall = "aa",
              chain = "both",
              method = "overlap")

dev.copy(pdf, paste0(output.fig3, "/Clonal_overlap_Stim_only.pdf"))
dev.off()

clonalOverlap(CD8.seurat.tcr.filt,
              cloneCall = "aa",
              chain = "both",
              method = "overlap")

dev.copy(pdf, paste0(output.fig3, "/Clonal_overlap.pdf"))
dev.off()






# Plot bargraph of clonetype frequency per cluster

Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$cloneType

cells.clonetype.per.clust<- table(CD8.seurat.tcr@meta.data$cloneType, CD8.seurat.tcr@meta.data$Clusters_l1)

cells.cluster <- colSums(cells.clonetype.per.clust)

output.df <- cells.clonetype.per.clust*NA

for(i in 1:length(cells.cluster)){
output.df[,i] <- (cells.clonetype.per.clust[,i]/cells.cluster[i]) * 100
}


colSums(output.df)

clonetype.cols <- c("salmon4", "salmon", "orange", "cyan3", "Lightblue")
names(clonetype.cols) <- c("Hyperexpanded", "Large", "Medium", "Small", "Single")



# Plot freq of sample per cluster
barplot(output.df, 
        las = 2,
        legend = T,
        col = clonetype.cols, 
        cex.names = 0.5)

dev.copy(pdf, paste0(output.fig3, "/Freq_of_clonetype_per_cluster.pdf"))
dev.off()



# Overview of clonotype overlap 

UMAPPlot(CD8.seurat.tcr, 
         group.by = "Clonotype_overlap")


Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$Clonotype_overlap

overlap.status <- levels(Idents(CD8.seurat.tcr))



for(i in seq_along(overlap.status)){
  print(scCustomize::Cluster_Highlight_Plot(CD8.seurat.tcr, 
                                            cluster_name = paste0(overlap.status[i]), 
                                            highlight_color = "navy",
                                            background_color = "lightgray",
                                            pt.size = 1) + 
          NoLegend() + 
          ggtitle(paste0(overlap.status[i])))
  
  dev.copy(pdf, paste0(output.fig3, "/UMAP_highlighted_clonotype_", overlap.status[i], ".pdf"))
  dev.off()
}











# subset data for shared clonotypes only
Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$Clonotype_overlap
temp.seurat <- subset(CD8.seurat.tcr, idents = "Shared")
Idents(temp.seurat) <- temp.seurat@meta.data$Clusters_l1


scRepertoire::clonalNetwork(temp.seurat, 
                                    reduction = "umap", 
                                    identity = "Clusters_l1",
                                    filter.proportion = 0.05,
                                    filter.identity = "Stimulated_1",
                                    cloneCall = "aa") + ggtitle("Shared clonetypes")

dev.copy(pdf, "temp.pdf")
dev.off()




























seurat.list <- expression2List(CD8.seurat.tcr.filt, split.by = "Clusters_l1")


clonalProportion(seurat.list,
                 split = split.var,
                 cloneCall = "aa") 


clonalHomeostasis(tcr.clusters, 
                  cloneType = homeostasis.var,
                  cloneCall = "aa") + 
  theme(axis.text.x = element_text(angle = 90))



# for clonalProportion split
split.var <- c(1, 5, 10, 20, 50, 500, 5000, 1e+05)

# for clonalHomeostasis split
homeostasis.var <- c(Rare = 0.001, 
                     very_small = 0.002,
                     Small = 0.005, 
                     Medium = 0.01, 
                     Large = 0.05, 
                     very_Large = 0.1, 
                     Hyperexpanded = 1)






UMAPPlot(object = CD8.seurat.tcr,
         group.by = "Clusters_l1",
         label = TRUE, 
         cols = CD8.cols,
         pt.size = 1,
         label.size = 6)  + NoLegend()








# visualise clusters
p1 <- UMAPPlot(CD8.seurat.tcr, 
         group.by = "Clusters_l1",
         label = T) + NoLegend() + ggtitle("")

# clusters to plot
clust.ids <- as.character(unique(CD8.seurat.tcr@meta.data$Clusters_l1))

# plot all clusters
for(i in seq_along(clust.ids)){

  p2 <- scRepertoire::clonalNetwork(US.seurat, 
                                    reduction = "umap", 
                                    identity = "Clusters_l1",
                                    filter.proportion = 0.01,
                                    filter.identity = clust.ids[i],
                                    cloneCall = "aa") + ggtitle("US")
  
  p3 <- scRepertoire::clonalNetwork(Stim.seurat, 
                                    reduction = "umap", 
                                    identity = "Clusters_l1",
                                    filter.proportion = 0.01,
                                    filter.identity = clust.ids[i],
                                    cloneCall = "aa") + ggtitle("Stim")
  

  print(p1 + p2/p3)
  
 ggsave(paste0("output_tcr/figures/clonalnetworks/ClonalNetwork_", clust.ids[i], ".pdf"))

  
}


clonalOverlay(CD8.seurat.tcr, reduction = "umap", 
              freq.cutpoint = 1) + 
                guides(color = "none")



scRepertoire::clonalNetwork(CD8.seurat.tcr, 
                                    reduction = "umap", 
                                    identity = "Clusters_l1",
                                    filter.proportion = 0.01,
                                    filter.identity = "Cytotoxic",
                                    cloneCall = "aa") + ggtitle("Stim")


scRepertoire::clonalNetwork(CD8.seurat.tcr, 
                                    reduction = "umap", 
                                    identity = "Clusters_l1",
                                    #filter.proportion = 0.01,
                                    filter.clones =  1,
                                    cloneCall = "aa") + ggtitle("Stim")



CD8.seurat.tcr@meta.data$cloneType

occupiedscRepertoire(CD8.seurat.tcr, x.axis = "Clusters_l1", proportion = T)

?occupiedscRepertoire
CD8.seurat.tcr@meta.data

temp.seurat <- clusterTCR(seurat.list,
                          chain = "TRB",
                          group.by = "group", 
                          sequence = "aa", 
                          threshold = 0.85)

DimPlot(seurat, group.by = "TRB_cluster") + 
  NoLegend()
?clusterTCR


head(CD8.seurat.tcr@meta.data)
StartracDiversity(CD8.seurat.tcr,
                  type = "condition", 
            sample = "group", by = "overall")




clonotypeBias(CD8.seurat.tcr, cloneCall = "aa", split.by = "group", group.by = "Clusters_l1",
n.boots = 20, min.expand =10)

### Misc code for exploring functions of scRep package

seurat.list <- expression2List(CD8.seurat.tcr, split.by = "Clusters_l1")
length(seurat.list) #now listed by cluster









scRepertoire::clonalNetwork(US.seurat, 
                                    reduction = "umap", 
                                    identity = "Clusters_l1",
                                    filter.proportion = 0.01,
                                    filter.identity = "Cytotoxic",
                                    cloneCall = "aa") + ggtitle("US")





```


### Heatmap of top clonotypes
```{r Heatmap_top_clonotypes}

library(gplots)
library(RColorBrewer)
########################
# Heatmap functions
########################

hclust.func <- function(x){
  hclust(x, method = 'ward.D2')
}

dist.func <- function(x){
  dist(x,method = 'binary')
}

col.palette <- colorRampPalette(brewer.pal(8, "RdYlBu"))(600)





# Get data of how many cells in each cluster are assigned to what clonotypes
clonotype.data <- table(CD8.seurat.tcr@meta.data$CTaa, CD8.seurat.tcr@meta.data$Clusters_l1)


# Plot the top 10 clonotypes
aa.keep <- sort(rowSums(clonotype.data), decreasing = TRUE)[1:50]

keep.vec <- rownames(clonotype.data) %in% names(aa.keep)

plot.data <- clonotype.data[keep.vec, ]


heatmap.2(as.matrix(log2(plot.data+1)), 
          scale = "none",
          trace = "none",
          ColSideColors = CD8.cols,
          Colv = TRUE,
          cexRow = 0.8, 
          margins = c(8, 15),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette))     

dev.copy(pdf, paste0(output.fig3, "/Heatmap_top_50_clonotypes.pdf"))
dev.off()







############################################
# Plot top clonotypes in each cluster 
############################################

for(i in seq_along(colnames(clonotype.data))){
  temp <- as.data.frame(clonotype.data)
  
  temp <- temp %>%
    dplyr::filter(Var2 == paste0(colnames(clonotype.data)[i])) %>%
    dplyr::top_n(10)
  
  clonotypes.to.plot <- temp$Var1
  
  keep.vec <- rownames(clonotype.data) %in% clonotypes.to.plot
  
  plot.data <- clonotype.data[keep.vec, ]
  
  heatmap.2(as.matrix(log2(plot.data+1)), 
            scale = "none",
            trace = "none",
            ColSideColors = CD8.cols,
            Colv = FALSE,
            cexRow = 0.8, 
            margins = c(8, 15),
            cexCol = 0.8, 
            distfun = dist.func,
            hclustfun = hclust.func,
            srtCol = 90,
            col = rev(col.palette))
  
  dev.copy(pdf, paste0(output.fig3, "/Heatmap_top_10_clonotypes_in_", colnames(clonotype.data)[i], ".pdf"))
  dev.off()
  
}




```






