---
title: "HNSCC_Transcription_Factor_analysis"
author: "Dillon Corvino"
date: "03/02/2020"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

### Environment

```{r Environment_setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Load packages and create output directories
source("R/Load_packages.R", local = knitr::knit_global())
source("R/Setup.R", local = knitr::knit_global())

# Establish colour scheme
source("R/Colour_scheme_variable.R", local = knitr::knit_global())

# Load custom functions
source("R/scRNAseq_function.R", local = knitr::knit_global())
source("R/annotate_seurat_heatmap_function.R", local = knitr::knit_global())
source("R/Enhanced_volcano_custom_defaults_function.R", local = knitr::knit_global())



# pipeline variables
quick.load <- TRUE
long.compute <- FALSE

# load saved seurat object with cluster annotations and imputation performed
if(quick.load){
  seurat.combined <- readRDS("Exported_RDS_files/seurat_combined.rds")
}

```

## Transcription factor analysis {.tabset}

### Progeny Analysis

```{r Progeny_analysis}

# Create output directory
if(!dir.exists("output/figures/GO_analysis/Progeny")){
  dir.create("output/figures/GO_analysis/Progeny", 
             recursive = TRUE)}

## We compute the Progeny activity scores and add them to our Seurat object
## as a new assay called Progeny. 

DefaultAssay(seurat.combined) <- "RNA"
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters


seurat.combined <- progeny(seurat.combined, 
                           scale = FALSE, 
                           organism = "Human", 
                           top = 500,
                           perm = 1,
                           return_assay = TRUE)


## We can now directly apply Seurat functions in our Progeny scores. 
## For instance, we scale the pathway activity scores. 
seurat.combined <- Seurat::ScaleData(seurat.combined, assay = "progeny")

## We transform Progeny scores into a data frame to better handling the results
progeny_scores_df <-
    as.data.frame(t(GetAssayData(seurat.combined, slot = "scale.data",
        assay = "progeny"))) %>%
    rownames_to_column("Cell") %>%
    gather(Pathway, Activity, -Cell)


## We create a data frame with the specification of the cells that belong to 
## each cluster to match with the Progeny scores.
CellsClusters <- data.frame(Cell = names(Idents(seurat.combined)),
                            CellType = as.character(Idents(seurat.combined)),
                            stringsAsFactors = FALSE)




## We match Progeny scores with the cell clusters.
progeny_scores_df <- inner_join(progeny_scores_df, CellsClusters)

## We summarize the Progeny scores by cellpopulation
summarized_progeny_scores <- progeny_scores_df %>%
  group_by(Pathway, CellType) %>%
  summarise(avg = mean(Activity), std = sd(Activity))


## We prepare the data for the plot
summarized_progeny_scores_df <- summarized_progeny_scores %>%
  dplyr::select(-std) %>%
  spread(Pathway, avg) %>%
  data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE)

paletteLength = 100
myColor = colorRampPalette(c("Darkblue", "white","red"))(paletteLength)

progenyBreaks = c(seq(min(summarized_progeny_scores_df), 0,
                      length.out=ceiling(paletteLength/2) + 1),
                  seq(max(summarized_progeny_scores_df)/paletteLength,
                      max(summarized_progeny_scores_df),
                      length.out=floor(paletteLength/2)))


progeny_hmap = pheatmap(t(summarized_progeny_scores_df[,-1]),fontsize=14,
                        fontsize_row = 10,
                        color=myColor, breaks = progenyBreaks,
                        main = "PROGENy (500)", angle_col = 45,
                        treeheight_col = 0,  border_color = NA)

dev.copy(pdf, "output/figures/GO_analysis/Progeny/Progeny_heatmap.pdf")
dev.off()

unique(progeny_scores_df$Pathway)

DefaultAssay(seurat.combined) <- "RNA"

```

### Searching for TFs

```{r TF_analysis}

# Create output dir
if(!dir.exists("output/figures/Transcription_factors")){
  dir.create("output/figures/Transcription_factors", 
             recursive = T)}


# Searching for TFs
TF.ref.list <- read.delim("Data/TFcheckpoint.txt") # http://www.tfcheckpoint.org/index.php/browse

dim(TF.ref.list) # 1020, 11
colnames(TF.ref.list)
head(TF.ref.list)

Cluster.markers <- read.csv("output/tables/DEG_Clusters/Cluster_markers_unfiltered.csv", row.names = 1)



# Up-regulated TFs
sig.genes <- Cluster.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0.5 | avg_logFC < -0.5) %>%
  dplyr::select(gene)

sig.genes <- unique(sig.genes$gene)
TF.logic <- sig.genes %in% TF.ref.list$Gene_symbol

TF.DEG <- sig.genes[TF.logic]




########################################
# Plot TFs differentially regulated 
########################################


seurat.combined.small <- subset(seurat.combined, downsample = 300)

average.seurat <- AverageExpression(seurat.combined,
                                    assay = "integrated",
                                    slot = "data",
                                    verbose = TRUE,
                                    return.seurat = TRUE)

DoHeatmap(average.seurat, 
          features = TF.DEG, 
          draw.lines = FALSE, 
          raster = FALSE)

dev.copy(pdf, "output/figures/Transcription_factors/Heatmap_of_TF_diff_expressed.pdf")
dev.off()



########################################################
# Plot TF heatmap but only annotate specific genes 
########################################################

# Source custom function // loaded in custom functions chunk 
#source("R/annotate_seurat_heatmap.R")

# Colour annotation for column labels 
names(clust.cols) <- clust.names

genes.id <- c("BACH2", "EOMES", "KLF2", "ZEB2", "IRF4", "TCF7", "JUNB", "ATF4", "IRF7", "STAT1")

# Plot heatmap averaged values
annotated.heatmap(seurat.combined, 
                  cluster.id = "seurat_clusters",
                  assay.use = "integrated",
                  average.expression = TRUE,
                  goi = TF.DEG, 
                  genes.to.label = genes.id, 
                  col_order = clust.names, 
                  col.colours = clust.cols, 
                  range.val = c(-2, 0, 2))

dev.copy(pdf, "output/figures/Large_Summary_heatmaps/Heatmap_top_up_genes_bycluster_highlighted_geneIDs_averaged.pdf")
dev.off()


# Plot heatmap downsampled single cell values
annotated.heatmap(seurat.combined, 
                  cluster.id = "seurat_clusters",
                  assay.use = "integrated",
                  average.expression = FALSE,
                  goi = top.genes, 
                  genes.to.label = genes.id, 
                  col_order = clust.names, 
                  col.colours = clust.cols, 
                  range.val = c(-2, 0, 2))

dev.copy(pdf, "output/figures/Large_Summary_heatmaps/Heatmap_top_up_genes_bycluster_highlighted_geneIDs.pdf")
dev.off()



```


### RcisTarget

```{r TF_regulation}

# Create output directory 
if(!dir.exists("output/figures/RcisTarget")){
  dir.create("output/figures/RcisTarget",
             recursive = TRUE)}

if(!dir.exists("output/tables/RcisTarget")){
  dir.create("output/tables/RcisTarget",
             recursive = TRUE)}


###################################
# Generate input gene lists 
###################################

Cluster.markers <- read.csv("output/tables/DEG_Clusters/Cluster_markers_unfiltered.csv", row.names = 1)


UP.genes <- Cluster.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0.5) %>%
  dplyr::select(gene)

UP.genes <- unique(UP.genes$gene)

DN.genes <- Cluster.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC < -0.5) %>%
  dplyr::select(gene)

DN.genes <- unique(DN.genes$gene)

geneLists <- list(UP = UP.genes, 
                  DN = DN.genes)



Clust.genes <- Cluster.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_logFC > 0.5) %>%
  dplyr::select(gene)

unique(Clust.genes$cluster)


geneLists <- list(Naive_like_1_CM = Clust.genes[Clust.genes$cluster == "Naive_like_1_CM", ]$gene, 
                  Naive_like_2_SC = Clust.genes[Clust.genes$cluster == "Naive_like_2_SC", ]$gene, 
                  Naive_like_3 = Clust.genes[Clust.genes$cluster == "Naive_like_3", ]$gene, 
                  Cytotoxic = Clust.genes[Clust.genes$cluster == "Cytotoxic", ]$gene,
                  Type_I_IFN = Clust.genes[Clust.genes$cluster == "Type_I_IFN", ]$gene, 
                  Stimulated_1 = Clust.genes[Clust.genes$cluster == "Stimulated_1", ]$gene,
                  Stimulated_exhausted = Clust.genes[Clust.genes$cluster == "Stimulated_exhausted", ]$gene, 
                  Exhausted_1 = Clust.genes[Clust.genes$cluster == "Exhausted_1", ]$gene,
                  Exhausted_2 = Clust.genes[Clust.genes$cluster == "Exhausted_2", ]$gene,
                  TRM = Clust.genes[Clust.genes$cluster == "TRM", ]$gene,
                  gd_T_g9d2 = Clust.genes[Clust.genes$cluster == "gd_T_g9d2", ]$gene,
                  gd_T_non_g9d2 = Clust.genes[Clust.genes$cluster == "gd_T_non_g9d2", ]$gene,
                  MAIT = Clust.genes[Clust.genes$cluster == "MAIT", ]$gene,
                  Proliferative = Clust.genes[Clust.genes$cluster == "Proliferative", ]$gene)







# Available motif databased - in order of narrow -> large search space
# available datasets can be found at = https://resources.aertslab.org/cistarget/

# Download database
#featherURL <- "https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg38/refseq_r80/mc9nr/gene_based/hg38__refseq-r80__10kb_up_and_down_tss.mc9nr.feather" 

#download.file(featherURL, destfile=basename(featherURL)) # saved in current dir


# Read in database
motifRankings <- importRankings("Data/RcisTarget_feather_files/hg38__refseq-r80__10kb_up_and_down_tss.mc9nr.feather")

# Load mouse motif annotations
data("motifAnnotations_hgnc")

TF.vec <- unique(motifAnnotations_hgnc$TF)


# perform analysis
motifEnrichmentTable_wGenes <- cisTarget(geneLists, 
                                         motifRankings,
                                         motifAnnot = motifAnnotations_hgnc, 
                                         #highlightTFs = c("EOMES", "TBX21", "BATF3"),
                                         nCores = 6)




# Add motif logo to table
motifEnrichmentTable_wGenes_wLogo <- addLogo(motifEnrichmentTable_wGenes)


 
# to view output data 
resultsSubset <- motifEnrichmentTable_wGenes_wLogo[1:50,]

datatable(resultsSubset[,-c("enrichedGenes", "TF_lowConf"), with=FALSE], 
          escape = FALSE, # To show the logo
          filter="top", options=list(pageLength=10))



# Write data to file 
write.csv(motifEnrichmentTable_wGenes_wLogo, "output/tables/RcisTarget/motif_enrichment_table_withLogo_10kb.csv")


# Extract all high confidence transcription factors for each geneset
anotatedTfs <- lapply(split(motifEnrichmentTable_wGenes$TF_highConf,
                            motifEnrichmentTable_wGenes$geneSet),
                      function(x) {
                        genes <- gsub(" \\(.*\\). ", "; ", x, fixed=FALSE)
                        genesSplit <- unique(unlist(strsplit(genes, "; ")))
                        return(genesSplit)
                        })
                      

fix.list.length <- function(input.list){
  temp <- unlist(lapply(input.list, length))
  x.max <- max(temp)
  
  output.list <- lapply(input.list, function(x){
    
    remain <- x.max - length(x)
    return(paste0(c(x, rep(NA, remain))))
  })
  
  return(output.list)
}

anotatedTfs.export <- fix.list.length(anotatedTfs)  

# Write data to file 
write.csv(anotatedTfs.export, "output/tables/RcisTarget/Transcription_factors_annotated_to_motifs_with_highConf.csv")
rm(anotatedTfs.export)


# Filter predicted TFs based on those that are also significantly DEGs

index.vec <- names(geneLists)
DEG.predicted.TFs <- list()

for(i in 1:length(index.vec)){
  
  DEG.var <- geneLists[[paste0(index.vec[i])]]
  TF.var <- anotatedTfs[[paste0(index.vec[i])]]
  DEG.predicted.TFs[[paste0(index.vec[i])]]<- DEG.var[DEG.var %in% TF.var]
  names(DEG.predicted.TFs)[i] <- paste0(index.vec[i])
  print(DEG.predicted.TFs)
}




DEG.predicted.TFs.export <- fix.list.length(DEG.predicted.TFs)  

# Write data to file 
write.csv(DEG.predicted.TFs.export, "output/tables/RcisTarget/Transcription_factors_predicted_which_are_also_DEG.csv")
rm(DEG.predicted.TFs.export)






# Plot heatmaps of DEG predicted TFs

Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters

seurat.combined.small <- subset(seurat.combined, downsample = 300)

average.seurat <- AverageExpression(seurat.combined,
                                    assay = "integrated",
                                    slot = "data",
                                    verbose = TRUE,
                                    return.seurat = TRUE)

x <- unique(unlist(DEG.predicted.TFs))

DoHeatmap(average.seurat, 
          features = x, 
          draw.lines = FALSE, 
          raster = FALSE)

dev.copy(pdf, "output/figures/RcisTarget/Heatmap_predicted_TFs_dif_expressed.pdf")
dev.off()



```


