---
title: "HNSCC_scTCRseq_dataset_preprocessing"
author: "Dillon Corvino"
date: "03/02/2020"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---


Built with R version `r getRversion()`

## Setup {.tabset}

### Environment
```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Load packages
#source("R/Load_packages.R", local = knitr::knit_global())

# Create output directories & load custom functions & colour scheme
source("R/Setup.R", local = knitr::knit_global())

# pipeline variables
quick.load <- FALSE
long.compute <- TRUE
do.plot <- TRUE

# Load scRNAseq dataset
seurat.combined <- readRDS("Exported_RDS_files/seurat_combined.rds")
DefaultAssay(seurat.combined) <- "RNA"

```

## Formatting {.tabset}

### Format data with scRepertoire
```{r Format_data_with_scRep}

  # Ensure set to working dir
setwd(working.dir)

#################################
# Read in and format data
#################################

# Load library
library("scRepertoire")

# read in contig files
S1.contig <- read.csv("Data/TCR_data/Sample1_filtered_contig_annotations.csv", stringsAsFactors = FALSE)
S2.contig <- read.csv("Data/TCR_data/Sample2_filtered_contig_annotations.csv", stringsAsFactors = FALSE)
S3.contig <- read.csv("Data/TCR_data/Sample3_filtered_contig_annotations.csv", stringsAsFactors = FALSE)
S4.contig <- read.csv("Data/TCR_data/Sample4_filtered_contig_annotations.csv", stringsAsFactors = FALSE)

# create list of contigs
contig.list <- list(S1.contig, S2.contig,
                    S3.contig, S4.contig)

# Combine contig list
combined.contig <- combineTCR(contig.list, 
                              samples = c("S1", "S2", "S3", "S4"), 
                              ID = c("US", "US", "Stim", "Stim"), 
                              cells = "T-AB",
                              removeNA = TRUE,
                              removeMulti = FALSE, 
                              filterMulti = TRUE) # take the top 2 expressed chains for cells with multiple chains 

# on average, filteringMulti results in approx. 200 more cells for each sample than outright removing cells

# Need to amend BarcodeID to match with seurat object
combined.contig$S1_US$barcode <- gsub("US_", "", combined.contig$S1_US$barcode)
combined.contig$S2_US$barcode <- gsub("US_", "", combined.contig$S2_US$barcode)
combined.contig$S3_Stim$barcode <- gsub("Stim_", "", combined.contig$S3_Stim$barcode)
combined.contig$S4_Stim$barcode <- gsub("Stim_", "", combined.contig$S4_Stim$barcode)


# Get metadata from seurat object
meta.data <- seurat.combined[[]]

# Custom function made by Nick called combineMeta

# Combined metadata with TCR info
meta.tcr <- combineMeta(combined.contig,
                        meta = meta.data, 
                        cloneTypes = c(None = 0,
                                       Single = 1, 
                                       Small = 5,
                                       Medium = 10, 
                                       Large = 20,
                                       Hyperexpanded = 150))



# Combine seurat object with contig info combineExpression
seurat.tcr <- combineExpression(combined.contig,
                                seurat.combined,
                                cloneCall = "aa",
                                cloneTypes = c(None = 0,
                                               Single = 1,
                                               Small = 5,
                                               Medium = 10,
                                               Large = 20,
                                               Hyperexpanded = 150),
                                groupBy = "none", # Alternative frequencies are calculated in next code Chunk
                                filterNA = TRUE) # should seurat object be subsetted to remove NA values


# Write out data for future use in scRep package
saveRDS(combined.contig, "Exported_RDS_files/combined_contig.rds")


# Max freq of a clonotype is 127, therefore reset cloneTypes accordingly

# Seurat object went from 11658 samples to 6625 samples 
# Therefore 56% of original dataset remains 


# Remove frequency column calculated by combineExpression function
logic.vec <- grepl("Frequency", colnames(seurat.tcr@meta.data))
seurat.tcr@meta.data[ ,logic.vec] <- NULL
head(seurat.tcr@meta.data)

# Ensure seurat object cloneType var is a factor with correct levels
seurat.tcr@meta.data$cloneType <- factor(seurat.tcr@meta.data$cloneType, 
                                         levels = c("Hyperexpanded (20 < X <= 150)",
                                                    "Large (10 < X <= 20)", 
                                                    "Medium (5 < X <= 10)",
                                                    "Small (1 < X <= 5)", 
                                                    "Single (0 < X <= 1)"))

# remove unneeded objects
rm(contig.list, S1.contig, S2.contig, S3.contig, S4.contig, meta.data, meta.tcr)

```


### Subsetting metadata to designate TRAV/J/C and TRBV/J/C genes
```{r Subset_CTaa_by_GeneID}

# Note, changed the order so that gene call is extracted into metadata before filtering for clone size 
# as such, there is a single cell with two TRA calls. this is messing with subsetting. 
#  therefore, remove this cell S2_CTTAGGAGTGACAAAT, from exhausted_2 cluster.


library(stringr)


cell.ids <- rownames(seurat.tcr@meta.data)
remove.logic <- grepl("^S2_CTTAGGAGTGACAAAT-1$", cell.ids)
sum(remove.logic)
cell.ids <- cell.ids[!remove.logic]


seurat.tcr <- subset(seurat.tcr, cells = cell.ids)




# Split the CTgene string at "." 
temp.list <- str_split(seurat.tcr@meta.data$CTgene, "\\.")

# First and second elements of the list are V and J calls
seurat.tcr@meta.data$TRAV <- sapply(temp.list, "[[", 1)
seurat.tcr@meta.data$TRAJ <- sapply(temp.list, "[[", 2)

# Third element is the TRAC and TRBV combined - therefore split this by "_"
temp.list.2 <- sapply(temp.list, "[[", 3)
temp.list.2 <- str_split(temp.list.2, "_")

# First element is TRAC and second is TRBV gene call
seurat.tcr@meta.data$TRAC <- sapply(temp.list.2, "[[", 1)
seurat.tcr@meta.data$TRBV <- sapply(temp.list.2, "[[", 2)

# 4th and 5th elements of original list are TRBJ and TRBC calls respectively
seurat.tcr@meta.data$TRBJ <- sapply(temp.list, "[[", 4)
seurat.tcr@meta.data$TRBC <- sapply(temp.list, "[[", 6)

# remove unneeded variables 
rm(temp.list, 
   temp.list.2)

```



### Filter dataset
```{r Filter_data}

####################################
# Global filtering of TCR dataset
####################################

# Keeps downstream analysis consistent
# Unfiltered data = seurat.tcr
# Filtered data = seurat.tcr.filt


################################
# Define filtering parameters 
################################

min.clonesize <- 3

clonosize.per.clust <- FALSE # Should the min.clonesize use total dataset clonesize (FALSE) or cluster-wise clonesize (TRUE)

remove.subsets <- c("gd_T_g9d2", 
                    "gd_T_non_g9d2", 
                    "MAIT")

remove.self.ribbons <- FALSE # term used later in circos plot generation


###################
# Filter dataset
###################

Idents(seurat.tcr) <- seurat.tcr@meta.data$seurat_clusters

# Remove non-alpha/beta T cells and invariant subsets
clust.var <- as.character(unique(Idents(seurat.tcr)))
clust.var <- clust.var[!clust.var %in% remove.subsets]

seurat.tcr.filt <- subset(seurat.tcr, idents = clust.var)





# Remove small clonotypes (i.e, <= min.clonesize) 
if(clonosize.per.clust){
  # Filter min clone size in a per condition specific manner i.e must hit threshold within the condition not the entire dataset
  clonotypes <- table(seurat.tcr.filt@meta.data$CTaa, seurat.tcr.filt@meta.data$condition)
  keep.clonotypes <- rownames(clonotypes)[clonotypes[ , "US"] > min.clonesize | clonotypes[ , "Stim"] > min.clonesize]
}else{
  clonotypes <- table(seurat.tcr.filt@meta.data$CTaa)
  keep.clonotypes <- names(clonotypes)[clonotypes > min.clonesize]
}

Idents(seurat.tcr.filt) <- seurat.tcr.filt@meta.data$CTaa

seurat.tcr.filt <- subset(seurat.tcr.filt, idents = keep.clonotypes)

#########################
# Filtering statistics
#########################

# Initial statistics 
initial.n <- nrow(seurat.tcr@meta.data)
init.clone.n <- length(unique(seurat.tcr@meta.data$CTaa))

# Filt values
filt.n <- nrow(seurat.tcr.filt@meta.data)
filt.clone.n <- length(unique(seurat.tcr.filt@meta.data$CTaa))


# Print info on filtering
print(paste0("# of cells in dataset before filtering = ", initial.n, " and # of clonotypes = ", init.clone.n))
print(paste0("# of cells in dataset after filtering = ", filt.n, " and # of clonotypes = ", filt.clone.n))
print(paste0("# of cells removed in filtering process = ", initial.n - filt.n))

# Remove unneeded variables 
rm(initial.n, 
   init.clone.n, 
   filt.n,
   filt.clone.n, 
   clonotypes, 
   keep.clonotypes, 
   clust.var)

```




### Count clonotype frequencies and normalise
```{r Count_and_normalise_clonotype_frequencies}

####################################################################################
# Need to count clonotype occurrence per condition and then normalise this value 
####################################################################################

# create temp meta.data df and append cell_ID as column (lost in dplyr workflow)
meta.data.temp <- seurat.tcr.filt@meta.data
meta.data.temp$cell_ID <- rownames(seurat.tcr.filt@meta.data)

corner(meta.data.temp, c = ncol(meta.data.temp))


# Get total number of cells in each condition
US.cell.n <- table(meta.data.temp$condition)[1] # 3,184 
Stim.cell.n <- table(meta.data.temp$condition)[2] # 3,441
total.cell.n <- US.cell.n + Stim.cell.n # 6,625 

#dim(meta.data.temp) // sanity check of total cell number




####################################################################################
# Count occurence of clonotypes across entire dataset 
####################################################################################
# Also: Normalise n() by # of cells within dataset
meta.data.temp <- meta.data.temp %>%
  dplyr::group_by(CTaa) %>%
  dplyr::mutate(Total_clonotype_n = n()) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(Total_clonotype_freq = (Total_clonotype_n/total.cell.n)*100) %>%
  dplyr::ungroup()

####################################################################################
# Count occurence of clonotypes across entire condition 
####################################################################################
# Also: Normalise n() by # of cells within condition
meta.data.temp <- meta.data.temp %>%
  dplyr::group_by(CTaa, condition) %>%
  dplyr::mutate(Clone_conditionwise_n = n()) %>%
  dplyr::group_by(condition) %>%
  dplyr::mutate(Clone_conditionwise_freq = case_when(condition == "US" ~ (Clone_conditionwise_n/US.cell.n)*100, 
                                                     condition == "Stim" ~ (Clone_conditionwise_n/Stim.cell.n)*100)) %>%
  dplyr::ungroup()

####################################################################################
# Count occurence of clonotypes within each cluster per condition 
####################################################################################
# Also: Normalise n() by # of cells within condition
meta.data.temp <- meta.data.temp %>%
  dplyr::group_by(CTaa, condition, seurat_clusters) %>%
  dplyr::mutate(Clone_Clust_conditionwise_n = n()) %>%
  dplyr::group_by(condition) %>%
  dplyr::mutate(Clone_Clust_conditionwise_freq = case_when(condition == "US" ~ (Clone_Clust_conditionwise_n/US.cell.n)*100, 
                                                           condition == "Stim" ~ (Clone_Clust_conditionwise_n/Stim.cell.n)*100)) %>%
  dplyr::ungroup()


######################################
# Add new metadata to seurat object
######################################

#########
# Total 
#########

# Add total clonotype n value
seurat.tcr.filt <- AddMetaData(object = seurat.tcr.filt, 
                          col.name = "Total_clonotype_n",
                          metadata = meta.data.temp$Total_clonotype_n)

# Add total clonotype freq
seurat.tcr.filt <- AddMetaData(object = seurat.tcr.filt, 
                          col.name = "Total_clonotype_freq",
                          metadata = meta.data.temp$Total_clonotype_freq)

##################
# Condition-wise 
##################

# Add clone condition-wise n value
seurat.tcr.filt <- AddMetaData(object = seurat.tcr.filt, 
                          col.name = "Clone_conditionwise_n",
                          metadata = meta.data.temp$Clone_conditionwise_n)

# Add Clone condition-wise freq
seurat.tcr.filt <- AddMetaData(object = seurat.tcr.filt, 
                          col.name = "Clone_conditionwise_freq",
                          metadata = meta.data.temp$Clone_conditionwise_freq)

################################
# Cluster and Condition-wise 
################################

# Add Clone Clust and condition-wise n value
seurat.tcr.filt <- AddMetaData(object = seurat.tcr.filt, 
                          col.name = "Clone_Clust_conditionwise_n",
                          metadata = meta.data.temp$Clone_Clust_conditionwise_n)

# Add Clone Clust and condition-wise Freq
seurat.tcr.filt <- AddMetaData(object = seurat.tcr.filt, 
                          col.name = "Clone_Clust_conditionwise_freq",
                          metadata = meta.data.temp$Clone_Clust_conditionwise_freq)


#############################################
# Scale variables for downstream plotting
#############################################
scale.var <- c(0, 10)

# Scale total clonotype number
seurat.tcr.filt@meta.data$Total_clonotype_scaled <- scales::rescale(seurat.tcr.filt@meta.data$Total_clonotype_n, 
                                                               to = scale.var)


# Scale clonotype per condition freq
seurat.tcr.filt@meta.data$Clone_conditionwise_scaled <- scales::rescale(seurat.tcr.filt@meta.data$Clone_conditionwise_freq, 
                                                                   to = scale.var)

# Scale clonotype per cluster per condition freq
seurat.tcr.filt@meta.data$Clone_Clust_conditionwise_scaled <- scales::rescale(seurat.tcr.filt@meta.data$Clone_Clust_conditionwise_freq, 
                                                                         to = scale.var)



rm(meta.data.temp)

```

### Clonotype Overlap
```{r Clonotype_Overlap}

Idents(seurat.tcr.filt) <- seurat.tcr.filt@meta.data$seurat_clusters

# Get vector of clonotypes found in US cells
US.clonotypes <- seurat.tcr.filt@meta.data %>%
  dplyr::filter(condition == "US") %>%
  dplyr::select(CTaa)

US.clonotypes <- unique(US.clonotypes$CTaa)

length(US.clonotypes) # 236 clonotypes

# Get vector of clonotypes found in US cells
Stim.clonotypes <- seurat.tcr.filt@meta.data %>%
  dplyr::filter(condition == "Stim") %>%
  dplyr::select(CTaa)

Stim.clonotypes <- unique(Stim.clonotypes$CTaa)

length(Stim.clonotypes) # 249 clonotypes


# Get shared and unique clonotypes 
shared.clonotypes <- US.clonotypes[US.clonotypes %in% Stim.clonotypes]
unique.US <- US.clonotypes[!US.clonotypes %in% Stim.clonotypes]
unique.Stim <- Stim.clonotypes[!Stim.clonotypes %in% US.clonotypes]


####################
# Add to metadata
####################
meta.data.vec <- seurat.tcr.filt@meta.data %>%
  dplyr::mutate(Clonotype_overlap = case_when(CTaa %in% shared.clonotypes ~ "Shared", 
                                              CTaa %in% unique.US ~ "US", 
                                              CTaa %in% unique.Stim ~ "Stim")) %>%
  dplyr::select(Clonotype_overlap)


seurat.tcr.filt <- AddMetaData(seurat.tcr.filt,
                               meta.data.vec$Clonotype_overlap,
                               "Clonotype_overlap")


# Remove large objects no longer required
rm(shared.clonotypes, 
   unique.US, 
   unique.Stim, 
   US.clonotypes, 
   Stim.clonotypes)

```



### Export tcr seurat object
```{r export_tcr_RDS}

saveRDS(seurat.tcr, file = "Exported_RDS_files/seurat_tcr.rds")
saveRDS(seurat.tcr.filt, file = "Exported_RDS_files/seurat_tcr_filt.rds")

```

