---
title: "HNSCC_scTCRseq_TCR_gene_expression"
author: "Dillon Corvino"
date: "03/02/2020"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---


Built with R version `r getRversion()`

## Setup {.tabset}

### Environment
```{r Environment_setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Load packages
source("R/Load_packages.R", local = knitr::knit_global())

# Create output directories & load custom functions & colour scheme
source("R/Setup.R", local = knitr::knit_global())

# pipeline variables
quick.load <- TRUE
long.compute <- FALSE
do.plot <- FALSE


# load saved seurat object with cluster annotations and imputation performed
if(quick.load){
  seurat.combined <- readRDS("Exported_RDS_files/seurat_combined.rds")
  DefaultAssay(seurat.combined) <- "integrated"
  
  seurat.tcr <- readRDS("Exported_RDS_files/seurat_tcr.rds")
  DefaultAssay(seurat.tcr) <- "integrated"
  
  seurat.tcr.filt <- readRDS("Exported_RDS_files/seurat_tcr_filt.rds")
  DefaultAssay(seurat.tcr.filt) <- "integrated"
  
}

```


## TCR analysis - based on scRNAseq expression {.tabset}

### Expression of TCR genes
```{r TCR_gene_expression}

# Set wd
setwd(working.dir)

# Create output directories
if(!dir.exists("output_tcr/figures/TCR_gene_expression/VlnPlot")){
  dir.create("output_tcr/figures/TCR_gene_expression/VlnPlot",
             recursive = T)}

# set and move to output.dir
output.dir <- "output_tcr/figures/TCR_gene_expression"
setwd(output.dir)

# set idents
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters

# set assay 
DefaultAssay(seurat.combined) <- "RNA"


# Find TCR genes to plot
genes <- rownames(seurat.combined)

pattern <- c("TRAV", "TRBV", "TRAC", "TRBC", "TRG", "TRD")

TCR.genes <- unique(grep(paste0(pattern, collapse = "|"), 
                         genes,
                         value = TRUE))

length(TCR.genes)


for(i in 1:length(TCR.genes)){
  print(paste0("Plotting ", TCR.genes[i]))
  
  skip.iteration <- FALSE
  tryCatch(print(VlnPlot(seurat.combined, 
                         paste0(TCR.genes[i]),
                         log = T,
                         group.by = "seurat_clusters",
                         cols = clust.cols,
                         pt.size = 0)),
           error = function(e){skip.iteration <- TRUE})
  if(skip.iteration){next}
  
  
  dev.copy(pdf, paste0("VlnPlot/VlnPlot_", TCR.genes[i], ".pdf"))
  dev.off()
  
}


##################################################
# Plot heatmap of gene expression for TCR genes
##################################################



# get average expression 
Avg.seurat <- AverageExpression(seurat.combined,
                                assay = "integrated",
                                slot = "data",
                                verbose = TRUE,
                                return.seurat = TRUE)



# TCRA genes 
pattern <- c("TRAV", "TRAC")

TCR.genes <- unique(grep(paste0(pattern, collapse = "|"), 
                         genes,
                         value = TRUE))

TCR.genes <- sort(TCR.genes, decreasing = TRUE)

DoHeatmap(Avg.seurat, 
          features = TCR.genes,
          draw.lines = FALSE,
          group.colors = clust.cols,
          raster = FALSE)

dev.copy(pdf, "Heatmap_TCRA_genes.pdf")
dev.off()


# TCRB genes
pattern <- c("TRBV", "TRBC")

TCR.genes <- unique(grep(paste0(pattern, collapse = "|"), 
                         genes,
                         value = TRUE))

TCR.genes <- sort(TCR.genes, decreasing = TRUE)

DoHeatmap(Avg.seurat, 
          features = TCR.genes,
          draw.lines = FALSE,
          group.colors = clust.cols,
          raster = FALSE)

dev.copy(pdf, "Heatmap_TCRB_genes.pdf")
dev.off()




# TCRG and TCRD genes
pattern <- c("TRG", "TRD")

TCR.genes <- unique(grep(paste0(pattern, collapse = "|"), 
                         genes,
                         value = TRUE))

TCR.genes <- sort(TCR.genes, decreasing = TRUE)

DoHeatmap(Avg.seurat, 
          features = TCR.genes,
          draw.lines = FALSE,
          group.colors = clust.cols,
          raster = FALSE)

dev.copy(pdf, "Heatmap_TCRG_TCRD_genes_all_clusts.pdf")
dev.off()



#################################################
# Heatmaps with just G/D and MAIT clusters
#################################################

# Subset data for just GD and MAIT clusters 
clust.pattern <- c("gd_T_non_g9d2", "gd_T_g9d2", "MAIT")
Avg.seurat <- subset(Avg.seurat, idents = clust.pattern)

# Plot all TCRG and TCRD genes across just G/D and MAIT clusters
pattern <- c("TRG", "TRD")

TCR.genes <- unique(grep(paste0(pattern, collapse = "|"), 
                         genes,
                         value = TRUE))

TCR.genes <- sort(TCR.genes, decreasing = TRUE)


keep.index <- grepl(paste0(clust.pattern, collapse = "|"), names(clust.cols))

DoHeatmap(Avg.seurat, 
          features = TCR.genes,
          group.colors = clust.cols[keep.index],
          draw.lines = FALSE,
          raster = FALSE)

dev.copy(pdf, "Heatmap_TCRG_TCRD_genes_gd_MAIT_clusts.pdf")
dev.off()


# Plot all specific G/D TCR and KLRB1 gene expression
TCR.genes <- c("TRGV9", "TRDV2", "TRGC1", "TRGC2", "TRDC", "TRAV1-2", "KLRB1")


DoHeatmap(Avg.seurat, 
          features = TCR.genes,
          draw.lines = FALSE,
          group.colors = clust.cols[keep.index],
          raster = FALSE)

dev.copy(pdf, "Heatmap_TCRG9_TCRD2_KLRB1_expression_gd_MAIT_clusts.pdf")
dev.off()


# Remove unneeded object
rm(Avg.seurat)

# return to working directory
setwd(working.dir)

```

### Expression of TCR genes // IMPUTED
```{r TCR_gene_expression_imputed}

# Set wd
setwd(working.dir)

# Create output directories
if(!dir.exists("output_tcr/figures/TCR_gene_expression/Imputed/VlnPlot")){
  dir.create("output_tcr/figures/TCR_gene_expression/Imputed/VlnPlot",
             recursive = T)}

# set and move to output.dir
output.dir <- "output_tcr/figures/TCR_gene_expression/Imputed"
setwd(output.dir)

# set idents
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters

# set assay 
DefaultAssay(seurat.combined) <- "alra"


# Find TCR genes to plot
genes <- rownames(seurat.combined)

pattern <- c("TRAV", "TRBV", "TRAC", "TRBC", "TRG", "TRD")

TCR.genes <- unique(grep(paste0(pattern, collapse = "|"), 
                         genes,
                         value = TRUE))

length(TCR.genes)


for(i in 1:length(TCR.genes)){
  print(paste0("Plotting ", TCR.genes[i]))
  
  skip.iteration <- FALSE
  tryCatch(print(VlnPlot(seurat.combined, 
                         paste0(TCR.genes[i]),
                         log = TRUE,
                         cols = clust.cols,
                         group.by = "seurat_clusters",
                         pt.size = 0)),
           error = function(e){skip.iteration <- TRUE})
  if(skip.iteration){next}
  
  
  dev.copy(pdf, paste0("VlnPlot/VlnPlot_", TCR.genes[i], "_imputed.pdf"))
  dev.off()
  
}


##################################################
# Plot heatmap of gene expression for TCR genes
##################################################



# get average expression 
Avg.seurat <- AverageExpression(seurat.combined,
                                assay = "alra",
                                slot = "data",
                                verbose = TRUE,
                                return.seurat = TRUE)



# TCRA genes 
pattern <- c("TRAV", "TRAC")

TCR.genes <- unique(grep(paste0(pattern, collapse = "|"), 
                         genes,
                         value = TRUE))

TCR.genes <- sort(TCR.genes, decreasing = TRUE)

DoHeatmap(Avg.seurat, 
          features = TCR.genes,
          draw.lines = FALSE,
          group.colors = clust.cols,
          raster = FALSE)

dev.copy(pdf, "Heatmap_TCRA_genes_imputed.pdf")
dev.off()


# TCRB genes
pattern <- c("TRBV", "TRBC")

TCR.genes <- unique(grep(paste0(pattern, collapse = "|"), 
                         genes,
                         value = TRUE))

TCR.genes <- sort(TCR.genes, decreasing = TRUE)

DoHeatmap(Avg.seurat, 
          features = TCR.genes,
          draw.lines = FALSE,
          group.colors = clust.cols,
          raster = FALSE)

dev.copy(pdf, "Heatmap_TCRB_genes_imputed.pdf")
dev.off()




# TCRG and TCRD genes
pattern <- c("TRG", "TRD")

TCR.genes <- unique(grep(paste0(pattern, collapse = "|"), 
                         genes,
                         value = TRUE))

TCR.genes <- sort(TCR.genes, decreasing = TRUE)

DoHeatmap(Avg.seurat, 
          features = TCR.genes,
          draw.lines = FALSE,
          group.colors = clust.cols,
          raster = FALSE)

dev.copy(pdf, "Heatmap_TCRG_TCRD_genes_all_clusts_imputed.pdf")
dev.off()



#################################################
# Heatmaps with just G/D and MAIT clusters
#################################################

# Subset data for just GD and MAIT clusters 
clust.pattern <- c("gd_T_non_g9d2", "gd_T_g9d2", "MAIT")
Avg.seurat <- subset(Avg.seurat, idents = clust.pattern)

# Plot all TCRG and TCRD genes across just G/D and MAIT clusters
pattern <- c("TRG", "TRD")

TCR.genes <- unique(grep(paste0(pattern, collapse = "|"), 
                         genes,
                         value = TRUE))

TCR.genes <- sort(TCR.genes, decreasing = TRUE)

keep.index <- grepl(paste0(clust.pattern, collapse = "|"), names(clust.cols))

DoHeatmap(Avg.seurat, 
          features = TCR.genes,
          draw.lines = FALSE,
          group.colors = clust.cols[keep.index],
          raster = FALSE)

dev.copy(pdf, "Heatmap_TCRG_TCRD_genes_gd_MAIT_clusts_imputed.pdf")
dev.off()


# Plot all specific G/D TCR and KLRB1 gene expression
TCR.genes <- c("TRGV9", "TRDV2", "TRGC1", "TRGC2", "TRDC", "TRAV1-2", "KLRB1")


DoHeatmap(Avg.seurat, 
          features = TCR.genes,
          draw.lines = FALSE,
          group.colors = clust.cols[keep.index],
          raster = FALSE)

dev.copy(pdf, "Heatmap_TCRG9_TCRD2_KLRB1_expression_gd_MAIT_clusts_imputed.pdf")
dev.off()


# Remove unneeded object
rm(Avg.seurat)

# return to working directory
setwd(working.dir)

```

