---
title: "HNSCC_scTCRseq_Tracking_stimulatable_clones"
author: "Dillon Corvino"
date: "03/02/2020"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---



Built with R version `r getRversion()`

## Setup {.tabset}

### Environment
```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
#setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Load packages
source("R/Load_packages.R", local = knitr::knit_global())

# Create output directories & load custom functions & colour scheme
source("R/Setup.R", local = knitr::knit_global())

# pipeline variables
quick.load <- TRUE
long.compute <- TRUE
do.plot <- TRUE

# load saved seurat object with cluster annotations and imputation performed
if(quick.load){
  seurat.combined <- readRDS("Exported_RDS_files/seurat_combined.rds")
  DefaultAssay(seurat.combined) <- "integrated"
  
  seurat.tcr <- readRDS("Exported_RDS_files/seurat_tcr.rds")
  DefaultAssay(seurat.tcr) <- "integrated"
  
  seurat.tcr.filt <- readRDS("Exported_RDS_files/seurat_tcr_filt.rds")
  DefaultAssay(seurat.tcr.filt) <- "integrated"
  
}

```



## Identifying stimulatable clonotypes {.tabset}

### Evaluating Exhausted clones for the ability to be stimulated
```{r Evaluating_exhausted_clones_for_stimulation}

# set working dir
setwd(working.dir)

# Create output directories
if(!dir.exists("output_tcr/figures/Exhausted_restimability")){
  dir.create("output_tcr/figures/Exhausted_restimability",
             recursive = TRUE)}

output.dir <- "output_tcr/figures/Exhausted_restimability"
setwd(output.dir)

# Save output dir as variable for future use
Exhaust.dir <- output.dir

# Variable info 
query.clusters <- c("Stimulated_exhausted", "Exhausted_1", "Exhausted_2")
source.clusters <- c("Exhausted_1", "Exhausted_2")
target.clusters <- c("Stimulated_exhausted")

# Clone size minimum 
clonesize.min <- 3


# Filter dataset by clonesize minimum
clonotypes <- seurat.tcr.filt@meta.data %>%
  dplyr::filter(Total_clonotype_n > clonesize.min) %>%
  dplyr::select(CTaa, seurat_clusters, condition, Clonotype_overlap)


# Get clonotypes found in Exhausted US subsets that are shared with Stim
Ex.clonotypes <- clonotypes %>%
  dplyr::filter(Clonotype_overlap == "Shared" & condition == "US") %>%
  dplyr::filter(seurat_clusters %in% source.clusters) %>%
  dplyr::select(CTaa)

Ex.clonotypes <- unique(Ex.clonotypes$CTaa)
length(Ex.clonotypes) # 93 clonotypes


# Within Stimulated condition filter dataset by Ex.clonotypes
Query.clonotypes <- clonotypes %>%
  dplyr::filter(Clonotype_overlap == "Shared" & condition == "Stim") %>%
  dplyr::filter(seurat_clusters %in% query.clusters) %>%
  dplyr::filter(CTaa %in% Ex.clonotypes) %>%
  dplyr::select(seurat_clusters, CTaa)

length(Query.clonotypes$CTaa) 
length(unique(Query.clonotypes$CTaa)) # 89 clonotypes // diff in number is due to some clonotypes which have a single cell occuring in US Exhausted cluster but not present in stimulated condition exhausted cluster or Sim_exh clusters.


# To clean up analysis - generated an external function called Clonotype_distribution_function.R

clonotype.freq <- Get.distribution(Query.clonotypes, 
                                   query.clusters = query.clusters,
                                   source.clusters = source.clusters,
                                   target.clusters = target.clusters,
                                   rename.source = "Exhausted",
                                   rename.target = "Stim_Exhausted_1",
                                   threshold.rowval = 3)



# Plot clonotype.freq
barplot(t(clonotype.freq), beside = T, las = 2, legend = T)

dev.copy(pdf, "Barplot_clonotype_freq_distribution.pdf")
dev.off()


# Select clones using freq stimulated
stim.minimum <- 40
unstim.max <- 20

stimulatable.clones <- rownames(clonotype.freq)[clonotype.freq[,"Stim_Exhausted_1"] >= stim.minimum]
unstimulatable.clones <- rownames(clonotype.freq)[clonotype.freq[,"Stim_Exhausted_1"] <= unstim.max]

length(unique(stimulatable.clones)) # 39 clones
length(unique(unstimulatable.clones)) # 26 clones

# Visualise threshold chosen
barplot(t(clonotype.freq),
        beside = TRUE,
        las = 2,
        legend = TRUE) +
  abline(h = c(stim.minimum, unstim.max),
         col = c("Blue", "Red"),
         lwd = "3")

dev.copy(pdf, "Barplot_clonotype_freq_distribution_threshold.pdf")
dev.off()


# Reduce seurat object to these clones in US Exhausted clusters only
clonotypes <- seurat.tcr.filt@meta.data %>%
  dplyr::filter(CTaa %in% c(unstimulatable.clones, stimulatable.clones)) %>%
  dplyr::filter(seurat_clusters %in% source.clusters) %>%
  dplyr::filter(condition == "US") %>%
  dplyr::select(CTaa, seurat_clusters, condition, Clonotype_overlap)

dim(clonotypes) # 338 cells
unique(clonotypes$seurat_clusters)
length(unique(clonotypes$CTaa)) # 65 clones


# Subset seurat object
keep.cells <- rownames(clonotypes)

seurat.exhausted <- subset(seurat.tcr.filt, cells = keep.cells)

# Annotate metadata in seurat object 
meta.data.temp <- seurat.exhausted@meta.data$CTaa
names(meta.data.temp) <- rownames(seurat.exhausted@meta.data)
logic.vec <- meta.data.temp %in% unstimulatable.clones
meta.data.temp[logic.vec] <- "Unstimable"

logic.vec <- meta.data.temp %in% stimulatable.clones
meta.data.temp[logic.vec] <- "Stimable"



seurat.exhausted <- AddMetaData(object = seurat.exhausted, 
                                col.name = "Stim_potential",
                                metadata = meta.data.temp)

table(seurat.exhausted@meta.data$Stim_potential)


# Plot data
DimPlot(seurat.exhausted, 
        pt.size = 2, 
        group.by = "Stim_potential", 
        split.by = "group")


# Perform dif gene expression
Idents(seurat.exhausted) <- seurat.exhausted@meta.data$Stim_potential

DefaultAssay(seurat.exhausted) <- "integrated"

Stim.potential.markers <- FindMarkers(seurat.exhausted, 
                                      ident.1 = "Stimable", 
                                      ident.2 = "Unstimable",
                                      logfc.threshold = 0.25, 
                                      min.pct = 0.1, 
                                      test.use = "wilcox")

Stim.potential.markers <- Stim.potential.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::arrange(desc(avg_log2FC))

dim(Stim.potential.markers)

Stim.potential.markers %>%
  dplyr::top_n(10, avg_log2FC)

Stim.potential.markers %>%
  dplyr::top_n(-10, avg_log2FC)


write.csv(Stim.potential.markers, 
          paste0("DEG_restimability_", stim.minimum, "_", unstim.max, ".csv"))


# Perform UMAP dim reduction 
seurat.exhausted <- RunUMAP(object = seurat.exhausted,
                            reduction = "pca",
                            reduction.key = "Subset_UMAP",
                            dims = 1:20,
                            umap.method = "uwot",
                            n.neighbors = 30, # 5 to 50
                            min.dist = 0.3, # Sensible values are in the range 0.001 to 0.5
                            seed.use = 42)


UMAPPlot(seurat.exhausted, group.by = "group")
UMAPPlot(seurat.exhausted, group.by = "Stim_potential")



#############################################
# Append metadata to larger seurat object
#############################################

# Create metadata for unstimable and stimable clonotypes
meta.data.temp <- seurat.tcr.filt@meta.data$CTaa
names(meta.data.temp) <- rownames(seurat.tcr.filt@meta.data)
logic.vec <- meta.data.temp %in% unstimulatable.clones
meta.data.temp[logic.vec] <- "Unstimable"

logic.vec <- meta.data.temp %in% stimulatable.clones
meta.data.temp[logic.vec] <- "Stimable"

logic.vec <- meta.data.temp %in% c("Unstimable", "Stimable")
meta.data.temp[!logic.vec] <- "Other"

unique(meta.data.temp)


seurat.tcr.filt <- AddMetaData(object = seurat.tcr.filt, 
                          col.name = "Exhausted_Stim_potential",
                          metadata = meta.data.temp)


UMAPPlot(seurat.tcr.filt, 
         group.by = "Exhausted_Stim_potential", 
         pt.size = 2,
         split.by = "condition",
         cols = c("#BEBEBE33", "Blue", "Red"))

dev.copy(pdf, "UMAP_exhausted_stim_potential.pdf")
dev.off()


# Remove un-needed variables
rm(seurat.exhausted)

# return to working directory 
setwd(working.dir)

```

### Evaluating Cytotoxic and Type I IFN clones for the ability to be stimulated
```{r Evaluating_Cytotoxic_and_Type_I_IFN_clones_for_stimulation}

# set working dir
setwd(working.dir)

# Create output directories
if(!dir.exists("output_tcr/figures/Cytotoxic_TypeI_restimability")){
  dir.create("output_tcr/figures/Cytotoxic_TypeI_restimability",
             recursive = TRUE)}

output.dir <- "output_tcr/figures/Cytotoxic_TypeI_restimability"
setwd(output.dir)

# Save output dir as variable for future use
CytoIFN.dir <- output.dir


# Variable info 
query.clusters <- c("Stimulated_1", "Cytotoxic", "Type_I_IFN")
source.clusters <- c("Cytotoxic", "Type_I_IFN")
target.clusters <- c("Stimulated_1")

# Clone size minimum 
clonesize.min <- 3


# Filter dataset by clonesize minimum
clonotypes <- seurat.tcr.filt@meta.data %>%
  dplyr::filter(Total_clonotype_n > clonesize.min) %>%
  dplyr::select(CTaa, seurat_clusters, condition, Clonotype_overlap)


# Get clonotypes found in Cytotoxic or Type I IFN US subsets that are shared with Stim
CytoIFN.clonotypes <- clonotypes %>%
  dplyr::filter(Clonotype_overlap == "Shared" & condition == "US") %>%
  dplyr::filter(seurat_clusters %in% source.clusters) %>%
  dplyr::select(CTaa)

CytoIFN.clonotypes <- unique(CytoIFN.clonotypes$CTaa)
length(CytoIFN.clonotypes) # 103 clonotypes


# Within Stimulated condition filter dataset by Ex.clonotypes
Query.clonotypes <- clonotypes %>%
  dplyr::filter(Clonotype_overlap == "Shared" & condition == "Stim") %>%
  dplyr::filter(seurat_clusters %in% query.clusters) %>%
  dplyr::filter(CTaa %in% CytoIFN.clonotypes) %>%
  dplyr::select(seurat_clusters, CTaa)

length(Query.clonotypes$CTaa) 
length(unique(Query.clonotypes$CTaa)) # 97 clonotypes




# To clean up analysis - generated an external function 

clonotype.freq <- Get.distribution(Query.clonotypes, 
                                   query.clusters = query.clusters,
                                   source.clusters = source.clusters,
                                   target.clusters = target.clusters,
                                   rename.source = "CytoIFN",
                                   rename.target = "Stim_1",
                                   threshold.rowval = 3)


# Plot clonotype.freq
barplot(t(clonotype.freq), beside = T, las = 2, legend = T)

dev.copy(pdf, "Barplot_clonotype_freq_distribution.pdf")
dev.off()


# Select clones using freq stimulated
stim.minimum <- 40
unstim.max <- 20

stimulatable.clones <- rownames(clonotype.freq)[clonotype.freq[,"Stim_1"] >= stim.minimum]
unstimulatable.clones <- rownames(clonotype.freq)[clonotype.freq[, "Stim_1"] <= unstim.max]

length(unique(stimulatable.clones)) # 23 clones
length(unique(unstimulatable.clones)) # 21 clones

# Visualise threshold chosen
barplot(t(clonotype.freq),
        beside = TRUE,
        las = 2,
        legend = TRUE) +
  abline(h = c(stim.minimum, unstim.max),
         col = c("Blue", "Red"),
         lwd = "3")

dev.copy(pdf, "Barplot_clonotype_freq_distribution_threshold.pdf")
dev.off()


# Reduce seurat object to these clones in US Exhausted clusters only
clonotypes <- seurat.tcr.filt@meta.data %>%
  dplyr::filter(CTaa %in% c(unstimulatable.clones, stimulatable.clones)) %>%
  dplyr::filter(seurat_clusters %in% source.clusters) %>%
  dplyr::filter(condition == "US") %>%
  dplyr::select(CTaa, seurat_clusters, condition, Clonotype_overlap)

dim(clonotypes) # 287 cells
unique(clonotypes$seurat_clusters)
length(unique(clonotypes$CTaa)) # 44 clones


# Subset seurat object
keep.cells <- rownames(clonotypes)

seurat.cytoIFN <- subset(seurat.tcr.filt, cells = keep.cells)

# Annotate metadata in seurat object 
meta.data.temp <- seurat.cytoIFN@meta.data$CTaa
names(meta.data.temp) <- rownames(seurat.cytoIFN@meta.data)
logic.vec <- meta.data.temp %in% unstimulatable.clones
meta.data.temp[logic.vec] <- "Unstimable"

logic.vec <- meta.data.temp %in% stimulatable.clones
meta.data.temp[logic.vec] <- "Stimable"



seurat.cytoIFN <- AddMetaData(object = seurat.cytoIFN, 
                                col.name = "Stim_potential",
                                metadata = meta.data.temp)

table(seurat.cytoIFN@meta.data$Stim_potential)


# Plot data
DimPlot(seurat.cytoIFN, 
        pt.size = 2, 
        group.by = "Stim_potential", 
        split.by = "group")


# Perform dif gene expression
Idents(seurat.cytoIFN) <- seurat.cytoIFN@meta.data$Stim_potential

DefaultAssay(seurat.cytoIFN) <- "integrated"

Stim.potential.markers <- FindMarkers(seurat.cytoIFN, 
                                      ident.1 = "Stimable", 
                                      ident.2 = "Unstimable",
                                      logfc.threshold = 0.25, 
                                      min.pct = 0.1, 
                                      test.use = "wilcox")


Stim.potential.markers <- Stim.potential.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::arrange(desc(avg_log2FC))

Stim.potential.markers %>%
  dplyr::top_n(10, avg_log2FC)

Stim.potential.markers %>%
  dplyr::top_n(-10, avg_log2FC)


write.csv(Stim.potential.markers, 
          paste0("DEG_restimability_", stim.minimum, "_", unstim.max, ".csv"))


# Perform UMAP dim reduction 
seurat.cytoIFN <- RunUMAP(object = seurat.cytoIFN,
                            reduction = "pca",
                            reduction.key = "Subset_UMAP",
                            dims = 1:20,
                            umap.method = "uwot",
                            n.neighbors = 30, # 5 to 50
                            min.dist = 0.3, # Sensible values are in the range 0.001 to 0.5
                            seed.use = 42)


UMAPPlot(seurat.cytoIFN, group.by = "group")
UMAPPlot(seurat.cytoIFN, group.by = "Stim_potential")



#############################################
# Append metadata to larger seurat object
#############################################

# Create metadata for unstimable and stimable clonotypes
meta.data.temp <- seurat.tcr.filt@meta.data$CTaa
names(meta.data.temp) <- rownames(seurat.tcr.filt@meta.data)
logic.vec <- meta.data.temp %in% unstimulatable.clones
meta.data.temp[logic.vec] <- "Unstimable"

logic.vec <- meta.data.temp %in% stimulatable.clones
meta.data.temp[logic.vec] <- "Stimable"

logic.vec <- meta.data.temp %in% c("Unstimable", "Stimable")
meta.data.temp[!logic.vec] <- "Other"

unique(meta.data.temp)


seurat.tcr.filt <- AddMetaData(object = seurat.tcr.filt, 
                          col.name = "CytoIFN_Stim_potential",
                          metadata = meta.data.temp)


UMAPPlot(seurat.tcr.filt, 
         group.by = "CytoIFN_Stim_potential", 
         pt.size = 2,
         split.by = "condition",
         cols = c("#BEBEBE33", "Blue", "Red"))

dev.copy(pdf, "UMAP_CytoIFN_stim_potential.pdf")
dev.off()


# Remove un-needed variables
rm(seurat.cytoIFN)

# return to working directory 
setwd(working.dir)

```

### Any clonotype movement
```{r Any_clonotype_movement}

# set working dir
setwd(working.dir)

# Create output directories
if(!dir.exists("output_tcr/figures/Clonotype_movement")){
  dir.create("output_tcr/figures/Clonotype_movement",
             recursive = TRUE)}

output.dir <- "output_tcr/figures/Clonotype_movement"
setwd(output.dir)

# Save output dir as variable for future use
TotalMovement.dir <- output.dir


# Variable info 
query.clusters <- c("Stimulated_1", "Cytotoxic", "Type_I_IFN", "Stimulated_exhausted", "Exhausted_1", "Exhausted_2")
source.clusters <- c("Cytotoxic", "Type_I_IFN", "Exhausted_1", "Exhausted_2")
target.clusters <- c("Stimulated_1", "Stimulated_exhausted")

# Clone size minimum 
clonesize.min <- 3


# Filter dataset by clonesize minimum
clonotypes <- seurat.tcr.filt@meta.data %>%
  dplyr::filter(Total_clonotype_n > clonesize.min) %>%
  dplyr::select(CTaa, seurat_clusters, condition, Clonotype_overlap)


# Get clonotypes found in Cytotoxic or Type I IFN US subsets that are shared with Stim
CytoIFN.clonotypes <- clonotypes %>%
  dplyr::filter(Clonotype_overlap == "Shared" & condition == "US") %>%
  dplyr::filter(seurat_clusters %in% source.clusters) %>%
  dplyr::select(CTaa)

CytoIFN.clonotypes <- unique(CytoIFN.clonotypes$CTaa)
length(CytoIFN.clonotypes) # 103 clonotypes


# Within Stimulated condition filter dataset by Ex.clonotypes
Query.clonotypes <- clonotypes %>%
  dplyr::filter(Clonotype_overlap == "Shared" & condition == "Stim") %>%
  dplyr::filter(seurat_clusters %in% query.clusters) %>%
  dplyr::filter(CTaa %in% CytoIFN.clonotypes) %>%
  dplyr::select(seurat_clusters, CTaa)

length(Query.clonotypes$CTaa) 
length(unique(Query.clonotypes$CTaa)) # 97 clonotypes




# To clean up analysis - generated an external function called Get.distribution

clonotype.freq <- Get.distribution(Query.clonotypes, 
                                   query.clusters = query.clusters,
                                   source.clusters = source.clusters,
                                   target.clusters = target.clusters,
                                   rename.source = "CytoIFN_or_Exhaust",
                                   rename.target = "Stim",
                                   threshold.rowval = 3)


# Plot clonotype.freq
barplot(t(clonotype.freq), beside = T, las = 2, legend = T)

dev.copy(pdf, "Barplot_clonotype_freq_distribution.pdf")
dev.off()


# Select clones using freq stimulated
stim.minimum <- 40
unstim.max <- 20

stimulatable.clones <- rownames(clonotype.freq)[clonotype.freq[,"Stim"] >= stim.minimum]
unstimulatable.clones <- rownames(clonotype.freq)[clonotype.freq[, "Stim"] <= unstim.max]

length(unique(stimulatable.clones)) # 64 clones
length(unique(unstimulatable.clones)) # 48 clones

# Visualise threshold chosen
barplot(t(clonotype.freq),
        beside = TRUE,
        las = 2,
        legend = TRUE) +
  abline(h = c(stim.minimum, unstim.max),
         col = c("Blue", "Red"),
         lwd = "3")

dev.copy(pdf, "Barplot_clonotype_freq_distribution_threshold.pdf")
dev.off()


# Reduce seurat object to these clones in US Exhausted clusters only
clonotypes <- seurat.tcr.filt@meta.data %>%
  dplyr::filter(CTaa %in% c(unstimulatable.clones, stimulatable.clones)) %>%
  dplyr::filter(seurat_clusters %in% source.clusters) %>%
  dplyr::filter(condition == "US") %>%
  dplyr::select(CTaa, seurat_clusters, condition, Clonotype_overlap)

dim(clonotypes) # 634 cells
unique(clonotypes$seurat_clusters)
length(unique(clonotypes$CTaa)) # 112 clones


# Subset seurat object
keep.cells <- rownames(clonotypes)

seurat.cytoIFN <- subset(seurat.tcr.filt, cells = keep.cells)

# Annotate metadata in seurat object 
meta.data.temp <- seurat.cytoIFN@meta.data$CTaa
names(meta.data.temp) <- rownames(seurat.cytoIFN@meta.data)
logic.vec <- meta.data.temp %in% unstimulatable.clones
meta.data.temp[logic.vec] <- "Unstimable"

logic.vec <- meta.data.temp %in% stimulatable.clones
meta.data.temp[logic.vec] <- "Stimable"



seurat.cytoIFN <- AddMetaData(object = seurat.cytoIFN, 
                                col.name = "Stim_potential",
                                metadata = meta.data.temp)

table(seurat.cytoIFN@meta.data$Stim_potential)


# Plot data
DimPlot(seurat.cytoIFN, 
        pt.size = 2, 
        group.by = "Stim_potential", 
        split.by = "group")


# Perform dif gene expression
Idents(seurat.cytoIFN) <- seurat.cytoIFN@meta.data$Stim_potential

DefaultAssay(seurat.cytoIFN) <- "integrated"

Stim.potential.markers <- FindMarkers(seurat.cytoIFN, 
                                      ident.1 = "Stimable", 
                                      ident.2 = "Unstimable",
                                      logfc.threshold = 0.25, 
                                      min.pct = 0.1, 
                                      test.use = "wilcox")

Stim.potential.markers <- Stim.potential.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::arrange(desc(avg_log2FC))

Stim.potential.markers %>%
  dplyr::top_n(10, avg_log2FC)

Stim.potential.markers %>%
  dplyr::top_n(-10, avg_log2FC)


write.csv(Stim.potential.markers, 
          paste0("DEG_restimability_", stim.minimum, "_", unstim.max, ".csv"))


# Perform UMAP dim reduction 
seurat.cytoIFN <- RunUMAP(object = seurat.cytoIFN,
                            reduction = "pca",
                            reduction.key = "Subset_UMAP",
                            dims = 1:20,
                            umap.method = "uwot",
                            n.neighbors = 30, # 5 to 50
                            min.dist = 0.3, # Sensible values are in the range 0.001 to 0.5
                            seed.use = 42)


UMAPPlot(seurat.cytoIFN, group.by = "group")
UMAPPlot(seurat.cytoIFN, group.by = "Stim_potential")



#############################################
# Append metadata to larger seurat object
#############################################

# Create metadata for unstimable and stimable clonotypes
meta.data.temp <- seurat.tcr.filt@meta.data$CTaa
names(meta.data.temp) <- rownames(seurat.tcr.filt@meta.data)
logic.vec <- meta.data.temp %in% unstimulatable.clones
meta.data.temp[logic.vec] <- "Unstimable"

logic.vec <- meta.data.temp %in% stimulatable.clones
meta.data.temp[logic.vec] <- "Stimable"

logic.vec <- meta.data.temp %in% c("Unstimable", "Stimable")
meta.data.temp[!logic.vec] <- "Other"

unique(meta.data.temp)


seurat.tcr.filt <- AddMetaData(object = seurat.tcr.filt, 
                          col.name = "Total_Stim_potential",
                          metadata = meta.data.temp)


UMAPPlot(seurat.tcr.filt, 
         group.by = "Total_Stim_potential", 
         pt.size = 2,
         split.by = "condition",
         cols = c("#BEBEBE33", "Blue", "Red"))

dev.copy(pdf, "UMAP_total_stim_potential.pdf")
dev.off()


# Remove un-needed variables
rm(seurat.cytoIFN)

# return to working directory 
setwd(working.dir)

```

### Vis DEGs responsible for clonotype movement
```{r Visualise_Clonotype_movement_DEGs}

# Create a variable of stim potential plus condition
stim.var <- c("Exhausted_Stim_potential", "CytoIFN_Stim_potential", "Total_Stim_potential")


for(i in seq_along(stim.var)){
  
  seurat.tcr.filt@meta.data[[paste0(stim.var[i], "_condition")]] <- paste0(seurat.tcr.filt@meta.data[[stim.var[i]]],
                                                                           "_",
                                                                           seurat.tcr.filt@meta.data$condition)
  
  
  seurat.tcr.filt@meta.data[[paste0(stim.var[i], "_condition")]]  <- factor(seurat.tcr.filt@meta.data[[paste0(stim.var[i], "_condition")]] , 
                                                                            levels = c("Unstimable_US", "Unstimable_Stim",
                                                                                       "Stimable_US", "Stimable_Stim",
                                                                                       "Other_US", "Other_Stim"))
  
}




path.var <- c(Exhaust.dir, CytoIFN.dir, TotalMovement.dir)
path.var <- paste0(path.var, "/")
stim.name.var <- paste0(stim.var, "_condition")

# Set assay to imputed
DefaultAssay(seurat.tcr.filt) <- "alra"

# or some unknown reason, this loop sometimes fails and sometimes works. 
# I can't reliabily reproduce the problem, and running each element outside the loop works perfectly
# running small loop over first data input lines also works perfectly every time
# I CANT WORK OUT WHAT THE HELL IS HAPPENING 
# Therefore, just run the loop a couple times till it works

for(i in seq_along(path.var)){
  

  # Read in DEG data
  
  file.name.var <- list.files(path = path.var[i],
                              pattern = ".csv")
  
  print(paste0(file.name.var))

  DEG.df <- read.csv(paste0(path.var[i], file.name.var))
  

  # Set idents to stim.name.var 
  Idents(seurat.tcr.filt) <- seurat.tcr.filt@meta.data[[stim.name.var[i]]]
  
  
  # Subset seurat.object by just stim and unstim classified clones/cells
  temp.seurat <- subset(seurat.tcr.filt, ident = c("Unstimable_US", "Stimable_US",
                                                   "Unstimable_Stim", "Stimable_Stim"))
  
  # Get average gene expression
  average.seurat <- AverageExpression(temp.seurat,
                                      assay = "alra",
                                      slot = "data",
                                      verbose = TRUE,
                                      return.seurat = TRUE)
  
  
  # Subset DEG list for top DEGs
  goi <- DEG.df %>% 
    filter(avg_log2FC > 1 | avg_log2FC < -1)
  
  # Remove TCR genes 
  pattern.var <- c("TRA", "TRB")
  logic.var <- grepl(paste0(pattern.var, collapse = "|"), goi$X)
  goi <- goi$X[!logic.var]
  
  
  # Generate heatmap of top DEGs
  
  print(DoHeatmap(average.seurat, 
                  features = c(goi), 
                  raster = FALSE,
                  draw.lines = FALSE))
  
  dev.copy(pdf, paste0(path.var[i], "Heatmap_top_DEGs.pdf"))
  dev.off()
  
  
  
  
  # Create VlnPlots for top DEGs
  Vlnplot.dir <- paste0(path.var[i], "VlnPlots")
  
  if(!dir.exists(Vlnplot.dir)){
    dir.create(Vlnplot.dir)
  }
  
  
  for(j in seq_along(goi)){
    
    print(VlnPlot(temp.seurat, 
                  goi[j], 
          pt.size = 0.1))
    
    dev.copy(pdf, paste0(Vlnplot.dir, "/Vlnplot_", goi[j], ".pdf"))
    dev.off()
  }
  
  
}


```

### TCR gene usage categorised by Restim status
```{r Restim_status_gene_usage}

########################
# Heatmap functions
########################

hclust.func <- function(x){
  hclust(x, method = 'ward.D2')
}

dist.func <- function(x){
  dist(x,method = 'binary')
}

col.palette <- colorRampPalette(brewer.pal(8, "RdYlBu"))(600)





path.var <- c(Exhaust.dir, CytoIFN.dir, TotalMovement.dir)
path.var <- paste0(path.var, "/")
stim.var <- c("Exhausted_Stim_potential", "CytoIFN_Stim_potential", "Total_Stim_potential")


for(i in seq_along(path.var)){

# Create output path
  heatmap.dir <- paste0(path.var[i], "TCR_usage")
  
  if(!dir.exists(heatmap.dir)){
    dir.create(heatmap.dir,
               recursive = TRUE)
  }
  

######################################################
# Plot V/J gene usage as a log2() of cell number
######################################################

# TRA-V gene usage

x <- table(seurat.tcr.filt@meta.data$TRAV, seurat.tcr.filt@meta.data[[stim.var[i]]])

print(heatmap.2(as.matrix(log2(x+1)), 
          scale = "column",
          trace = "none",
          Colv = FALSE,
          cexRow = 0.5, 
          margins = c(8, 5),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette)))     

dev.copy(pdf, paste0(heatmap.dir, "/", "Heatmap_TRA_V_usage.pdf"))
dev.off()

# TRA-J gene usage

x <- table(seurat.tcr.filt@meta.data$TRAJ, seurat.tcr.filt@meta.data[[stim.var[i]]])

print(heatmap.2(as.matrix(log2(x+1)), 
          scale = "column",
          trace = "none",
          Colv = FALSE,
          cexRow = 0.5, 
          margins = c(8, 5),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette)))     

dev.copy(pdf, paste0(heatmap.dir, "/", "Heatmap_TRA_J_usage.pdf"))
dev.off()


# TRB-V gene usage

x <- table(seurat.tcr.filt@meta.data$TRBV, seurat.tcr.filt@meta.data[[stim.var[i]]])

print(heatmap.2(as.matrix(log2(x+1)), 
          scale = "column",
          trace = "none",
          Colv = FALSE,
          cexRow = 0.5, 
          margins = c(8, 5),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette)))     

dev.copy(pdf, paste0(heatmap.dir, "/", "Heatmap_TRB_V_usage.pdf"))
dev.off()



# TRB-J gene usage
x <- table(seurat.tcr.filt@meta.data$TRBJ, seurat.tcr.filt@meta.data[[stim.var[i]]])


print(heatmap.2(as.matrix(log2(x+1)), 
          scale = "column",
          trace = "none",
          Colv = FALSE,
          cexRow = 0.5, 
          margins = c(8, 5),
          cexCol = 0.8, 
          distfun = dist.func,
          hclustfun = hclust.func,
          srtCol = 90,
          col = rev(col.palette)))     

dev.copy(pdf, paste0(heatmap.dir, "/", "Heatmap_TRB_J_usage.pdf"))
dev.off()

}

```



### Export tcr seurat object
```{r export_RDS}

  saveRDS(seurat.tcr, file = "Exported_RDS_files/seurat_tcr.rds")
  saveRDS(seurat.tcr.filt, file = "Exported_RDS_files/seurat_tcr_filt.rds")
  
```

