---
title: "Immune Landscape of Viral- and Carcinogen-Driven Head and Neck Cancer"
author: "DillonCorvino"
date: "22/06/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

## Dataset {.tabset}

### Dataset info
```{r Dataset_details}

# Manuscript title: "Immune Landscape of Viral- and Carcinogen-Driven Head and Neck Cancer"
# Authors: Anthony R. Cillo, Cornelius H.L. Ku€rten, Tracy Tabib, ..., Tullia C. Bruno, Robert L. Ferris, Dario A.A. Vignali
# Journal: Immunity
# Year: 2020
# DOI: 10.1016/j.immuni.2019.11.014
## Abstract: Head and neck squamous cell carcinoma (HNSCC) arises through exposure to environmental carcino- gens or malignant transformation by human papillomavirus (HPV). Here, we assessed the transcriptional profiles of 131,224 single cells from peripheral and intra-tumoral immune populations from patients with HPV– and HPV+ HNSCC and healthy donors. Immune cells within tumors of HPV and HPV HNSCC displayed a spectrum of transcriptional signatures, with helper CD4+ T cells and B cells being relatively divergent and CD8+ T cells and CD4+ regulatory T cells being relatively similar. Transcriptional results were contextualized through multispectral immuno- fluorescence analyses and evaluating putative cell- cell communication based on spatial proximity. These analyses defined a gene expression signature associated with CD4+ T follicular helper cells that is associated with longer progression-free survival in HNSCC patients. The datasets and analytical ap- proaches herein provide a resource for the further study of the impact of immune cells on viral- and carcinogen-induced cancers.


# Total of 63 samples were analyzed by scRNAseq. 
# 26 HNSCC patients (18 HPVNeg and 8 HPVPos)
# HNSCC patients have paired PBMC and TIL samples
# 6 PBMC from healthy donors
# 5 healthy donor tonsils 

```

## Set up {.tabset}

### Environment
```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)


knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Load packages
source("R/Load_packages.R", local = knitr::knit_global())

# Create output directories & load custom functions & colour scheme
source("R/Setup.R", local = knitr::knit_global())


# pipeline variables
quick.load <- FALSE
CD8.only <- TRUE # Should Cillo dataset be subsetted to only use cells designated as CD8


```

### Reading in data
```{r reading_data}

############################################################
# Data comes from Cillo et al., 2020 Immunity
# GSE139324
############################################################

###################################
# Read and formatt TIL samples 
###################################

# Load files
named.vect <- list.dirs("Data/Cillo_et_al_data/HNSCC_TILS/")
named.vect <- named.vect[-1]
names(named.vect) <- gsub("HNSCC_TILS/", "", named.vect)

HNSCC.TILs <- Read10X(data.dir = named.vect)

# 26 TIL samples from HNSCC patients


# Format barcode IDs to allow matching with cillo metadata
colnames(HNSCC.TILs)
colnames(HNSCC.TILs) <- gsub("Data/Cillo_et_al_data//", "", colnames(HNSCC.TILs))
colnames(HNSCC.TILs) <- gsub("-1$", "", colnames(HNSCC.TILs))
colnames(HNSCC.TILs)


# Read in metadata provided by Anthony Cillo 
meta.data <- read.delim("Data/Cillo_et_al_data/Cillo_HNSCC_cell_cluster_annotations.21.01.15.tsv")


head(meta.data)
sort(unique(meta.data$overall_clusterID))
unique(meta.data$patient_id)
unique(meta.data$overall_clusterID)
unique(meta.data$cell_type)
unique(meta.data$sample_origin)
unique(meta.data$cd8_cells_clusterID)

# Filter metadata for TIL annotations 
meta.data %>% 
  dplyr::filter(sample_origin == "TIL") -> TIL.metadata

# format metadata
sum(duplicated(TIL.metadata$cell_barcode))

# Need to append PT.ID to make barcode unique 
head(TIL.metadata)

as.factor(TIL.metadata$patient_id) # 26 levels 

TIL.metadata$Pt_ID <- gsub("HNSCC_", "Pt", TIL.metadata$patient_id)

head(TIL.metadata)
as.factor(TIL.metadata$Pt_ID)

# Attach Pt_ID to barcode_ID to make it unique 
TIL.metadata$barcode <- paste0(TIL.metadata$Pt_ID, "_", TIL.metadata$cell_barcode)

sum(duplicated(TIL.metadata$barcode))

rownames(TIL.metadata) <- TIL.metadata$barcode



# Initialize the Seurat object with the raw (non-normalized data).  Keep all
# genes expressed in >= 3 cells. Keep all cells with at
# least 200 detected genes
seurat.Cillo.TIL <- CreateSeuratObject(counts = HNSCC.TILs,
                                       meta.data = TIL.metadata,
                                       min.cells = 3,
                                       min.features = 200)


seurat.Cillo.TIL 
# 23,389 genes 60,923 cells

seurat.Cillo.TIL@meta.data



if(CD8.only){
  
  unique(seurat.Cillo.TIL@meta.data$cell_type)
  
  sum(is.na(seurat.Cillo.TIL@meta.data$cell_type)) # 5,329 cells have no cell type origin 
  
  
  Idents(seurat.Cillo.TIL) <- seurat.Cillo.TIL@meta.data$cell_type
  
  seurat.Cillo.TIL <- subset(seurat.Cillo.TIL, 
                             idents = "cd8.cells")
  
  seurat.Cillo.TIL # 18,737 cells and 23,389 genes
  
  # Remove cells that do not have a CD8.cluster ID
  na.logic <- is.na(seurat.Cillo.TIL@meta.data$cd8_cells_clusterID)
  keep.cells <- rownames(seurat.Cillo.TIL@meta.data[!na.logic, ])
  
  seurat.Cillo.TIL <- subset(seurat.Cillo.TIL, 
                             cells = keep.cells)
}
  

```
  
## QC and Normalisation {.tabset}
  
### QC & Normalisation
```{r QC_and_normalisation}

#  Get percentage Mitochondria gene expression
seurat.Cillo.TIL[["percent.mito"]] <- PercentageFeatureSet(seurat.Cillo.TIL, pattern = "^MT-")


##################################
# Plot mitochondria percentage
##################################

# nFeature & nCount & Mito 
VlnPlot(object = seurat.Cillo.TIL, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"))

# Mito 
VlnPlot(seurat.Cillo.TIL, features = "percent.mito", y.max = 20)


########################################
# Count vs Mito & Count vs Feature
########################################

plot1 <- FeatureScatter(seurat.Cillo.TIL, feature1 = "nCount_RNA", feature2 = "percent.mito")
plot2 <- FeatureScatter(seurat.Cillo.TIL, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
CombinePlots(plots = list(plot1, plot2))


############################
# Structure and metadata 
############################

head(seurat.Cillo.TIL@meta.data)
str(seurat.Cillo.TIL@meta.data)

###################
#  Filter cells
###################

# Seurat recommends removing cells with <200 or >2,500 genes or >10% mito genes

Cillo.TIL.filt <- subset(seurat.Cillo.TIL,
                         subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mito < 10)

# Post-filtering metrics

seurat.Cillo.TIL # 23,389 features across 60,923 samples
Cillo.TIL.filt # 23,389 features across 55,272 samples

# Therefore
60923 - 55272
# 5,651 cells were deleted

##########################################
# Visualise QC metrics after filtering
##########################################

# Before Filtering
VlnPlot(object = seurat.Cillo.TIL, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), pt.size = 0)

# After filtering
VlnPlot(object = Cillo.TIL.filt, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), pt.size = 0)

################################################
# Normalise, scale & find variable genes 
################################################

# Normalise
Cillo.TIL.filt <- NormalizeData(object = Cillo.TIL.filt,
                                assay = "RNA",
                                normalization.method = "LogNormalize",
                                scale.factor = 10000,
                                verbose = TRUE)

# Find variable genes
Cillo.TIL.filt <- FindVariableFeatures(object = Cillo.TIL.filt,
                                       verbose = TRUE)

# Scale data
Cillo.TIL.filt <- ScaleData(object = Cillo.TIL.filt,
                            assay = "RNA",
                            vars.to.regress = c("nCount_RNA", "percent.mito"),
                            model.use = "linear",
                            do.scale = TRUE,
                            do.center = TRUE,
                            verbose = TRUE)


```
  


## Dim reduction {.tabset}
  
### Dim reduction
```{r Dimreduction}

DefaultAssay(Cillo.TIL.filt) <- "RNA"


# Run PCA
Cillo.TIL.filt <- RunPCA(Cillo.TIL.filt, 
                            npcs = 30, 
                            verbose = TRUE)


# Run tSNE
Cillo.TIL.filt <- RunTSNE(object = Cillo.TIL.filt,
                             reduction = "pca",
                             dims = 1:20,
                             seed.use = 42)


# Run UMAP
Cillo.TIL.filt <- RunUMAP(object = Cillo.TIL.filt,
                          reduction = "pca",
                          dims = 1:20,
                          umap.method = "uwot",
                          n.neighbors = 50, 
                          min.dist = 0.8, 
                          seed.use = 42)

  
```
  
### Visualise UMAP and TSNE plots
```{r visualise_dims}

if(CD8.only){
  
  # Plot UMAP
  UMAPPlot(Cillo.TIL.filt,
           label = T, 
           order = T, 
           pt.size = 1,
           group.by = "cd8_cells_clusterID")
  
  # Plot TSNE
  TSNEPlot(Cillo.TIL.filt,
           label = T, 
           order = T, 
           pt.size = 1,
           group.by = "cd8_cells_clusterID")
  
}else{
  
  # Plot UMAP
  UMAPPlot(Cillo.TIL.filt,
           label = T, 
           order = T, 
           pt.size = 1,
           group.by = "cell_type")
  
  # Plot TSNE
  TSNEPlot(Cillo.TIL.filt,
           label = T, 
           order = T, 
           pt.size = 1,
           group.by = "cell_type")
  
}



```
 
### Add Cillo FItSNE data
```{r add_cillo_FItSNE}



meta.data <- read.csv("Data/Cillo_et_al_data/Archive_older_potentially_incorrect_metadata/Cillo_HNSCC_cell_annotations_with_FItSNE.tsv", 
                      sep = "\t")


if(CD8.only){
  
  # filter metadata
  meta.data <- meta.data %>%
    dplyr::filter(cell_type == "cd8.cells" & sample_origin == "TIL")
  
  
  # Make barcode unique 
  x <- gsub("HNSCC_", "Pt", meta.data$sample)
  x <- gsub("_TIL", "", x)
  
  meta.data$Pt <- x
  
  meta.data$barcode <- paste0(meta.data$Pt, "_", meta.data$cell_barcode)
  
  
  # check barcode is now unique 
  sum(duplicated(meta.data$barcode))
  length(meta.data$barcode) # 19,194 this is larger than dataset probably because data has been filtered  
  
  
  keep.logic <- meta.data$barcode %in% Cillo.TIL.filt@meta.data$barcode
  meta.data <- meta.data[keep.logic, ]
  
  
  x <- match(meta.data$barcode, Cillo.TIL.filt@meta.data$barcode)
  
  Cillo.TIL.filt@meta.data$FItSNE_1 <- meta.data$FItSNE_1[x]
  Cillo.TIL.filt@meta.data$FItSNE_2 <- meta.data$FItSNE_2[x]
  
  keep.vec <- grepl("FItSNE_", colnames(Cillo.TIL.filt@meta.data))
  
  cillo_tsne <- Cillo.TIL.filt@meta.data[,keep.vec]
  cillo_tsne <- as.matrix(cillo_tsne)
  
  # Add data as a dim reduction
  Cillo.TIL.filt[["FItSNE"]] <- CreateDimReducObject(embeddings = cillo_tsne,
                                                     key = "FItSNE_",
                                                     assay = DefaultAssay(Cillo.TIL.filt))
  
  # We can now use this as you would any other dimensional reduction in all downstream functions
  DimPlot(Cillo.TIL.filt, 
          reduction = "FItSNE",
          pt.size = 0.5, 
          group.by = "cd8_cells_clusterID")
  
  
}else{
  
  # filter metadata
  meta.data <- meta.data %>%
    dplyr::filter(sample_origin == "TIL")
  
  
  # Make barcode unique 
  x <- gsub("HNSCC_", "Pt", meta.data$sample)
  x <- gsub("_TIL", "", x)
  
  meta.data$Pt <- x
  
  meta.data$barcode <- paste0(meta.data$Pt, "_", meta.data$cell_barcode)
  
  
  # check barcode is now unique 
  sum(duplicated(meta.data$barcode))
  length(meta.data$barcode)   
  
  
  keep.logic <- meta.data$barcode %in% Cillo.TIL.filt@meta.data$barcode
  meta.data <- meta.data[keep.logic, ]
  
  
  x <- match(meta.data$barcode, Cillo.TIL.filt@meta.data$barcode)
  
  Cillo.TIL.filt@meta.data$FItSNE_1 <- meta.data$FItSNE_1[x]
  Cillo.TIL.filt@meta.data$FItSNE_2 <- meta.data$FItSNE_2[x]
  
  keep.vec <- grepl("FItSNE_", colnames(Cillo.TIL.filt@meta.data))
  
  cillo_tsne <- Cillo.TIL.filt@meta.data[,keep.vec]
  cillo_tsne <- as.matrix(cillo_tsne)
  
  # Add data as a dim reduction
  Cillo.TIL.filt[["FItSNE"]] <- CreateDimReducObject(embeddings = cillo_tsne,
                                                     key = "FItSNE_",
                                                     assay = DefaultAssay(Cillo.TIL.filt))
  
  # We can now use this as you would any other dimensional reduction in all downstream functions
  DimPlot(Cillo.TIL.filt, 
          reduction = "FItSNE",
          pt.size = 0.5, 
          group.by = "cell_type")
  
}


```
  
### Export seurat object
```{r export_seurat}
  
# Write dataset to fil
if(CD8.only){
  saveRDS(Cillo.TIL.filt, file = "Exported_RDS_files/Cillo_TIL_CD8_only_filt.rds")
}else{
  saveRDS(Cillo.TIL.filt, file = "Exported_RDS_files/Cillo_TIL_filt.rds")
}

```
  
  
  