---
title: "Immune Landscape of Viral- and Carcinogen-Driven Head and Neck Cancer"
author: "DillonCorvino"
date: "22/06/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

## Dataset {.tabset}

### Dataset info
```{r Dataset_details}

# Manuscript title: "Immune Landscape of Viral- and Carcinogen-Driven Head and Neck Cancer"
# Authors: Anthony R. Cillo, Cornelius H.L. Ku€rten, Tracy Tabib, ..., Tullia C. Bruno, Robert L. Ferris, Dario A.A. Vignali
# Journal: Immunity
# Year: 2020
# DOI: 10.1016/j.immuni.2019.11.014
## Abstract: Head and neck squamous cell carcinoma (HNSCC) arises through exposure to environmental carcino- gens or malignant transformation by human papillomavirus (HPV). Here, we assessed the transcriptional profiles of 131,224 single cells from peripheral and intra-tumoral immune populations from patients with HPV– and HPV+ HNSCC and healthy donors. Immune cells within tumors of HPV and HPV HNSCC displayed a spectrum of transcriptional signatures, with helper CD4+ T cells and B cells being relatively divergent and CD8+ T cells and CD4+ regulatory T cells being relatively similar. Transcriptional results were contextualized through multispectral immuno- fluorescence analyses and evaluating putative cell- cell communication based on spatial proximity. These analyses defined a gene expression signature associated with CD4+ T follicular helper cells that is associated with longer progression-free survival in HNSCC patients. The datasets and analytical ap- proaches herein provide a resource for the further study of the impact of immune cells on viral- and carcinogen-induced cancers.


# Total of 63 samples were analyzed by scRNAseq. 
# 26 HNSCC patients (18 HPVNeg and 8 HPVPos)
# HNSCC patients have paired PBMC and TIL samples
# 6 PBMC from healthy donors
# 5 healthy donor tonsils 

```

## Set up {.tabset}

### Environment
```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)


knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Load packages
source("R/Load_packages.R", local = knitr::knit_global())

# Create output directories & load custom functions & colour scheme
source("R/Setup.R", local = knitr::knit_global())


# pipeline variables
quick.load <- FALSE

```

```{r load_datasets}

# Load BALD dataset
HNSCC.Bald <- readRDS("Exported_RDS_files/seurat_combined.rds")
Cillo.TIL.filt <- readRDS("Exported_RDS_files/Cillo_TIL_CD8_only_filt.rds")






 # Cillo dataset reports 
  # CD8 TILS most frequent in cluster 1-4 and PBMC and tonsil 5-8
  # Although at least 50% of the cluster is TIL derived in clusters 1-7
  
  # cluster 1 = cycling 
  # cluster 2 = IFN
  # cluster 3 & 4 = exhaustion 
  # cluster 5 = quiescent and low effector function 
  # cluster 6 = naive or memory 
  # cluster 7 = early activation 
  # cluster 8 = effector function 




```


### Integrate data
  
```{r Integrate_Bald_and_Cillo_Datasets}

# rename metadata columns to match accross datasets 
colnames(HNSCC.Bald@meta.data)
colnames(Cillo.TIL.filt@meta.data)


head(HNSCC.Bald@meta.data)
head(Cillo.TIL.filt@meta.data)

# Original ident column 
HNSCC.Bald@meta.data$orig.ident <- "Bald"
Cillo.TIL.filt@meta.data$orig.ident <- "Cillo"

# Create metadata for dataset origin 
HNSCC.Bald@meta.data$dataset <- "Bald"
Cillo.TIL.filt@meta.data$dataset <- "Cillo"


# Condition and sample origin
colnames(HNSCC.Bald@meta.data)[grepl("^condition$", colnames(HNSCC.Bald@meta.data))] <- "condition_origin"
colnames(Cillo.TIL.filt@meta.data)[grepl("^sample_origin$", colnames(Cillo.TIL.filt@meta.data))] <- "condition_origin"

# Pt_ID and group
colnames(HNSCC.Bald@meta.data)[grepl("^group$", colnames(HNSCC.Bald@meta.data))] <- "sample_ID"
colnames(Cillo.TIL.filt@meta.data)[grepl("^Pt_ID$", colnames(Cillo.TIL.filt@meta.data))] <- "sample_ID"


# Cluster ID
colnames(HNSCC.Bald@meta.data)[grepl("^seurat_clusters$", colnames(HNSCC.Bald@meta.data))] <- "clusterID"
colnames(Cillo.TIL.filt@meta.data)[grepl("^cd8_cells_clusterID$", colnames(Cillo.TIL.filt@meta.data))] <- "clusterID"


#########################################
# remove unnecessary metadata columns 
#########################################

# Cillo
colnames(Cillo.TIL.filt@meta.data)

pattern.vec <- c("cell_barcode", "barcode", "percent.mito")
keep.cols <- !grepl(paste0(pattern.vec, collapse = "|"), colnames(Cillo.TIL.filt@meta.data))

Cillo.TIL.filt@meta.data <- Cillo.TIL.filt@meta.data[ ,keep.cols]

colnames(Cillo.TIL.filt@meta.data)

# Bald
colnames(HNSCC.Bald@meta.data)

pattern.vec <- c("integrated_snn_res.0.4", "seurat_clusters_old", "seurat_clusters_new", "percent.mito", "condition_clust")
keep.cols <- !grepl(paste0(pattern.vec, collapse = "|"), colnames(HNSCC.Bald@meta.data))

HNSCC.Bald@meta.data <- HNSCC.Bald@meta.data[ ,keep.cols]
colnames(HNSCC.Bald@meta.data)




# Set activate assay to RNA
DefaultAssay(HNSCC.Bald) <- "RNA"
DefaultAssay(Cillo.TIL.filt) <- "RNA"



# Create list of seurat objects
seurat.list <- list(Bald = HNSCC.Bald,
                    Cilo = Cillo.TIL.filt)

# Find Integration Anchors
immune.anchors <- FindIntegrationAnchors(object.list = seurat.list, 
                                         anchor.features = 2000, 
                                         scale = TRUE, 
                                         normalization.method = "LogNormalize", 
                                         reduction = "cca", 
                                         l2.norm = TRUE,
                                         dims = 1:30,
                                         k.anchor = 5,
                                         k.filter = 200,
                                         k.score = 30,
                                         max.features = 200,
                                         nn.method = "rann",
                                         eps = 0,
                                         verbose = TRUE)

# Integrate dataset
Bald.Cillo.seurat <- IntegrateData(anchorset = immune.anchors,
                                   new.assay.name = "integrated",
                                   normalization.method = "LogNormalize",
                                   dims = 1:30,
                                   k.weight = 100,
                                   sd.weight = 1,
                                   verbose = TRUE)


# Basic overview of data
Bald.Cillo.seurat # 29,244 cells and 29,957 genes

HNSCC.Bald # 11,658 cells and 20,295 genes
Cillo.TIL.filt # 17,586 cells and 23,389 genes

# 11658 + 17586


# Remove unnecessary data
rm(HNSCC.Bald, 
   Cillo.TIL.filt,
   immune.anchors, 
   seurat.list)
  
```
  
  
## Dim reduction {.tabset}
  
### Dim reduction
```{r Dimreduction}

DefaultAssay(Bald.Cillo.seurat) <- "integrated"

# Scale data
Bald.Cillo.seurat <- ScaleData(object = Bald.Cillo.seurat,
                               assay = "integrated",
                               model.use = "linear",
                               do.scale = TRUE,
                               do.center = TRUE,
                               verbose = TRUE)

# Run PCA
Bald.Cillo.seurat <- RunPCA(Bald.Cillo.seurat, 
                            npcs = 30, 
                            verbose = TRUE)


# Run tSNE
Bald.Cillo.seurat <- RunTSNE(object = Bald.Cillo.seurat,
                             reduction = "pca",
                             dims = 1:20,
                             seed.use = 42)


# Run UMAP
# Based on optimisation - n.neighbors = 50 & min.dist = 0.8
Bald.Cillo.seurat <- RunUMAP(object = Bald.Cillo.seurat,
                             reduction = "pca",
                             dims = 1:20,
                             umap.method = "uwot",
                             n.neighbors = 50,
                             min.dist = 0.8,
                             seed.use = 42)

  
```
  
### Metadata formatting
```{r metadata_formatting}

# Make explicit dataset metadata column
Bald.Cillo.seurat@meta.data$dataset <- Bald.Cillo.seurat@meta.data$orig.ident

# Remove un-needed metadata columns 
pattern.vec <- c("UMAP_1", "UMAP_2", "PCA_1", "PCA_2", "patient_id")
remove.index <- grepl(paste0(pattern.vec, collapse = "|"), colnames(Bald.Cillo.seurat@meta.data))

Bald.Cillo.seurat@meta.data[ , remove.index] <- NULL
  
```
  
### Export seurat object
```{r export_seurat}
  
# Save files 
saveRDS(Bald.Cillo.seurat, file = "Exported_RDS_files/Bald.Cillo.seurat.rds")
  
```
  
### Bald vs Cillo dataset overlaps
 
```{r Visualise_overlap}

# Make output directory
if(!dir.exists("output/figures/Cillo_Bald_overlap")){
  dir.create("output/figures/Cillo_Bald_overlap",
             recursive = TRUE)}

head(Bald.Cillo.seurat@meta.data)

# UMAP
UMAPPlot(Bald.Cillo.seurat,
           label = TRUE, 
           order = TRUE, 
           pt.size = 1,
           group.by = "dataset")

dev.copy(pdf, "output/figures/Cillo_Bald_overlap/UMAP_dataset.pdf")
dev.off()

UMAPPlot(Bald.Cillo.seurat,
         label = TRUE, 
         order = TRUE, 
         pt.size = 1,
         split.by = "dataset",
         group.by = "condition_origin")

dev.copy(pdf, "output/figures/Cillo_Bald_overlap/UMAP_cell_origin.pdf")
dev.off()

UMAPPlot(Bald.Cillo.seurat,
         label = TRUE, 
         order = TRUE, 
         pt.size = 1,
         split.by = "dataset",
         group.by = "clusterID") +
  NoLegend()

dev.copy(pdf, "output/figures/Cillo_Bald_overlap/UMAP_cluster_ID.pdf")
dev.off()


# TSNE
TSNEPlot(Bald.Cillo.seurat,
         label = TRUE, 
         order = TRUE, 
         pt.size = 1,
         group.by = "dataset")

dev.copy(pdf, "output/figures/Cillo_Bald_overlap/TSNE_dataset.pdf")
dev.off()


TSNEPlot(Bald.Cillo.seurat,
         label = TRUE, 
         order = TRUE, 
         pt.size = 1,
         split.by = "dataset",
         group.by = "condition_origin")

dev.copy(pdf, "output/figures/Cillo_Bald_overlap/TSNE_cell_origin.pdf")
dev.off()

TSNEPlot(Bald.Cillo.seurat,
         label = TRUE, 
         order = TRUE, 
         pt.size = 1,
         split.by = "dataset",
         group.by = "clusterID") + 
  NoLegend()

dev.copy(pdf, "output/figures/Cillo_Bald_overlap/TSNE_cluster_ID.pdf")
dev.off()

  
  


```

  
  
  