---
title: "Immune Landscape of Viral- and Carcinogen-Driven Head and Neck Cancer"
author: "DillonCorvino"
date: "22/06/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

## Dataset {.tabset}

### Dataset info
```{r Dataset_details}

# Manuscript title: "Immune Landscape of Viral- and Carcinogen-Driven Head and Neck Cancer"
# Authors: Anthony R. Cillo, Cornelius H.L. Ku€rten, Tracy Tabib, ..., Tullia C. Bruno, Robert L. Ferris, Dario A.A. Vignali
# Journal: Immunity
# Year: 2020
# DOI: 10.1016/j.immuni.2019.11.014
## Abstract: Head and neck squamous cell carcinoma (HNSCC) arises through exposure to environmental carcino- gens or malignant transformation by human papillomavirus (HPV). Here, we assessed the transcriptional profiles of 131,224 single cells from peripheral and intra-tumoral immune populations from patients with HPV– and HPV+ HNSCC and healthy donors. Immune cells within tumors of HPV and HPV HNSCC displayed a spectrum of transcriptional signatures, with helper CD4+ T cells and B cells being relatively divergent and CD8+ T cells and CD4+ regulatory T cells being relatively similar. Transcriptional results were contextualized through multispectral immuno- fluorescence analyses and evaluating putative cell- cell communication based on spatial proximity. These analyses defined a gene expression signature associated with CD4+ T follicular helper cells that is associated with longer progression-free survival in HNSCC patients. The datasets and analytical ap- proaches herein provide a resource for the further study of the impact of immune cells on viral- and carcinogen-induced cancers.


# Total of 63 samples were analyzed by scRNAseq. 
# 26 HNSCC patients (18 HPVNeg and 8 HPVPos)
# HNSCC patients have paired PBMC and TIL samples
# 6 PBMC from healthy donors
# 5 healthy donor tonsils 

```

## Set up {.tabset}

### Environment
```{r Environment_setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  eval = TRUE, 
  tidy = TRUE
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory

start.time <- Sys.time()

###################
# Install packages
###################

pkgs <- c("remedy", "Seurat", "dplyr", 
          "rstudioapi", "cowplot", "ggplot2",
          "grid", "gridExtra", "styler", "stringr")


for(i in 1:length(pkgs)){
  if(!require(pkgs[i], character.only = T)){
    install.packages(pkgs[i])
    require(pkgs[i], character.only = T)
  }else{
    require(pkgs[i], character.only = T)
  }
}

pkgs <- c("gplots")

for(i in 1:length(pkgs)){
  if(!require(pkgs[i], character.only = T)){
    BiocManager::install(pkgs[i])
    require(pkgs[i], character.only = T)
  }else{
    require(pkgs[i], character.only = T)
  }
}


# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
working.dir <- getwd()

# since moving script from local to github - I want to adjust work dir to be main github dir - therefore 
setwd("..")


# create output directories
if(!dir.exists("Exported_RDS_files")){dir.create("Exported_RDS_files", recursive = T)}

# Read in functions
source("~/Documents/Work/Sciebo/Scripts/scRNAseq_Functions.R")

```

### Reading in data
```{r reading_data}

############################################################
# Data comes from Cillo et al., 2020 Immunity
# GSE139324
############################################################

###################################
# Read and formatt TIL samples 
###################################

# Load files
named.vect <- list.dirs("Data/Cillo_et_al_data/HNSCC_TILS/")
named.vect <- named.vect[-1]
names(named.vect) <- gsub("HNSCC_TILS/", "", named.vect)

HNSCC.TILs <- Read10X(data.dir = named.vect)

# 26 TIL samples from HNSCC patients


# Format barcode IDs to allow matching with cillo metadata
colnames(HNSCC.TILs)
colnames(HNSCC.TILs) <- gsub("Data/Cillo_et_al_data//", "", colnames(HNSCC.TILs))
colnames(HNSCC.TILs) <- gsub("-1$", "", colnames(HNSCC.TILs))
colnames(HNSCC.TILs)


# Read in metadata provided by Anthony Cillo 
meta.data.df <- read.delim("Data/Cillo_et_al_data/Cillo_HNSCC_cell_annotations_with_FItSNE.tsv")

head(meta.data.df)
sort(unique(meta.data.df$clusterID))
as.factor(meta.data.df$sample)
as.factor(meta.data.df$sample_origin)

# Filter metadata for TIL annotations 
meta.data.df %>% 
  dplyr::filter(sample_origin == "TIL") -> TIL.metadata

# format metadata
sum(duplicated(TIL.metadata$cell_barcode))

# Need to append PT.ID to make barcode unique 
head(TIL.metadata)
as.factor(TIL.metadata$sample) # 26 levels 

TIL.metadata$Pt_ID <- gsub("HNSCC_", "Pt", TIL.metadata$sample)

TIL.metadata$Pt_ID <- gsub("_.*", "", TIL.metadata$Pt_ID)

head(TIL.metadata)
as.factor(TIL.metadata$Pt_ID)

# Attach Pt_ID to barcode_ID to make it unique 
TIL.metadata$barcode <- paste0(TIL.metadata$Pt_ID, "_", TIL.metadata$cell_barcode)

sum(duplicated(TIL.metadata$barcode))

rownames(TIL.metadata) <- TIL.metadata$barcode



# Initialize the Seurat object with the raw (non-normalized data).  Keep all
# genes expressed in >= 3 cells. Keep all cells with at
# least 200 detected genes
seurat.object <- CreateSeuratObject(
  counts = HNSCC.TILs,
  meta.data = TIL.metadata,
  min.cells = 3,
  min.features = 200
)


seurat.object 
# 23,389 genes 60,923 cells

seurat.object@meta.data




###################################
# Read and format HNSCC PBMC samples 
###################################

named.vect <- list.dirs("Data/Cillo_et_al_data/HNSCC_PBMC/")
named.vect <- named.vect[-1]
names(named.vect) <- gsub("HNSCC_PBMC/", "", named.vect)

HNSCC.PBMC <- Read10X(data.dir = named.vect)

# Format barcode IDs to allow matching with cillo metadata
colnames(HNSCC.PBMC)
colnames(HNSCC.PBMC) <- gsub("Data/Cillo_et_al_data//", "", colnames(HNSCC.PBMC))
colnames(HNSCC.PBMC) <- gsub("-1$", "", colnames(HNSCC.PBMC))
colnames(HNSCC.PBMC)

sort(unique(meta.data.df$clusterID))


temp <- 
  meta.data.df %>%
  filter(cell_type == "cd8.cells") %>%
  group_by(clusterID) %>%
  summarise(total = n())

sum(temp$total)



```

## QC and Normalisation {.tabset}

### QC & Normalisation
```{r QC_and_normalisation}

# QC metrics commonly used
# The number of unique genes detected in each cell.
# Low-quality cells or empty droplets will often have very few genes
# Cell doublets or multiplets may exhibit an aberrantly high gene count

# Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)

# The percentage of reads that map to the mitochondrial genome
# Low-quality / dying cells often exhibit extensive mitochondrial contamination


################################################
#  Get percentage Mitochondria gene expression
################################################

seurat.object[["percent.mito"]] <- PercentageFeatureSet(seurat.object, pattern = "^MT-")



##################################
# Plot mitochondria percentage
##################################

# nFeature & nCount & Mito 
VlnPlot(object = seurat.object, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"))

# Mito 
VlnPlot(seurat.object, features = "percent.mito", y.max = 20)


########################################
# Count vs Mito & Count vs Feature
########################################

plot1 <- FeatureScatter(seurat.object, feature1 = "nCount_RNA", feature2 = "percent.mito")
plot2 <- FeatureScatter(seurat.object, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
CombinePlots(plots = list(plot1, plot2))


############################
# Structure and metadata 
############################

head(seurat.object@meta.data)
str(seurat.object@meta.data)

###################
#  Filter cells
###################

# Seurat recommends removing cells with <200 or >2,500 genes or >10% mito genes

seurat.object.filt <- subset(seurat.object,
                             subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mito < 10)

# Post-filtering metrics

seurat.object # 23,389 features across 60,923 samples
seurat.object.filt # 23,389 features across 55,272 samples

# Therefore
60923 - 55272
# 5,651 cells were deleted

##########################################
# Visualise QC metrics after filtering
##########################################

# Before Filtering
VlnPlot(object = seurat.object, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), pt.size = 0)

# After filtering
VlnPlot(object = seurat.object.filt, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), pt.size = 0)

################################################
# Normalise, scale & find variable genes 
################################################

# Normalise
seurat.object.filt <- NormalizeData(object = seurat.object.filt,
                                    assay = "RNA",
                                    normalization.method = "LogNormalize",
                                    scale.factor = 10000,
                                    verbose = TRUE)

# Find variable genes
seurat.object.filt <- FindVariableFeatures(object = seurat.object.filt,
                                           verbose = TRUE)

# Scale data
seurat.object.filt <- ScaleData(object = seurat.object.filt,
                                assay = "RNA",
                                vars.to.regress = c("nCount_RNA", "percent.mito"),
                                model.use = "linear",
                                do.scale = TRUE,
                                do.center = TRUE,
                                verbose = TRUE)


rm(seurat.object, HNSCC.TILs, HNSCC.PBMC)

```


## Dim reduction {.tabset}

### Dim reduction
```{r Dimreduction}


# Run PCA
seurat.object.filt <- RunPCA(seurat.object.filt, 
                             npcs = 30, 
                             verbose = TRUE)


# Run tSNE
seurat.object.filt <- RunTSNE(object = seurat.object.filt,
                              reduction = "pca",
                              dims = 1:20,
                              seed.use = 42)


# Run UMAP
# Based on optimisation - n.neighbors = 50 & min.dist = 0.8
seurat.object.filt <- RunUMAP(object = seurat.object.filt,
                              reduction = "pca",
                              dims = 1:20,
                              umap.method = "uwot",
                              n.neighbors = 50,
                              min.dist = 0.8,
                              seed.use = 42)

```
 
 
### Export seurat object
```{r export_seurat}

saveRDS(seurat.object.filt, "Exported_RDS_files/Cillo_et_al_seurat_object_filt.rds")

```

 
 
 
 
 
 

 
 
