---
title: "HNSCC_scRNAseq_analysis"
author: "Dillon Corvino"
date: "03/02/2020"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

### Environment

```{r Environment_setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Load packages and create output directories
source("R/Load_packages.R", local = knitr::knit_global())
source("R/Setup.R", local = knitr::knit_global())

# Establish colour scheme
source("R/Colour_scheme_variable.R", local = knitr::knit_global())

# Load custom functions
source("R/scRNAseq_function.R", local = knitr::knit_global())
source("R/annotate_seurat_heatmap_function.R", local = knitr::knit_global())
source("R/Enhanced_volcano_custom_defaults_function.R", local = knitr::knit_global())


# pipeline variables
quick.load <- TRUE
long.compute <- FALSE

# load saved seurat object with cluster annotations and imputation performed
if(quick.load){
  seurat.combined <- readRDS("Exported_RDS_files/seurat_combined.rds")
}



# TO DO 
# Make initialisation chunck like this one for major r.markdown sections 
# split up by pre-processing 
# DEG 
# Visualisation 
# etc - maybe smaller analysis like singlecellhaystack is a sep .R script 
# combine code for tcr analysis nick wrote 
# split tcr analysis rmark down in a similar manner 
# other datasets use functions and set up script from this as wel


# DEG analysis has some incomplete chunks 
```


## GSEA analysis {.tabset}

### Scillus

```{r Scillus_GSEA}

source("R/Scillus_GO_analysis.R", local = knitr::knit_global())

```


### Clust module analysis

```{r Clust_module_analysis}

# Python program 
# output data for use in Clust analysis 

Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters

seurat.combined.small <- subset(seurat.combined, downsample = 30)

output.df <- GetAssayData(seurat.combined.small, assay = "integrated", slot = "data")

colnames(output.df) <- seurat.combined.small@meta.data$seurat_clusters

colnames(output.df)




average.seurat <- AverageExpression(seurat.combined, assays = "integrated", return.seurat = TRUE)

output.df <- GetAssayData(average.seurat, assay = "integrated", slot = "data")

remove.index <- grepl(paste0(c("TRM", "gd_T", "Prolif", "MAIT"), collapse = "|"), colnames(output.df))
output.df <- output.df[,!remove.index]
colnames(output.df)

write.table(output.df, file="output/tables/Clust_matrix.tsv", quote=FALSE, sep='\t', col.names = NA)

```



## Additional analysis {.tabset}

### Cell Cycle & CytoTrace & Augur analysis

```{r Additional_analysis}

if(long.compute){
  source("R/CellCycle_analysis.R", local = knitr::knit_global())
  source("R/CytoTrace_analysis.R", local = knitr::knit_global())
  source("R/Augur_analysis.R", local = knitr::knit_global())
}
```














## Type I IFN cluster analysis {.tabset} // INCOMPLETE 

### Plotting top DEGs in Type I IFN cluster
```{r Type_I_IFN_cluster_vis}

# Aim 
# Visualise gene expression of the Type I IFN cluster

# Generate output directories
if(!dir.exists("output/figures/Type_I_IFN")){
  dir.create("output/figures/Type_I_IFN", 
             recursive = TRUE)
}

# Set assay to integrated
DefaultAssay(seurat.combined) <- "integrated"

# Set Idents 
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters

# Calculate DEGs found in Type I IFN vs everything else 
TypeI.markers <- FindMarkers(seurat.combined,
                             ident.1 = "Type_I_IFN",
                             assay = "integrated", 
                             only.pos = TRUE)

write.csv(TypeI.markers, "output/tables/Type_I_IFN_upregulated_markers.csv")

Top.DEGs <- 
  TypeI.markers %>% 
  dplyr::filter(p_val_adj < 0.05) %>%
  top_n(20, avg_logFC) %>%
  rownames()

# Add transcription factors identified in next code chunk 
Top.DEGs <- unique(c(Top.DEGs, "IRF7", "STAT1", "PLSCR1", "SP110"))

for(i in seq_along(Top.DEGs)){
  
  print(plot_density(seurat.combined,
                     Top.DEGs[i], 
                     pal = "inferno"))
  
  dev.copy(pdf, paste0("output/figures/Type_I_IFN/Density_UMAP_", Top.DEGs[i], ".pdf"))
  dev.off()
}


```


### Type I IFN cluster Transcription factors

```{r TF_analysis}

# Create output dir
# Generate output directories
if(!dir.exists("output/figures/Type_I_IFN")){
  dir.create("output/figures/Type_I_IFN", 
             recursive = TRUE)
}


# Searching for TFs
TF.ref.list <- read.delim("Data/TFcheckpoint.txt") # http://www.tfcheckpoint.org/index.php/browse

dim(TF.ref.list) # 1020, 11
colnames(TF.ref.list)
head(TF.ref.list)

TypeI.markers <- read.csv("output/tables/Type_I_IFN_upregulated_markers.csv", row.names = 1)

# Up-regulated TFs
sig.genes <- TypeI.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  rownames()

sig.genes <- unique(sig.genes)

TF.logic <- sig.genes %in% TF.ref.list$Gene_symbol

TF.DEG <- sig.genes[TF.logic]


########################################
# Plot TFs differentially regulated 
########################################

Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters

seurat.combined.small <- subset(seurat.combined, downsample = 300)

average.seurat <- AverageExpression(seurat.combined,
                                    assay = "integrated",
                                    slot = "data",
                                    verbose = TRUE,
                                    return.seurat = TRUE)

DoHeatmap(average.seurat, 
          features = TF.DEG, 
          draw.lines = FALSE, 
          raster = FALSE) + NoLegend()

dev.copy(pdf, "output/figures/Type_I_IFN/Heatmap_TFs_upregulated_in_Type_I_IFN_average.pdf")
dev.off()


DoHeatmap(seurat.combined.small, 
          features = TF.DEG, 
          draw.lines = TRUE, 
          raster = FALSE) + NoLegend()

dev.copy(pdf, "output/figures/Type_I_IFN/Heatmap_TFs_upregulated_in_Type_I_IFN.pdf")
dev.off()


```


### Creation and establishment of Type I IFN signature 
```{r Type_I_signature}

# Aim: 
# Establish a signature that can identify Type I IFN cluster in this dataset and be applied to external validation datasets 
# Goal is to have a minimal gene signature that can accurately and specifically identify Type I IFN cells


# Set assay to integrated
DefaultAssay(seurat.combined) <- "integrated"

# Set Idents 
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters

# Calculate DEGs found in Type I IFN vs everything else 
TypeI.markers <- FindMarkers(seurat.combined,
                             ident.1 = "Type_I_IFN",
                             assay = "integrated", 
                             only.pos = FALSE)


# Downsample 
seurat.combined.small <- subset(seurat.combined, downsample = 300)

# Get average expression
average.seurat <- AverageExpression(seurat.combined,
                                    assay = "integrated",
                                    slot = "data",
                                    verbose = TRUE,
                                    return.seurat = TRUE)



# Number of genes that are significantly DEGs

# Total 
TypeI.markers %>% 
  dplyr::filter(p_val_adj < 0.05) %>%
  nrow() # 343 genes 

# Up-regulated only 
TypeI.markers %>% 
  dplyr::filter(p_val_adj < 0.05 & avg_log2FC > 0) %>%
  nrow() # 126 genes 

# Down-regulated only 
TypeI.markers %>% 
  dplyr::filter(p_val_adj < 0.05 & avg_log2FC < 0) %>%
  nrow() # 217 genes 








# Iterate over generating a gene signature of various sizes

n.val <- c(10, 20, 30, 40, 50, 60, 70, 80, 90, 100)
n.val <- c(1:15)

output.table <- matrix(nrow = length(n.val), ncol = 2)
colnames(output.table) <- c("nval", "AUCval")
output.table[,"nval"] <- n.val



for(i in seq_along(n.val)){


  # Generate signature using the top x genes in Type I cluster
  IFN.sig <- 
    TypeI.markers %>% 
    dplyr::filter(p_val_adj < 0.05) %>%
    dplyr::top_n(n.val[i], avg_log2FC) %>%
    rownames()
  
  
# Plot heatmap of gene expression within signature
DoHeatmap(average.seurat,
          features = IFN.sig, 
          draw.lines = FALSE) + ggtitle(paste0("Signature using ", n.val[i])) + NoLegend()



# Create and calculate module score using gene signature 
seurat.combined <- AddModuleScore(seurat.combined, 
                                  features = list(IFN.sig), 
                                  assay = "integrated", 
                                  name = "IFNsig",
                                  ctrl = 80,
                                  seed = 42)


# Visualise gene signature module score on defined clusters
VlnPlot(seurat.combined, "IFNsig1",
        pt.size = 0, 
        assay = "integrated", 
        slot = "data",
        cols = clust.cols)


temp <- seurat.combined@meta.data %>%
  summarize(mean = mean(IFNsig1), 
            n = n(), 
            median = median(IFNsig1),
            qs = quantile(IFNsig1, c(0.75)))

threshold.val <- temp$qs


temp <- seurat.combined@meta.data %>%
  dplyr::mutate(classification = case_when(IFNsig1 <= threshold.val ~ "Other", 
                                          IFNsig1 > threshold.val ~ "IFN_clust"))




temp <- temp %>%
  mutate(response_var = case_when(seurat_clusters == "Type_I_IFN" ~ 1,
                                  TRUE ~ 0), 
         prediction_var = case_when(classification == "IFN_clust" ~ 1, 
                                    TRUE ~ 0))



roc_obj <- roc(response = temp$response_var,
               predictor = temp$prediction_var, 
               ci = TRUE, 
               ci.alpha = 0.9,
               stratified = FALSE,
               plot = TRUE, 
               auc.polygon = TRUE, 
               max.auc.polygon = TRUE, 
               grid = TRUE, 
               print.auc = TRUE, 
               show.thres = TRUE)



print(paste0("for a signature of nval = ", n.val[i], " AUC = ", roc_obj$auc))
print(threshold.val)


output.table[i,"AUCval"] <- roc_obj$auc


}


output.table



# AUC hits limit at 14 genes. but top 10 genes is AUC = 0.9
# therefore use top 10 upregulated genes as signature and 75% signature threshold 

  IFN.sig <- 
    TypeI.markers %>% 
    dplyr::filter(p_val_adj < 0.05) %>%
    dplyr::top_n(10, avg_log2FC) %>%
    rownames()

write.csv(IFN.sig, "output/tables/TypeI_IFN_sig.csv")





##
































# Iterate over generating a gene signature of various sizes

n.val <- c(10, 20, 30, 40, 50, 60, 70, 80, 90, 100)

for(i in seq_along(n.val)){

  # Generate signature using the top x genes in Type I cluster
  IFN.sig <- 
    TypeI.markers %>% 
    dplyr::filter(p_val_adj < 0.05) %>%
    dplyr::top_n(n.val[i], avg_log2FC) %>%
    rownames()
  
  
# Plot heatmap of gene expression within signature
DoHeatmap(average.seurat,
          features = IFN.sig, 
          draw.lines = FALSE) + ggtitle(paste0("Signature using ", n.val[i])) + NoLegend()



# Create and calculate module score using gene signature 
seurat.combined <- AddModuleScore(seurat.combined, 
                                  features = list(IFN.sig), 
                                  assay = "integrated", 
                                  name = "IFNsig",
                                  ctrl = 80,
                                  seed = 42)


# Visualise gene signature module score on defined clusters
VlnPlot(seurat.combined, "IFNsig1",
        pt.size = 0, 
        assay = "integrated", 
        slot = "data",
        cols = clust.cols)


temp <- seurat.combined@meta.data %>%
  group_by(seurat_clusters) %>%
  summarize(mean = mean(IFNsig1), 
            n = n(), 
            median = median(IFNsig1),
            qs = quantile(IFNsig1, c(0.75)))


temp <- temp %>%
  filter(seurat_clusters != "Type_I_IFN") %>%
  summarize(max = max(qs))


print(paste0("threshold max = ", signif(temp$max)))

# threshold = 0.4



temp <- seurat.combined@meta.data %>%
  dplyr::mutate(classification = case_when(IFNsig1 <= temp$max ~ "Other", 
                                          IFNsig1 > temp$max ~ "IFN_clust"))





temp <- temp %>%
  mutate(response_var = case_when(seurat_clusters == "Type_I_IFN" ~ 1,
                                  TRUE ~ 0), 
         prediction_var = case_when(classification == "IFN_clust" ~ 1, 
                                    TRUE ~ 0))



roc_obj <- roc(response = temp$response_var,
               predictor = temp$prediction_var, 
               ci = TRUE, 
               ci.alpha = 0.9,
               stratified = FALSE,
               plot = TRUE, 
               auc.polygon = TRUE, 
               max.auc.polygon = TRUE, 
               grid = TRUE, 
               print.auc = TRUE, 
               show.thres = TRUE)


#sens.ci <- ci.se(roc_obj)

#plot(sens.ci, type = "shape", col = "lightblue")
#plot(sens.ci, type = "bars")

print(paste0("for a signature of ", n.val[i], roc_obj$auc))

}



write.csv(Top.UP.genes, "output/tables/TypeI_IFN_sig.csv")


```

### Type I signature validation
```{r Type_I_signature_validation_Cillo_dataset}


# IFN signature 
IFN.sig <- read.csv("output/tables/TypeI_IFN_sig.csv")
IFN.sig <- IFN.sig$x


# Read in validation dataset
Cillo_et_al <- readRDS("Exported_RDS_files/Cillo_TIL_CD8_only_filt.rds")


Idents(Cillo_et_al) <- Cillo_et_al@meta.data$cd8_cells_clusterID


IFN.sig.function <- function(input.seurat, 
                             IFN.sig,
                             validation.cluster = NULL, 
                             validation.cluster.column = NULL){

  
  
  # Add module score
  input.seurat <- AddModuleScore(input.seurat, 
                              features = list(IFN.sig), 
                              name = "IFNsig",
                              ctrl = 80,
                              seed = 42)
  
  # Calculate 75% quantile for threshold
  stats.vals <- input.seurat@meta.data %>%
    summarize(mean = mean(IFNsig1), 
              n = n(), 
              median = median(IFNsig1),
              qs = quantile(IFNsig1, c(0.75)))
  
  
threshold.val <- stats.vals$qs


input.seurat <- input.seurat@meta.data %>%
  dplyr::mutate(classification = case_when(IFNsig1 <= threshold.val ~ "Other", 
                                          IFNsig1 > threshold.val ~ "Predicted_IFN_clust"))



if(!is.null(validation.cluster)){

    output.temp <- input.seurat %>%
    mutate(response_var = case_when(get(validation.cluster.column) == get("validation.cluster") ~ 1,
                                    TRUE ~ 0), 
           prediction_var = case_when(classification == "Predicted_IFN_clust" ~ 1, 
                                      TRUE ~ 0))
  

roc_obj <- roc(response = output.temp$response_var,
               predictor = output.temp$prediction_var, 
               ci = TRUE, 
               ci.alpha = 0.9,
               stratified = FALSE,
               plot = TRUE, 
               auc.polygon = TRUE, 
               max.auc.polygon = TRUE, 
               grid = TRUE, 
               print.auc = TRUE, 
               show.thres = TRUE)




}


return(input.seurat)



}




# cd8_cells_clusterID = 2 is IFN sig cluster in cillo dataset

IFN.sig.function(Cillo_et_al, 
                 IFN.sig, 
                 validation.cluster = "2", 
                 validation.cluster.column = "cd8_cells_clusterID")






VlnPlot(Cillo_et_al, 
        feature = "IFNsig1",
        group.by = "cd8_cells_clusterID",
        pt.size = 0)


rm(Cillo_et_al)

```



```{r Sade_Feldman_IFN_signature}



# IFN signature 
IFN.sig <- read.csv("output/tables/TypeI_IFN_sig.csv")
IFN.sig <- IFN.sig$x


# Read in dataset
#Sade_Feldman <- readRDS("Exported_RDS_files/Sade_Feldman_CD8_only.rds")

Sade_Feldman <- readRDS("Exported_RDS_files/Sade_Feldman_all_cells.rds")

head(Sade_Feldman@meta.data)

Idents(Sade_Feldman) <- Sade_Feldman@meta.data$cluster_named





x <- IFN.sig.function(Sade_Feldman, 
                 IFN.sig)

head(x)


Sade_Feldman <- AddMetaData(Sade_Feldman, 
                            metadata = x$classification, 
                            col.name = "classification")


Sade_Feldman <- AddMetaData(Sade_Feldman, 
                            metadata = x$IFNsig1, 
                            col.name = "IFNsig1")


Sade_Feldman@meta.data


VlnPlot(Sade_Feldman, 
        feature = "IFNsig1",
        group.by = "cluster_named",
        pt.size = 0)

UMAPPlot(Sade_Feldman, 
         group.by = "classification", 
         pt.size = 1)

TSNEPlot(Sade_Feldman, 
         group.by = "classification", 
         pt.size = 1)


TSNEPlot(Sade_Feldman, 
         group.by = "cluster_named", 
         pt.size = 1)


#rm(Sade_Feldman_CD8)
```



















## Inhibitory receptor expression analysis {.tabset} // INCOMPLETE 

### Inhibitory receptor co-expression
```{r inhibitory_receptor_coexpression}

# Generate output directories
if(!dir.exists("output/figures/IR_expression")){
  dir.create("output/figures/IR_expression", 
             recursive = TRUE)
}


Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters
DefaultAssay(seurat.combined) <- "integrated"

seurat.combined.small <- subset(seurat.combined, downsample = 300)

inhibitory.receptors <- c("HAVCR2", "LAG3", "PDCD1", "TIGIT", "ENTPD1", "TOX", 
                          "CTLA4", "VSIR", "TNFRSF9", "PRF1", "IFNG", "GZMA", "GZMB")

DoHeatmap(seurat.combined.small, 
          inhibitory.receptors)


inhibitory.receptors <- c("HAVCR2", "LAG3", "PDCD1", "TIGIT", "TOX", "CTLA4", "ENTPD1")

for(i in seq_along(inhibitory.receptors)){
  
  
  print(Nebulosa::plot_density(seurat.combined, 
                               inhibitory.receptors[i], 
                               pal = "inferno"))
  
  dev.copy(pdf, paste0("output/figures/IR_expression/Density_UMAP_", inhibitory.receptors[i], ".pdf"))
  dev.off()
  
  
}


# Plot 

markers <- c("HAVCR2", "LAG3", "PDCD1", "ENTPD1")

z <- Nebulosa::plot_density(seurat.combined, 
                       , 
                       joint = TRUE, 
                       pal = "inferno")


z[[5]]




```



### Secretome // Unfinished chunk

```{r Secretome}

# set working dir
setwd(working.dir)

# Create output directories
if(!dir.exists("output/figures/Secretome")){
  dir.create("output/figures/Secretome",
             recursive = T)}

output.dir <- "output/figures/Secretome"
setwd(output.dir)

# Read in data
chemo.cytokines <- read.csv(paste0(working.dir, "/Data/Secretome/ChemoCytokines.csv"))
receptors <- read.csv(paste0(working.dir, "/Data/Secretome/receptors.csv"))


# Set idents
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters


# Identify which of the query genes are in the database
query.genes <- unique(c(chemo.cytokines$GeneID, receptors$GeneID))

logic.vec <- rownames(seurat.combined) %in% query.genes
genes.to.test <- rownames(seurat.combined)[logic.vec]


# Run DEG test on all genes in query set that can be found in database
Query.DEG <- FindAllMarkers(seurat.combined,
                            assay = "RNA",
                            features = genes.to.test,
                            only.pos = FALSE,
                            min.diff.pct = 0.1,
                            logfc.threshold = 0,
                            return.thresh = 0.01)


# Filter Query.DEG for sig and arrange by LogFC within groups for nicer plotting
genes.plot <- Query.DEG %>% 
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::group_by(cluster) %>%
  dplyr::arrange(desc(avg_logFC), .by_group = TRUE) %>%
  dplyr::select(gene)

genes.plot <- unique(genes.plot$gene)


# Get average expression
average.seurat <- AverageExpression(seurat.combined,
                                    assay = "RNA",
                                    slot = "data",
                                    verbose = TRUE,
                                    return.seurat = TRUE)


# Plot all genes DEG from query list
DoHeatmap(average.seurat, 
          features = genes.plot, 
          size = 4, 
          angle = 90, 
          draw.lines = FALSE,
          raster = FALSE)

dev.copy(pdf, "All_sig_genes.pdf")
dev.off()



# Plot just chemokine and cytokine genes
logic.vec <- genes.plot %in% receptors$GeneID
chemo.cytokine.genes <- genes.plot[!logic.vec]

DoHeatmap(average.seurat, 
          features = chemo.cytokine.genes, 
          size = 4, 
          angle = 90, 
          draw.lines = FALSE,
          raster = FALSE)

dev.copy(pdf, "Chemo_cytokine_genes.pdf")
dev.off()


# Plot just receptor genes
logic.vec <- genes.plot %in% receptors$GeneID
receptor.genes <- genes.plot[logic.vec]

DoHeatmap(average.seurat, 
          features = receptor.genes, 
          size = 4, 
          angle = 90, 
          draw.lines = FALSE,
          raster = FALSE)

dev.copy(pdf, "Receptor_genes.pdf")
dev.off()


# Return to working dir
setwd(working.dir)

# remove unneeded variables
rm(receptor.genes, genes.plot, receptors, chemo.cytokine.genes, average.seurat, Query.DEG)
```














## Misc analysis {.tabset}

### Milo Differential abundance analysis Knn graph // Incomplete

```{r Differential_abundance_analysis}

#devtools::install_github("MikeDMorgan/miloR") 

#BiocManager::install("edgeR")
#BiocManager::install("limma")
 
library(miloR)
library(SingleCellExperiment)
library(scater)
library(dplyr)
library(patchwork)




library(Seurat)

seurat.combined@meta.data$sample <- seurat.combined@meta.data$group

logic.vec <- seurat.combined@meta.data$sample == "S3"
seurat.combined@meta.data$sample[logic.vec] <- "S1"

logic.vec <- seurat.combined@meta.data$sample == "S4"
seurat.combined@meta.data$sample[logic.vec] <- "S2"

seurat.combined.sce <- as.SingleCellExperiment(seurat.combined)
seurat.combined.sce.milo <- Milo(seurat.combined.sce)
rm(seurat.combined.sce)

length(unique(seurat.combined@meta.data$seurat_clusters))

traj_milo <- buildGraph(seurat.combined.sce.milo, k = 14, d = 20)

traj_milo


traj_milo <- makeNhoods(traj_milo, prop = 0.5, k = 14, d=30, refined = TRUE)

plotNhoodSizeHist(traj_milo)

traj_milo <- countCells(traj_milo,
                        meta.data = data.frame(colData(traj_milo)), 
                        sample = "group")


head(nhoodCounts(traj_milo))



traj_design <- data.frame(colData(traj_milo))[,c("group", "condition")]


traj_design <- distinct(traj_design)

traj_design


traj_milo <- calcNhoodDistance(traj_milo, d=20)

da_results <- testNhoods(traj_milo,
                         design = ~ condition,
                         design.df = traj_design)



da_results %>%
  arrange(- SpatialFDR) %>%
  head()

da_results %>%
  arrange(SpatialFDR) %>%
  head()


traj_milo <- buildNhoodGraph(traj_milo)



plotUMAP(traj_milo) + 
  plotNhoodGraphDA(traj_milo, da_results, alpha=0.8) +
  plot_layout(guides="collect")




DimPlot(seurat.combined, reduction = "pca", split.by = "condition")




```


















## Script info {.tabset}

### Run time

```{r code_runtime, eval = TRUE}

# calculate compute time for compiling code so far
compute.time <- Sys.time()
print(compute.time-start.time)


```

## Code archive {.tabset}


## Cell-classification model {.tabset}

### scPred
```{r scPred_model_creation}

# Set default assay
DefaultAssay(seurat.combined) <- "integrated"

# Set idents
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters

#keep.idents <- unique(seurat.combined@meta.data$seurat_clusters)

#pattern.remove <- c("gd_T_g9d2", "gd_T_non_g9d2", "TRM", "MAIT", "Proliferative")
#index.val <- grepl(paste0(pattern.remove, collapse = "|"), keep.idents)
#keep.idents <- keep.idents[!index.val]


# subset dataset to include only Type I IFN cells 
#temp.seurat <- subset(seurat.combined, ident = keep.idents)

#temp.seurat <- RunPCA(temp.seurat, 
#                      npcs = 30, 
#                      verbose = TRUE)



#UMAPPlot(temp.seurat)



# Get feature space
reference.obj <- getFeatureSpace(seurat.combined,
                                 pvar = "seurat_clusters",
                                 reduction = "pca")


# Train model - in parallel
num.cores <- 5
cl <- makePSOCKcluster(num.cores) 
registerDoParallel(cl)


reference.obj <- trainModel(reference.obj, 
                            model = "svmRadial", 
                            allowParallel = TRUE)

stopCluster(cl)



# Access probabilities 
get_probabilities(reference.obj) %>% head()

# Access scpred object slot 
get_scpred(reference.obj)

# Plot prediction probability
plot_probabilities(reference.obj)




################################################
# Load Cillo_et_al dataset (TIL data only)
################################################
# Processed in Cillo_et_al script 

Cillo_et_al <- readRDS("Exported_RDS_files/Cillo_et_al_seurat_object_filt.rds")

Cillo_et_al <- scPredict(Cillo_et_al, reference.obj)

crossTab(Cillo_et_al, "cell_type", "scpred_prediction", output = "prop")


Cillo_et_al@meta.data$cell_type



DimPlot(Cillo_et_al, 
        group.by = "scpred_prediction", 
        reduction = "umap")




DimPlot(Cillo_et_al, 
        group.by = "scpred_prediction", 
        reduction = "scpred")

Cillo_et_al <- RunUMAP(Cillo_et_al,
                       reduction = "scpred",
                       dims = 1:30)

DimPlot(Cillo_et_al,
        group.by = "scpred_prediction", 
        label = TRUE, 
        repel = TRUE, 
        cols = col.vec)


col.vec <- rep("Grey", length(unique(Cillo_et_al@meta.data$scpred_prediction)))

col.vec[c(14)] <- c("Red")


DimPlot(Cillo_et_al,
        group.by = "clusterID", 
        label = TRUE, 
        repel = TRUE)



x <- get_classifiers(reference.obj)[["Type_I_IFN"]]



head(seurat.object.filt@meta.data)

DimPlot(Cillo_et_al, 
        group.by = "cell_type", 
        reduction = "umap")


DimPlot(Cillo_et_al, 
        group.by = "clusterID", 
        reduction = "umap", 
        cols = col.vec)


col.vec <- rep("Grey", length(unique(seurat.object.filt@meta.data$clusterID)))

col.vec[c(2, 4)] <- c("Red", "Blue")


# clusters 2 and 4 IFN?



Cluster.markers.pos <- read.csv("output/tables/DEG_Clusters/Cluster_markers_unfiltered_pos.csv")

colnames(Cluster.markers.pos)


temp <- Cluster.markers.pos %>%
  dplyr::filter(cluster == "Type_I_IFN") %>%
  dplyr::top_n(10, avg_logFC) %>%
  dplyr::pull(gene)

gene 



```


### Align with reference dataset

```{r align_with_ref_dataset}




#remotes::install_github("satijalab/seurat", ref = "release/4.0.0")
#remotes::install_github("jlmelville/uwot")
#remotes::install_github("mojaveazure/seurat-disk")


library(Seurat)
library(SeuratDisk)
library(ggplot2)
library(patchwork)



reference <- LoadH5Seurat("Data/pbmc_multimodal.h5seurat")

Idents(reference) <- reference@meta.data$celltype.l3
length(unique(reference@meta.data$celltype.l3))

reference <- subset(reference, downsample = 1000)


DimPlot(reference, 
        reduction = "wnn.umap", 
        group.by = "celltype.l3", 
        label = TRUE, 
        label.size = 3, 
        repel = TRUE) + 
  NoLegend()

Idents(reference) <- reference@meta.data$celltype.l1
unique(reference@meta.data$celltype.l1)

reference <- subset(reference, idents = c("CD8 T", "NK", "other T"))

DimPlot(reference, 
        reduction = "wnn.umap", 
        group.by = "celltype.l3", 
        label = TRUE, 
        label.size = 3, 
        repel = TRUE) + 
  NoLegend()


query.data <- SCTransform(seurat.combined, verbose = FALSE)


query.data@assays$SCT@data
head(x = rownames(x = query.data))


library(SeuratData)
InstallData('pbmc3k')

pbmc3k <- SCTransform(pbmc3k, verbose = FALSE)

DefaultAssay(pbmc3k) <- "RNA"
reference




anchors <- FindTransferAnchors(
  reference = reference,
  query = pbmc3k,
  project.query = FALSE,
  npcs = NULL,
  normalization.method = "SCT",
  reduction = "pcaproject",
  dims = 1:20)


?FindTransferAnchors


```




