---
title: "Merge_vs_integrate"
author: "DillonCorvino"
date: "12/17/2020"
output: html_document
---
Built with R version `{r} getRversion()`

## Setup {.tabset}

### Dataset information

```{r Dataset_Info}

# HNSCC scRNAseq/scTCRseq dataset
# Cells are HNSCC TILs sorted on Lymphocytes/Live/CD3+/CD4-/CD8+
# Human samples
# Each sample is a pool of 4 patients
# samples 1 and 2 are unstimulated 
# samples 3 and 4 are stimulated 
# Patients in sample 1 match patients in sample 3 and sample 2 pairs with 4
# Data acquired was transcript expression, ADT (antibody expression), and TCRA & B sequences
# This analysis uses just Transcript and TCR data


# General information
# CellRanger was used for sequence alignment/QC/counts/TCR calls - performed by Ross (QIMR, Brisbane, Aus)
# DC was provided the output from CellRanger (counts matrix) and used this as input for analysis within this script

```

### Environment

```{r Environment_setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory

# pipeline variables
start.time <- Sys.time()
quick.load <- TRUE
long.compute <- TRUE

###################
# Install packages
###################

pkgs <- c("remedy", "dplyr", "rstudioapi",
          "cowplot", "ggplot2", "grid", "gridExtra",
          "styler", "stringr", "inlmisc", "RColorBrewer",
          "readxl", "devtools", "tidyverse", "hdf5r", "scales",
          "useful", "renv", "pROC")

for(i in 1:length(pkgs)){
  if(!require(pkgs[i], character.only = T)){
    install.packages(pkgs[i])
    require(pkgs[i], character.only = T)
  }else{
    require(pkgs[i], character.only = T)
  }
}

pkgs <- c("gplots", "fgsea", "biomaRt", "clusterProfiler", 
          "GSEABase", "org.Hs.eg.db", "pcaMethods",
          "SingleCellExperiment", "batchelor", 
          "DelayedArray", "DelayedMatrixStats",
          "limma", "SummarizedExperiment", "SingleR",
          #"celldex",
          "progeny", "RcisTarget",
          "doMC", "doRNG", "DT", "visNetwork", "readr", "pheatmap", "tibble")

for(i in 1:length(pkgs)){
  if(!require(pkgs[i], character.only = T)){
    BiocManager::install(pkgs[i])
    require(pkgs[i], character.only = T)
  }else{
    require(pkgs[i], character.only = T)
  }
}

#####################
# Github packages
#####################

# library("devtools")

# usethis::browse_github_pat()
# usethis::edit_r_environ()
# GITHUB_PAT = "d8207153aef7b295cdf66eb1e1b2a2ed38b0ca18"
# R_MAX_VSIZE = 30Gb


# Scillus 
#devtools::install_github("xmc811/Scillus", ref = "development")
library("Scillus")

# Nebulosa for density plotting
#devtools::install_github("powellgenomicslab/Nebulosa")
library("Nebulosa")

# Volcano
#devtools::install_github('kevinblighe/EnhancedVolcano')
library("EnhancedVolcano")

# Clustifyr
#devtools::install_github("rnabioco/clustifyr")
library("clustifyr")

# ComplexHeatmap
#install_github("jokergoo/ComplexHeatmap")
library("ComplexHeatmap")

# scPred and requirements
#devtools::install_github("immunogenomics/harmony")
#devtools::install_github("powellgenomicslab/scPred")
library("scPred")
library("magrittr")
library("doParallel")

# Cytotrace
# Download source file from - https://cytotrace.stanford.edu/
#devtools::install_local("PATH/TO/DIRECTORY/CytoTRACE_0.3.3.tar.gz")
library("CytoTRACE")

# circlize
#devtools::install_github("jokergoo/circlize")
library("circlize")


######################
# Cell-cell interaction
######################

# celltalker
#devtools::install_github("arc85/celltalker")
library("celltalker")

# iTalk
#devtools::install_github("Coolgenome/iTALK", build_vignettes = TRUE)
library("iTALK")

# singleCellHaystack
#remotes::install_github("alexisvdb/singleCellHaystack")
library("singleCellHaystack")


######################
# Seurat
######################

#devtools::install_github('satijalab/seurat-data')
#devtools::install_github('satijalab/seurat-wrappers')
library("SeuratWrappers")


######################
# Trajectory analysis
######################

# Trajectory analysis packages are loaded when required

# Velocity 
#devtools::install_github("velocyto-team/velocyto.R")

# Dyno
#devtools::install_github("dynverse/dyno")

# Monocle
#devtools::install_github('cole-trapnell-lab/leidenbase')
#devtools::install_github('cole-trapnell-lab/monocle3')

#.rs.restartR()


####################
# Colour scheme
####################

Condition.cols <- c("turquoise", "red")

clust.cols <- c("#E41A1C", # Naive_like_1_CM
                "#A6761D", # Naive_like_2_SC
                "#8DA0CB", # Naive_like_3
                "#666666", # Cytotoxic
                "#A6D854", # Type_I_IFN
                "#984EA3", # Stimulated_1
                "#1B9E77", # Stimulated_exhausted
                "#D95F02", # Exhausted_1
                "#7570B3", # Exhausted_2
                "#E7298A", # TRM
                "#E6AB02", # gd_T_g9d2
                "#8DD3C7", # gd_T_non_g9d2
                "#FF7F00", # MAIT 
                "#E78AC3") # Proliferative

clust.names <- c("Naive_like_1_CM",
                 "Naive_like_2_SC",
                 "Naive_like_3",
                 "Cytotoxic",
                 "Type_I_IFN",
                 "Stimulated_1",
                 "Stimulated_exhausted",
                 "Exhausted_1",
                 "Exhausted_2",
                 "TRM",
                 "gd_T_g9d2",
                 "gd_T_non_g9d2",
                 "MAIT",
                 "Proliferative")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
working.dir <- getwd()

# since moving script from local to github - I want to adjust work dir to be main github dir - therefore 
setwd("..")

# create output directories
if(!dir.exists("Exported_RDS_files")){dir.create("Exported_RDS_files", recursive = T)}
if(!dir.exists("output")){dir.create("output", recursive = T)}
if(!dir.exists("output/figures")){dir.create("output/figures", recursive = T)}
if(!dir.exists("output/tables")){dir.create("output/tables", recursive = T)}
if(!dir.exists("output/QC")){dir.create("output/QC", recursive = T)}


# load saved seurat object with cluster annotations and imputation performed
if(quick.load){
  seurat.combined <- readRDS("Exported_RDS_files/seurat_combined.rds")
}


```

### Custom Functions
```{r script_custom_functions}

# Load some custom functions
source("R/scRNAseq_Functions.R")

# Export function
save.data.frame.function <- function(df, path, title){
    
    x <- as.data.frame(as.matrix(df))
    
    write.table(x, 
                paste0(path, title, ".txt"),
                sep = "\t",
                quote = FALSE)
  }
  


# Heatmap annotation
source("R/annotate_seurat_heatmap.R")

#################
# Volcano Plots
#################

# edited enhancedvolcano() function to have the default variables for aesthetics I want
source("R/Enhanced_volcano_custom_defaults.R")

clean.data <- function(input.data, group.id){
  
  output.data <- input.data %>%
    dplyr::filter(cluster == paste0(group.id))
  
  rownames(output.data) <- output.data$gene
  
  output.data <- output.data %>%
    dplyr::select(avg_logFC, p_val_adj)
  
  colnames(output.data) <- c("logFC", "FDR")
  return(output.data)
  
}



colour.points <- function(volcano.data, 
                          increase.col = "Red",
                          decreased.col = "Blue",
                          FDR.cutoff = 0.05,
                          logFC.cutoff = 0.25){
  
  # set the base colour
  keyvals <- rep('grey50', nrow(volcano.data))
  
  # set the base name/label as 'NS'
  names(keyvals) <- rep('NS', nrow(volcano.data))
  
  # modify keyvals for vars meeting FDR and LogFC threshold
  
  # Increased
  keyvals[which(volcano.data$logFC > logFC.cutoff & volcano.data$FDR < FDR.cutoff)] <- increase.col
  names(keyvals)[which(volcano.data$logFC > logFC.cutoff & volcano.data$FDR < FDR.cutoff)] <- 'Increased'
  
  # Decreased
  keyvals[which(volcano.data$logFC < -logFC.cutoff & volcano.data$FDR < FDR.cutoff)] <- decreased.col
  names(keyvals)[which(volcano.data$logFC < -logFC.cutoff & volcano.data$FDR < FDR.cutoff)] <- 'Decreased'
  
  return(keyvals)
  
}




```


### Reproducibility
```{r Reproducibility}

# Only run once to initialise 
#renv::init()

# Run snapshot to update renv.lock file 
#renv::snapshot()


# use to restore environment 
#renv::restore()


```

### Reading data

```{r reading_data}

# Load dataset from output generated by Ross, file is annotated as filtered_feature_bc_matrix

Sample1.data <- Read10X(data.dir = "Data/10X_GEX_data/Sample1_US/")
Sample2.data <- Read10X(data.dir = "Data/10X_GEX_data/Sample2_US/")
Sample3.data <- Read10X(data.dir = "Data/10X_GEX_data/Sample3_Stim/")
Sample4.data <- Read10X(data.dir = "Data/10X_GEX_data/Sample4_Stim/")


Sample1.seurat <- CreateSeuratObject(counts = Sample1.data,
                                     min.cells = 3,
                                     min.features = 200)

Sample2.seurat <- CreateSeuratObject(counts = Sample2.data,
                                     min.cells = 3,
                                     min.features = 200)

Sample3.seurat <- CreateSeuratObject(counts = Sample3.data,
                                     min.cells = 3,
                                     min.features = 200)

Sample4.seurat <- CreateSeuratObject(counts = Sample4.data,
                                     min.cells = 3,
                                     min.features = 200)

# Add metadata 
Sample1.seurat@meta.data$group <- "S1"
Sample2.seurat@meta.data$group <- "S2"
Sample3.seurat@meta.data$group <- "S3"
Sample4.seurat@meta.data$group <- "S4"

```

### Merge Seurat objects

```{r merge_seurat}


######################
# Merge samples
######################

# Unstimulated samples 1 & 2
US.seurat <- merge(x = Sample1.seurat, 
                   y = Sample2.seurat,
                   add.cell.ids = c("S1", "S2"), 
                   merge.data = FALSE, # Should normed data be merged as well or just raw data slot
                   project = "Unstimulated")

x <- ncol(Sample1.seurat) + ncol(Sample2.seurat)
paste0("expected number of cells after merge is ", ncol(US.seurat) == x)

Sample1.seurat # 14,695 genes and 3,042 cells
Sample2.seurat # 14,706 genes and 3,536 cells
US.seurat # 15,428 genes and 6,578 cells

head(US.seurat@meta.data)


# Stimulated samples 3 & 4
Stim.seurat <- merge(x = Sample3.seurat, 
                     y = Sample4.seurat,
                     add.cell.ids = c("S3", "S4"), 
                     merge.data = FALSE, # Should normed data be merged as well or just raw data slot
                     project = "Stimulated")

x <- ncol(Sample3.seurat) + ncol(Sample4.seurat)
paste0("expected number of cells after merge is ", ncol(Stim.seurat) == x)

Sample3.seurat # 14,740 genes and 3,885 cells
Sample4.seurat # 14,926 genes and 3,279 cells
Stim.seurat # 15,619 genes and 7,164 cells

head(Stim.seurat@meta.data)

# Add a condition metadata column
US.seurat@meta.data$condition <- "US"
Stim.seurat@meta.data$condition <- "Stim"





total.seurat <- merge(x = US.seurat, 
                   y = Stim.seurat,
                   merge.data = FALSE, # Should normed data be merged as well or just raw data slot
                   project = "merged_seurat")







```

### Remove large variables that are no longer needed

```{r Remove_unneeded_variables}

rm(Sample1.data, 
   Sample2.data, 
   Sample3.data, 
   Sample4.data,
   Sample1.seurat, 
   Sample2.seurat,
   Sample3.seurat,
   Sample4.seurat)


```

## QC and Normalisation {.tabset}

### QC & Normalisation

```{r QC_and_normalisation}


################################################
#  Get percentage Mitochondria gene expression
################################################

# total
total.seurat[["percent.mito"]] <- PercentageFeatureSet(total.seurat, pattern = "^MT-")


# total
total.seurat.filt <- subset(total.seurat,
                            subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mito < 10)

# Post-filtering metrics

total.seurat # 16,295 genes across 13,751 cells
total.seurat.filt # 16,295 genes across 11,827 cells


################################################
# Normalise, scale & find variable genes 
################################################


#################
# total
#################

# Normalise
total.seurat.filt <- NormalizeData(
  object = total.seurat.filt,
  assay = "RNA",
  normalization.method = "LogNormalize",
  scale.factor = 10000,
  verbose = TRUE
)

# Find variable genes
total.seurat.filt <- FindVariableFeatures(
  object = total.seurat.filt,
  verbose = TRUE
)

# Scale data
total.seurat.filt <- ScaleData(
  object = total.seurat.filt,
  assay = "RNA",
  vars.to.regress = c("nCount_RNA", "percent.mito"),
  model.use = "linear",
  do.scale = TRUE,
  do.center = TRUE,
  verbose = TRUE
)



###############################
# Remove unneeded variables
###############################

rm(US.seurat, Stim.seurat, total.seurat)


```
