---
title: "Defining T Cell States Associated with Response to Checkpoint Immunotherapy in Melanoma"
author: "DillonCorvino"
date: "17/06/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Note, this code recalculates tSNE and UMAP dims, in the input_data folder the original manuscript dim reductions (tSNE) can be used instead
# Clean code 
# Read in tSNE dims supplied with dataset
# what is DN_DP_plate?



## Dataset {.tabset}

### Dataset info
```{r Dataset_details}

# Manuscript title: "Defining T Cell States Associated with Response to Checkpoint Immunotherapy in Melanoma"
# Authors: Moshe Sade-Feldman, Keren Yizhak, Stacey L. Bjorgaard, ..., Ryan J. Sullivan, Gad Getz, Nir Hacohen
# Journal: Cell
# Year: 2018
# DOI: https://doi.org/10.1016/j.cell.2018.10.038
## Abstract: Treatment of cancer has been revolutionized by immune checkpoint blockade therapies. Despite the high rate of response in advanced melanoma, the majority of patients succumb to disease. To identify factors associated with success or fail- ure of checkpoint therapy, we profiled transcrip- tomes of 16,291 individual immune cells from 48 tumor samples of melanoma patients treated with checkpoint inhibitors. Two distinct states of CD8+ T cells were defined by clustering and associated with patient tumor regression or progression. A single transcription factor, TCF7, was visualized within CD8+ T cells in fixed tumor samples and pre- dicted positive clinical outcome in an independent cohort of checkpoint-treated patients. We delin- eated the epigenetic landscape and clonality of these T cell states and demonstrated enhanced antitumor immunity by targeting novel combina- tions of factors in exhausted cells. Our study of immune cell transcriptomes from tumors demon- strates a strategy for identifying predictors, mech- anisms, and targets for enhancing checkpoint immunotherapy.

## Cluster annotations from manuscript 
# G1 = B cell
# G2 = Plasma cell
# G3 = Monocyte/macrophage 
# G4 = DCs
# G5 = Lymphocytes
# G6 = Exhausted CD8+ T cells
# G7 = regulatory T cells
# G8 = Cytotoxicity (Lymphocytes)
# G9 = Exhausted/HS CD8+ T cell
# G10 = Memory T cells
# G11 = Lymphocytes exhausted/cell cycle


# CD8 cell cluster annotations from manuscript 
# 1 = Exhaustion / Cell cycle
# 2 = Exhaustion / heat shock protein
# 3 = Exhaustion
# 4 = Memory / Effector 
# 5 = Early activated cells
# 6 = Memory / Effector


# Smart seq 2 scRNAseq


# 32 patients
# Anti-PD1 therapy 
# 16,291 individual immune cells 
# 48 tumor samples of melanoma patients treated with checkpoint inhibitors 

# B = baseline
# P = post 
# NR = non-responder
# R = responder

# Num of samples 
# B NR = 10
# B R = 9
# P NR = 21
# P R = 8

```

## Set up {.tabset}

### Environment
```{r Environment_setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  eval = TRUE, 
  tidy = TRUE
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory

start.time <- Sys.time()
quick.load <- FALSE


###################
# Install packages
###################

pkgs <- c("remedy", "Seurat", "dplyr", 
          "rstudioapi", "cowplot", "ggplot2",
          "grid", "gridExtra", "styler", "stringr", "data.table")


for(i in 1:length(pkgs)){
  if(!require(pkgs[i], character.only = T)){
    install.packages(pkgs[i])
    require(pkgs[i], character.only = T)
  }else{
    require(pkgs[i], character.only = T)
  }
}

pkgs <- c("gplots", "fgsea", "biomaRt", 
          "clusterProfiler", "GSEABase", "org.Hs.eg.db",
          "org.Mm.eg.db")

for(i in 1:length(pkgs)){
  if(!require(pkgs[i], character.only = T)){
    BiocManager::install(pkgs[i])
    require(pkgs[i], character.only = T)
  }else{
    require(pkgs[i], character.only = T)
  }
}

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")


# create output directories
if(!dir.exists("Exported_RDS_files")){dir.create("Exported_RDS_files", recursive = T)}



```

### Reading data
```{r reading_data}
# Load the dataset
scRNAseq.data <- fread("Data/Melanoma_Sade_Feldman_Cell_2018/tpm_all_scp.txt")
scRNAseq.data <- as.data.frame(scRNAseq.data)

rownames(scRNAseq.data) <- scRNAseq.data$GENE
scRNAseq.data <- scRNAseq.data[, -"GENE"]
colnames(scRNAseq.data)
rownames(scRNAseq.data)


metadata.allcells <- fread("Data/Melanoma_Sade_Feldman_Cell_2018/cells_all_scp.txt")
metadata.allcells <- as.data.frame(metadata.allcells)

metadata.allcells <- metadata.allcells[-1, ]
colnames(metadata.allcells)
rownames(metadata.allcells)  <- metadata.allcells$NAME

metadata.allcells <- metadata.allcells[ , -1]



# Initialize the Seurat object with the raw (non-normalized data).  Keep all
# genes expressed in >= 3 cells. Keep all cells with at
# least 200 detected genes
Sade.Feldman.seurat <- CreateSeuratObject(counts = scRNAseq.data,
                                    meta.data = metadata.allcells,
                                    min.cells = 3,
                                    min.features = 200)

Sade.Feldman.seurat # 12,364 genes and 16,291 cells



```


## QC and Normalisation {.tabset}

### QC & Normalisation
```{r QC_and_normalisation}

# QC metrics commonly used
# The number of unique genes detected in each cell.
# Low-quality cells or empty droplets will often have very few genes
# Cell doublets or multiplets may exhibit an aberrantly high gene count

# Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)

# The percentage of reads that map to the mitochondrial genome
# Low-quality / dying cells often exhibit extensive mitochondrial contamination


################################################
#  Get percentage Mitochondria gene expression
################################################

Sade.Feldman.seurat[["percent.mito"]] <- PercentageFeatureSet(Sade.Feldman.seurat, pattern = "^MT-")



##################################
# Plot mitochondria percentage
##################################

# nFeature & nCount & Mito 
VlnPlot(object = Sade.Feldman.seurat, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"))

# Mito 
VlnPlot(Sade.Feldman.seurat, features = "percent.mito", y.max = 20)


########################################
# Count vs Mito & Count vs Feature
########################################

plot1 <- FeatureScatter(Sade.Feldman.seurat, feature1 = "nCount_RNA", feature2 = "percent.mito")
plot2 <- FeatureScatter(Sade.Feldman.seurat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
CombinePlots(plots = list(plot1, plot2))


############################
# Structure and metadata 
############################

head(Sade.Feldman.seurat@meta.data)
str(Sade.Feldman.seurat@meta.data)

###################
#  Filter cells
###################

# Seurat recommends removing cells with <200 or >2,500 genes or >10% mito genes

Sade.Feldman.seurat.filt <- subset(Sade.Feldman.seurat,
                             subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mito < 10)

# Remove DN_DP_Plates cells
Idents(Sade.Feldman.seurat.filt) <- Sade.Feldman.seurat.filt@meta.data$response

Sade.Feldman.seurat.filt <- subset(Sade.Feldman.seurat.filt,
                             idents = c("R", "NR"))


# Post-filtering metrics

Sade.Feldman.seurat # 12,364 genes across 16,291 cells
Sade.Feldman.seurat.filt # 12,364 genes across 12,670 cells

# Therefore
16291 - 12670
# 3621 cells were deleted

##########################################
# Visualise QC metrics after filtering
##########################################

# Before Filtering
VlnPlot(object = Sade.Feldman.seurat, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"))

# After filtering
VlnPlot(object = Sade.Feldman.seurat.filt, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"))


################################################
# Normalise, scale & find variable genes 
################################################

# Normalise
Sade.Feldman.seurat.filt <- NormalizeData(object = Sade.Feldman.seurat.filt,
                                    assay = "RNA",
                                    normalization.method = "LogNormalize",
                                    scale.factor = 10000,
                                    verbose = TRUE)

# Find variable genes
Sade.Feldman.seurat.filt <- FindVariableFeatures(object = Sade.Feldman.seurat.filt,
                                                 verbose = TRUE)

# Scale data
Sade.Feldman.seurat.filt <- ScaleData(object = Sade.Feldman.seurat.filt,
                                      assay = "RNA",
                                      vars.to.regress = c("nCount_RNA", "percent.mito"),
                                      model.use = "linear",
                                      do.scale = TRUE,
                                      do.center = TRUE,
                                      verbose = TRUE)



########################################
# Create dataset of just CD8 T cells
########################################

levels(as.factor(Sade.Feldman.seurat.filt@meta.data$isCD8))
prop.table(table(Sade.Feldman.seurat.filt@meta.data$isCD8))* 100

Idents(Sade.Feldman.seurat.filt) <- Sade.Feldman.seurat.filt@meta.data$isCD8

seurat.CD8 <- subset(Sade.Feldman.seurat.filt,
                     idents = c("TRUE"))

levels(as.factor(seurat.CD8@meta.data$isCD8))


###############################
# Remove unneeded variables
###############################

rm(Sade.Feldman.seurat, scRNAseq.data, metadata.allcells)


```

### Renaming Idents
```{r Renaming_idents}


###############
# All cells
###############

# set idents
Idents(Sade.Feldman.seurat.filt) <- Sade.Feldman.seurat.filt@meta.data$cluster_all

Sade.Feldman.seurat.filt <- RenameIdents(object = Sade.Feldman.seurat.filt,
                                   "G01" = "B_cell",
                                   "G02" = "Plasma_cell",
                                   "G03" = "Monocyte_macrophage",
                                   "G04" = "DCs",
                                   "G05" = "Lymphocytes",
                                   "G06" = "Exhausted_CD8",
                                   "G07" = "regulatory_T",
                                   "G08" = "Cytotoxicity_Lymphocytes",
                                   "G09" = "Exhausted_HS_CD8",
                                   "G10" = "Memory_T",
                                   "G11" = "Lymphocytes_exhausted_cell_cycle")


# reset levels of factor variable 
Sade.Feldman.seurat.filt@active.ident <- factor(Sade.Feldman.seurat.filt@active.ident, 
                                          levels = c("Lymphocytes",
                                                     "Cytotoxicity_Lymphocytes", 
                                                     "Exhausted_CD8", 
                                                     "Exhausted_HS_CD8",
                                                     "Memory_T",
                                                     "regulatory_T",
                                                     "B_cell",
                                                     "Plasma_cell",
                                                     "Monocyte_macrophage", 
                                                     "DCs", 
                                                     "Lymphocytes_exhausted_cell_cycle"))

# save ident as metadata variable
Sade.Feldman.seurat.filt@meta.data$cluster_named <- Idents(Sade.Feldman.seurat.filt)



###############
# CD8 only
###############

# set idents
Idents(seurat.CD8) <- seurat.CD8@meta.data$cluster_cd8


seurat.CD8 <- RenameIdents(object = seurat.CD8,
                           "1" = "Exhaustion_Cellcycle",
                           "2" = "Exhaustion_heatshock",
                           "3" = "Exhaustion",
                           "4" = "Memory_effector",
                           "5" = "Early_activated",
                           "6" = "Memory_effector")



levels(as.factor(seurat.CD8@active.ident))
                           
# save ident as metadata variable
seurat.CD8@meta.data$CD8_clusts_named <- Idents(seurat.CD8)


```
 
 
 
 
## Clustering and dim reduction {.tabset}

### PCA and cluster identificaton
```{r Cluster_data}

##############
# Run PCA
##############

# All cells
Sade.Feldman.seurat.filt <- RunPCA(Sade.Feldman.seurat.filt, 
                             npcs = 30, 
                             verbose = TRUE)

# CD8 only
seurat.CD8 <- RunPCA(seurat.CD8, 
                     npcs = 30, 
                     verbose = TRUE)

```

### Run Dim reduction
```{r Dimreduction}

########################
# Run tSNE
########################

# All cells
Sade.Feldman.seurat.filt <- RunTSNE(object = Sade.Feldman.seurat.filt,
                              reduction = "pca",
                              dims = 1:20,
                              seed.use = 42)

# CD8 only
seurat.CD8 <- RunTSNE(object = seurat.CD8,
                      reduction = "pca",
                      dims = 1:20,
                      seed.use = 42)


########################
# Run UMAP
########################

# All cells || After running UMAP optimisation results: n.neighbors = 45, min.dist = 0.5
Sade.Feldman.seurat.filt <- RunUMAP(object = Sade.Feldman.seurat.filt,
                              reduction = "pca",
                              dims = 1:20,
                              umap.method = "uwot",
                              n.neighbors = 45, # 5 to 50
                              min.dist = 0.5, # Sensible values are in the range 0.001 to 0.5
                              seed.use = 42)

# CD8 only
seurat.CD8 <- RunUMAP(object = seurat.CD8,
                      reduction = "pca",
                      dims = 1:20,
                      umap.method = "uwot",
                      n.neighbors = 45, # 5 to 50
                      min.dist = 0.5, # Sensible values are in the range 0.001 to 0.5
                      seed.use = 42)



#################################
# Save data seurat objects
#################################

# All cells
saveRDS(Sade.Feldman.seurat.filt, file = "Exported_RDS_files/Sade_Feldman_all_cells.rds")

# CD8 only
saveRDS(seurat.CD8, file = "Exported_RDS_files/Sade_Feldman_CD8_only.rds")


```
 
### Visualise dim reductions
```{r Vis_tSNE_UMAP}

##############
# All cells
##############

# Set idents
Idents(Sade.Feldman.seurat.filt) <- Sade.Feldman.seurat.filt@meta.data$cluster_named

# tSNE
TSNEPlot(object = Sade.Feldman.seurat.filt,
         label = FALSE)


# UMAP
UMAPPlot(object = Sade.Feldman.seurat.filt,
         label = FALSE)


# UMAP labelled
UMAPPlot(object = Sade.Feldman.seurat.filt,
         label = TRUE) + 
  NoLegend()


##############
# CD8 only
##############

# Set idents
Idents(seurat.CD8) <- seurat.CD8@meta.data$CD8_clusts_named

# tSNE
TSNEPlot(object = seurat.CD8,
         label = FALSE)

# UMAP
UMAPPlot(object = seurat.CD8,
         label = FALSE)


```
 
